--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -39,6 +_,22 @@
  */
 public interface Entity extends Metadatable, CommandSender, Nameable, PersistentDataHolder, net.kyori.adventure.text.event.HoverEventSource<net.kyori.adventure.text.event.HoverEvent.ShowEntity>, net.kyori.adventure.sound.Sound.Emitter, DataComponentView { // Paper
 
+    // Sakura start - entity pushed by fluid api
+    /**
+     * Gets if the entity will be pushed by fluid.
+     *
+     * @return if this entity can be pushed by fluid.
+     */
+    boolean isPushedByFluid();
+
+    /**
+     * Sets if the entity will be pushed by fluid.
+     *
+     * @param state whether entity should be pushed by fluid
+     */
+    void setPushedByFluid(boolean state);
+    // Sakura end - entity pushed by fluid api
+
     /**
      * Gets the entity's current position
      *
