--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -71,6 +_,12 @@
 @NullMarked
 public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
 
+    // Sakura start - entity tracking range modifier
+    double getTrackingRangeModifier();
+
+    void setTrackingRangeModifier(final double modifier);
+    // Sakura end - entity tracking range modifier
+
     // Paper start
     @Override
     default net.kyori.adventure.identity.Identity identity() {
