--- a/src/main/java/org/bukkit/entity/SulfurCube.java
+++ b/src/main/java/org/bukkit/entity/SulfurCube.java
@@ -13,7 +_,7 @@
  * Represents a Sulfur Cube.
  */
 @NullMarked
-public interface SulfurCube extends AbstractCubeMob, Shearable, Bucketable, Ageable {
+public interface SulfurCube extends AbstractCubeMob, Shearable, Bucketable, Ageable, me.samsuik.sakura.entity.CannonEntity { // Sakura - cannon entity api
 
     /**
      * Gets the amount of ticks until this sulfur cube explode.
