--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -54,6 +_,17 @@
     public int spawnRange = 4;
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
+    // Sakura start - configure default mob spawner properties
+    public BaseSpawner() {
+        this.minSpawnDelay = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.minSpawnDelay;
+        this.maxSpawnDelay = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.maxSpawnDelay;
+        this.spawnCount = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.spawnCount;
+        this.maxNearbyEntities = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.maxNearbyEntities;
+        this.requiredPlayerRange = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.requiredPlayerRange;
+        this.spawnRange = me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.mobSpawnerDefaults.spawnRange;
+    }
+    // Sakura start - configure default mob spawner properties
+
     public void setEntityId(final EntityType<?> type, final @Nullable Level level, final RandomSource random, final BlockPos pos) {
         this.getOrCreateNextSpawnData(level, random, pos).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(type).toString());
         this.spawnPotentials = WeightedList.of(); // CraftBukkit - SPIGOT-3496, MC-92282
@@ -89,7 +_,7 @@
         tickDelay = level.paperConfig().tickRates.mobSpawner;
         if (tickDelay == -1) { return; } // If disabled
         // Paper end - Configurable mob spawner tick rate
-        if (this.isNearPlayer(level, pos) && level.isSpawnerBlockEnabled()) {
+        if ((!level.sakuraConfig().environment.mobSpawner.requireNearbyPlayer || this.isNearPlayer(level, pos)) && level.isSpawnerBlockEnabled()) { // Sakura - configure mob spawner behaviour
             if (this.spawnDelay < -tickDelay) { // Paper - Configurable mob spawner tick rate
                 this.delay(level, pos);
             }
@@ -129,7 +_,7 @@
                                 if (!customSpawnRules.isValidPosition(spawnBlockPos, level)) {
                                     continue;
                                 }
-                            } else if (!SpawnPlacements.checkSpawnRules(entityType.get(), level, EntitySpawnReason.SPAWNER, spawnBlockPos, level.getRandom())) {
+                            } else if (level.sakuraConfig().environment.mobSpawner.checkSpawnConditions && !SpawnPlacements.checkSpawnRules(entityType.get(), level, EntitySpawnReason.SPAWNER, spawnBlockPos, level.getRandom())) { // Sakura - configure mob spawner behaviour
                                 continue;
                             }
 
@@ -157,12 +_,19 @@
                                 return;
                             }
 
-                            int nearBy = level.getEntities(
-                                    EntityTypeTest.forExactClass(entity.getClass()),
-                                    new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
-                                    EntitySelector.NO_SPECTATORS
-                                )
-                                .size();
+                            // Sakura start - configure mob spawner behaviour
+                            final int nearBy;
+                            if (level.sakuraConfig().environment.mobSpawner.ignoreEntityLimit) {
+                                nearBy = 0;
+                            } else {
+                                nearBy = level.getEntities(
+                                        EntityTypeTest.forExactClass(entity.getClass()),
+                                        new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
+                                        EntitySelector.NO_SPECTATORS
+                                    )
+                                    .size();
+                            }
+                            // Sakura end - configure mob spawner behaviour
                             if (nearBy >= this.maxNearbyEntities) {
                                 this.delay(level, pos);
                                 return;
@@ -170,8 +_,7 @@
 
                             entity.snapTo(entity.getX(), entity.getY(), entity.getZ(), random.nextFloat() * 360.0F, 0.0F);
                             if (entity instanceof Mob mob) {
-                                if (nextSpawnData.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(level, EntitySpawnReason.SPAWNER)
-                                    || !mob.checkSpawnObstruction(level)) {
+                                if (level.sakuraConfig().environment.mobSpawner.checkSpawnConditions && (nextSpawnData.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(level, EntitySpawnReason.SPAWNER) || !mob.checkSpawnObstruction(level))) { // Sakura - configure mob spawner behaviour
                                     continue;
                                 }
 
