--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -115,7 +_,7 @@
     public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
     public final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
-    private boolean tickingBlockEntities;
+    public boolean tickingBlockEntities;
     public final Thread thread;
     private final boolean isDebug;
     private int skyDarken;
@@ -165,6 +_,18 @@
         return this.paperConfig;
     }
     // Paper end - add paper world config
+    // Sakura start - sakura configuration files
+    private final me.samsuik.sakura.configuration.WorldConfiguration sakuraConfig;
+    public final me.samsuik.sakura.configuration.WorldConfiguration sakuraConfig() {
+        return this.sakuraConfig;
+    }
+    // Sakura end - sakura configuration files
+    // Sakura start - local config api
+    private final me.samsuik.sakura.configuration.local.LocalConfiguration localConfig = new me.samsuik.sakura.configuration.local.LocalConfiguration(this);
+    public final me.samsuik.sakura.configuration.local.LocalConfiguration localConfig() {
+        return this.localConfig;
+    }
+    // Sakura end - local config api
 
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     private int tileTickPosition;
@@ -833,6 +_,7 @@
         org.bukkit.World.Environment environment, // Paper
         java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Spigot - create per world config
         io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, // Paper - create paper world config
+        java.util.function.Supplier<me.samsuik.sakura.configuration.WorldConfiguration> sakuraWorldConfigCreator, // Sakura - sakura configuration files
         java.util.concurrent.Executor executor // Paper - Anti-Xray
     ) {
         // Paper start - getblock optimisations - cache world height/sections
@@ -846,6 +_,7 @@
         // Paper end - getblock optimisations - cache world height/sections
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.sakuraConfig = sakuraWorldConfigCreator.get(); // Sakura - sakura configuration files
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, generator, biomeProvider, environment);
 
