--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -54,6 +_,8 @@
     public float yield;
     // CraftBukkit end
     public boolean excludeSourceFromDamage = true; // Paper - Allow explosions to damage source
+    private final boolean consistentExplosionRadius; // Sakura - consistent explosion radius
+    protected final boolean brokenPaperBehaviour; // Sakura - broken paper explosion behaviour
     // Paper start - collisions optimisations
     private static final double[] CACHED_RAYS;
     static {
@@ -99,7 +_,7 @@
     protected ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[] directMappedBlockCache;
     protected BlockPos.MutableBlockPos mutablePos;
 
-    private ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache getOrCacheExplosionBlock(final int x, final int y, final int z,
+    protected ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache getOrCacheExplosionBlock(final int x, final int y, final int z,
                                                                                                     final long key, final boolean calculateResistance) {
         ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache ret = this.blockCache.get(key);
         if (ret != null) {
@@ -320,6 +_,8 @@
         this.yield = this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F;
         this.yield = Double.isFinite(this.yield) ? this.yield : 0; // Paper - Don't allow infinite default yields
         // Paper end - add yield
+        this.consistentExplosionRadius = level.localConfig().at(this.center).consistentExplosionRadius; // Sakura - consistent explosion radius
+        this.brokenPaperBehaviour = level.localConfig().at(this.center).brokenPaperExplosionBehaviour; // Sakura - broken paper explosion behaviour
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(final @Nullable Entity source) {
@@ -408,8 +_,7 @@
             final double incZ = CACHED_RAYS[ray + 2];
 
             ray += 3;
-
-            float power = this.radius * (0.7F + this.level.getRandom().nextFloat() * 0.6F);
+            float power = this.radius * (0.7F + (this.consistentExplosionRadius ? 0.7F : this.level.random.nextFloat()) * 0.6F); // Sakura - consistent explosion radius
 
             do {
                 final int blockX = Mth.floor(currX);
@@ -605,7 +_,20 @@
             this.level.getBlockState(pos).onExplosionHit(this.level, pos, this, (stackx, position) -> addOrAppendStack(stacks, stackx, position));
         }
 
+        // Sakura start - configure explosions dropping items
+        if (!this.level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+            targetBlocks.clear();
+        }
+        // Sakura end - configure explosions dropping items
+
+        // Sakura start - configure items that can that drop from explosions
+        final Set<net.minecraft.world.item.Item> explosionItemDrops = this.level.sakuraConfig().entity.items.explosionItemDrops.items;
+        final boolean whitelist = this.level.sakuraConfig().entity.items.explosionItemDrops.whitelistOverBlacklist;
         for (ServerExplosion.StackCollector stack : stacks) {
+            if (explosionItemDrops.contains(stack.stack.getItem()) != whitelist) {
+                continue;
+            }
+            // Sakura end - configure items that can that drop from explosions
             Block.popResource(this.level, stack.pos, stack.stack);
         }
     }
@@ -733,7 +_,7 @@
 
     // Paper start - Optimize explosions
     protected float getBlockDensity(Vec3 vec3d, Entity entity) {
-        if (!this.level.paperConfig().environment.optimizeExplosions) {
+        if (!this.brokenPaperBehaviour) { // Sakura - broken paper explosion behaviour
             return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
         }
         CacheKey key = new CacheKey(this, entity.getBoundingBox());
