--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -60,6 +_,7 @@
     public float yield;
     // CraftBukkit end
     public boolean excludeSourceFromDamage = true; // Paper - Allow explosions to damage source
+    private final boolean consistentRadius; // Sakura - consistent explosion radius
     // Paper start - collisions optimisations
     private static final double[] CACHED_RAYS;
     static {
@@ -323,6 +_,7 @@
         this.damageSource = damageSource == null ? level.damageSources().explosion(this) : damageSource;
         this.damageCalculator = damageCalculator == null ? this.makeDamageCalculator(source) : damageCalculator;
         this.yield = this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F; // CraftBukkit
+        this.consistentRadius = level.localConfig().config(BlockPos.containing(this.center)).consistentRadius; // Sakura - consistent explosion radius
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity entity) {
@@ -411,8 +_,7 @@
             final double incZ = CACHED_RAYS[ray + 2];
 
             ray += 3;
-
-            float power = this.radius * (0.7F + this.level.random.nextFloat() * 0.6F);
+            float power = this.radius * (0.7F + (this.consistentRadius ? 0.7F : this.level.random.nextFloat()) * 0.6F); // Sakura - consistent explosion radius
 
             do {
                 final int blockX = Mth.floor(currX);
@@ -619,7 +_,20 @@
                 .onExplosionHit(this.level, blockPos, this, (itemStack, blockPos1) -> addOrAppendStack(list, itemStack, blockPos1));
         }
 
+        // Sakura start - configure explosions dropping items
+        if (!this.level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+            list.clear();
+        }
+        // Sakura end - configure explosions dropping items
+
+        // Sakura start - configure items that can that drop from explosions
+        final Set<net.minecraft.world.item.Item> explosionItemDrops = this.level.sakuraConfig().entity.items.explosionItemDrops.items;
+        final boolean whitelist = this.level.sakuraConfig().entity.items.explosionItemDrops.whitelistOverBlacklist;
         for (ServerExplosion.StackCollector stackCollector : list) {
+            if (explosionItemDrops.contains(stackCollector.stack.getItem()) != whitelist) {
+                continue;
+            }
+            // Sakura start - configure items that can that drop from explosions
             Block.popResource(this.level, stackCollector.pos, stackCollector.stack);
         }
     }
