--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSulfurCube.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSulfurCube.java
@@ -17,6 +_,18 @@
         super(server, entity);
     }
 
+    // Sakura start - height parity api
+    @Override
+    public final void setHeightParity(final boolean heightParity) {
+        this.getHandle().heightParity = heightParity;
+    }
+
+    @Override
+    public final boolean getHeightParity() {
+        return this.getHandle().heightParity;
+    }
+    // Sakura end - height parity api
+
     @Override
     public net.minecraft.world.entity.monster.cubemob.SulfurCube getHandle() {
         return (net.minecraft.world.entity.monster.cubemob.SulfurCube) this.entity;
