--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -254,6 +_,18 @@
         this.firstPlayed = System.currentTimeMillis();
     }
 
+    // Sakura start - entity tracking range modifier
+    @Override
+    public final double getTrackingRangeModifier() {
+        return this.getHandle().trackingRangeModifier * 100.0;
+    }
+
+    @Override
+    public final void setTrackingRangeModifier(final double modifier) {
+        this.getHandle().trackingRangeModifier = modifier / 100.0;
+    }
+    // Sakura end - entity tracking range modifier
+
     @Override
     public ServerPlayer getHandle() {
         return (ServerPlayer) this.entity;
