--- a/net/minecraft/world/entity/projectile/AbstractThrownPotion.java
+++ b/net/minecraft/world/entity/projectile/AbstractThrownPotion.java
@@ -39,6 +_,26 @@
         super(entityType, x, y, z, level, item);
     }
 
+    // Sakura start - configure potion mechanics
+    @Override
+    public void shoot(final double x, final double y, final double z, final float speed, final float divergence) { // may be overridden by plugins
+        super.shoot(x, y, z, speed, divergence);
+
+        final net.minecraft.world.phys.Vec3 movement = this.getDeltaMovement();
+        final double moveX = movement.x * this.level().sakuraConfig().entity.thrownPotion.horizontalSpeed;
+        final double moveY = movement.y * this.level().sakuraConfig().entity.thrownPotion.verticalSpeed;
+        final double moveZ = movement.z * this.level().sakuraConfig().entity.thrownPotion.horizontalSpeed;
+
+        this.setDeltaMovement(moveX, moveY, moveZ);
+    }
+
+    @Override
+    protected boolean checkLeftOwner() { // may be overridden by plugins
+        return this.level().sakuraConfig().entity.thrownPotion.allowBreakingInsideEntities && this.tickCount >= 5
+            || super.checkLeftOwner();
+    }
+    // Sakura end - configure potion mechanics
+
     @Override
     protected double getDefaultGravity() {
         return 0.05;
