--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -46,10 +_,17 @@
     private @Nullable Entity lastDeflectedBy;
     protected boolean hitCancelled = false; // CraftBukkit
 
+    // Sakura start - configure potion mechanics
+    protected boolean breakInsideOwner() {
+        return false;
+    }
+    // Sakura end - configure potion mechanics
+
     protected Projectile(final EntityType<? extends Projectile> type, final Level level) {
         super(type, level);
     }
 
+
     protected void setOwner(final @Nullable EntityReference<Entity> owner) {
         this.owner = owner;
     }
@@ -425,7 +_,7 @@
                 }
             }
             // Paper end - Cancel hit for vanished entities
-            return owner == null || this.leftOwner || !owner.isPassengerOfSameVehicle(entity);
+            return owner == null || this.leftOwner || this.breakInsideOwner() || !owner.isPassengerOfSameVehicle(entity); // Sakura - configure potion mechanics
         }
     }
 
