--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -51,9 +_,15 @@
             vec3 = hitResult.getLocation();
         }
 
-        HitResult entityHitResult = getEntityHitResult(
-            level, source, pos, vec3, source.getBoundingBox().expandTowards(deltaMovement).inflate(1.0), filter, margin
-        );
+        // Sakura start - configure potion mechanics
+        final AABB movementAABB = source.getBoundingBox().expandTowards(deltaMovement).inflate(1.0);
+        final HitResult entityHitResult;
+        if (level.sakuraConfig().entity.thrownPotion.allowBreakingInsideEntities && source instanceof AbstractThrownPotion) {
+            entityHitResult = getEntityHitResult(source, pos, vec3, movementAABB, filter, margin);
+        } else {
+            entityHitResult = getEntityHitResult(level, source, pos, vec3, movementAABB, filter, margin);
+        }
+        // Sakura end - configure potion mechanics
         if (entityHitResult != null) {
             hitResult = entityHitResult;
         }
