--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -75,7 +_,17 @@
             to = hitResult.getLocation();
         }
 
-        HitResult entityHit = getEntityHitResult(level, source, from, to, source.getBoundingBox().expandTowards(delta).inflate(1.0), matching, entityMargin);
+        // Sakura start - configure potion mechanics
+        final AABB collisionBox = source.getBoundingBox().expandTowards(delta).inflate(1.0);
+        final HitResult entityHit;
+        if (level.sakuraConfig().entity.thrownPotion.allowBreakingInsideEntities
+            && source instanceof net.minecraft.world.entity.projectile.throwableitemprojectile.AbstractThrownPotion) {
+            entityHit = getEntityHitResult(source, from, to, collisionBox, matching, entityMargin);
+        } else {
+            entityHit = getEntityHitResult(level, source, from, to, collisionBox, matching, entityMargin);
+        }
+        // Sakura end - configure potion mechanics
+
         if (entityHit != null) {
             hitResult = entityHit;
         }
