--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -46,10 +_,17 @@
     private @Nullable Entity lastDeflectedBy;
     protected boolean hitCancelled = false; // CraftBukkit
 
+    // Sakura start - configure potion mechanics
+    protected boolean breakInsideOwner() {
+        return false;
+    }
+    // Sakura end - configure potion mechanics
+
     protected Projectile(final EntityType<? extends Projectile> type, final Level level) {
         super(type, level);
     }
 
+
     protected void setOwner(final @Nullable EntityReference<Entity> owner) {
         this.owner = owner;
     }
@@ -161,7 +_,20 @@
         if (!sourceMovement.isFinite()) {
             sourceMovement = Vec3.ZERO;
         }
-        if (!source.level().paperConfig().misc.disableRelativeProjectileVelocity) {
+
+        // Sakura start - relative potion velocity
+        boolean relativeVelocity = true;
+        if (source.level().paperConfig().misc.disableRelativeProjectileVelocity) {
+            relativeVelocity = false;
+        }
+
+        if (source.level().sakuraConfig().entity.thrownPotion.disableRelativePotionVelocity
+            && this instanceof net.minecraft.world.entity.projectile.throwableitemprojectile.AbstractThrownPotion) {
+            relativeVelocity = false;
+        }
+
+        if (relativeVelocity) {
+            // Sakura end - relative potion velocity
         this.setDeltaMovement(this.getDeltaMovement().add(sourceMovement.x, source.onGround() ? 0.0 : sourceMovement.y, sourceMovement.z));
         }
         // Paper end - allow disabling relative velocity
@@ -425,7 +_,7 @@
                 }
             }
             // Paper end - Cancel hit for vanished entities
-            return owner == null || this.leftOwner || !owner.isPassengerOfSameVehicle(entity);
+            return owner == null || this.leftOwner || this.breakInsideOwner() || !owner.isPassengerOfSameVehicle(entity); // Sakura - configure potion mechanics
         }
     }
 
