--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -13,6 +_,12 @@
 public abstract class ThrowableProjectile extends Projectile {
     private static final float MIN_CAMERA_DISTANCE_SQUARED = 12.25F;
 
+    // Sakura start - configure enderpearl mechanics
+    protected net.minecraft.world.level.ClipContext.Block getClipType() {
+        return net.minecraft.world.level.ClipContext.Block.COLLIDER;
+    }
+    // Sakura end - configure enderpearl mechanics
+
     protected ThrowableProjectile(final EntityType<? extends ThrowableProjectile> type, final Level level) {
         super(type, level);
     }
@@ -47,7 +_,7 @@
         this.handleFirstTickBubbleColumn();
         this.applyGravity();
         this.applyInertia();
-        HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
+        HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity, this.getClipType()); // Sakura - configure enderpearl mechanics
         Vec3 newPosition;
         if (result.getType() != HitResult.Type.MISS) {
             newPosition = result.getLocation();
