--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -39,6 +_,15 @@
         super(EntityType.ENDER_PEARL, owner, level, item);
     }
 
+    // Sakura start - enderpearls use outline for collision
+    @Override
+    protected final net.minecraft.world.level.ClipContext.Block getClipType() {
+        return this.level().sakuraConfig().entity.enderPearl.useOutlineForCollision
+            ? net.minecraft.world.level.ClipContext.Block.OUTLINE
+            : super.getClipType();
+    }
+    // Sakura end - enderpearls use outline for collision
+
     @Override
     public Item getDefaultItem() {
         return Items.ENDER_PEARL;
@@ -103,6 +_,15 @@
             if (owner != null && isAllowedToTeleportOwner(owner, serverLevel)) {
                 Vec3 vec3 = this.oldPosition();
                 if (owner instanceof ServerPlayer serverPlayer) {
+                    // Sakura start - prevent ender pearls teleporting inside blocks
+                    if (this.level().sakuraConfig().entity.enderPearl.preventTeleportingInsideBlocks && serverPlayer.insideAnyCollisionAt(serverLevel, vec3)) {
+                        vec3 = Vec3.atBottomCenterOf(this.blockPosition()); // teleportation position
+                        if (serverPlayer.insideAnyCollisionAt(serverLevel, vec3)) {
+                            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                            return;
+                        }
+                    }
+                    // Sakura end - prevent ender pearls teleporting inside blocks
                     if (serverPlayer.connection.isAcceptingMessages()) {
                         // CraftBukkit start
                         // Store pre teleportation position as the teleport has been moved up.
