--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -118,6 +_,21 @@
             if (owner != null && isAllowedToTeleportOwner(owner, serverLevel)) {
                 Vec3 vec3 = this.oldPosition();
                 if (owner instanceof ServerPlayer serverPlayer) {
+                    // Sakura start - prevent ender pearls teleporting inside blocks
+                    if (this.level().sakuraConfig().entity.enderPearl.preventTeleportingInsideBlocks && serverPlayer.insideAnyCollisionAt(serverLevel, vec3)) {
+                        for (int offset = 0;; ++offset) {
+                            vec3 = Vec3.atBottomCenterOf(BlockPos.containing(vec3.subtract(0, offset, 0))); // teleportation position
+                            if (!serverPlayer.insideAnyCollisionAt(serverLevel, vec3)) {
+                                break;
+                            }
+
+                            if (offset == 1) {
+                                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                                return;
+                            }
+                        }
+                    }
+                    // Sakura end - prevent ender pearls teleporting inside blocks
                     if (serverPlayer.connection.isAcceptingMessages()) {
                         // CraftBukkit start
                         // Store pre teleportation position as the teleport has been moved up.
@@ -163,6 +_,11 @@
                     if (entity != null) {
                         entity.resetFallDistance();
                     }
+                    // Sakura start - prevent ender pearls teleporting inside blocks
+                    if (entity instanceof net.minecraft.world.entity.player.Player player) {
+                        player.updatePlayerPose();
+                    }
+                    // Sakura end - prevent ender pearls teleporting inside blocks
 
                     this.playSound(serverLevel, vec3);
                 }
@@ -181,6 +_,13 @@
             return entity.canUsePortal(true);
         }
     }
+
+    // Sakura start - enderpearls use outline for collision
+    @Override
+    protected net.minecraft.world.level.ClipContext.Block getClipType() {
+        return this.level().sakuraConfig().entity.enderPearl.useOutlineForCollision ? net.minecraft.world.level.ClipContext.Block.OUTLINE : super.getClipType();
+    }
+    // Sakura end - enderpearls use outline for collision
 
     @Override
     public void tick() {
