--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -22,6 +_,12 @@
         this.setPos(x, y, z);
     }
 
+    // Sakura start - enderpearls use outline for collision
+    protected net.minecraft.world.level.ClipContext.Block getClipType() {
+        return net.minecraft.world.level.ClipContext.Block.COLLIDER;
+    }
+    // Sakura end - enderpearls use outline for collision
+
     @Override
     public boolean shouldRenderAtSqrDistance(final double distance) {
         if (this.tickCount < 2 && distance < 12.25) {
@@ -47,7 +_,7 @@
         this.handleFirstTickBubbleColumn();
         this.applyGravity();
         this.applyInertia();
-        HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
+        HitResult hitResultOnMoveVector = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity, this.getClipType()); // Sakura - enderpearls use outline for collision
         Vec3 newPosition;
         if (result.getType() != HitResult.Type.MISS) {
             newPosition = result.getLocation();
