--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -548,6 +_,8 @@
         }
     }
     // Paper end - optimise entity tracker
+    public boolean pushedByFluid = true; // Sakura - entity pushed by fluid api
+    public boolean cannonEntity = false; // Sakura - add cannon entity field
 
     public Entity(final EntityType<?> type, final Level level) {
         this.type = type;
@@ -1264,8 +_,12 @@
                     }
                 }
 
-                float blockSpeedFactor = this.getBlockSpeedFactor();
-                this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
+                // Sakura start - only apply block speed factor while moving
+                if (movement.x != 0.0 || movement.z != 0.0) {
+                    float blockSpeedFactor = this.getBlockSpeedFactor();
+                    this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
+                }
+                // Sakura end - only apply block speed factor while moving
                 profiler.pop();
             }
         }
@@ -4308,7 +_,7 @@
     }
 
     public boolean isPushedByFluid() {
-        return true;
+        return this.pushedByFluid; // Sakura - entity pushed by fluid api
     }
 
     public static double getViewScale() {
