--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -405,6 +_,7 @@
         return this.dimensions.makeBoundingBox(x, y, z);
     }
     // Paper end
+    public boolean pushedByFluid = true; // Sakura - entity pushed by fluid api
 
     public Entity(EntityType<?> entityType, Level level) {
         this.type = entityType;
@@ -3749,7 +_,7 @@
     }
 
     public boolean isPushedByFluid() {
-        return true;
+        return this.pushedByFluid; // Sakura - entity pushed by fluid api
     }
 
     public static double getViewScale() {
