--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -534,6 +_,7 @@
         }
     }
     // Paper end - optimise entity tracker
+    public boolean pushedByFluid = true; // Sakura - entity pushed by fluid api
 
     public Entity(EntityType<?> type, Level level) {
         this.type = type;
@@ -4294,7 +_,7 @@
     }
 
     public boolean isPushedByFluid() {
-        return true;
+        return this.pushedByFluid; // Sakura - entity pushed by fluid api
     }
 
     public static double getViewScale() {
