--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -757,7 +_,7 @@
     protected final void serverAiStep() {
         this.noActionTime++;
         // Paper start - Allow nerfed mobs to jump and float
-        if (!this.aware) {
+        if (!this.aware || this.level().sakuraConfig().entity.disableMobAi) { // Sakura - add option to disable entity ai
             if (this.goalFloat != null) {
                 if (this.goalFloat.canUse()) this.goalFloat.tick();
                 this.getJumpControl().tick();
