--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -151,6 +_,25 @@
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
+    // Sakura start - simplify nerfed mob logic
+    private boolean isAwareOnGroundOrSwimming() {
+        return this.aware
+            || this.verticalCollision
+            || this.isInWater()
+            || this.getDeltaMovement().y > 0.0;
+    }
+
+    @Override
+    protected boolean updateInWaterStateAndDoFluidPushing() {
+        return this.isAwareOnGroundOrSwimming() && super.updateInWaterStateAndDoFluidPushing();
+    }
+
+    @Override
+    public boolean isInWall() {
+        return this.isAwareOnGroundOrSwimming() && super.isInWall();
+    }
+    // Sakura end - simplify nerfed mob logic
+
     protected Mob(final EntityType<? extends Mob> type, final Level level) {
         super(type, level);
         this.goalSelector = new GoalSelector();
@@ -164,6 +_,11 @@
         if (level instanceof ServerLevel) {
             this.registerGoals();
         }
+        // Sakura start - add disable mob ai
+        if (this.level().sakuraConfig().entity.disableMobAi) {
+            this.aware = false;
+        }
+        // Sakura end - end disable mob ai
     }
 
     protected void registerGoals() {
@@ -216,6 +_,7 @@
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
+        if (!this.aware) return; // Sakura - fix nerfed mobs ticking goals while inactive
         if (this.goalSelector.inactiveTick()) {
             this.goalSelector.tick();
         }
