--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -164,6 +_,11 @@
         if (level instanceof ServerLevel) {
             this.registerGoals();
         }
+        // Sakura start - add disable mob ai
+        if (this.level().sakuraConfig().entity.disableMobAi) {
+            this.aware = false;
+        }
+        // Sakura end - end disable mob ai
     }
 
     // CraftBukkit start
@@ -221,6 +_,7 @@
     @Override
     public void inactiveTick() {
         super.inactiveTick();
+        if (!this.aware) return; // Sakura - fix nerfed mobs ticking goals while inactive
         if (this.goalSelector.inactiveTick()) {
             this.goalSelector.tick();
         }
