--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -310,12 +_,18 @@
 
     public void fillBiomesFromNoise(BiomeResolver biomeResolver, Climate.Sampler climateSampler, int x, int y, int z) {
         PalettedContainer<Holder<Biome>> palettedContainer = this.biomes.recreate();
+        Holder<Biome> biome = null; // Sakura - calculate biome noise for each chunk section
         int i = 4;
 
         for (int i1 = 0; i1 < 4; i1++) {
             for (int i2 = 0; i2 < 4; i2++) {
                 for (int i3 = 0; i3 < 4; i3++) {
-                    palettedContainer.getAndSetUnchecked(i1, i2, i3, biomeResolver.getNoiseBiome(x + i1, y + i2, z + i3, climateSampler));
+                    // Sakura start - calculate biome noise once for each chunk section
+                    if (biome == null || !me.samsuik.sakura.configuration.GlobalConfiguration.get().environment.calculateBiomeNoiseOncePerChunkSection) {
+                        biome = biomeResolver.getNoiseBiome(x + i1, y + i2, z + i3, climateSampler);
+                    }
+                    palettedContainer.getAndSetUnchecked(i1, i2, i3, biome);
+                    // Sakura end - calculate biome noise once for each chunk section
                 }
             }
         }
