--- a/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
+++ b/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
@@ -418,6 +_,7 @@
             this.contactDamages.clear();
             this.knockbackModifier = SulfurCubeArchetype.DEFAULT_KNOCKBACK_MODIFIERS;
             this.soundSettings = SulfurCubeArchetype.DEFAULT_SOUND_SETTINGS;
+            this.cannonEntity = false; // Sakura - add cannon entity field
 
             for (SulfurCubeArchetype archetype : this.matchingArchetypes(current)) {
                 if (archetype.buoyant()) {
@@ -426,6 +_,7 @@
 
                 if (archetype.explosion().isPresent()) {
                     this.explosionData = archetype.explosion();
+                    this.cannonEntity = true; // Sakura - add cannon entity field
                 }
 
                 if (archetype.contactDamage().isPresent()) {
