--- /dev/null
+++ b/me/samsuik/sakura/SakuraFeatureHooks.java
@@ -1,0 +_,27 @@
+package me.samsuik.sakura;
+
+import io.papermc.paper.configuration.type.Duration;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.Identifier;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.level.Level;
+import org.bukkit.command.Command;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+public final class SakuraFeatureHooks {
+    public static void setupCommands(final Map<String, Command> commands) {
+    }
+
+    public static void setDurableMaterialExpiration(final Identifier worldIdentifier, final Duration duration) {
+    }
+
+    private static Optional<Level> getLevel(final Identifier worldIdentifier) {
+        final ResourceKey<Level> dimensionKey = ResourceKey.create(Registries.DIMENSION, worldIdentifier);
+        final Level level = MinecraftServer.getServer().getLevel(dimensionKey);
+        return Optional.ofNullable(level);
+    }
+}
