--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -300,6 +_,12 @@
         ).isValid();
     }
     // Paper end
+    // Sakura start - local config api
+    @Override
+    public final me.samsuik.sakura.configuration.local.LocalConfigurationAccessor localConfig() {
+        return this.getHandle().localConfig();
+    }
+    // Sakura end - local config api
 
     private static final Random rand = new Random();
 
