--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -292,6 +_,12 @@
         ).isValid();
     }
     // Paper end
+    // Sakura start - local config and property storage
+    @Override
+    public final me.samsuik.sakura.local.storage.LocalStorageHandler getStorageHandler() {
+        return this.getHandle().localConfig();
+    }
+    // Sakura end - local config and property storage
 
     private static final Random rand = new Random();
 
