--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -277,6 +_,11 @@
         this.server.spark.registerCommandBeforePlugins(this.server); // Paper - spark
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
+        // Sakura start - sakura configuration files
+        sakuraConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        sakuraConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        me.samsuik.sakura.command.SakuraCommands.registerCommands(this);
+        // Sakura end - sakura configuration files
 
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
