From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ca1c6c4c36046fd71a0495c2cb23912d5a34909a..fceb393ba0bca6f6ce0daa3b379dfecafc5c5863 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2450,6 +2450,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             handle.expToDrop = data.getIntOr("expToDrop", 0);
             handle.keepLevel = data.getBooleanOr("keepLevel", false);
         });
+        // Sakura start - client visibility settings; load from nbt
+        input.child("sakura").ifPresent(sakura -> {
+            this.getHandle().visibilitySettings.loadData(sakura);
+        });
+        // Sakura end - client visibility settings; load from nbt
     }
 
     public void setExtraData(ValueOutput output) {
@@ -2471,6 +2476,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         paper.putLong("LastLogin", handle.loginTime);
         paper.putLong("LastSeen", System.currentTimeMillis());
         // Paper end
+        // Sakura start - client visibility settings; save to nbt
+        final ValueOutput sakura = output.child("sakura");
+        this.getHandle().visibilitySettings.saveData(sakura);
+        // Sakura end - client visibility settings; save to nbt
     }
 
     @Override
@@ -3175,6 +3184,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         return this.getHandle().allowsListing();
     }
 
+    // Sakura start - client visibility settings; api
+    @Override
+    public final me.samsuik.sakura.player.visibility.VisibilitySettings getVisibility() {
+        return this.getHandle().visibilitySettings;
+    }
+    // Sakura end - client visibility settings; api
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
