From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 13f6d27edf12628a9e3d7891ad0573aeffd1bcde..9c3d6bad289f11d7c450485557f9d9deb63887b3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2383,6 +2383,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             handle.expToDrop = data.getIntOr("expToDrop", 0);
             handle.keepLevel = data.getBooleanOr("keepLevel", false);
         });
+        // Sakura start - client visibility settings; load from nbt
+        CompoundTag sakuraTag = tag.getCompoundOrEmpty("sakura");
+        this.getHandle().visibilitySettings.loadData(sakuraTag);
+        // Sakura end - client visibility settings; load from nbt
     }
 
     public void setExtraData(CompoundTag tag) {
@@ -2412,6 +2416,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         paper.putLong("LastLogin", handle.loginTime);
         paper.putLong("LastSeen", System.currentTimeMillis());
         // Paper end
+        // Sakura start - client visibility settings; save to nbt
+        CompoundTag sakuraTag = tag.getCompoundOrEmpty("sakura");
+        this.getHandle().visibilitySettings.saveData(sakuraTag);
+        tag.put("sakura", sakuraTag);
+        // Sakura end - client visibility settings; save to nbt
     }
 
     @Override
@@ -3072,6 +3081,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return this.getHandle().allowsListing();
     }
 
+    // Sakura start - client visibility settings; api
+    @Override
+    public final me.samsuik.sakura.player.visibility.VisibilitySettings getVisibility() {
+        return this.getHandle().visibilitySettings;
+    }
+    // Sakura end - client visibility settings; api
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
