From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 09e9006aca1803c95938f288919fd8143647c8ef..d1aa3ccbea754abab8ec939875adb5d384e29e97 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2401,6 +2401,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
                 handle.keepLevel = data.getBoolean("keepLevel");
             }
         }
+
+        // Sakura start - client visibility settings; load from nbt
+        if (nbttagcompound.contains("sakura", 10)) {
+            CompoundTag sakuraTag = nbttagcompound.getCompound("sakura");
+            this.getHandle().visibilitySettings.loadData(sakuraTag);
+        }
+        // Sakura end - client visibility settings; load from nbt
     }
 
     public void setExtraData(CompoundTag nbttagcompound) {
@@ -2430,6 +2437,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         paper.putLong("LastLogin", handle.loginTime);
         paper.putLong("LastSeen", System.currentTimeMillis());
         // Paper end
+        // Sakura start - client visibility settings; save to nbt
+        CompoundTag sakuraTag = nbttagcompound.getCompound("sakura");
+        this.getHandle().visibilitySettings.saveData(sakuraTag);
+        nbttagcompound.put("sakura", sakuraTag);
+        // Sakura end - client visibility settings; save to nbt
     }
 
     @Override
@@ -3088,6 +3100,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return this.getHandle().allowsListing();
     }
 
+    // Sakura start - client visibility settings; api
+    @Override
+    public final me.samsuik.sakura.player.visibility.VisibilitySettings getVisibility() {
+        return this.getHandle().visibilitySettings;
+    }
+    // Sakura end - client visibility settings; api
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
