From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 9 Sep 2023 18:39:15 +0100
Subject: [PATCH] Merge Cannon Entities


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
index e77b06c3ef9b534a8f4f160d077cf4339ee7622d..a33e91a5b35109415b8140721469217c336dfc00 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
@@ -25,6 +25,27 @@ public class CraftFallingBlock extends CraftEntity implements FallingBlock {
         return this.getHandle().heightParity;
     }
     // Sakura end - falling block height parity api
+    // Sakura start - merge cannon entities
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeLevel getMergeLevel() {
+        return this.getHandle().getMergeEntityData().mergeLevel;
+    }
+
+    @Override
+    public final void setMergeLevel(me.samsuik.sakura.entity.merge.MergeLevel level) {
+        this.getHandle().getMergeEntityData().mergeLevel = level;
+    }
+
+    @Override
+    public final int getStacked() {
+        return this.getHandle().getMergeEntityData().count;
+    }
+
+    @Override
+    public final void setStacked(int stacked) {
+        this.getHandle().getMergeEntityData().count = stacked;
+    }
+    // Sakura end - merge cannon entities
 
     @Override
     public FallingBlockEntity getHandle() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
index eea78db7f12f5976077663a9ff0c0e188ffa8849..2e7803bd1f3411e75d8a51e03276b00469496f3b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
@@ -12,6 +12,28 @@ public class CraftTNTPrimed extends CraftEntity implements TNTPrimed {
         super(server, entity);
     }
 
+    // Sakura start - merge cannon entities
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeLevel getMergeLevel() {
+        return this.getHandle().getMergeEntityData().mergeLevel;
+    }
+
+    @Override
+    public final void setMergeLevel(me.samsuik.sakura.entity.merge.MergeLevel level) {
+        this.getHandle().getMergeEntityData().mergeLevel = level;
+    }
+
+    @Override
+    public final int getStacked() {
+        return this.getHandle().getMergeEntityData().count;
+    }
+
+    @Override
+    public final void setStacked(int stacked) {
+        this.getHandle().getMergeEntityData().count = stacked;
+    }
+    // Sakura end - merge cannon entities
+
     @Override
     public PrimedTnt getHandle() {
         return (PrimedTnt) this.entity;
