--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -46,6 +_,11 @@
     private float progressO = 0.0F;
     private long lastTicked;
     private int deathTicks;
+    // Sakura start - cache moving block entity collision shape
+    private VoxelShape collisionShape = Shapes.empty();
+    private @org.jspecify.annotations.Nullable Direction shapeDirection = null;
+    private float shapeProgress = Float.MIN_VALUE;
+    // Sakura end - cache moving block entity collision shape
 
     public PistonMovingBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityTypes.PISTON, worldPosition, blockState);
@@ -368,6 +_,18 @@
     }
 
     public VoxelShape getCollisionShape(final BlockGetter level, final BlockPos pos) {
+        // Sakura start - cache moving block entity collision shape
+        final Direction direction = NOCLIP.get();
+        if (this.progress == this.shapeProgress && direction == this.shapeDirection) {
+            return this.collisionShape;
+        } else {
+            this.shapeProgress = this.progress;
+            this.shapeDirection = direction;
+            return this.collisionShape = this.createCollisionShape(level, pos);
+        }
+    }
+    private VoxelShape createCollisionShape(final BlockGetter level, final BlockPos pos) {
+        // Sakura end - cache moving block entity collision shape
         VoxelShape pistonHeadShape;
         if (!this.extending && this.isSourcePiston && this.movedState.getBlock() instanceof PistonBaseBlock) {
             pistonHeadShape = this.movedState.setValue(PistonBaseBlock.EXTENDED, true).getCollisionShape(level, pos);
