--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -389,6 +_,14 @@
         for (int i = toPush.size() - 1; i >= 0; i--) {
             // Paper start - fix a variety of piston desync dupes
             boolean allowDesync = io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication;
+            // Sakura start - configure tnt duplication
+            if (level.sakuraConfig().technical.allowTntDuplication) {
+                final BlockState movedState = toPushShapes.get(i);
+                if (movedState.is(Blocks.TNT) || movedState.getBlock() instanceof net.minecraft.world.level.block.BaseCoralPlantTypeBlock) {
+                    allowDesync = true;
+                }
+            }
+            // Sakura end - configure tnt duplication
             BlockPos oldPos = toPush.get(i);
             BlockPos pos = oldPos;
             BlockState blockState = allowDesync ? level.getBlockState(oldPos) : null;
