--- a/src/main/java/org/bukkit/craftbukkit/util/ApiVersion.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/ApiVersion.java
@@ -17,7 +_,7 @@
 
     static {
         versions = new HashMap<>();
-        CURRENT = getOrCreateVersion("1.21.10");
+        CURRENT = getOrCreateVersion("1.21.9");
         FLATTENING = getOrCreateVersion("1.13");
         FIELD_NAME_PARITY = getOrCreateVersion("1.20.5");
         ABSTRACT_COW = getOrCreateVersion("1.21.5");
