--- a/net/minecraft/world/entity/monster/Enderman.java
+++ b/net/minecraft/world/entity/monster/Enderman.java
@@ -87,6 +_,7 @@
     private static final UniformInt PERSISTENT_ANGER_TIME = TimeUtil.rangeOfSeconds(20, 39);
     private long persistentAngerEndTime;
     private @Nullable EntityReference<LivingEntity> persistentAngerTarget;
+    private long lastTeleportAttemptTime; // Sakura - fix endermen teleporting causing lag
 
     public Enderman(final EntityType<? extends Enderman> type, final Level level) {
         super(type, level);
@@ -301,6 +_,14 @@
 
     @VisibleForTesting
     public boolean teleport(final double x, final double y, final double z) {
+        // Sakura start - fix endermen teleporting causing lag
+        if (this.tickCount == this.lastTeleportAttemptTime) {
+            return false;
+        }
+
+        this.lastTeleportAttemptTime = this.tickCount;
+        // Sakura end - fix endermen teleporting causing lag
+
         if (this.isPassenger()) {
             return false;
         }
