From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 9 Sep 2023 18:39:15 +0100
Subject: [PATCH] Merge Cannon Entities


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
index f32f1cc5ed1bf429b45bfe06c97520edd1b87ef4..9f6371864cef8cc6cc69b65b56bed9bfcd1ea245 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
@@ -25,6 +25,27 @@ public class CraftFallingBlock extends CraftEntity implements FallingBlock {
         return this.getHandle().heightParity;
     }
     // Sakura end - falling block height parity api
+    // Sakura start - merge cannon entities
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeLevel getMergeLevel() {
+        return this.getHandle().getMergeEntityData().mergeLevel;
+    }
+
+    @Override
+    public final void setMergeLevel(final me.samsuik.sakura.entity.merge.MergeLevel mergeLevel) {
+        this.getHandle().getMergeEntityData().mergeLevel = mergeLevel;
+    }
+
+    @Override
+    public final int getStacked() {
+        return this.getHandle().getMergeEntityData().count;
+    }
+
+    @Override
+    public final void setStacked(final int stacked) {
+        this.getHandle().getMergeEntityData().count = stacked;
+    }
+    // Sakura end - merge cannon entities
 
     @Override
     public FallingBlockEntity getHandle() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
index 4a62bd61ad3fcd59433e6cb7ddba3af39d714fef..d243c2b92747ff20c275089f4bf247996c528536 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
@@ -13,6 +13,28 @@ public class CraftTNTPrimed extends CraftEntity implements TNTPrimed {
         super(server, entity);
     }
 
+    // Sakura start - merge cannon entities
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeLevel getMergeLevel() {
+        return this.getHandle().getMergeEntityData().mergeLevel;
+    }
+
+    @Override
+    public final void setMergeLevel(final me.samsuik.sakura.entity.merge.MergeLevel mergeLevel) {
+        this.getHandle().getMergeEntityData().mergeLevel = mergeLevel;
+    }
+
+    @Override
+    public final int getStacked() {
+        return this.getHandle().getMergeEntityData().count;
+    }
+
+    @Override
+    public final void setStacked(final int stacked) {
+        this.getHandle().getMergeEntityData().count = stacked;
+    }
+    // Sakura end - merge cannon entities
+
     @Override
     public PrimedTnt getHandle() {
         return (PrimedTnt) this.entity;
