From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 3 May 2024 15:04:31 +0100
Subject: [PATCH] Specialised Explosions


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java b/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
index ece6db7b9a0dfd535141c0c756947c4898140503..fbe351a877aa0313d522b8b01fc06bcf36d567af 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
@@ -42,6 +42,16 @@ public final class IteratorSafeOrderedReferenceSet<E> {
         this.listElements = (E[])Array.newInstance(arrComponent, arrayCapacity);
     }
 
+    // Sakura start - specialised explosions; add indexOf and rawGet methods
+    public E rawGet(final int index) {
+        return this.listElements[index];
+    }
+
+    public int indexOf(final E element) {
+        return this.indexMap.getInt(element);
+    }
+    // Sakura end - specialised explosions; add indexOf and rawGet methods
+
     // includes null (gravestone) elements
     public E[] getListRaw() {
         return this.listElements;
