--- a/net/minecraft/world/level/material/Fluid.java
+++ b/net/minecraft/world/level/material/Fluid.java
@@ -72,6 +_,12 @@
 
     protected abstract boolean canBeReplacedWith(FluidState state, final BlockGetter level, final BlockPos pos, Fluid other, Direction direction);
 
+    // Sakura start - lava flow speed api
+    public int getTickDelay(final Level world, final BlockPos pos) {
+        return this.getTickDelay(world);
+    }
+    // Sakura end - lava flow speed api
+
     protected abstract Vec3 getFlow(BlockGetter level, BlockPos pos, FluidState fluidState);
 
     public abstract int getTickDelay(LevelReader level);
