--- a/net/minecraft/world/level/material/Fluid.java
+++ b/net/minecraft/world/level/material/Fluid.java
@@ -74,6 +_,12 @@
 
     protected abstract Vec3 getFlow(BlockGetter level, BlockPos pos, FluidState fluidState);
 
+    // Sakura start - lava flow speed api
+    public int sakura$getTickDelay(final Level world, final BlockPos pos) {
+        return this.getTickDelay(world);
+    }
+    // Sakura end - lava flow speed api
+
     public abstract int getTickDelay(LevelReader level);
 
     protected boolean isRandomlyTicking() {
