--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -331,6 +_,7 @@
     public volatile boolean abnormalExit; // Paper - Improved watchdog support
     public volatile Thread shutdownThread; // Paper - Improved watchdog support
     public final io.papermc.paper.configuration.PaperConfigurations paperConfigurations; // Paper - add paper configuration files
+    public final me.samsuik.sakura.configuration.SakuraConfigurations sakuraConfigurations; // Sakura - sakura configuration files
     public boolean isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
     private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
@@ -510,6 +_,17 @@
         LOGGER.warn("Performed emergency save");
     }
     // Paper end - rewrite chunk system
+    // Sakura start - track tick information
+    private final me.samsuik.sakura.tps.TickInformationCollector tickInformationCollector = new me.samsuik.sakura.tps.TickInformationCollector();
+
+    public final me.samsuik.sakura.tps.ServerTickInformation latestTickInformation() {
+        return this.tickInformationCollector.latestTickInformation();
+    }
+
+    public final it.unimi.dsi.fastutil.objects.ObjectImmutableList<me.samsuik.sakura.tps.ServerTickInformation> tickHistory(final long from, final long to) {
+        return this.tickInformationCollector.collect(from, to);
+    }
+    // Sakura end - track tick information
 
     public MinecraftServer(
         // CraftBukkit start
@@ -575,6 +_,10 @@
         Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.util.ServerShutdownThread(this));
         // CraftBukkit end
         this.paperConfigurations = services.paper().configurations(); // Paper - add paper configuration files
+        // Sakura start - sakura configuration files
+        final java.nio.file.Path sakuraConfigDirPath = ((java.io.File) options.valueOf("sakura-settings-directory")).toPath();
+        this.sakuraConfigurations = me.samsuik.sakura.configuration.SakuraConfigurations.setup(sakuraConfigDirPath);
+        // Sakura end - sakura configuration files
     }
 
     protected abstract boolean initServer() throws IOException;
@@ -1333,6 +_,13 @@
 
                 this.currentTickStart = tickStart;
                 ++MinecraftServer.currentTick;
+                // Sakura start - track tick information
+                final int roundedTickRate = (int) Math.ceil(this.tickRateManager.tickrate());
+                final int sampleRate = Math.clamp(roundedTickRate, 20, 200);
+                if (MinecraftServer.currentTick % sampleRate == 0) {
+                    this.tickInformationCollector.levelData(this.levels.values(), getTPS(this.tickTimes1s, this.tickRateManager.nanosecondsPerTick()));
+                }
+                // Sakura end - track tick information
                 // Paper end - improve tick loop
 
                 boolean sprinting = thisTickNanos == 0L;
@@ -1347,6 +_,7 @@
                     this.processPacketsAndTick(sprinting);
                     ProfilerFiller profiler = Profiler.get();
                     profiler.push("nextTickWait");
+                    this.tickInformationCollector.tickDuration((System.nanoTime() - tickStart) / 1_000_000L); // Sakura - track tick information
                     this.mayHaveDelayedTasks = true;
                     this.delayedTasksMaxNextTickTimeNanos = Math.max(Util.getNanos() + thisTickNanos, this.nextTickTimeNanos);
                     this.startMeasuringTaskExecutionTime();
