From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 9 Feb 2025 17:52:59 +0000
Subject: [PATCH] Track block changes and level tick scheduler


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c6efa1eefd5d47ada2caddda8c9e542571397a15..4dfd1c9bdd43d748449ea941bdb59aa54fb96d59 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1836,6 +1836,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             profilerFiller.pop();
             profilerFiller.pop();
             serverLevel.explosionDensityCache.clear(); // Paper - Optimize explosions
+            serverLevel.levelTickScheduler.tick(); // Sakura - level tick and block change listeners
         }
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 0832a9edc81d6fbf87801dd6ab97c39c3ebbbea3..6cefbd63f3390ef373c4de91313d54200b91022e 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -830,6 +830,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return chunk != null ? chunk.getNoiseBiome(x, y, z) : this.getUncachedNoiseBiome(x, y, z);
     }
     // Paper end - optimise random ticking
+    // Sakura start - track block changes and tick scheduler
+    public final me.samsuik.sakura.scheduler.LevelTickScheduler levelTickScheduler = new me.samsuik.sakura.scheduler.LevelTickScheduler(this);
+    public final me.samsuik.sakura.block_change.BlockStateChangeTracker blockStateChangeTracker = new me.samsuik.sakura.block_change.BlockStateChangeTracker(this);
+    // Sakura end - track block changes and tick scheduler
 
     protected Level(
         WritableLevelData levelData,
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index de59760d27280555a334bda4f436164568cffbd6..b3d7a39894a5c86898289f463c343a5547e04508 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -136,6 +136,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         return this.getBlockStateFinal(x, y, z);
     }
     // Paper end - get block chunk optimisation
+    // Sakura start - track block changes and tick scheduler
+    private final me.samsuik.sakura.block_change.ChunkBlockStateWatcher blockStateWatcher = new me.samsuik.sakura.block_change.ChunkBlockStateWatcher();
+
+    public final me.samsuik.sakura.block_change.ChunkBlockStateWatcher getBlockStateWatcher() {
+        return this.blockStateWatcher;
+    }
+    // Sakura end - track block changes and tick scheduler
 
     public LevelChunk(Level level, ChunkPos pos) {
         this(level, pos, UpgradeData.EMPTY, new LevelChunkTicks<>(), new LevelChunkTicks<>(), 0L, null, null, null);
@@ -173,6 +180,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.debug = !empty && this.level.isDebug();
         this.defaultBlockState = empty ? VOID_AIR_BLOCKSTATE : AIR_BLOCKSTATE;
         // Paper end - get block chunk optimisation
+        this.blockStateWatcher.load(this); // Sakura - track block changes and tick scheduler
     }
 
     public LevelChunk(ServerLevel level, ProtoChunk chunk, @Nullable LevelChunk.PostLoadProcessor postLoad) {
@@ -423,6 +431,8 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                 if (!section.getBlockState(i, i1, i2).is(block)) {
                     return null;
                 } else {
+                    this.blockStateWatcher.blockChange(this.level, pos, state, blockState); // Sakura - track block changes and tick scheduler
+
                     if (!this.level.isClientSide() && (flags & 512) == 0 && (!this.level.captureBlockStates || block instanceof net.minecraft.world.level.block.BaseEntityBlock)) { // CraftBukkit - Don't place while processing the BlockPlaceEvent, unless it's a BlockContainer. Prevents blocks such as TNT from activating when cancelled.
                         state.onPlace(this.level, pos, blockState, flag1);
                     }
