From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 9 Feb 2025 17:52:59 +0000
Subject: [PATCH] Track block changes and level tick scheduler


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 2752a4f8785eb668f50da00f5efc131ac229833c..ddf084d7e329fecd4693978f9c74ab4ae662687f 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1853,6 +1853,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             profilerFiller.pop();
             profilerFiller.pop();
             serverLevel.explosionDensityCache.clear(); // Paper - Optimize explosions
+            serverLevel.levelTickScheduler.tick(); // Sakura - level tick and block change listeners
         }
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 915672bdcd90c559f183e586e8cd9302f2044de2..187d20d5b22b61f44e70f81ece59ac29d3240ed2 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -829,6 +829,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return chunk != null ? chunk.getNoiseBiome(x, y, z) : this.getUncachedNoiseBiome(x, y, z);
     }
     // Paper end - optimise random ticking
+    // Sakura start - track block changes and tick scheduler
+    public final me.samsuik.sakura.scheduler.LevelTickScheduler levelTickScheduler = new me.samsuik.sakura.scheduler.LevelTickScheduler(this);
+    public final me.samsuik.sakura.block_change.BlockStateChangeTracker blockStateChangeTracker = new me.samsuik.sakura.block_change.BlockStateChangeTracker(this);
+    // Sakura end - track block changes and tick scheduler
 
     protected Level(
         WritableLevelData levelData,
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 3acc1374a7ef968d88e9f566ce7b812fb8d580af..ce0d7f70a6467c30e5bd23e39786a46875dcfc81 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -135,6 +135,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         return this.getBlockStateFinal(x, y, z);
     }
     // Paper end - get block chunk optimisation
+    // Sakura start - track block changes and tick scheduler
+    private final me.samsuik.sakura.block_change.ChunkBlockStateWatcher blockStateWatcher = new me.samsuik.sakura.block_change.ChunkBlockStateWatcher();
+
+    public final me.samsuik.sakura.block_change.ChunkBlockStateWatcher getBlockStateWatcher() {
+        return this.blockStateWatcher;
+    }
+    // Sakura end - track block changes and tick scheduler
 
     public LevelChunk(Level level, ChunkPos pos) {
         this(level, pos, UpgradeData.EMPTY, new LevelChunkTicks<>(), new LevelChunkTicks<>(), 0L, null, null, null);
@@ -172,6 +179,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.debug = !empty && this.level.isDebug();
         this.defaultBlockState = empty ? VOID_AIR_BLOCKSTATE : AIR_BLOCKSTATE;
         // Paper end - get block chunk optimisation
+        this.blockStateWatcher.load(this); // Sakura - track block changes and tick scheduler
     }
 
     public LevelChunk(ServerLevel level, ProtoChunk chunk, LevelChunk.@Nullable PostLoadProcessor postLoad) {
@@ -421,6 +429,8 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                 if (!section.getBlockState(i, i1, i2).is(block)) {
                     return null;
                 } else {
+
+                    this.blockStateWatcher.blockChange(this.level, pos, state, blockState); // Sakura - track block changes and tick scheduler
                     if (!this.level.isClientSide() && (flags & Block.UPDATE_SKIP_ON_PLACE) == 0 && (!this.level.captureBlockStates || block instanceof net.minecraft.world.level.block.BaseEntityBlock)) { // CraftBukkit - Don't place while processing the BlockPlaceEvent, unless it's a BlockContainer. Prevents blocks such as TNT from activating when cancelled.
                         state.onPlace(this.level, pos, blockState, flag1);
                     }
