From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings


diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index 617a4bf9479364f2e234c7c73ba77806f8b8c679..f7e8320006171fadde78afaa4c89373292ab13de 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -8,6 +8,6 @@ import org.jspecify.annotations.NullMarked;
 @NullMarked
 public final class ScheduleSakuraTasks {
     public static void schedule(final LevelTickScheduler scheduler) {
-
+        scheduler.endless((tick, level, cb) -> level.explosionPositions.clear(), 0); // Sakura - client visibility settings
     }
 }
diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 42bfd4b3b626ef66b0f2ef8438add5342b367800..1125b25fe9dd130f4666ea37245acdb9ae1f2004 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -31,7 +31,7 @@ public class ClientboundLevelChunkPacketData {
     private static final int TWO_MEGABYTES = 2097152; // Paper - diff on change - if this changes, update PacketEncoder
     private final Map<Heightmap.Types, long[]> heightmaps;
     private final byte[] buffer;
-    private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
+    public final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData; // Sakura - client visibility settings; private -> public; ATs for some reason don't work here
     // Paper start - Handle oversized block entities in chunks
     private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new java.util.ArrayList<>();
     private static final int BLOCK_ENTITY_LIMIT = Integer.getInteger("Paper.excessiveTELimit", 750);
@@ -153,7 +153,7 @@ public class ClientboundLevelChunkPacketData {
         return this.heightmaps;
     }
 
-    private static class BlockEntityInfo {
+    public static class BlockEntityInfo { // Sakura - client visibility settings; package-protected -> public; ATs for some reason don't work here
         public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLevelChunkPacketData.BlockEntityInfo> STREAM_CODEC = StreamCodec.ofMember(
             ClientboundLevelChunkPacketData.BlockEntityInfo::write, ClientboundLevelChunkPacketData.BlockEntityInfo::new
         );
@@ -162,7 +162,7 @@ public class ClientboundLevelChunkPacketData {
         );
         private final int packedXZ;
         private final int y;
-        private final BlockEntityType<?> type;
+        public final BlockEntityType<?> type; // Sakura - client visibility settings; private -> public; ATs for some reason don't work here
         private final @Nullable CompoundTag tag;
 
         private BlockEntityInfo(final int packedXZ, final int y, final BlockEntityType<?> type, final @Nullable CompoundTag tag) {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index b1a81e0a6b3d0cde7f7bf67a4c542abc5f1392d4..d6a7b003334b9cb867eab70dd94f1dfb5b0401a9 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -171,6 +171,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         return null; // Paper - rewrite chunk system
     }
     // Paper end
+    private final it.unimi.dsi.fastutil.longs.Long2IntMap minimalEntities; // Sakura - client visibility settings; minimal tnt/sand
 
     public ChunkMap(
         final ServerLevel level,
@@ -232,6 +233,10 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         );
         this.setServerViewDistance(serverViewDistance);
         this.worldGenContext = new WorldGenContext(level, generator, structureManager, this.lightEngine, null, this::setChunkUnsaved); // Paper - rewrite chunk system
+        // Sakura start - client visibility settings; minimal tnt/sand
+        this.minimalEntities = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap();
+        this.minimalEntities.defaultReturnValue(Integer.MIN_VALUE);
+        // Sakura end - client visibility settings; minimal tnt/sand
     }
 
     private void setChunkUnsaved(final ChunkPos chunkPos) {
@@ -1053,6 +1058,8 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 tracker.serverEntity.sendChanges();
             }
         }
+
+        this.minimalEntities.clear(); // Sakura - client visibility settings; minimal tnt/sand
     }
     // Paper end - optimise entity tracker
 
@@ -1293,6 +1300,32 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             return !this.seenBy.isEmpty();
         }
         // Paper end - optimise entity tracker
+        // Sakura start - client visibility settings; entity visibility
+        private boolean checkEntityVisibility(final ServerPlayer player) {
+            final Entity entity = this.entity;
+            final me.samsuik.sakura.player.visibility.VisibilitySettings settings = player.visibilitySettings;
+            if (!settings.playerModified() || !(entity.isPrimedTNT || entity.isFallingBlock)) {
+                return true;
+            }
+
+            final me.samsuik.sakura.player.visibility.VisibilityType type = entity.isPrimedTNT
+                ? me.samsuik.sakura.player.visibility.VisibilityTypes.TNT
+                : me.samsuik.sakura.player.visibility.VisibilityTypes.SAND;
+            final me.samsuik.sakura.player.visibility.VisibilityState state = settings.get(type);
+
+            if (state == me.samsuik.sakura.player.visibility.VisibilityState.MINIMAL) {
+                final long key = entity.blockPosition().asLong() ^ entity.getType().hashCode();
+                final long visibleEntity = ChunkMap.this.minimalEntities.get(key);
+                if (visibleEntity != Integer.MIN_VALUE) {
+                    return entity.getId() == visibleEntity;
+                } else {
+                    ChunkMap.this.minimalEntities.put(key, entity.getId());
+                }
+            }
+
+            return state != me.samsuik.sakura.player.visibility.VisibilityState.OFF;
+        }
+        // Sakura end - client visibility settings; entity visibility
 
         public TrackedEntity(final Entity entity, final int range, final int updateInterval, final boolean trackDelta) {
             Objects.requireNonNull(ChunkMap.this);
@@ -1382,6 +1415,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                     && this.entity.broadcastToPlayer(player)
                     && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x(), this.entity.chunkPosition().z());
                 // Paper end - Configurable entity tracking range by Y
+                visibleToPlayer &= this.checkEntityVisibility(player); // Sakura start - client visibility settings; entity visibility
                 // CraftBukkit start - respect vanish API
                 if (visibleToPlayer && !player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper - only consider hits
                     visibleToPlayer = false;
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 898a0059586dbc6750a4b3968e5f4ae31aa87121..7a1b62df9f2c60898546ee44e182151e62b28663 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -603,6 +603,22 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.playerTickingChunks.remove((LevelChunk)chunkHolder.getCurrentChunk());
     }
     // Paper end - chunk tick iteration
+    // Sakura start - client visibility settings
+    public final LongSet explosionPositions = new it.unimi.dsi.fastutil.longs.LongOpenHashSet();
+
+    public final boolean checkExplosionVisibility(final Vec3 position, final ServerPlayer player) {
+        final me.samsuik.sakura.player.visibility.VisibilitySettings settings = player.visibilitySettings;
+        if (settings.isDisabled(me.samsuik.sakura.player.visibility.VisibilityTypes.EXPLOSIONS)) {
+            return false;
+        } else if (settings.isToggled(me.samsuik.sakura.player.visibility.VisibilityTypes.EXPLOSIONS)) {
+            final BlockPos blockPosition = BlockPos.containing(position);
+            final long encodedPosition = blockPosition.asLong();
+            return this.explosionPositions.add(encodedPosition);
+        }
+
+        return true;
+    }
+    // Sakura end - client visibility settings
 
     public ServerLevel(
         final MinecraftServer server,
@@ -2011,6 +2027,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         for (ServerPlayer player : this.players) {
             if (player.distanceToSqr(center) < 4096.0) {
                 Optional<Vec3> playerKnockback = Optional.ofNullable(explosion.getHitPlayers().get(player));
+                // Sakura start - client visibility settings; let players toggle explosion particles
+                if (!this.checkExplosionVisibility(center, player)) {
+                    if (playerKnockback.isPresent() && !this.paperConfig().environment.disableExplosionKnockback) {
+                        player.connection.send(new net.minecraft.network.protocol.game.ClientboundPlayerPositionPacket(
+                            Short.MIN_VALUE - 2048,
+                            new net.minecraft.world.entity.PositionMoveRotation(
+                                Vec3.ZERO,
+                                playerKnockback.get(),
+                                0.0f,
+                                0.0f
+                            ),
+                            net.minecraft.world.entity.Relative.ALL
+                        ));
+                    }
+
+                    continue;
+                }
+                // Sakura end - client visibility settings; let players toggle explosion particles
                 player.connection.send(new ClientboundExplodePacket(center, r, blockCount, playerKnockback, explosionParticle, explosionSound, blockParticles));
             }
         }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index eb807af00ba960cfb0f8345eed56f1f62cdc3178..dc17919f1bd149ba2b965461ac63738029413a3e 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -485,6 +485,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     // Paper end - rewrite chunk system
     public double trackingRangeModifier = 1.0; // Sakura - entity tracking range modifier api
+    public final me.samsuik.sakura.player.visibility.PlayerVisibilitySettings visibilitySettings = new me.samsuik.sakura.player.visibility.PlayerVisibilitySettings(); // Sakura - client visibility settings
 
     public ServerPlayer(final MinecraftServer server, final ServerLevel level, final GameProfile gameProfile, final ClientInformation clientInformation) {
         super(level, gameProfile);
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 7ed50fc00c58c7781c958df1d828a67adf696f3e..29222a62c8fb76fb3f2a8245879c29e7a4795be0 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -57,6 +57,60 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     public @Nullable String playerBrand;
     public final java.util.Set<String> pluginMessagerChannels;
     // Paper end - retain certain values
+    // Sakura start - client visibility settings
+    private boolean filterExtraPackets(final Packet<?> packet, final net.minecraft.server.level.ServerPlayer player) {
+        final me.samsuik.sakura.player.visibility.VisibilitySettings settings = player.visibilitySettings;
+        if (settings.isToggled(me.samsuik.sakura.player.visibility.VisibilityTypes.SPAWNERS)) {
+            if (packet instanceof net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket bedPacket) {
+                return bedPacket.getType() == net.minecraft.world.level.block.entity.BlockEntityType.MOB_SPAWNER;
+            }
+        }
+
+        return false;
+    }
+
+    private boolean filterOrModifyPacket(final Packet<?> packet, final net.minecraft.server.level.ServerPlayer player) {
+        final me.samsuik.sakura.player.visibility.VisibilitySettings settings = player.visibilitySettings;
+        final java.util.List<Packet<?>> extraPackets = packet.getExtraPackets();
+        if (extraPackets != null) {
+            // block entity data is sent as an extra packet
+            extraPackets.removeIf(extraPacket -> this.filterExtraPackets(extraPacket, player));
+        }
+
+        if (settings.isToggled(me.samsuik.sakura.player.visibility.VisibilityTypes.SPAWNERS)) {
+            net.minecraft.network.protocol.game.ClientboundLevelChunkPacketData chunkData = null;
+            if (packet instanceof net.minecraft.network.protocol.game.ClientboundLevelChunkWithLightPacket chunkPacket) {
+                chunkData = chunkPacket.getChunkData();
+            } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundLevelChunkPacketData chunkPacket) {
+                chunkData = chunkPacket;
+            }
+
+            if (chunkData != null) {
+                chunkData.blockEntitiesData.removeIf(blockEntityInfo -> blockEntityInfo.type == net.minecraft.world.level.block.entity.BlockEntityType.MOB_SPAWNER);
+            }
+
+            if (packet instanceof net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket bedPacket && bedPacket.getType() == net.minecraft.world.level.block.entity.BlockEntityType.MOB_SPAWNER) {
+                return true;
+            }
+        }
+
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundBlockEventPacket bePacket) {
+            if (settings.isToggled(me.samsuik.sakura.player.visibility.VisibilityTypes.PISTONS)) {
+                return bePacket.getBlock() instanceof net.minecraft.world.level.block.piston.PistonBaseBlock;
+            }
+        }
+        
+        return false;
+    }
+
+    private net.minecraft.server.level.ServerPlayer getPlayer() {
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            return gamePacketListener.getPlayer();
+        } else {
+            return this.connection.getPlayer();
+        }
+    }
+    // Sakura end - client visibility settings
 
     public ServerCommonPacketListenerImpl(final MinecraftServer server, final Connection connection, final CommonListenerCookie cookie) {
         this.server = server;
@@ -311,6 +365,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket defaultSpawnPositionPacket && this instanceof ServerGamePacketListenerImpl serverGamePacketListener) {
             serverGamePacketListener.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(defaultSpawnPositionPacket.respawnData().pos(), serverGamePacketListener.getPlayer().level());
         }
+        // Sakura start - client visibility settings
+        final net.minecraft.server.level.ServerPlayer player = this.getPlayer();
+        if (player != null && player.visibilitySettings.playerModified() && this.filterOrModifyPacket(packet, player)) {
+            return;
+        }
+        // Sakura end - client visibility settings
         // CraftBukkit end
         if (packet.isTerminal()) {
             this.close();
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 5b779bd2a5d836bebbc31718df6e9c735f12e41e..7133038f1fb0bb712416deffc8e6a8d4d09861c0 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3355,6 +3355,7 @@ public class ServerGamePacketListenerImpl
 
                         event.setCancelled(cancelled);
                         net.minecraft.world.inventory.AbstractContainerMenu oldContainer = this.player.containerMenu; // SPIGOT-1224
+                        me.samsuik.sakura.player.gui.FeatureGui.clickEvent(event); // Sakura - client visibility settings
                         this.cserver.getPluginManager().callEvent(event);
                         if (this.player.containerMenu != oldContainer) {
                             this.player.containerMenu.resumeRemoteUpdates();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 05ddf3debd87adc3fb29caa8bac058932bedbe3a..1dfef3c5e5e5cbd49cf299a09fe63b7ffceaed01 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -549,6 +549,10 @@ public abstract class Entity
     }
     // Paper end - optimise entity tracker
     public boolean pushedByFluid = true; // Sakura - entity pushed by fluid api
+    // Sakura start - client visibility settings
+    public boolean isPrimedTNT;
+    public boolean isFallingBlock;
+    // Sakura end - client visibility settings
 
     public Entity(final EntityType<?> type, final Level level) {
         this.type = type;
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index a6c3f7be4b1c77cd9ebb0ebc25203f943eb96aad..c40487da0e8971b3704d461d2b44df57f843c691 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -106,6 +106,7 @@ public class FallingBlockEntity extends Entity {
         super(type, level);
         this.dropItem = level.sakuraConfig().cannons.sand.dropItems; // Sakura - configure falling blocks dropping items
         this.heightParity = level.sakuraConfig().cannons.mechanics.fallingBlockParity; // Sakura - configure cannon mechanics
+        this.isFallingBlock = true; // Sakura - client visibility settings
     }
 
     public FallingBlockEntity(final Level level, final double x, final double y, final double z, final BlockState blockState) {
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 8c5b676a9ab97f0fe038996ba8801d1d34d3e549..ff4634479269a44e85302cc45c316da74d5319e4 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -64,6 +64,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
         this.blocksBuilding = true;
+        this.isPrimedTNT = true; // Sakura - client visibility settings
     }
 
     public PrimedTnt(final Level level, final double x, final double y, final double z, final @Nullable LivingEntity owner) {
