From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings


diff --git a/me/samsuik/sakura/SakuraFeatureHooks.java b/me/samsuik/sakura/SakuraFeatureHooks.java
index 16580d8e41124782eee7f072fdb66497d0ac2603..03fbd9470083bc947adafa7117e645336e31680d 100644
--- a/me/samsuik/sakura/SakuraFeatureHooks.java
+++ b/me/samsuik/sakura/SakuraFeatureHooks.java
@@ -1,10 +1,18 @@
 package me.samsuik.sakura;
 
+import me.samsuik.sakura.command.subcommand.FPSCommand;
+import me.samsuik.sakura.command.subcommand.VisualCommand;
+import me.samsuik.sakura.player.visibility.VisibilityTypes;
 import org.bukkit.command.Command;
 
 import java.util.Map;
 
 public final class SakuraFeatureHooks {
     public static void setupCommands(final Map<String, Command> commands) {
+        // Sakura start - client visibility settings
+        commands.put("fps", new FPSCommand("fps"));
+        commands.put("tntvisibility", new VisualCommand(VisibilityTypes.TNT, "tnttoggle"));
+        commands.put("sandvisibility", new VisualCommand(VisibilityTypes.SAND, "sandtoggle"));
+        // Sakura end - client visibility settings
     }
 }
\ No newline at end of file
diff --git a/me/samsuik/sakura/command/subcommand/FPSCommand.java b/me/samsuik/sakura/command/subcommand/FPSCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..1147138817683351ac37ada4a7f5078060cc7151
--- /dev/null
+++ b/me/samsuik/sakura/command/subcommand/FPSCommand.java
@@ -0,0 +1,20 @@
+package me.samsuik.sakura.command.subcommand;
+
+import me.samsuik.sakura.command.PlayerOnlySubCommand;
+import me.samsuik.sakura.player.visibility.VisibilityGui;
+import org.bukkit.entity.Player;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class FPSCommand extends PlayerOnlySubCommand {
+    private final VisibilityGui visibilityGui = new VisibilityGui();
+
+    public FPSCommand(final String name) {
+        super(name);
+    }
+
+    @Override
+    public void execute(final Player player, final String[] args) {
+        this.visibilityGui.showTo(player);
+    }
+}
diff --git a/me/samsuik/sakura/command/subcommand/VisualCommand.java b/me/samsuik/sakura/command/subcommand/VisualCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..6450f2013eb71b956fcdfb2aa8ba00503ceea121
--- /dev/null
+++ b/me/samsuik/sakura/command/subcommand/VisualCommand.java
@@ -0,0 +1,31 @@
+package me.samsuik.sakura.command.subcommand;
+
+import me.samsuik.sakura.command.PlayerOnlySubCommand;
+import me.samsuik.sakura.configuration.GlobalConfiguration;
+import me.samsuik.sakura.player.visibility.VisibilitySettings;
+import me.samsuik.sakura.player.visibility.VisibilityState;
+import me.samsuik.sakura.player.visibility.VisibilityType;
+import org.bukkit.entity.Player;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.Arrays;
+
+@NullMarked
+public final class VisualCommand extends PlayerOnlySubCommand {
+    private final VisibilityType type;
+
+    public VisualCommand(final VisibilityType type, final String... aliases) {
+        super(type.key() + "visibility");
+        this.setAliases(Arrays.asList(aliases));
+        this.type = type;
+    }
+
+    @Override
+    public void execute(final Player player, final String[] args) {
+        final VisibilitySettings settings = player.getVisibility();
+        final VisibilityState state = settings.toggle(type);
+
+        final String stateName = (state == VisibilityState.ON) ? "Enabled" : "Disabled";
+        player.sendMessage(GlobalConfiguration.get().messages.fpsSettingChangeComponent(this.type.key(), stateName));
+    }
+}
diff --git a/me/samsuik/sakura/player/visibility/FilterOrModifyPackets.java b/me/samsuik/sakura/player/visibility/FilterOrModifyPackets.java
new file mode 100644
index 0000000000000000000000000000000000000000..fff2220e351e925307fa55a43d7f98252037a0bd
--- /dev/null
+++ b/me/samsuik/sakura/player/visibility/FilterOrModifyPackets.java
@@ -0,0 +1,96 @@
+package me.samsuik.sakura.player.visibility;
+
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
+import net.minecraft.network.protocol.game.ClientboundBlockEventPacket;
+import net.minecraft.network.protocol.game.ClientboundLevelChunkPacketData;
+import net.minecraft.network.protocol.game.ClientboundLevelChunkWithLightPacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.item.FallingBlockEntity;
+import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraft.world.level.block.piston.PistonBaseBlock;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.List;
+
+@NullMarked
+public final class FilterOrModifyPackets {
+    public static boolean filterExplosion(
+        final Vec3 explosionPos,
+        final ServerPlayer player,
+        final long gameTime,
+        final MinimalVisibility minimalVisibility
+    ) {
+        final VisibilitySettings settings = player.visibilitySettings;
+        final VisibilityState state = settings.get(VisibilityTypes.EXPLOSIONS);
+
+        if (state == VisibilityState.MINIMAL && minimalVisibility.filterExplosion(explosionPos, gameTime)) {
+            return false;
+        }
+
+        return state != VisibilityState.ON;
+    }
+
+    public static boolean canSendEntityUpdates(
+        final Entity entity,
+        final ServerPlayer player,
+        final MinimalVisibility minimalVisibility
+    ) {
+        final VisibilitySettings settings = player.visibilitySettings;
+        if (!settings.playerModified() || !(entity instanceof PrimedTnt || entity instanceof FallingBlockEntity)) {
+            return true;
+        }
+
+        final VisibilityType type = entity instanceof PrimedTnt ? VisibilityTypes.TNT : VisibilityTypes.SAND;
+        final VisibilityState state = settings.get(type);
+        if (state == VisibilityState.MINIMAL && minimalVisibility.filterEntity(entity)) {
+            return true;
+        }
+
+        return state != VisibilityState.OFF;
+    }
+
+    private static boolean filterExtraPackets(final Packet<?> packet, final ServerPlayer player) {
+        final VisibilitySettings settings = player.visibilitySettings;
+        if (settings.isToggled(VisibilityTypes.SPAWNERS) && packet instanceof final ClientboundBlockEntityDataPacket bedPacket) {
+            return bedPacket.getType() == BlockEntityType.MOB_SPAWNER;
+        }
+
+        return false;
+    }
+
+    public static boolean filterOrModifyPacket(final Packet<?> packet, final ServerPlayer player) {
+        final VisibilitySettings settings = player.visibilitySettings;
+        final List<Packet<?>> extraPackets = packet.getExtraPackets();
+        if (extraPackets != null) {
+            // block entity data is sent as an extra packet
+            extraPackets.removeIf(extraPacket -> filterExtraPackets(extraPacket, player));
+        }
+
+        if (settings.isToggled(VisibilityTypes.SPAWNERS)) {
+            ClientboundLevelChunkPacketData chunkData = null;
+            if (packet instanceof final ClientboundLevelChunkWithLightPacket chunkPacket) {
+                chunkData = chunkPacket.getChunkData();
+            } else if (packet instanceof final ClientboundLevelChunkPacketData chunkPacket) {
+                chunkData = chunkPacket;
+            }
+
+            if (chunkData != null) {
+                chunkData.blockEntitiesData.removeIf(blockEntity -> blockEntity.type == BlockEntityType.MOB_SPAWNER);
+            }
+
+            if (packet instanceof final ClientboundBlockEntityDataPacket bedPacket && bedPacket.getType() == BlockEntityType.MOB_SPAWNER) {
+                return true;
+            }
+        }
+
+        if (packet instanceof final ClientboundBlockEventPacket bePacket && settings.isToggled(VisibilityTypes.PISTONS)) {
+            return bePacket.getBlock() instanceof PistonBaseBlock;
+        }
+
+        return false;
+    }
+}
diff --git a/me/samsuik/sakura/player/visibility/MinimalVisibility.java b/me/samsuik/sakura/player/visibility/MinimalVisibility.java
new file mode 100644
index 0000000000000000000000000000000000000000..c7b81da60c8d05406042d020cc95c17e895fb8a6
--- /dev/null
+++ b/me/samsuik/sakura/player/visibility/MinimalVisibility.java
@@ -0,0 +1,44 @@
+package me.samsuik.sakura.player.visibility;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+import java.time.Duration;
+
+@NullMarked
+public final class MinimalVisibility {
+    private final Cache<Long, Entity> minimalEntities = CacheBuilder.newBuilder()
+        .expireAfterAccess(Duration.ofSeconds(5))
+        .maximumSize(Short.MAX_VALUE)
+        .build();
+    private final ObjectOpenHashSet<BlockPos> explosions = new ObjectOpenHashSet<>();
+    private long lastExplosion;
+
+    public boolean filterEntity(final Entity entity) {
+        final BlockPos blockPos = entity.blockPosition();
+        final long posAndType = blockPos.asLong() ^ entity.getType().hashCode();
+        final Entity visibleEntity = this.minimalEntities.getIfPresent(posAndType);
+
+        if (visibleEntity != null && visibleEntity.isAlive() && blockPos.equals(visibleEntity.blockPosition())) {
+            return entity == visibleEntity;
+        } else {
+            this.minimalEntities.put(posAndType, entity);
+            return false;
+        }
+    }
+
+    public boolean filterExplosion(final Vec3 explosionPos, final long gameTime) {
+        if (this.lastExplosion != gameTime) {
+            this.explosions.clear();
+            this.lastExplosion = gameTime;
+        }
+
+        final BlockPos explosionBlockPos = BlockPos.containing(explosionPos);
+        return this.explosions.add(explosionBlockPos);
+    }
+}
diff --git a/me/samsuik/sakura/player/visibility/PlayerVisibilitySettings.java b/me/samsuik/sakura/player/visibility/PlayerVisibilitySettings.java
new file mode 100644
index 0000000000000000000000000000000000000000..5239478c397169f82206fcaefb77f71d2d83596d
--- /dev/null
+++ b/me/samsuik/sakura/player/visibility/PlayerVisibilitySettings.java
@@ -0,0 +1,62 @@
+package me.samsuik.sakura.player.visibility;
+
+import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import net.minecraft.world.level.storage.ValueInput;
+import net.minecraft.world.level.storage.ValueOutput;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class PlayerVisibilitySettings implements VisibilitySettings {
+    private static final String VISIBILITY_SETTINGS_TAG = "client_visibility_settings";
+    private final Reference2ObjectMap<VisibilityType, VisibilityState> visibilityStates = new Reference2ObjectOpenHashMap<>();
+
+    @Override
+    public VisibilityState get(final VisibilityType type) {
+        final VisibilityState state = this.visibilityStates.get(type);
+        //noinspection ConstantValue
+        return state != null ? state : type.getDefault();
+    }
+
+    @Override
+    public VisibilityState set(final VisibilityType type, final VisibilityState state) {
+        if (type.isDefault(state)) {
+            this.visibilityStates.remove(type);
+        } else {
+            this.visibilityStates.put(type, state);
+        }
+        return state;
+    }
+
+    @Override
+    public VisibilityState currentState() {
+        final int modifiedCount = this.visibilityStates.size();
+        if (modifiedCount == 0) {
+            return VisibilityState.ON;
+        } else if (modifiedCount != VisibilityTypes.types().size()) {
+            return VisibilityState.MODIFIED;
+        } else {
+            return VisibilityState.OFF;
+        }
+    }
+
+    @Override
+    public boolean playerModified() {
+        return !this.visibilityStates.isEmpty();
+    }
+
+    public void loadData(final ValueInput input) {
+        input.child(VISIBILITY_SETTINGS_TAG).ifPresent(settings -> {
+            for (final VisibilityType type : VisibilityTypes.types()) {
+                final String typeKey = type.key();
+                final String stateName = settings.getStringOr(typeKey, type.getDefault().name());
+                this.set(type, VisibilityState.valueOf(stateName));
+            }
+        });
+    }
+
+    public void saveData(final ValueOutput output) {
+        final ValueOutput settings = output.child(VISIBILITY_SETTINGS_TAG);
+        this.visibilityStates.forEach((type, state) -> settings.putString(type.key(), state.name()));
+    }
+}
diff --git a/me/samsuik/sakura/player/visibility/VisibilityGui.java b/me/samsuik/sakura/player/visibility/VisibilityGui.java
new file mode 100644
index 0000000000000000000000000000000000000000..8c1b3eb16b924141302e9f1bf2460363c035037a
--- /dev/null
+++ b/me/samsuik/sakura/player/visibility/VisibilityGui.java
@@ -0,0 +1,100 @@
+package me.samsuik.sakura.player.visibility;
+
+import it.unimi.dsi.fastutil.ints.IntArrayFIFOQueue;
+import me.samsuik.sakura.configuration.GlobalConfiguration;
+import me.samsuik.sakura.player.gui.FeatureGui;
+import me.samsuik.sakura.player.gui.FeatureGuiInventory;
+import me.samsuik.sakura.player.gui.components.ItemButton;
+import me.samsuik.sakura.player.gui.components.ItemSwitch;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.Inventory;
+import org.jspecify.annotations.NullMarked;
+
+import static me.samsuik.sakura.player.gui.ItemStackUtil.itemWithBlankName;
+
+@NullMarked
+public final class VisibilityGui extends FeatureGui {
+    private static final NamespacedKey TOGGLE_BUTTON_KEY = new NamespacedKey("sakura", "toggle_button");
+    private static final NamespacedKey MENU_ITEMS_KEY = new NamespacedKey("sakura", "menu_items");
+
+    public VisibilityGui() {
+        super(45, Component.text("FPS Settings"));
+    }
+
+    @Override
+    protected void fillInventory(final Inventory inventory) {
+        for (int slot = 0; slot < inventory.getSize(); ++slot) {
+            // x, y from top left of the inventory
+            final int x = slot % 9;
+            final int y = slot / 9;
+            // from center
+            final int rx = x - 4;
+            final int ry = y - 2;
+
+            if (Math.sqrt(rx * rx + ry * ry) <= 3.25) {
+                inventory.setItem(slot, itemWithBlankName(GlobalConfiguration.get().fps.material));
+            } else if (x % 8 == 0) {
+                inventory.setItem(slot, itemWithBlankName(Material.BLACK_STAINED_GLASS_PANE));
+            } else {
+                inventory.setItem(slot, itemWithBlankName(Material.WHITE_STAINED_GLASS_PANE));
+            }
+        }
+    }
+
+    @Override
+    protected void afterFill(final Player player, final FeatureGuiInventory inventory) {
+        final VisibilitySettings settings = player.getVisibility();
+        final IntArrayFIFOQueue slots = this.availableSlots();
+        this.updateToggleButton(settings, player, inventory);
+
+        for (final VisibilityType type : VisibilityTypes.types()) {
+            final VisibilityState state = settings.get(type);
+            final int index = type.states().indexOf(state);
+            final int slot = slots.dequeueInt();
+
+            // Switch between the visibility states
+            final ItemSwitch itemSwitch = new ItemSwitch(
+                VisibilityGuiItems.GUI_ITEMS.get(type),
+                slot, index,
+                (_, clickedInventory) -> {
+                    settings.cycle(type);
+                    this.updateToggleButton(settings, player, clickedInventory);
+                }
+            );
+
+            inventory.addComponent(itemSwitch, MENU_ITEMS_KEY);
+        }
+    }
+
+    private void updateToggleButton(final VisibilitySettings settings, final Player player, final FeatureGuiInventory inventory) {
+        inventory.removeComponents(TOGGLE_BUTTON_KEY);
+        final VisibilityState settingsState = settings.currentState();
+        final ItemButton button = new ItemButton(
+            VisibilityGuiItems.TOGGLE_BUTTON_ITEMS.get(settingsState),
+            (2 * 9) + 8,
+            (_, clickedInventory) -> {
+                settings.toggleAll();
+                inventory.removeAllComponents();
+                this.afterFill(player, clickedInventory);
+            }
+        );
+
+        inventory.addComponent(button, TOGGLE_BUTTON_KEY);
+    }
+
+    private IntArrayFIFOQueue availableSlots() {
+        final IntArrayFIFOQueue slots = new IntArrayFIFOQueue();
+        for (int row = 1; row < 4; ++row) {
+            for (int column = 3; column < 6; ++column) {
+                if ((column + row) % 2 == 0) {
+                    slots.enqueue((row * 9) + column);
+                }
+            }
+        }
+
+        return slots;
+    }
+}
diff --git a/me/samsuik/sakura/player/visibility/VisibilityGuiItems.java b/me/samsuik/sakura/player/visibility/VisibilityGuiItems.java
new file mode 100644
index 0000000000000000000000000000000000000000..b7895fdde6fbc7693ebe5746a37dcd464194af2a
--- /dev/null
+++ b/me/samsuik/sakura/player/visibility/VisibilityGuiItems.java
@@ -0,0 +1,54 @@
+package me.samsuik.sakura.player.visibility;
+
+import it.unimi.dsi.fastutil.objects.ObjectImmutableList;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import me.samsuik.sakura.player.gui.ItemStackUtil;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import org.apache.commons.lang3.StringUtils;
+import org.bukkit.Material;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.Locale;
+
+@NullMarked
+public final class VisibilityGuiItems {
+    static final Reference2ObjectMap<VisibilityType, ObjectImmutableList<ItemStack>> GUI_ITEMS = new Reference2ObjectOpenHashMap<>();
+    static final Reference2ObjectMap<VisibilityState, ItemStack> TOGGLE_BUTTON_ITEMS = new Reference2ObjectOpenHashMap<>();
+
+    static {
+        final Reference2ObjectMap<VisibilityType, ItemStack> items = new Reference2ObjectOpenHashMap<>();
+        items.put(VisibilityTypes.TNT, ItemStackUtil.itemWithName(Material.TNT, Component.text("Tnt", NamedTextColor.RED)));
+        items.put(VisibilityTypes.SAND, ItemStackUtil.itemWithName(Material.SAND, Component.text("Sand", NamedTextColor.YELLOW)));
+        items.put(VisibilityTypes.EXPLOSIONS, ItemStackUtil.itemWithName(Material.COBWEB, Component.text("Explosions", NamedTextColor.WHITE)));
+        items.put(VisibilityTypes.SPAWNERS, ItemStackUtil.itemWithName(Material.SPAWNER, Component.text("Spawners", NamedTextColor.DARK_GRAY)));
+        items.put(VisibilityTypes.PISTONS, ItemStackUtil.itemWithName(Material.PISTON, Component.text("Pistons", NamedTextColor.GOLD)));
+
+        for (final VisibilityType type : VisibilityTypes.types()) {
+            final ItemStack item = items.get(type);
+            final ObjectImmutableList<ItemStack> stateItems = type.states().stream()
+                .map(state -> createItemForState(item, state))
+                .collect(ObjectImmutableList.toList());
+            GUI_ITEMS.put(type, stateItems);
+        }
+
+        TOGGLE_BUTTON_ITEMS.put(VisibilityState.ON, ItemStackUtil.itemWithName(Material.GREEN_STAINED_GLASS_PANE, Component.text("Enabled", NamedTextColor.GREEN)));
+        TOGGLE_BUTTON_ITEMS.put(VisibilityState.MODIFIED, ItemStackUtil.itemWithName(Material.MAGENTA_STAINED_GLASS_PANE, Component.text("Modified", NamedTextColor.LIGHT_PURPLE)));
+        TOGGLE_BUTTON_ITEMS.put(VisibilityState.OFF, ItemStackUtil.itemWithName(Material.RED_STAINED_GLASS_PANE, Component.text("Disabled", NamedTextColor.RED)));
+    }
+
+    private static String lowerCaseThenCapitalise(final String name) {
+        final String lowerCaseName = name.toLowerCase(Locale.ENGLISH);
+        return StringUtils.capitalize(lowerCaseName);
+    }
+
+    private static ItemStack createItemForState(final ItemStack in, final VisibilityState state) {
+        final String capitalisedName = lowerCaseThenCapitalise(state.name());
+        final Component component = Component.text(" | " + capitalisedName, NamedTextColor.GRAY);
+        final ItemStack itemCopy = in.clone();
+        itemCopy.editMeta(meta -> meta.itemName(meta.itemName().append(component)));
+        return itemCopy;
+    }
+}
diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 42bfd4b3b626ef66b0f2ef8438add5342b367800..1125b25fe9dd130f4666ea37245acdb9ae1f2004 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -31,7 +31,7 @@ public class ClientboundLevelChunkPacketData {
     private static final int TWO_MEGABYTES = 2097152; // Paper - diff on change - if this changes, update PacketEncoder
     private final Map<Heightmap.Types, long[]> heightmaps;
     private final byte[] buffer;
-    private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
+    public final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData; // Sakura - client visibility settings; private -> public; ATs for some reason don't work here
     // Paper start - Handle oversized block entities in chunks
     private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new java.util.ArrayList<>();
     private static final int BLOCK_ENTITY_LIMIT = Integer.getInteger("Paper.excessiveTELimit", 750);
@@ -153,7 +153,7 @@ public class ClientboundLevelChunkPacketData {
         return this.heightmaps;
     }
 
-    private static class BlockEntityInfo {
+    public static class BlockEntityInfo { // Sakura - client visibility settings; package-protected -> public; ATs for some reason don't work here
         public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLevelChunkPacketData.BlockEntityInfo> STREAM_CODEC = StreamCodec.ofMember(
             ClientboundLevelChunkPacketData.BlockEntityInfo::write, ClientboundLevelChunkPacketData.BlockEntityInfo::new
         );
@@ -162,7 +162,7 @@ public class ClientboundLevelChunkPacketData {
         );
         private final int packedXZ;
         private final int y;
-        private final BlockEntityType<?> type;
+        public final BlockEntityType<?> type; // Sakura - client visibility settings; private -> public; ATs for some reason don't work here
         private final @Nullable CompoundTag tag;
 
         private BlockEntityInfo(final int packedXZ, final int y, final BlockEntityType<?> type, final @Nullable CompoundTag tag) {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index b1a81e0a6b3d0cde7f7bf67a4c542abc5f1392d4..87441a0345943ec843e55505f859b456765c8aab 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1382,6 +1382,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                     && this.entity.broadcastToPlayer(player)
                     && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x(), this.entity.chunkPosition().z());
                 // Paper end - Configurable entity tracking range by Y
+                visibleToPlayer = visibleToPlayer && me.samsuik.sakura.player.visibility.FilterOrModifyPackets.canSendEntityUpdates(entity, player, ChunkMap.this.level.minimalVisibility); // Sakura start - client visibility settings; entity visibility
                 // CraftBukkit start - respect vanish API
                 if (visibleToPlayer && !player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper - only consider hits
                     visibleToPlayer = false;
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 898a0059586dbc6750a4b3968e5f4ae31aa87121..30ec8cd216cce3cbde2d540cda3c0ed47bd66097 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2011,6 +2011,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         for (ServerPlayer player : this.players) {
             if (player.distanceToSqr(center) < 4096.0) {
                 Optional<Vec3> playerKnockback = Optional.ofNullable(explosion.getHitPlayers().get(player));
+                // Sakura start - client visibility settings; let players toggle explosion particles
+                if (me.samsuik.sakura.player.visibility.FilterOrModifyPackets.filterExplosion(center, player, this.getGameTime(), this.minimalVisibility)) {
+                    if (playerKnockback.isPresent() && !this.paperConfig().environment.disableExplosionKnockback) {
+                        player.connection.send(new net.minecraft.network.protocol.game.ClientboundPlayerPositionPacket(
+                            Short.MIN_VALUE - 2048,
+                            new net.minecraft.world.entity.PositionMoveRotation(
+                                Vec3.ZERO,
+                                playerKnockback.get(),
+                                0.0f,
+                                0.0f
+                            ),
+                            net.minecraft.world.entity.Relative.ALL
+                        ));
+                    }
+
+                    continue;
+                }
+                // Sakura end - client visibility settings; let players toggle explosion particles
                 player.connection.send(new ClientboundExplodePacket(center, r, blockCount, playerKnockback, explosionParticle, explosionSound, blockParticles));
             }
         }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index eb807af00ba960cfb0f8345eed56f1f62cdc3178..dc17919f1bd149ba2b965461ac63738029413a3e 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -485,6 +485,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     // Paper end - rewrite chunk system
     public double trackingRangeModifier = 1.0; // Sakura - entity tracking range modifier api
+    public final me.samsuik.sakura.player.visibility.PlayerVisibilitySettings visibilitySettings = new me.samsuik.sakura.player.visibility.PlayerVisibilitySettings(); // Sakura - client visibility settings
 
     public ServerPlayer(final MinecraftServer server, final ServerLevel level, final GameProfile gameProfile, final ClientInformation clientInformation) {
         super(level, gameProfile);
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 7ed50fc00c58c7781c958df1d828a67adf696f3e..f816ceb53ab72c22cabe782f3f20af43733090e1 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -311,6 +311,17 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket defaultSpawnPositionPacket && this instanceof ServerGamePacketListenerImpl serverGamePacketListener) {
             serverGamePacketListener.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(defaultSpawnPositionPacket.respawnData().pos(), serverGamePacketListener.getPlayer().level());
         }
+        // Sakura start - client visibility settings
+        final net.minecraft.server.level.ServerPlayer player = this instanceof final ServerGamePacketListenerImpl gamePacketListener
+            ? gamePacketListener.getPlayer()
+            : this.connection.getPlayer();
+        //noinspection ConstantValue
+        if (player != null
+            && player.visibilitySettings.playerModified()
+            && me.samsuik.sakura.player.visibility.FilterOrModifyPackets.filterOrModifyPacket(packet, player)) {
+            return;
+        }
+        // Sakura end - client visibility settings
         // CraftBukkit end
         if (packet.isTerminal()) {
             this.close();
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index aa10a8cad6e24340d637988dedfa5765dfa9d870..4ec0bda340ca569d4bbf68b498317190f655d610 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -843,6 +843,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return this.blockStateRevisionCounter.get();
     }
     // Sakura end - implement state watchers and tick scheduler
+    public final me.samsuik.sakura.player.visibility.MinimalVisibility minimalVisibility = new me.samsuik.sakura.player.visibility.MinimalVisibility(); // Sakura - client visibility settings
 
     protected Level(
         final WritableLevelData levelData,
