From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 8 May 2026 21:39:04 +0100
Subject: [PATCH] Optimise explosions


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 8e5d6cc4cf66be6e5dcf87b0736555221166a277..6087d66f6ba9c1389dd7997f53cf867f07a6744d 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -77,6 +77,12 @@ public final class ChunkEntitySlices {
         this.allEntities.removeEntity(entity, sectionIndex);
     }
     // Sakura end - optimise cannon entities
+    // Sakura start - optimise explosions
+    public Entity[] getSectionEntities(final int sectionY) {
+        final int sectionIndex = sectionY - this.minSection;
+        return this.allEntities.getSectionEntities(sectionIndex);
+    }
+    // Sakura end - optimise explosions
 
     public ChunkEntitySlices(final Level world, final int chunkX, final int chunkZ, final FullChunkStatus status,
                              final ChunkData chunkData, final int minSection, final int maxSection) { // inclusive, inclusive
@@ -492,6 +498,17 @@ public final class ChunkEntitySlices {
         private final BasicEntityList<Entity>[] entitiesBySection;
         private int count;
 
+        // Sakura start - optimise explosions
+        public Entity[] getSectionEntities(final int sectionIndex) {
+            final BasicEntityList<Entity> entityList = this.entitiesBySection[sectionIndex];
+            if (entityList != null) {
+                return entityList.storage;
+            } else {
+                return new Entity[0];
+            }
+        }
+        // Sakura end - optimise explosions
+
         public EntityCollectionBySection(final ChunkEntitySlices slices) {
             this.slices = slices;
 
diff --git a/me/samsuik/sakura/explosion/ExplosionToSend.java b/me/samsuik/sakura/explosion/ExplosionToSend.java
new file mode 100644
index 0000000000000000000000000000000000000000..41ebc4274994426fdee1ec66981f8c50d459d2b1
--- /dev/null
+++ b/me/samsuik/sakura/explosion/ExplosionToSend.java
@@ -0,0 +1,8 @@
+package me.samsuik.sakura.explosion;
+
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public record ExplosionToSend(Vec3 center, int blockCount) {
+}
diff --git a/me/samsuik/sakura/explosion/ProtectionPluginProbeEvent.java b/me/samsuik/sakura/explosion/ProtectionPluginProbeEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..4e5a86f568ea0560465b25c64ecd6e086a2c53f0
--- /dev/null
+++ b/me/samsuik/sakura/explosion/ProtectionPluginProbeEvent.java
@@ -0,0 +1,35 @@
+package me.samsuik.sakura.explosion;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import org.bukkit.ExplosionResult;
+import org.bukkit.Location;
+import org.bukkit.block.Block;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.entity.EntityExplodeEvent;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.List;
+
+@NullMarked
+public final class ProtectionPluginProbeEvent extends EntityExplodeEvent {
+    public ProtectionPluginProbeEvent(final Entity entity) {
+        super(entity, entity.getLocation(), probeList(entity), 0.0f, ExplosionResult.DESTROY);
+    }
+
+    @Override
+    public boolean callEvent() {
+        return super.callEvent() && !this.blockList().isEmpty();
+    }
+
+    private static List<Block> probeList(final Entity entity) {
+        final Location location = entity.getLocation();
+        final int maxHeight = location.getWorld().getMaxHeight();
+
+        // TNT protection plugins look below blocks for any whitelisted blocks.
+        // If the plugin finds one of those blocks then it clears the block list.
+        // We are using an out of bounds block to try probe for these plugins.
+        final List<Block> list = new ObjectArrayList<>();
+        list.add(location.set(0.0, maxHeight, 0.0).getBlock());
+        return list;
+    }
+}
diff --git a/me/samsuik/sakura/explosion/SortExplosionRays.java b/me/samsuik/sakura/explosion/SortExplosionRays.java
new file mode 100644
index 0000000000000000000000000000000000000000..3985d6977ba9b128a9f88fbbe654a934d878033f
--- /dev/null
+++ b/me/samsuik/sakura/explosion/SortExplosionRays.java
@@ -0,0 +1,46 @@
+package me.samsuik.sakura.explosion;
+
+import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * Sort explosion rays for better cache utilisation.
+ * <pre>
+ *   x +   Vanilla     Sorted
+ * z @ z      8           5
+ * - x      6   7       6   4
+ *        4   @   5   7   @   3
+ *          2   3       8   2
+ *            1           1
+ * </pre>
+ */
+@NullMarked
+public final class SortExplosionRays {
+    private static final Comparator<double[]> EXPLOSION_RAY_COMPARATOR = Comparator.comparingDouble(vec -> {
+        final double sign = Math.signum(vec[0]);
+        final double dir = (sign - 1) / 2;
+        return sign + 8 + vec[2] * dir;
+    });
+
+    public static double[] sortExplosionRays(final DoubleArrayList rayCoords) {
+        final List<double[]> explosionRays = new ArrayList<>();
+        for (int index = 0; index < rayCoords.size(); index += 3) {
+            final double[] vector = new double[3];
+            rayCoords.getElements(index, vector, 0, 3);
+            explosionRays.add(vector);
+        }
+
+        rayCoords.clear();
+        explosionRays.sort(EXPLOSION_RAY_COMPARATOR);
+
+        final double[] rays = new double[explosionRays.size() * 3];
+        for (int i = 0; i < explosionRays.size() * 3; i++) {
+            rays[i] = explosionRays.get(i / 3)[i % 3];
+        }
+        return rays;
+    }
+}
diff --git a/me/samsuik/sakura/explosion/density/ExplosionEntityKey.java b/me/samsuik/sakura/explosion/density/ExplosionEntityKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..1a12f3495e54aabdf7907ba9dd37accdf9bbea39
--- /dev/null
+++ b/me/samsuik/sakura/explosion/density/ExplosionEntityKey.java
@@ -0,0 +1,26 @@
+package me.samsuik.sakura.explosion.density;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class ExplosionEntityKey {
+    public static int encode(final Vec3 center, final Entity entity) {
+        return encode(center, entity.blockPosition(), entity.getType());
+    }
+
+    public static int encode(final Vec3 center, final BlockPos entityBlockPos, final EntityType<?> type) {
+        int key        = Mth.floor(center.x());
+        key = 31 * key + Mth.floor(center.y());
+        key = 31 * key + Mth.floor(center.z());
+        key = 31 * key + entityBlockPos.getX();
+        key = 31 * key + entityBlockPos.getY();
+        key = 31 * key + entityBlockPos.getZ();
+        key = 31 * key + type.hashCode();
+        return key;
+    }
+}
diff --git a/me/samsuik/sakura/explosion/density/ExplosionObstructionCache.java b/me/samsuik/sakura/explosion/density/ExplosionObstructionCache.java
new file mode 100644
index 0000000000000000000000000000000000000000..61df4306c7abef4dbe68786721ed9055d32c0f35
--- /dev/null
+++ b/me/samsuik/sakura/explosion/density/ExplosionObstructionCache.java
@@ -0,0 +1,55 @@
+package me.samsuik.sakura.explosion.density;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2FloatOpenHashMap;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@SuppressWarnings("DataFlowIssue")
+@NullMarked
+public final class ExplosionObstructionCache {
+    public static final float UNKNOWN_SEEN_PERCENT = -1.0f;
+
+    private final Int2ObjectOpenHashMap<KnownObstruction> knownObstructions = new Int2ObjectOpenHashMap<>();
+    private final Object2FloatOpenHashMap<PaperObstructionCacheKey> paperObstructionCache = new Object2FloatOpenHashMap<>();
+    private long lastBlockRevision;
+
+    public @Nullable KnownObstruction getKnownObstruction(final int explosionEntityKey, final long blockRevision) {
+        if (this.lastBlockRevision != blockRevision) {
+            this.knownObstructions.clear();
+        }
+
+        return this.knownObstructions.get(explosionEntityKey);
+    }
+
+    public void addKnownObstruction(
+        final int explosionEntityKey,
+        final Vec3 center,
+        final AABB entityBox,
+        final float seenPercent,
+        final long blockRevision
+    ) {
+        this.knownObstructions.put(explosionEntityKey, new KnownObstruction(center, entityBox, seenPercent));
+        this.lastBlockRevision = blockRevision;
+    }
+
+    public float getPaperObstruction(final PaperObstructionCacheKey cacheKey) {
+        return this.paperObstructionCache.getOrDefault(cacheKey, UNKNOWN_SEEN_PERCENT);
+    }
+
+    public void addPaperObstruction(final PaperObstructionCacheKey cacheKey, final float seenPercent) {
+        this.paperObstructionCache.put(cacheKey, seenPercent);
+    }
+
+    public void tick(final long gameTime) {
+        this.knownObstructions.clear();
+        this.paperObstructionCache.clear();
+
+        if (gameTime % 200 == 0) {
+            this.knownObstructions.trim();
+            this.paperObstructionCache.trim();
+        }
+    }
+}
diff --git a/me/samsuik/sakura/explosion/density/KnownObstruction.java b/me/samsuik/sakura/explosion/density/KnownObstruction.java
new file mode 100644
index 0000000000000000000000000000000000000000..1087594cb544ee3d935737d779aad58a4982f3bb
--- /dev/null
+++ b/me/samsuik/sakura/explosion/density/KnownObstruction.java
@@ -0,0 +1,50 @@
+package me.samsuik.sakura.explosion.density;
+
+import me.samsuik.sakura.utils.AABBUtil;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.HitResult;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class KnownObstruction {
+    private AABB explosionBox;
+    private AABB entityBox;
+    private final float seenPercent;
+
+    public KnownObstruction(final Vec3 center, final AABB entityBox, final float seenPercent) {
+        this.explosionBox = new AABB(center, center);
+        this.entityBox = entityBox;
+        this.seenPercent = seenPercent;
+    }
+
+    public float getSeenPercent() {
+        return this.seenPercent;
+    }
+
+    public boolean isPartiallyObstructed() {
+        return this.seenPercent != 0.0f && this.seenPercent != 1.0f;
+    }
+
+    public HitResult.Type hitResult() {
+        return this.seenPercent == 1.0f ? HitResult.Type.MISS : HitResult.Type.BLOCK;
+    }
+
+    public boolean has(final Vec3 center, final AABB entityBox) {
+        return this.isExplosionKnown(center)
+            && AABBUtil.containsInclusive(this.entityBox, entityBox);
+    }
+
+    public boolean isExplosionKnown(final Vec3 center) {
+        return AABBUtil.containsInclusive(this.explosionBox, center);
+    }
+
+    public boolean isEntityPositionKnown(final Vec3 from) {
+        return AABBUtil.containsInclusive(this.entityBox, from);
+    }
+
+    public void expandKnownArea(final Vec3 center, final AABB entityBox) {
+        this.explosionBox = AABBUtil.expand(this.explosionBox, center);
+        this.entityBox    = this.entityBox.minmax(entityBox);
+    }
+}
diff --git a/me/samsuik/sakura/explosion/density/PaperObstructionCacheKey.java b/me/samsuik/sakura/explosion/density/PaperObstructionCacheKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..5ce6fd6f5fdf8772512db4fa9dcbfa90ad1072fa
--- /dev/null
+++ b/me/samsuik/sakura/explosion/density/PaperObstructionCacheKey.java
@@ -0,0 +1,12 @@
+package me.samsuik.sakura.explosion.density;
+
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public record PaperObstructionCacheKey(Vec3 explosionPos, Vec3 entityPos, float width, float height) {
+    public PaperObstructionCacheKey(final Vec3 center, final Entity entity) {
+        this(center, entity.position(), entity.getBbWidth(), entity.getBbHeight());
+    }
+}
diff --git a/me/samsuik/sakura/explosion/density/SeenPercentRays.java b/me/samsuik/sakura/explosion/density/SeenPercentRays.java
new file mode 100644
index 0000000000000000000000000000000000000000..ebd6c04e276b607077545f197bf326a7d8002e4f
--- /dev/null
+++ b/me/samsuik/sakura/explosion/density/SeenPercentRays.java
@@ -0,0 +1,74 @@
+package me.samsuik.sakura.explosion.density;
+
+import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class SeenPercentRays {
+    private static final AABB BLOCK_ENTITY_BOX = new AABB(Vec3.ZERO, new Vec3(0.98f, 0.98f, 0.98f));
+    public static final double[] SEEN_PERCENT_RAYS;
+
+    static {
+        SEEN_PERCENT_RAYS = rays(BLOCK_ENTITY_BOX).toDoubleArray();
+    }
+
+    private static DoubleArrayList rays(final AABB boundingBox) {
+        // Copied from Paper/Moonrise ServerExplosion
+        final double diffX = boundingBox.getXsize();
+        final double diffY = boundingBox.getYsize();
+        final double diffZ = boundingBox.getZsize();
+
+        final double incX = 1.0 / (diffX * 2.0 + 1.0);
+        final double incY = 1.0 / (diffY * 2.0 + 1.0);
+        final double incZ = 1.0 / (diffZ * 2.0 + 1.0);
+
+        final double offX = (1.0 - Math.floor(1.0 / incX) * incX) * 0.5;
+        final double offZ = (1.0 - Math.floor(1.0 / incZ) * incZ) * 0.5;
+
+        final int horizontalCount = (int) Math.floor(1.0 / incX);
+        final int verticalCount   = (int) Math.floor(1.0 / incY);
+
+        final DoubleArrayList rays = new DoubleArrayList();
+        for (int x = 0; x <= horizontalCount; x++) {
+            for (int y = 0; y <= verticalCount; y++) {
+                for (int z = 0; z <= horizontalCount; z++) {
+                    final double fromX = Math.fma(incX * x, diffX, offX);
+                    final double fromY = (incY * y) * diffY;
+                    final double fromZ = Math.fma(incZ * z, diffZ, offZ);
+
+                    if (isCornerOrCenter(x, y, z, horizontalCount, verticalCount)) {
+                        rays.add(0, fromX);
+                        rays.add(1, fromY);
+                        rays.add(2, fromZ);
+                    } else {
+                        rays.add(fromX);
+                        rays.add(fromY);
+                        rays.add(fromZ);
+                    }
+                }
+            }
+        }
+
+        return rays;
+    }
+
+    private static boolean isCornerOrCenter(
+        final int x,
+        final int y,
+        final int z,
+        final int horizontalCount,
+        final int verticalCount
+    ) {
+        if ((x == 0 || x == horizontalCount)
+            && (y == 0 || y == verticalCount)
+            && (z == 0 || z == horizontalCount)) {
+            return true;
+        }
+
+        final int horizontalCenter = horizontalCount / 2;
+        final int verticalCenter = verticalCount / 2;
+        return x == horizontalCenter && y == verticalCenter && z == horizontalCenter;
+    }
+}
diff --git a/me/samsuik/sakura/explosion/specialised/SpecialisedExplosion.java b/me/samsuik/sakura/explosion/specialised/SpecialisedExplosion.java
new file mode 100644
index 0000000000000000000000000000000000000000..ad5cf60b4a865c6089bf2dbca0325e717fdffa65
--- /dev/null
+++ b/me/samsuik/sakura/explosion/specialised/SpecialisedExplosion.java
@@ -0,0 +1,520 @@
+package me.samsuik.sakura.explosion.specialised;
+
+import ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet;
+import ca.spottedleaf.moonrise.common.util.WorldUtil;
+import ca.spottedleaf.moonrise.patches.chunk_system.level.entity.ChunkEntitySlices;
+import ca.spottedleaf.moonrise.patches.chunk_system.level.entity.EntityLookup;
+import ca.spottedleaf.moonrise.patches.collisions.CollisionUtil;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import me.samsuik.sakura.entity.EntityState;
+import me.samsuik.sakura.entity.SakuraEntity;
+import me.samsuik.sakura.entity.SakuraEntityController;
+import me.samsuik.sakura.explosion.density.*;
+import me.samsuik.sakura.utils.AABBUtil;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.Mth;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.entity.monster.cubemob.SulfurCube;
+import net.minecraft.world.entity.projectile.hurtingprojectile.windcharge.AbstractWindCharge;
+import net.minecraft.world.level.Explosion;
+import net.minecraft.world.level.ExplosionDamageCalculator;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.ServerExplosion;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+@NullMarked
+public abstract class SpecialisedExplosion<E extends Entity & SakuraEntity> extends ServerExplosion {
+    private static final double SAFE_DISPATCH_DISTANCE_SQR = 16.0 * 16.0;
+    private static final int ALL_DIRECTIONS = 0b111;
+    private static final int FOUND_ALL_BLOCKS = ALL_DIRECTIONS + 12;
+    private static final int REDUCED_SEEN_PERCENT_RAYS = 9 * 3;
+
+    private final E cause;
+    private final Vec3 originalPosition;
+    private int swinging = 0;
+    private boolean moved = false;
+    private Vec3 safeDispatchPosition;
+    private final List<Vec3> bufferedExplosions = new ObjectArrayList<>();
+    private AABB bounds;
+    private final Set<BlockPos> gameEvents = new ObjectOpenHashSet<>();
+    private final boolean interactWithBlocks;
+
+    public SpecialisedExplosion(
+        final ServerLevel level,
+        final E cause,
+        final @Nullable DamageSource damageSource,
+        final @Nullable ExplosionDamageCalculator damageCalculator,
+        final Vec3 center,
+        final float radius,
+        final boolean fire,
+        final Explosion.BlockInteraction blockInteraction
+    ) {
+        super(level, cause, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+        this.cause = cause;
+        this.originalPosition = center;
+        this.safeDispatchPosition = center;
+        this.bounds = new AABB(center, center);
+        this.interactWithBlocks = this.interactsWithBlocks();
+    }
+
+    public static ServerExplosion appropriateExplosion(
+        final ServerLevel level,
+        final @Nullable Entity source,
+        final @Nullable DamageSource damageSource,
+        final @Nullable ExplosionDamageCalculator damageCalculator,
+        final Vec3 center,
+        final float radius,
+        final boolean fire,
+        final Explosion.BlockInteraction blockInteraction
+    ) {
+        if (level.sakuraConfig().cannons.explosion.batchExplosions) {
+            if (source instanceof final PrimedTnt tnt) {
+                return new TntExplosion(level, tnt, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+            }
+
+            if (source instanceof final AbstractWindCharge windCharge) {
+                return new WindChargeExplosion(level, windCharge, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+            }
+        }
+
+        return new ServerExplosion(level, source, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+    }
+
+    protected abstract Vec3 explosionPosition(final E explodingEntity);
+
+    protected abstract void preventEntityExploding(final E explodableEntity);
+
+    protected abstract boolean continueExploding(final E explodingEntity);
+
+    private @Nullable E nextEntity(final E explodingEntity, final boolean source) {
+        final IteratorSafeOrderedReferenceSet<Entity> entities = this.level().entityTickList.entities;
+        entities.startRawIterator();
+
+        final int index = entities.indexOf(explodingEntity);
+        final int nextIndex = entities.advanceRawIterator(index);
+        if (nextIndex != Integer.MAX_VALUE) {
+            final Entity nextEntity = entities.rawGet(nextIndex);
+            explodingEntity.getSakuraEntityController().decouple();
+
+            if (nextEntity instanceof final SakuraEntity nextSakuraEntity
+                && nextSakuraEntity.samePropertiesAndState(explodingEntity, source ? 1 : 0)) {
+                entities.finishRawIterator();
+                return (E) nextEntity;
+            }
+        }
+
+        entities.finishRawIterator();
+        return null;
+    }
+
+    @Override
+    public final int explode() {
+        this.initialiseBlockCache(); // initialise the paper explosion block cache
+        final double doubleRadius = this.radius() * 2.0;
+
+        E explodingEntity = this.cause;
+        E nextEntity = this.nextEntity(explodingEntity, true);
+        while (true) {
+            final boolean lastExplosion = nextEntity == null;
+            final SakuraEntityController controller = explodingEntity.getSakuraEntityController();
+            final List<BlockPos> blocksToExplode = this.collectBlocksAndImpactEntities(
+                explodingEntity,
+                doubleRadius,
+                this.interactWithBlocks && this.swinging < FOUND_ALL_BLOCKS,
+                lastExplosion
+            );
+
+            final boolean destroyedBlocks = this.finalizeExplosion(blocksToExplode);
+            if (lastExplosion || !this.continueExploding(explodingEntity)) {
+                break;
+            } else {
+                // Find the next entity before the entity is removed from the nextTickingList
+                final EntityState entityState = controller.getEntityState();
+                explodingEntity = nextEntity;
+                nextEntity = this.nextEntity(explodingEntity, false);
+
+                entityState.apply(explodingEntity);
+
+                // The source velocity has to be calculated before nextExplosion as it clears the blockCache
+                this.impactEntity(explodingEntity, this.center, 1, doubleRadius);
+                this.nextExplosion(blocksToExplode, destroyedBlocks);
+
+                if (!this.tickExplodableEntity(explodingEntity, doubleRadius, destroyedBlocks)) {
+                    if (!this.bufferedExplosions.isEmpty()) {
+                        this.bufferedExplosions.removeLast();
+                    }
+
+                    this.locateAndImpactEntitiesInBounds(doubleRadius);
+                    return 0;
+                }
+            }
+        }
+
+        return 0;
+    }
+
+    private List<BlockPos> collectBlocksAndImpactEntities(
+        final E explodingEntity,
+        final double doubleRadius,
+        final boolean interactWithBlocks,
+        boolean dispatch
+    ) {
+        final Vec3 center = this.center;
+        if (interactWithBlocks && this.gameEvents.add(BlockPos.containing(center))) {
+            this.level().gameEvent(explodingEntity, GameEvent.EXPLODE, center);
+        }
+
+        // Force a dispatch if the explosion has "drifted" too far
+        // This is to reduce the size of the block cache and prevent abuse
+        if (center.distanceToSqr(this.safeDispatchPosition) > SAFE_DISPATCH_DISTANCE_SQR) {
+            dispatch = true;
+            this.safeDispatchPosition = center;
+            this.blockCache.clear();
+            Arrays.fill(this.directMappedBlockCache, null);
+        }
+
+        // Collect all the blocks to explode
+        final List<BlockPos> blocksToExplode = interactWithBlocks
+            ? this.calculateExplodedPositions()
+            : Collections.emptyList();
+
+        // Buffer explosions to reduce the amount of calculations and improve locality
+        this.bufferedExplosions.add(center);
+        this.bounds = AABBUtil.expand(this.bounds, center);
+
+        // Dispatch the buffered explosions
+        final SakuraEntityController controller = explodingEntity.getSakuraEntityController();
+        if (dispatch || !blocksToExplode.isEmpty() || controller.haveBlocksChangedSinceTicking(this.level())) {
+            this.locateAndImpactEntitiesInBounds(doubleRadius);
+            this.bounds = new AABB(center, center);
+            this.bufferedExplosions.clear();
+            this.gameEvents.clear();
+        }
+
+        return blocksToExplode;
+    }
+
+    private boolean finalizeExplosion(final List<BlockPos> blocksToExplode) {
+        this.wasCanceled = false;
+        final List<BlockPos> blocksDestroyed = new ObjectArrayList<>(blocksToExplode);
+        this.interactWithBlocks(blocksDestroyed);
+
+        if (!this.wasCanceled) {
+            // Packets are sent after the explosion
+            this.queueNotification(this.center, blocksDestroyed.size());
+        }
+
+        return !blocksDestroyed.isEmpty();
+    }
+
+    protected void nextExplosion(final List<BlockPos> blocksToExplode, final boolean destroyedBlocks) {
+        // Reuse the block cache between explosions. This can help a lot when searching for blocks and raytracing.
+        // This is disabled by default as it's incompatible with plugins that modify blocks in the explosion event.
+        if (this.level().sakuraConfig().cannons.explosion.reuseBlockCacheAcrossExplosions && !blocksToExplode.isEmpty() && !destroyedBlocks) {
+            this.markBlocksInCacheAsExplodable(blocksToExplode);
+        } else {
+            super.blockCache.clear();
+        }
+
+        Arrays.fill(this.directMappedBlockCache, null);
+
+        if (this.swinging >= ALL_DIRECTIONS) {
+            // Increment "swinging" if no blocks have been found, and it has swung in every direction.
+            // This is used to skip expensive exploded block calculations.
+            if (!destroyedBlocks && this.level().sakuraConfig().cannons.explosion.avoidRedundantBlockSearches) {
+                this.swinging++;
+            } else {
+                this.swinging = ALL_DIRECTIONS;
+            }
+        }
+    }
+
+    private void updateSwingingState(final Vec3 newMovement, final Vec3 oldMovement) {
+        for (final Direction.Axis axis : Direction.Axis.VALUES) {
+            final double current = newMovement.get(axis);
+            final double previous = oldMovement.get(axis);
+            if (current == previous || current * previous <= 0.0) {
+                this.swinging |= 1 << axis.ordinal();
+            }
+        }
+    }
+
+    private boolean tickExplodableEntity(final E explodableEntity, final double doubleRadius, final boolean destroyedBlocks) {
+        final SakuraEntityController controller = explodableEntity.getSakuraEntityController();
+        final Vec3 oldMovement = controller.getEntityState().movement();
+        final boolean hasMoved;
+
+        // Check if we have moved before applying the entity state
+        if (this.moved) {
+            hasMoved = true;
+        } else if (this.center.equals(cause.position())) {
+            hasMoved = false;
+        } else {
+            final double newMovementSqr = explodableEntity.getDeltaMovement().lengthSqr();
+            final double oldMovementSqr = oldMovement.lengthSqr();
+            final double maxExplosionRadiusSqr = Math.pow(doubleRadius + 1.0, 2.0);
+            hasMoved = oldMovementSqr <= maxExplosionRadiusSqr || newMovementSqr <= oldMovementSqr;
+        }
+
+        // Ticking is always required after destroying a block.
+        if (destroyedBlocks || hasMoved) {
+            this.preventEntityExploding(explodableEntity);
+            controller.tickEntity();
+
+            if (explodableEntity.isRemoved()) {
+                return false;
+            }
+
+            this.center = this.explosionPosition(explodableEntity);
+            this.moved |= !this.center.equals(this.originalPosition);
+        }
+
+        this.source = explodableEntity;
+        explodableEntity.discard();
+
+        // Update swinging state
+        if (this.moved) {
+            this.updateSwingingState(controller.getEntityState().movement(), oldMovement);
+        } else {
+            this.swinging = ALL_DIRECTIONS;
+        }
+
+        return true;
+    }
+
+    private void impactEntity(final Entity entity, final Vec3 explosionPos, final int potential, final double doubleRadius) {
+        if (entity == this.source || entity instanceof AbstractWindCharge) {
+            return;
+        }
+
+        // Fast path for cannon entities
+        if (entity.cannonEntity && !(entity instanceof SulfurCube)) {
+            this.impactCannonEntity(entity, explosionPos, potential, doubleRadius);
+        } else {
+            final AABB impactArea = new AABB(explosionPos, explosionPos).inflate(doubleRadius + 1.0);
+            for (int count = 0; count < potential; ++count) {
+                this.sakura$impactEntity(entity, explosionPos, impactArea, doubleRadius);
+            }
+        }
+    }
+
+    protected final void impactCannonEntity(final Entity entity, final Vec3 explosionPos, final int potential, final double radius) {
+        double distanceFromBottom = Math.sqrt(entity.distanceToSqr(explosionPos)) / radius;
+        if (distanceFromBottom <= 1.0) {
+            double x = entity.getX() - explosionPos.x;
+            double y = (entity instanceof PrimedTnt ? entity.getY() : entity.getEyeY()) - explosionPos.y;
+            double z = entity.getZ() - explosionPos.z;
+            double distance = Math.sqrt(x * x + y * y + z * z);
+
+            if (distance >= 1.0e-5) {
+                x /= distance;
+                y /= distance;
+                z /= distance;
+                final double density = this.getSeenPercentCached(explosionPos, entity);
+                final double exposure = (1.0D - distanceFromBottom) * density;
+                if (exposure == 0.0) {
+                    return;
+                }
+
+                x *= exposure;
+                y *= exposure;
+                z *= exposure;
+
+                this.applyEntityVelocity(entity, x, y, z, potential);
+            }
+        }
+    }
+
+    protected final void applyEntityVelocity(
+        final Entity entity,
+        final double x,
+        final double y,
+        final double z,
+        final int potential
+    ) {
+        final Vec3 movement = entity.getDeltaMovement();
+        double moveX = movement.x();
+        double moveY = movement.y();
+        double moveZ = movement.z();
+
+        for (int count = 0; count < potential; ++count) {
+            moveX += x;
+            moveY += y;
+            moveZ += z;
+        }
+
+        entity.setDeltaMovement(moveX, moveY, moveZ);
+        entity.needsSync = true;
+    }
+
+    protected final float getSeenPercentCached(final Vec3 center, final Entity entity) {
+        final Level level = this.level();
+        final AABB entityBox = entity.getBoundingBox();
+
+        final long blockRevision = level.getBlockStateRevisionCount();
+        final int explosionEntityKey = ExplosionEntityKey.encode(center, entity);
+        final KnownObstruction obstruction = level.obstructionCache.getKnownObstruction(explosionEntityKey, blockRevision);
+
+        if (obstruction != null && obstruction.has(center, entityBox)) {
+            return obstruction.getSeenPercent();
+        }
+
+        final PaperObstructionCacheKey cacheKey;
+        if (this.brokenPaperBehaviour) {
+            cacheKey = new PaperObstructionCacheKey(center, entity);
+
+            final float paperSeenPercent = level.obstructionCache.getPaperObstruction(cacheKey);
+            if (paperSeenPercent != ExplosionObstructionCache.UNKNOWN_SEEN_PERCENT) {
+                return paperSeenPercent;
+            }
+        } else {
+            cacheKey = null;
+        }
+
+        final KnownObstruction safeKnownObstruction;
+        if (obstruction != null && !obstruction.isPartiallyObstructed() && obstruction.isExplosionKnown(center)) {
+            safeKnownObstruction = obstruction;
+        } else {
+            safeKnownObstruction = null;
+        }
+
+        final float seenPercent = this.calculateSeenPercent(level, center, entity, safeKnownObstruction);
+        if (obstruction == null || obstruction.isPartiallyObstructed()) {
+            level.obstructionCache.addKnownObstruction(explosionEntityKey, center, entityBox, seenPercent, blockRevision);
+        } else if (seenPercent == obstruction.getSeenPercent()) {
+            obstruction.expandKnownArea(center, entityBox);
+        }
+
+        if (cacheKey != null) {
+            level.obstructionCache.addPaperObstruction(cacheKey, seenPercent);
+        }
+
+        return seenPercent;
+    }
+
+    private float calculateSeenPercent(
+        final Level level,
+        final Vec3 center,
+        final Entity entity,
+        final @Nullable KnownObstruction obstruction
+    ) {
+        // For now, we only support skipping ray casts on the edges for block-like entities
+        if (!level.sakuraConfig().cannons.explosion.reduceExposureRaycasts
+            || obstruction != null
+            || entity.getBbHeight() != 0.98f
+            || entity.getBbWidth()  != 0.98f) {
+            return this.getSeenFraction(center, entity, this.directMappedBlockCache, this.mutablePos, obstruction);
+        }
+
+        final CollisionUtil.LazyEntityCollisionContext context = new CollisionUtil.LazyEntityCollisionContext(entity);
+
+        final AABB entityBox = entity.getBoundingBox();
+        final double baseX = entityBox.minX;
+        final double baseY = entityBox.minY;
+        final double baseZ = entityBox.minZ;
+
+        final double[] rayCoords = SeenPercentRays.SEEN_PERCENT_RAYS;
+        int end = Math.min(REDUCED_SEEN_PERCENT_RAYS, rayCoords.length);
+
+        int totalRays = 0;
+        int missedRays = 0;
+
+        for (int rays = 0; rays < end; rays += 3) {
+            final double x = rayCoords[rays];
+            final double y = rayCoords[rays + 1];
+            final double z = rayCoords[rays + 2];
+            final Vec3 from = new Vec3(
+                baseX + x,
+                baseY + y,
+                baseZ + z
+            );
+
+            final int result = this.clipsAnything(from, center, context, this.directMappedBlockCache, this.mutablePos);
+            if ((result & ANY_COLLISION) == 0) {
+                missedRays++;
+            }
+
+            totalRays++;
+
+            if (missedRays != 0 && totalRays != missedRays || (result & MISSED_COLLISION) != 0) {
+                end = rayCoords.length;
+            }
+        }
+
+        return (float) missedRays / (float) totalRays;
+    }
+
+    private void locateAndImpactEntitiesInBounds(final double doubleRadius) {
+        final List<Vec3> bufferedExplosions = this.bufferedExplosions;
+        final AABB bounds = this.bounds;
+
+        final double maxDeviation = Math.max(bounds.getXsize(), Math.max(bounds.getYsize(), bounds.getZsize()));
+        final double effectiveDistanceSqr = Math.pow(doubleRadius + maxDeviation, 2.0);
+        final boolean stationary = maxDeviation == 0.0;
+        final Vec3 boundsCenter = bounds.getCenter();
+
+        this.forEachEntitySliceInBounds(bounds.inflate(effectiveDistanceSqr), entities -> {
+            for (int index = 0; index < entities.length; index++) {
+                final Entity entity = entities[index];
+                if (entity == null) break; // end of the entity slice
+
+                if (stationary) {
+                    this.impactEntity(entity, bufferedExplosions.getFirst(), bufferedExplosions.size(), doubleRadius);
+                } else {
+                    // Check if the entity is in range
+                    if (entity.distanceToSqr(boundsCenter) > effectiveDistanceSqr) {
+                        continue;
+                    }
+
+                    //noinspection ForLoopReplaceableByForEach
+                    for (int explosionIndex = 0; explosionIndex < bufferedExplosions.size(); ++explosionIndex) {
+                        this.impactEntity(entity, bufferedExplosions.get(explosionIndex), 1, doubleRadius);
+                    }
+                }
+
+                // Entities can be removed from the world mid-explosion
+                if (entities[index] != entity) {
+                    index--;
+                }
+            }
+        });
+    }
+
+    private void forEachEntitySliceInBounds(final AABB bounds, final Consumer<@Nullable Entity[]> sliceConsumer) {
+        final int minSection = WorldUtil.getMinSection(this.level());
+        final int maxSection = WorldUtil.getMaxSection(this.level());
+
+        final int minChunkX = Mth.floor(bounds.minX) >> 4;
+        final int minChunkY = Mth.clamp(Mth.floor(bounds.minY) >> 4, minSection, maxSection);
+        final int minChunkZ = Mth.floor(bounds.minZ) >> 4;
+        final int maxChunkX = Mth.floor(bounds.maxX) >> 4;
+        final int maxChunkY = Mth.clamp(Mth.floor(bounds.maxY) >> 4, minSection, maxSection);
+        final int maxChunkZ = Mth.floor(bounds.maxZ) >> 4;
+
+        final EntityLookup entityLookup = this.level().moonrise$getEntityLookup();
+        for (int chunkX = minChunkX; chunkX <= maxChunkX; ++chunkX) {
+            for (int chunkZ = minChunkZ; chunkZ <= maxChunkZ; ++chunkZ) {
+                final ChunkEntitySlices chunk = entityLookup.getChunk(chunkX, chunkZ);
+                if (chunk == null) {
+                    continue;
+                }
+
+                for (int chunkY = minChunkY; chunkY <= maxChunkY; ++chunkY) {
+                    sliceConsumer.accept(chunk.getSectionEntities(chunkY));
+                }
+            }
+        }
+    }
+}
diff --git a/me/samsuik/sakura/explosion/specialised/TntExplosion.java b/me/samsuik/sakura/explosion/specialised/TntExplosion.java
new file mode 100644
index 0000000000000000000000000000000000000000..3242025d231b5e492f918c779a5dd2b3dbd906ad
--- /dev/null
+++ b/me/samsuik/sakura/explosion/specialised/TntExplosion.java
@@ -0,0 +1,44 @@
+package me.samsuik.sakura.explosion.specialised;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.level.ExplosionDamageCalculator;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public final class TntExplosion extends SpecialisedExplosion<PrimedTnt> {
+    public TntExplosion(
+        final ServerLevel level,
+        final PrimedTnt cause,
+        final @Nullable DamageSource damageSource,
+        final @Nullable ExplosionDamageCalculator damageCalculator,
+        final Vec3 center,
+        final float radius,
+        final boolean fire,
+        final BlockInteraction blockInteraction
+    ) {
+        super(level, cause, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+    }
+
+    @Override
+    protected Vec3 explosionPosition(final PrimedTnt explodingEntity) {
+        return new Vec3(
+            explodingEntity.getX(),
+            explodingEntity.getY(0.0625),
+            explodingEntity.getZ()
+        );
+    }
+
+    @Override
+    protected void preventEntityExploding(final PrimedTnt explodableEntity) {
+        explodableEntity.canExplode = false;
+    }
+
+    @Override
+    protected boolean continueExploding(PrimedTnt explodingEntity) {
+        return true;
+    }
+}
diff --git a/me/samsuik/sakura/explosion/specialised/WindChargeExplosion.java b/me/samsuik/sakura/explosion/specialised/WindChargeExplosion.java
new file mode 100644
index 0000000000000000000000000000000000000000..02bdd15270c0f921aa5786c6468be74d75eb6674
--- /dev/null
+++ b/me/samsuik/sakura/explosion/specialised/WindChargeExplosion.java
@@ -0,0 +1,40 @@
+package me.samsuik.sakura.explosion.specialised;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.projectile.hurtingprojectile.windcharge.AbstractWindCharge;
+import net.minecraft.world.level.ExplosionDamageCalculator;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public final class WindChargeExplosion extends SpecialisedExplosion<AbstractWindCharge> {
+    public WindChargeExplosion(
+        final ServerLevel level,
+        final AbstractWindCharge cause,
+        final @Nullable DamageSource damageSource,
+        final @Nullable ExplosionDamageCalculator damageCalculator,
+        final Vec3 center,
+        final float radius,
+        final boolean fire,
+        final BlockInteraction blockInteraction
+    ) {
+        super(level, cause, damageSource, damageCalculator, center, radius, fire, blockInteraction);
+    }
+
+    @Override
+    protected Vec3 explosionPosition(final AbstractWindCharge explodingEntity) {
+        return this.center;
+    }
+
+    @Override
+    protected void preventEntityExploding(final AbstractWindCharge explodableEntity) {
+        explodableEntity.canExplode = false;
+    }
+
+    @Override
+    protected boolean continueExploding(final AbstractWindCharge explodingEntity) {
+        return !explodingEntity.getSakuraEntityController().haveBlocksChangedSinceTicking(this.level());
+    }
+}
diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index 72c00ba66881539cfec183deec9f37bf63e92ac9..c86ff7140402b407e04f456b73f8f0586ce72a57 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -9,5 +9,6 @@ import org.jspecify.annotations.NullMarked;
 public final class ScheduleSakuraTasks {
     public static void schedule(final LevelTickScheduler scheduler) {
         scheduler.endless((tick, level) -> level.sakuraEntityManager.tick(tick), 0);
+        scheduler.endless((tick, level) -> level.obstructionCache.tick(tick), 0);
     }
 }
diff --git a/me/samsuik/sakura/utils/AABBUtil.java b/me/samsuik/sakura/utils/AABBUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..ca9cdb3fe90f1f9224b5ff6afccb81943ff313a0
--- /dev/null
+++ b/me/samsuik/sakura/utils/AABBUtil.java
@@ -0,0 +1,30 @@
+package me.samsuik.sakura.utils;
+
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public final class AABBUtil {
+    public static boolean containsInclusive(final AABB box, final AABB otherBox) {
+        return box.minX <= otherBox.minX && box.maxX >= otherBox.maxX
+            && box.minY <= otherBox.minY && box.maxY >= otherBox.maxY
+            && box.minZ <= otherBox.minZ && box.maxZ >= otherBox.maxZ;
+    }
+
+    public static boolean containsInclusive(final AABB box, final Vec3 position) {
+        return box.minX <= position.x && box.maxX >= position.x
+            && box.minY <= position.y && box.maxY >= position.y
+            && box.minZ <= position.z && box.maxZ >= position.z;
+    }
+    
+    public static AABB expand(final AABB box, final Vec3 position) {
+        final double minX = Math.min(box.minX, position.x);
+        final double minY = Math.min(box.minY, position.y);
+        final double minZ = Math.min(box.minZ, position.z);
+        final double maxX = Math.max(box.maxX, position.x);
+        final double maxY = Math.max(box.maxY, position.y);
+        final double maxZ = Math.max(box.maxZ, position.z);
+        return new AABB(minX, minY, minZ, maxX, maxY, maxZ);
+    }
+}
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 4b2389e85b9d3c129a00cafa80910e1fd06b4aa3..cd5fa9f0ef914d42a2987bdbcde89dd0d24ab390 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2020,7 +2020,18 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             case STANDARD -> Explosion.BlockInteraction.DESTROY; // CraftBukkit - handle custom explosion type
         };
         Vec3 center = new Vec3(x, y, z);
-        ServerExplosion explosion = new ServerExplosion(this, source, damageSource, damageCalculator, center, r, fire, blockInteraction);
+        // Sakura start - optimise explosions
+        ServerExplosion explosion = me.samsuik.sakura.explosion.specialised.SpecialisedExplosion.appropriateExplosion(
+            this,
+            source,
+            damageSource,
+            damageCalculator,
+            center,
+            r,
+            fire,
+            blockInteraction
+        );
+        // Sakura end - optimise explosions
         if (configurator != null) configurator.accept(explosion); // Paper - Allow explosions to damage source
         this.sakuraEntityManager.decoupleAllTickingEntities(); // Sakura - optimise cannon entities
         int blockCount = explosion.explode();
@@ -2029,6 +2040,34 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             return explosion;
         }
         // CraftBukkit end
+        // Sakura start - optimise explosions
+        for (final me.samsuik.sakura.explosion.ExplosionToSend explosionToSend : explosion.getToSend()) {
+            this.notifyPlayersOfExplosion(
+                explosion,
+                explosionToSend.center(),
+                explosionToSend.blockCount(),
+                r,
+                smallExplosionParticles,
+                largeExplosionParticles,
+                blockParticles,
+                explosionSound
+            );
+        }
+
+        return explosion;
+    }
+
+    private void notifyPlayersOfExplosion(
+        final ServerExplosion explosion,
+        final Vec3 center,
+        final int blockCount,
+        float r,
+        final ParticleOptions smallExplosionParticles,
+        final ParticleOptions largeExplosionParticles,
+        final WeightedList<ExplosionParticleInfo> blockParticles,
+        final Holder<SoundEvent> explosionSound
+    ) {
+        // Sakura end - optimise explosions
         ParticleOptions explosionParticle = explosion.isSmall() ? smallExplosionParticles : largeExplosionParticles;
 
         for (ServerPlayer player : this.players) {
@@ -2056,7 +2095,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
         }
 
-        return explosion; // CraftBukkit
+        explosion.getHitPlayers().clear(); // Sakura - optimise explosions
     }
 
     private Explosion.BlockInteraction getDestroyType(final GameRule<Boolean> gameRule) {
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 12723eaed70c7172a31431854ed464dd5256f2f3..c6e777045d97bd800299872e1ccfc3d6a09fe32c 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -66,6 +66,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
 
     // Sakura start - optimise cannon entities
     private final me.samsuik.sakura.entity.SakuraEntityController sakuraEntityController;
+    public boolean canExplode = true; // Sakura - optimise explosions
 
     @Override
     public final me.samsuik.sakura.entity.SakuraEntityController getSakuraEntityController() {
@@ -206,7 +207,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
     }
 
     private void explode() {
-        if (this.level() instanceof ServerLevel level && level.getGameRules().get(GameRules.TNT_EXPLODES)) {
+        if (this.level() instanceof ServerLevel level && level.getGameRules().get(GameRules.TNT_EXPLODES) && this.canExplode) { // Sakura - optimise explosions
             // CraftBukkit start
             ExplosionPrimeEvent event = CraftEventFactory.callExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
             if (event.isCancelled()) {
diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java
index 9c466764d2008010684d7bae23b288d156487ad3..1ce2029f195465d383e8bafc459e7e8086a9077a 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java
@@ -33,6 +33,8 @@ public abstract class AbstractWindCharge extends AbstractHurtingProjectile imple
     public static final double JUMP_SCALE = 0.25;
 
     // Sakura start - optimise cannon entities
+    public boolean canExplode = true; // Sakura - optimise explosions
+
     @Override
     public final boolean canLink(final Entity entity, final int tickDifference) {
         return true;
diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/BreezeWindCharge.java b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/BreezeWindCharge.java
index 8477b19c67cafcc339719b97886dd77a346cbe3a..1cf14061e3eb7e09677e1ca0d67bcf1019756b3c 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/BreezeWindCharge.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/BreezeWindCharge.java
@@ -22,6 +22,7 @@ public class BreezeWindCharge extends AbstractWindCharge {
 
     @Override
     public void explode(final Vec3 position) {
+        if (!this.canExplode) { return; } // Sakura - optimise explosions
         // Paper start - Fire event for WindCharge explosions
         org.bukkit.event.entity.ExplosionPrimeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExplosionPrimeEvent(this, RADIUS, false);
         if (event.isCancelled()) {
diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java
index 0f952be087c52ba5de78d1f91aed1c160ec5788d..7f79b85f3d2983c8fa7f5b9cab551adbd4a1160d 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java
@@ -67,6 +67,7 @@ public class WindCharge extends AbstractWindCharge {
 
     @Override
     public void explode(final Vec3 position) {
+        if (!this.canExplode) { return; } // Sakura - optimise explosions
         // Paper start - Fire event for WindCharge explosions
         org.bukkit.event.entity.ExplosionPrimeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExplosionPrimeEvent(this, RADIUS, false);
         if (event.isCancelled()) {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 17ed97203d10eb2b3119247ee1a467b8ad3006e0..e4e6133147accd82e94bd1c188f94c4f9cd716a4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -846,6 +846,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     // Sakura end - implement state watchers and tick scheduler
     public final me.samsuik.sakura.player.visibility.MinimalVisibility minimalVisibility = new me.samsuik.sakura.player.visibility.MinimalVisibility(); // Sakura - client visibility settings
     public final me.samsuik.sakura.entity.SakuraEntityManager sakuraEntityManager = new me.samsuik.sakura.entity.SakuraEntityManager(); // Sakura - optimise cannon entities
+    public final me.samsuik.sakura.explosion.density.ExplosionObstructionCache obstructionCache = new me.samsuik.sakura.explosion.density.ExplosionObstructionCache(); // Sakura - optimise explosions
 
     protected Level(
         final WritableLevelData levelData,
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 344526b2ded83444eb266b6a2159195abb43cda8..d2de73e517bf7e40258a6eab62befa467bcf64e5 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -81,7 +81,7 @@ public class ServerExplosion implements Explosion {
             }
         }
 
-        CACHED_RAYS = rayCoords.toDoubleArray();
+        CACHED_RAYS = me.samsuik.sakura.explosion.SortExplosionRays.sortExplosionRays(rayCoords); // Sakura - optimise explosions
     }
 
     private static final int CHUNK_CACHE_SHIFT = 2;
@@ -140,7 +140,7 @@ public class ServerExplosion implements Explosion {
         return ret;
     }
 
-    protected boolean clipsAnything(final Vec3 from, final Vec3 to,
+    protected int clipsAnything(final Vec3 from, final Vec3 to, // Sakura - optimise explosions
                                   final ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.LazyEntityCollisionContext context,
                                   final ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[] blockCache,
                                   final BlockPos.MutableBlockPos currPos) {
@@ -150,9 +150,11 @@ public class ServerExplosion implements Explosion {
         final double adjZ = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON * (from.z - to.z);
 
         if (adjX == 0.0 && adjY == 0.0 && adjZ == 0.0) {
-            return false;
+            return MISSED_COLLISION; // Sakura - optimise explosions
         }
 
+        int result = 0; // Sakura - optimise explosions
+
         final double toXAdj = to.x - adjX;
         final double toYAdj = to.y - adjY;
         final double toZAdj = to.z - adjZ;
@@ -199,14 +201,14 @@ public class ServerExplosion implements Explosion {
                     (currZ & BLOCK_EXPLOSION_CACHE_MASK) << (BLOCK_EXPLOSION_CACHE_SHIFT + BLOCK_EXPLOSION_CACHE_SHIFT);
             ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cachedBlock = blockCache[cacheKey];
             if (cachedBlock == null || cachedBlock.key != key) {
-                blockCache[cacheKey] = cachedBlock = this.getOrCacheExplosionBlock(currX, currY, currZ, key, false);
+                blockCache[cacheKey] = cachedBlock = this.getOrCacheExplosionBlock(currX, currY, currZ, key, true); // Sakura - optimise explosions
             }
 
             final BlockState blockState = cachedBlock.blockState;
             if (blockState != null && !((ca.spottedleaf.moonrise.patches.collisions.block.CollisionBlockState)blockState).moonrise$emptyContextCollisionShape()) {
                 net.minecraft.world.phys.shapes.VoxelShape collision = cachedBlock.cachedCollisionShape;
                 if (collision == null) {
-                    collision = ((ca.spottedleaf.moonrise.patches.collisions.block.CollisionBlockState)blockState).moonrise$getConstantContextCollisionShape();
+                    collision = ((ca.spottedleaf.moonrise.patches.collisions.block.CollisionBlockState) blockState).moonrise$getConstantContextCollisionShape();
                     if (collision == null) {
                         collision = blockState.getCollisionShape(this.level, currPos, context);
                         if (!context.isDelegated()) {
@@ -219,13 +221,21 @@ public class ServerExplosion implements Explosion {
                     }
                 }
 
-                if (!collision.isEmpty() && collision.clip(from, to, currPos) != null) {
-                    return true;
+                // Sakura start - optimise explosions
+                if (!collision.isEmpty()) {
+                    if (collision.clip(from, to, currPos) != null) {
+                        return result | ANY_COLLISION;
+                    }
+
+                    if (!collision.moonrise$isFullBlock()) {
+                        result = MISSED_COLLISION;
+                    }
                 }
             }
 
             if (normalizedCurrX > 1.0 && normalizedCurrY > 1.0 && normalizedCurrZ > 1.0) {
-                return false;
+                return result;
+                // Sakura end - optimise explosions
             }
 
             // inc the smallest normalized coordinate
@@ -253,7 +263,8 @@ public class ServerExplosion implements Explosion {
 
     protected float getSeenFraction(final Vec3 source, final Entity target,
                                    final ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[] blockCache,
-                                   final BlockPos.MutableBlockPos blockPos) {
+                                   final BlockPos.MutableBlockPos blockPos,
+                                   final @org.jetbrains.annotations.Nullable me.samsuik.sakura.explosion.density.KnownObstruction obstruction) { // Sakura - optimise explosions
         final AABB boundingBox = target.getBoundingBox();
         final double diffX = boundingBox.maxX - boundingBox.minX;
         final double diffY = boundingBox.maxY - boundingBox.minY;
@@ -289,7 +300,11 @@ public class ServerExplosion implements Explosion {
                             Math.fma(dz, diffZ, offZ)
                     );
 
-                    if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
+                    // Sakura start - optimise explosions
+                    if (obstruction != null && obstruction.isEntityPositionKnown(from)) {
+                        missedRays += obstruction.getSeenPercent() == 1.0f ? 1 : 0;
+                    } else if ((this.clipsAnything(from, source, context, blockCache, blockPos) & ANY_COLLISION) == 0) {
+                        // Sakura end - optimise explosions
                         ++missedRays;
                     }
                 }
@@ -299,6 +314,40 @@ public class ServerExplosion implements Explosion {
         return (float)missedRays / (float)totalRays;
     }
     // Paper end - collisions optimisations
+    // Sakura start - optimise explosions
+    protected static final int ANY_COLLISION = 0b1;
+    protected static final int MISSED_COLLISION = 0b10;
+
+    private final List<me.samsuik.sakura.explosion.ExplosionToSend> explosionsToSend = new ObjectArrayList<>();
+
+    protected final void queueNotification(final Vec3 center, final int blockCount) {
+        this.explosionsToSend.add(new me.samsuik.sakura.explosion.ExplosionToSend(center, blockCount));
+    }
+
+    public final List<me.samsuik.sakura.explosion.ExplosionToSend> getToSend() {
+        return this.explosionsToSend;
+    }
+
+    protected final void initialiseBlockCache() {
+        this.blockCache = new it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<>();
+        this.chunkPosCache = new long[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
+        java.util.Arrays.fill(this.chunkPosCache, ChunkPos.INVALID_CHUNK_POS);
+        this.chunkCache = new net.minecraft.world.level.chunk.LevelChunk[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
+        this.directMappedBlockCache = new ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH];
+        this.mutablePos = new BlockPos.MutableBlockPos();
+    }
+
+    protected final void markBlocksInCacheAsExplodable(final List<BlockPos> explodedPositions) {
+        for (final BlockPos explodedPos : explodedPositions) {
+            final ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cache = this.blockCache.get(explodedPos.asLong());
+            // May be null if the blockCache is cleared then retrieved from the recent block cache
+            //noinspection ConstantValue
+            if (cache != null) {
+                cache.shouldExplode = null;
+            }
+        }
+    }
+    // Sakura end - optimise explosions
 
     public ServerExplosion(
         final ServerLevel level,
@@ -331,6 +380,16 @@ public class ServerExplosion implements Explosion {
     }
 
     public static float getSeenPercent(final Vec3 center, final Entity entity) {
+        // Sakura start - optimise explosions
+        return getSeenPercent(center, entity, null);
+    }
+
+    private static float getSeenPercent(
+        final Vec3 center,
+        final Entity entity,
+        final @org.jetbrains.annotations.Nullable me.samsuik.sakura.explosion.density.KnownObstruction obstruction
+    ) {
+        // Sakura end - optimise explosions
         AABB bb = entity.getBoundingBox();
         double xs = 1.0 / ((bb.maxX - bb.minX) * 2.0 + 1.0);
         double ys = 1.0 / ((bb.maxY - bb.minY) * 2.0 + 1.0);
@@ -348,8 +407,15 @@ public class ServerExplosion implements Explosion {
                         double y = Mth.lerp(yy, bb.minY, bb.maxY);
                         double z = Mth.lerp(zz, bb.minZ, bb.maxZ);
                         Vec3 from = new Vec3(x + xOffset, y, z + zOffset);
-                        if (entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType()
-                            == HitResult.Type.MISS) {
+                        // Sakura start - optimise explosions
+                        final net.minecraft.world.phys.HitResult.Type hitResult;
+                        if (obstruction != null && obstruction.isEntityPositionKnown(from)) {
+                            hitResult = obstruction.hitResult();
+                        } else {
+                            hitResult = entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                        }
+                        if (hitResult == HitResult.Type.MISS) {
+                            // Sakura end - optimise explosions
                             hits++;
                         }
 
@@ -394,6 +460,18 @@ public class ServerExplosion implements Explosion {
             initialCache = this.getOrCacheExplosionBlock(blockX, blockY, blockZ, key, true);
         }
 
+        // Sakura start - optimise explosions
+        if (!this.interactsWithBlocks() || initialCache.resistance > (this.radius * 1.3f)) {
+            return ret;
+        }
+
+        if (this.source != null
+            && this.level.sakuraConfig().cannons.explosion.optimiseProtectedRegions
+            && !new me.samsuik.sakura.explosion.ProtectionPluginProbeEvent(this.source.getBukkitEntity()).callEvent()) {
+            return ret;
+        }
+        // Sakura end - optimise explosions
+
         // only ~1/3rd of the loop iterations in vanilla will result in a ray, as it is iterating the perimeter of
         // a 16x16x16 cube
         // we can cache the rays and their normals as well, so that we eliminate the excess iterations / checks and
@@ -482,16 +560,31 @@ public class ServerExplosion implements Explosion {
             int z0 = Mth.floor(this.center.z - doubleRadius - 1.0);
             int z1 = Mth.floor(this.center.z + doubleRadius + 1.0);
             List<Entity> list = this.level.getEntities(this.excludeSourceFromDamage ? this.source : null, new AABB(x0, y0, z0, x1, y1, z1), entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
-
+            // Sakura start - optimise explosions
+            final AABB impactArea = new AABB(x0, y0, z0, x1, y1, z1);
             for (Entity entity : list) { // Paper - used in loop
+                this.sakura$impactEntity(entity, this.center, impactArea, doubleRadius);
+            }
+        }
+    }
+
+    protected final void sakura$impactEntity(
+        final Entity entity,
+        final Vec3 explosionPos,
+        final @Nullable AABB impactArea,
+        final double doubleRadius
+    ) {
+        {
+            if (entity.isAlive() && !entity.isSpectator()) {
+                // Sakura end - optimise explosions
                 if (!entity.ignoreExplosion(this)) {
-                    double dist = Math.sqrt(entity.distanceToSqr(this.center)) / doubleRadius;
+                    double dist = Math.sqrt(entity.distanceToSqr(explosionPos)) / doubleRadius; // Sakura - optimise explosions
                     if (!(dist > 1.0)) {
                         Vec3 entityOrigin = entity instanceof PrimedTnt ? entity.position() : entity.getEyePosition();
-                        Vec3 direction = entityOrigin.subtract(this.center).normalize();
+                        Vec3 direction = entityOrigin.subtract(explosionPos).normalize(); // Sakura - optimise explosions
                         boolean shouldDamageEntity = this.damageCalculator.shouldDamageEntity(this, entity);
                         float knockbackMultiplier = this.damageCalculator.getKnockbackMultiplier(entity);
-                        float exposure = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : this.getBlockDensity(this.center, entity); // Paper - Optimize explosions
+                        float exposure = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : this.getBlockDensity(explosionPos, entity); // Sakura - optimise explosions // Paper - Optimize explosions
                         if (shouldDamageEntity) {
                             // CraftBukkit start
 
@@ -502,7 +595,7 @@ public class ServerExplosion implements Explosion {
                             // - Damaging EnderDragon does nothing
                             // - EnderDragon hitbox always covers the other parts and is therefore always present
                             if (entity instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragonPart) {
-                                continue;
+                                return; // Sakura - optimise explosions
                             }
 
                             entity.lastDamageCancelled = false;
@@ -510,7 +603,7 @@ public class ServerExplosion implements Explosion {
                             if (entity instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon enderDragon) {
                                 for (net.minecraft.world.entity.boss.enderdragon.EnderDragonPart dragonPart : enderDragon.getSubEntities()) {
                                     // Calculate damage separately for each EntityComplexPart
-                                    if (list.contains(dragonPart)) {
+                                    if (impactArea != null && impactArea.intersects(dragonPart.getBoundingBox())) { // Sakura - optimise explosions
                                         dragonPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, dragonPart, exposure));
                                     }
                                 }
@@ -519,7 +612,7 @@ public class ServerExplosion implements Explosion {
                             }
 
                             if (entity.lastDamageCancelled) { // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Skip entity if damage event was cancelled
-                                continue;
+                                return; // Sakura - optimise explosions
                             }
                             // CraftBukkit end
                         }
@@ -541,7 +634,10 @@ public class ServerExplosion implements Explosion {
                         if (entity.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
                             projectile.setOwner(this.damageSource.getEntity());
                         } else if (entity instanceof Player player && !player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Option to disable explosion knockback
-                            this.hitPlayers.put(player, knockback);
+                            // Sakura start - optimise explosions; tally player velocity
+                            final Vec3 finalKnockback = knockback;
+                            this.hitPlayers.compute(player, (p, v) -> v != null ? v.add(finalKnockback) : finalKnockback);
+                            // Sakura end - optimise explosions; tally player velocity
                         }
 
                         entity.onExplosionHit(this.source);
@@ -557,7 +653,7 @@ public class ServerExplosion implements Explosion {
 
         // CraftBukkit start
         org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(this.center, this.level);
-        List<org.bukkit.block.Block> blockList = new ObjectArrayList<>();
+        List<org.bukkit.block.Block> blockList = new ObjectArrayList<>(fire ? ObjectArrayList.DEFAULT_INITIAL_CAPACITY : targetBlocks.size()); // Sakura - optimise explosions
         for (int i1 = targetBlocks.size() - 1; i1 >= 0; i1--) {
             org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, targetBlocks.get(i1));
             if (!bblock.getType().isAir()) {
@@ -661,6 +757,7 @@ public class ServerExplosion implements Explosion {
             this.createFire(toBlow);
         }
 
+        this.queueNotification(this.center, toBlow.size()); // Sakura - optimise explosions
         // Paper start - collision optimisations
         this.blockCache = null;
         this.chunkPosCache = null;
@@ -750,12 +847,12 @@ public class ServerExplosion implements Explosion {
     // Paper start - Optimize explosions
     protected float getBlockDensity(Vec3 vec3d, Entity entity) {
         if (!this.brokenPaperBehaviour) { // Sakura - broken paper explosion behaviour
-            return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+            return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos, null); // Sakura - optimise explosions // Paper - collision optimisations
         }
         CacheKey key = new CacheKey(this, entity.getBoundingBox());
         Float blockDensity = this.level.explosionDensityCache.get(key);
         if (blockDensity == null) {
-            blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+            blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos, null); // Sakura - optimise explosions // Paper - collision optimisations
             this.level.explosionDensityCache.put(key, blockDensity);
         }
 
