From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 19 Apr 2024 22:20:03 +0100
Subject: [PATCH] Optimise paper explosions


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 2dfd412344a0e57f25a08d9c65656a13113baa4e..a578b89a8653336a2e473589442385104b1e7164 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -319,6 +319,12 @@ public final class ChunkEntitySlices {
         return true;
     }
 
+    // Sakura start - optimise paper explosions; expose slice entity list
+    public Entity[] getSectionEntities(final int sectionY) {
+        return this.allEntities.getSectionEntities(sectionY);
+    }
+    // Sakura end - optimise paper explosions; expose slice entity list
+
     public void getHardCollidingEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate) {
         this.hardCollidingEntities.getEntities(except, box, into, predicate);
     }
@@ -517,6 +523,17 @@ public final class ChunkEntitySlices {
             }
         }
 
+        // Sakura start - optimise paper explosions; expose slice entity list
+        public Entity[] getSectionEntities(final int sectionY) {
+            final BasicEntityList<Entity> entityList = this.entitiesBySection[sectionY - this.slices.minSection];
+            if (entityList != null) {
+                return entityList.storage;
+            } else {
+                return new Entity[0];
+            }
+        }
+        // Sakura end - optimise paper explosions; expose slice entity list
+
         public void getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate) {
             if (this.count == 0) {
                 return;
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index e016e21522bcae271c9de72c8cec1b3affc85b45..b3838c018f69d8c9955680a23dc03f01000722f9 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -88,7 +88,7 @@ public class ServerExplosion implements Explosion {
             }
         }
 
-        CACHED_RAYS = rayCoords.toDoubleArray();
+        CACHED_RAYS = me.samsuik.sakura.explosion.SortExplosionRays.sortExplosionRays(rayCoords); // Sakura - optimise paper explosions
     }
 
     private static final int CHUNK_CACHE_SHIFT = 2;
@@ -400,6 +400,12 @@ public class ServerExplosion implements Explosion {
             initialCache = this.getOrCacheExplosionBlock(blockX, blockY, blockZ, key, true);
         }
 
+        // Sakura start - optimise paper explosions
+        if (!this.interactsWithBlocks() || initialCache.resistance > (this.radius * 1.3f)) {
+            return ret;
+        }
+        // Sakura end - optimise paper explosions
+
         // only ~1/3rd of the loop iterations in vanilla will result in a ray, as it is iterating the perimeter of
         // a 16x16x16 cube
         // we can cache the rays and their normals as well, so that we eliminate the excess iterations / checks and
@@ -562,7 +568,7 @@ public class ServerExplosion implements Explosion {
 
         // CraftBukkit start
         Location location = CraftLocation.toBukkit(this.center, this.level);
-        List<org.bukkit.block.Block> blockList = new ObjectArrayList<>();
+        List<org.bukkit.block.Block> blockList = new ObjectArrayList<>(fire ? ObjectArrayList.DEFAULT_INITIAL_CAPACITY : blocks.size()); // Sakura - optimise paper explosions
         for (int i1 = blocks.size() - 1; i1 >= 0; i1--) {
             org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, blocks.get(i1));
             if (!bblock.getType().isAir()) {
