From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 16 Aug 2023 22:34:49 +0100
Subject: [PATCH] Store Entity Data/State


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8399470539d95a3794088d1ff537dd8cdf73591f..a7686db66153a86c7cba67fbc04a90ec558c8e70 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -552,6 +552,21 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return flags;
     }
     // Sakura end - load chunks on movement
+    // Sakura start - store entity data/state
+    private me.samsuik.sakura.entity.EntityState entityState = null;
+
+    public final void storeEntityState() {
+        this.entityState = me.samsuik.sakura.entity.EntityState.of(this);
+    }
+
+    public final me.samsuik.sakura.entity.EntityState entityState() {
+        return this.entityState;
+    }
+
+    public final boolean compareState(final Entity to) {
+        return to.entityState() != null && to.entityState().comparePositionAndMotion(this);
+    }
+    // Sakura end - store entity data/state
 
     public Entity(EntityType<?> type, Level level) {
         this.type = type;
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 7c747732f229da6c24eaa0c19d9283effda8cd9b..443605c2b64f390d56af19e85076b5a4eb92103f 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1444,6 +1444,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     public <T extends Entity> void guardEntityTick(Consumer<T> action, T entity) {
         try {
+            entity.storeEntityState(); // Sakura - store entity data/state
             action.accept(entity);
         } catch (Throwable var6) {
             // Paper start - Prevent block entity and entity crashes
