From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 9 Sep 2023 18:39:15 +0100
Subject: [PATCH] Merge Cannon Entities


diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index f7e8320006171fadde78afaa4c89373292ab13de..d42b73fa20025d7a05171e301a115424dc1bbb9f 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -9,5 +9,6 @@ import org.jspecify.annotations.NullMarked;
 public final class ScheduleSakuraTasks {
     public static void schedule(final LevelTickScheduler scheduler) {
         scheduler.endless((tick, level, cb) -> level.explosionPositions.clear(), 0); // Sakura - client visibility settings
+        scheduler.endless((tick, level, cb) -> level.mergeHandler.expire(tick), 200); // Sakura - merge cannon entities
     }
 }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 7a1b62df9f2c60898546ee44e182151e62b28663..ace3db39a29bbc534553a95a702b763de5d69ab1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -883,6 +883,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+            final Entity[] previousEntity = new Entity[1]; // Sakura - merge cannon entities
             this.entityTickList
                 .forEach(
                     entity -> {
@@ -902,6 +903,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                     }
 
                                     profiler.push("tick");
+                                    // Sakura start - merge cannon entities
+                                    final Entity previous = previousEntity[0];
+                                    if (this.mergeHandler.tryMerge(entity, previous)) {
+                                        return;
+                                    } else {
+                                        previousEntity[0] = entity;
+                                    }
+                                    // Sakura end - merge cannon entities
                                     this.guardEntityTick(this::tickNonPassenger, entity);
                                     profiler.pop();
                                 }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index badf603e78d28b053ef34e962f5a09b298d9cd35..841b488a83391557f5a969889bd7a43db16e9e5f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -581,6 +581,27 @@ public abstract class Entity
         return to.entityState() != null && to.entityState().comparePositionAndMotion(this);
     }
     // Sakura end - store entity data/state
+    // Sakura start - merge cannon entities
+    public final void updateBukkitHandle(final Entity entity) {
+        if (this.bukkitEntity != null) {
+            this.bukkitEntity.setHandle(entity);
+        } else {
+            this.bukkitEntity = entity.getBukkitEntity();
+        }
+    }
+
+    public final long getPackedOriginPosition() {
+        if (this.origin != null) {
+            return BlockPos.asLong(
+                Mth.floor(this.origin.x()),
+                Mth.floor(this.origin.y()),
+                Mth.floor(this.origin.z())
+            );
+        } else {
+            return Long.MIN_VALUE;
+        }
+    }
+    // Sakura end - merge cannon entities
 
     public Entity(final EntityType<?> type, final Level level) {
         this.type = type;
@@ -5162,6 +5183,11 @@ public abstract class Entity
         if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
         this.levelCallback.onRemove(reason);
         this.onRemoval(reason);
+        // Sakura start - merge cannon entities
+        if (removalReason == RemovalReason.DISCARDED) {
+            this.level.mergeHandler.removeEntity(this);
+        }
+        // Sakura end - merge cannon entities
         // Paper start - Folia schedulers
         if (!(this instanceof ServerPlayer) && reason != RemovalReason.CHANGED_DIMENSION && !alreadyRemoved) {
             // Players need to be special cased, because they are regularly removed from the world
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 9100673b30c6e34b27e657c145461f829c953e4b..8cad3eaece77dc19f9b4cd1855470a8b6ba73703 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -53,7 +53,7 @@ import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class FallingBlockEntity extends Entity {
+public class FallingBlockEntity extends Entity implements me.samsuik.sakura.entity.merge.MergeableEntity { // Sakura - merge cannon entities
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final BlockState DEFAULT_BLOCK_STATE = Blocks.SAND.defaultBlockState();
     private static final int DEFAULT_TIME = 0;
@@ -74,12 +74,63 @@ public class FallingBlockEntity extends Entity {
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
     public boolean heightParity; // Sakura - falling block height parity api
 
+    // Sakura start - merge cannon entities
+    private final me.samsuik.sakura.entity.merge.MergeEntityData mergeData = new me.samsuik.sakura.entity.merge.MergeEntityData(this);
+
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeEntityData getMergeEntityData() {
+        return this.mergeData;
+    }
+
+    @Override
+    public final boolean isSafeToMergeInto(final me.samsuik.sakura.entity.merge.MergeableEntity entity, final boolean ticksLived) {
+        return entity instanceof FallingBlockEntity fbe
+            && fbe.blockState.equals(this.blockState)
+            && (!ticksLived || fbe.time - 1 == this.time);
+    }
+
+    @Override
+    public final void respawnEntity(int count) {
+        while (count-- >= 1) {
+            // Unlike PrimedTnt we have to try respawn each stacked entity
+            final FallingBlockEntity fallingBlock = new FallingBlockEntity(EntityType.FALLING_BLOCK, this.level());
+
+            // Try to stack the falling block
+            this.entityState().apply(fallingBlock);
+            fallingBlock.blockState = this.blockState;
+            fallingBlock.spawnReason = this.spawnReason;
+            fallingBlock.time = this.time - 1;
+            fallingBlock.tick();
+
+            // If you horizontal stack into a moving piston block this condition will be met.
+            if (!fallingBlock.isRemoved()) {
+                this.mergeData.count = count + 1;
+                fallingBlock.storeEntityState();
+                fallingBlock.entityState().apply(this);
+                break;
+            } else if (count == 0) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            }
+        }
+    }
+
+    @Override
+    public @Nullable ItemEntity spawnAtLocation(final ServerLevel level, final net.minecraft.world.level.ItemLike item) { // may be overridden by plugins
+        ItemEntity itemEntity = null;
+        for (int i = 0; i < this.mergeData.count; ++i) {
+            itemEntity = super.spawnAtLocation(level, item);
+        }
+        return itemEntity;
+    }
+    // Sakura end - merge cannon entities
+
     public FallingBlockEntity(final EntityType<? extends FallingBlockEntity> type, final Level level) {
         super(type, level);
         this.dropItem = level.sakuraConfig().cannons.sand.dropItems; // Sakura - configure falling blocks dropping items
         this.heightParity = level.sakuraConfig().cannons.mechanics.fallingBlockParity; // Sakura - configure cannon mechanics
         this.isFallingBlock = true; // Sakura - client visibility settings
         this.loadChunks = level.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
+        this.mergeData.mergeLevel = level.sakuraConfig().cannons.mergeLevel; // Sakura - merge cannon entities
     }
 
     public FallingBlockEntity(final Level level, final double x, final double y, final double z, final BlockState blockState) {
@@ -249,6 +300,7 @@ public class FallingBlockEntity extends Entity {
                                     return;
                                 }
                                 // CraftBukkit end
+                                if (this.tryToRespawnEntity()) return; // Sakura - merge cannon entities
                                 if (this.level().setBlock(pos, this.blockState, Block.UPDATE_ALL)) {
                                     serverLevel.getChunkSource()
                                         .chunkMap
@@ -355,6 +407,11 @@ public class FallingBlockEntity extends Entity {
 
         output.putBoolean("CancelDrop", this.cancelDrop);
         if (!this.autoExpire) output.putBoolean("Paper.AutoExpire", false); // Paper - Expand FallingBlock API
+        // Sakura start - merge cannon entities; save to nbt
+        if (this.mergeData.count != 1) {
+            output.putInt("merge_count", this.mergeData.count);
+        }
+        // Sakura end - merge cannon entities; save to nbt
     }
 
     @Override
@@ -369,6 +426,7 @@ public class FallingBlockEntity extends Entity {
         this.blockData = input.read("TileEntityData", CompoundTag.CODEC).map(blockData -> this.level().paperConfig().entities.spawning.filterBadTileEntityNbtFromFallingBlocks && this.blockState.getBlock() instanceof net.minecraft.world.level.block.GameMasterBlock ? null : blockData).map(CompoundTag::copy).orElse(null); // Paper - Filter bad block entity nbt data from falling blocks
         this.cancelDrop = input.getBooleanOr("CancelDrop", false);
         this.autoExpire = input.getBooleanOr("Paper.AutoExpire", true); // Paper - Expand FallingBlock API
+        this.mergeData.count = input.getIntOr("merge_count", 1); // Sakura - merge cannon entities; load from nbt
     }
 
     public void setHurtsEntities(final float damagePerDistance, final int damageMax) {
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 4b8895393e81349db56b7d3390af1d3753b11ca5..4a9df1e64388585e20cf2b424ee40866f47fddf1 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -34,7 +34,7 @@ import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
 // CraftBukkit end
 
-public class PrimedTnt extends Entity implements TraceableEntity {
+public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sakura.entity.merge.MergeableEntity { // Sakura - merge cannon entities
     private static final EntityDataAccessor<Integer> DATA_FUSE_ID = SynchedEntityData.defineId(PrimedTnt.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<BlockState> DATA_BLOCK_STATE_ID = SynchedEntityData.defineId(PrimedTnt.class, EntityDataSerializers.BLOCK_STATE);
     private static final short DEFAULT_FUSE_TIME = 80;
@@ -61,11 +61,48 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     public float explosionPower = 4.0F;
     public boolean isIncendiary = false; // CraftBukkit
 
+    // Sakura start - merge cannon entities
+    private final me.samsuik.sakura.entity.merge.MergeEntityData mergeData = new me.samsuik.sakura.entity.merge.MergeEntityData(this);
+
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeEntityData getMergeEntityData() {
+        return this.mergeData;
+    }
+
+    @Override
+    public final boolean isSafeToMergeInto(final me.samsuik.sakura.entity.merge.MergeableEntity entity, final boolean ticksLived) {
+        return entity instanceof PrimedTnt tnt
+            && tnt.getFuse() + 1 == this.getFuse()
+            // required to prevent issues with powdered snow
+            && (tnt.entityState().fallDistance() == this.fallDistance
+            || tnt.entityState().fallDistance() > 2.5f && this.fallDistance > 2.5f);
+    }
+
+    @Override
+    public final void respawnEntity(int count) {
+        final PrimedTnt tnt = new PrimedTnt(EntityType.TNT, this.level());
+        tnt.updateBukkitHandle(this); // update handle for plugins
+        while (count-- > 1) {
+            this.setFuse(100); // Prevent unwanted explosions while ticking
+
+            // Cause an explosion to affect this entity
+            tnt.setPos(this.position());
+            tnt.setDeltaMovement(this.getDeltaMovement());
+            this.entityState().apply(this);
+            tnt.explode();
+            this.storeEntityState();
+
+            this.tick();
+        }
+    }
+    // Sakura end - merge cannon entities
+
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
         this.blocksBuilding = true;
         this.isPrimedTNT = true; // Sakura - client visibility settings
         this.loadChunks = level.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
+        this.mergeData.mergeLevel = level.sakuraConfig().cannons.mergeLevel; // Sakura - merge cannon entities
     }
 
     public PrimedTnt(final Level level, final double x, final double y, final double z, final @Nullable LivingEntity owner) {
@@ -145,6 +182,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         if (fuse <= 0) {
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
             //this.discard();
+            this.tryToRespawnEntity(); // Sakura - merge cannon entities
             if (!this.level().isClientSide()) {
                 this.explode();
             }
@@ -204,6 +242,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         }
 
         EntityReference.store(this.owner, output, "owner");
+        // Sakura start - merge cannon entities; save to nbt
+        if (this.mergeData.count != 1) {
+            output.putInt("merge_count", this.mergeData.count);
+        }
+        // Sakura end - merge cannon entities; save to nbt
     }
 
     @Override
@@ -212,6 +255,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         this.setBlockState(input.read("block_state", BlockState.CODEC).orElse(DEFAULT_BLOCK_STATE));
         this.explosionPower = Mth.clamp(input.getFloatOr("explosion_power", 4.0F), 0.0F, 128.0F);
         this.owner = EntityReference.read(input, "owner");
+        this.mergeData.count = input.getIntOr("merge_count", 1); // Sakura - merge cannon entities; load from nbt
     }
 
     @Override
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index f7b688d89864ac6ee3b93f0baf595f14ec4bb180..12d1879590b97821f30e7a8a463b0bcd1e5024c2 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -843,6 +843,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return this.blockStateRevisionCounter.get();
     }
     // Sakura end - implement state watchers and tick scheduler
+    public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
 
     protected Level(
         final WritableLevelData levelData,
diff --git a/net/minecraft/world/level/block/BasePressurePlateBlock.java b/net/minecraft/world/level/block/BasePressurePlateBlock.java
index fef90ff995a211f117a28a388867ddc3047b6817..19d6bea66edcbeadc2cf4b015acb40862177f54b 100644
--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -100,7 +100,7 @@ public abstract class BasePressurePlateBlock extends Block {
     }
 
     private void checkPressed(final @Nullable Entity sourceEntity, final Level level, final BlockPos pos, final BlockState state, final int oldSignal) {
-        int signal = this.getSignalStrength(level, pos);
+        int signal = this.getSignalStrength(level, pos, oldSignal == 0); // Sakura - merge cannon entities
         boolean wasPressed = oldSignal > 0;
         boolean isPressed = signal > 0;
 
@@ -170,6 +170,12 @@ public abstract class BasePressurePlateBlock extends Block {
         // CraftBukkit end
     }
 
+    // Sakura start - merge cannon entities
+    protected int getSignalStrength(final Level world, final BlockPos pos, final boolean entityInside) {
+        return this.getSignalStrength(world, pos);
+    }
+    // Sakura end - merge cannon entities
+
     protected abstract int getSignalStrength(Level level, BlockPos pos);
 
     protected abstract int getSignalForState(BlockState state);
diff --git a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
index e26385eb25338e7f0770b9154a542273b651c2f4..ff9730987a1ede9aced4d5bb9e84deae3082b26d 100644
--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -39,6 +39,13 @@ public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
 
     @Override
     protected int getSignalStrength(final Level level, final BlockPos pos) {
+        // Sakura start - merge cannon entities
+        return this.getSignalStrength(level, pos, false);
+    }
+
+    @Override
+    protected final int getSignalStrength(final Level level, final BlockPos pos, final boolean entityInside) {
+        // Sakura end - merge cannon entities
         // CraftBukkit start
         // int min = Math.min(getEntityCount(level, TOUCH_AABB.move(pos), Entity.class), this.maxWeight);
         int count = 0;
@@ -54,7 +61,7 @@ public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
 
             // We only want to block turning the plate on if all events are cancelled
             if (!cancellable.isCancelled()) {
-                count++;
+                count += !entityInside && entity instanceof me.samsuik.sakura.entity.merge.MergeableEntity mergeEntity ? mergeEntity.getMergeEntityData().count : 1; // Sakura - merge cannon entities
             }
         }
         // CraftBukkit end
