From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 9 Sep 2023 18:39:15 +0100
Subject: [PATCH] Merge Cannon Entities


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 9ff91f08b8d32f90d799d7f0e63e4c216f56df51..ad61cba4160d634b5b3648525a63919dcd49e3d4 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -726,6 +726,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
         this.levelTickScheduler.repeatingTask(this.explosionPositions::clear, 0); // Sakura - client visibility settings
+        this.levelTickScheduler.repeatingTask(this.mergeHandler::expire, 200); // Sakura - merge cannon entities
     }
 
     // Paper start
@@ -849,6 +850,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+            final Entity[] previousEntity = new Entity[1]; // Sakura - merge cannon entities
             this.entityTickList
                 .forEach(
                     entity -> {
@@ -867,6 +869,15 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                                         entity.stopRiding();
                                     }
 
+                                    // Sakura start - merge cannon entities
+                                    final Entity previous = previousEntity[0];
+                                    if (this.mergeHandler.tryMerge(entity, previous)) {
+                                        return;
+                                    } else {
+                                        previousEntity[0] = entity;
+                                    }
+                                    // Sakura end - merge cannon entities
+
                                     profilerFiller.push("tick");
                                     this.guardEntityTick(this::tickNonPassenger, entity);
                                     profilerFiller.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a7686db66153a86c7cba67fbc04a90ec558c8e70..564588076a33e905f0b2452048fb2a31367ee0f7 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -567,6 +567,27 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return to.entityState() != null && to.entityState().comparePositionAndMotion(this);
     }
     // Sakura end - store entity data/state
+    // Sakura start - merge cannon entities
+    public final void updateBukkitHandle(final Entity entity) {
+        if (this.bukkitEntity != null) {
+            this.bukkitEntity.setHandle(entity);
+        } else {
+            this.bukkitEntity = entity.getBukkitEntity();
+        }
+    }
+
+    public final long getPackedOriginPosition() {
+        if (this.origin != null) {
+            return BlockPos.asLong(
+                Mth.floor(this.origin.x()),
+                Mth.floor(this.origin.y()),
+                Mth.floor(this.origin.z())
+            );
+        } else {
+            return Long.MIN_VALUE;
+        }
+    }
+    // Sakura end - merge cannon entities
 
     public Entity(EntityType<?> type, Level level) {
         this.type = type;
@@ -5230,6 +5251,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
         this.levelCallback.onRemove(removalReason);
         this.onRemoval(removalReason);
+        // Sakura start - merge cannon entities
+        if (removalReason == RemovalReason.DISCARDED) {
+            this.level.mergeHandler.removeEntity(this);
+        }
+        // Sakura end - merge cannon entities
         // Paper start - Folia schedulers
         if (!(this instanceof ServerPlayer) && removalReason != RemovalReason.CHANGED_DIMENSION && !alreadyRemoved) {
             // Players need to be special cased, because they are regularly removed from the world
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index a14e431948264c4b99be49669322f530119c3d87..afd6e062422943e053af594613beb09b40c8a00b 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -53,7 +53,7 @@ import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class FallingBlockEntity extends Entity {
+public class FallingBlockEntity extends Entity implements me.samsuik.sakura.entity.merge.MergeableEntity { // Sakura - merge cannon entities
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final BlockState DEFAULT_BLOCK_STATE = Blocks.SAND.defaultBlockState();
     private static final int DEFAULT_TIME = 0;
@@ -74,12 +74,63 @@ public class FallingBlockEntity extends Entity {
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
     public boolean heightParity; // Sakura - falling block height parity api
 
+    // Sakura start - merge cannon entities
+    private final me.samsuik.sakura.entity.merge.MergeEntityData mergeData = new me.samsuik.sakura.entity.merge.MergeEntityData(this);
+
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeEntityData getMergeEntityData() {
+        return this.mergeData;
+    }
+
+    @Override
+    public final boolean isSafeToMergeInto(final me.samsuik.sakura.entity.merge.MergeableEntity entity, final boolean ticksLived) {
+        return entity instanceof FallingBlockEntity fbe
+            && fbe.blockState.equals(this.blockState)
+            && (!ticksLived || fbe.time - 1 == this.time);
+    }
+
+    @Override
+    public final void respawnEntity(int count) {
+        while (count-- >= 1) {
+            // Unlike PrimedTnt we have to try respawn each stacked entity
+            final FallingBlockEntity fallingBlock = new FallingBlockEntity(EntityType.FALLING_BLOCK, this.level());
+
+            // Try to stack the falling block
+            this.entityState().apply(fallingBlock);
+            fallingBlock.blockState = this.blockState;
+            fallingBlock.spawnReason = this.spawnReason;
+            fallingBlock.time = this.time - 1;
+            fallingBlock.tick();
+
+            // If you horizontal stack into a moving piston block this condition will be met.
+            if (!fallingBlock.isRemoved()) {
+                this.mergeData.count = count + 1;
+                fallingBlock.storeEntityState();
+                fallingBlock.entityState().apply(this);
+                break;
+            } else if (count == 0) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            }
+        }
+    }
+
+    @Override
+    public @Nullable ItemEntity spawnAtLocation(final ServerLevel level, final net.minecraft.world.level.ItemLike item) { // may be overridden by plugins
+        ItemEntity itemEntity = null;
+        for (int i = 0; i < this.mergeData.count; ++i) {
+            itemEntity = super.spawnAtLocation(level, item);
+        }
+        return itemEntity;
+    }
+    // Sakura end - merge cannon entities
+
     public FallingBlockEntity(EntityType<? extends FallingBlockEntity> type, Level level) {
         super(type, level);
         this.dropItem = level.sakuraConfig().cannons.sand.dropItems; // Sakura - configure falling blocks dropping items
         this.heightParity = level.sakuraConfig().cannons.mechanics.fallingBlockParity; // Sakura - configure cannon mechanics
         this.isFallingBlock = true; // Sakura - client visibility settings
         this.loadChunks = level.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
+        this.mergeData.mergeLevel = level.sakuraConfig().cannons.mergeLevel; // Sakura - merge cannon entities
     }
 
     public FallingBlockEntity(Level level, double x, double y, double z, BlockState state) {
@@ -238,6 +289,7 @@ public class FallingBlockEntity extends Entity {
                                     return;
                                 }
                                 // CraftBukkit end
+                                if (this.tryToRespawnEntity()) return; // Sakura - merge cannon entities HEAD
                                 if (this.level().setBlock(blockPos, this.blockState, Block.UPDATE_ALL)) {
                                     serverLevel.getChunkSource()
                                         .chunkMap
@@ -344,6 +396,11 @@ public class FallingBlockEntity extends Entity {
 
         output.putBoolean("CancelDrop", this.cancelDrop);
         if (!this.autoExpire) output.putBoolean("Paper.AutoExpire", false); // Paper - Expand FallingBlock API
+        // Sakura start - merge cannon entities; save to nbt
+        if (this.mergeData.count != 1) {
+            output.putInt("merge_count", this.mergeData.count);
+        }
+        // Sakura end - merge cannon entities; save to nbt
     }
 
     @Override
@@ -358,6 +415,7 @@ public class FallingBlockEntity extends Entity {
         this.blockData = input.read("TileEntityData", CompoundTag.CODEC).map(blockData -> this.level().paperConfig().entities.spawning.filterBadTileEntityNbtFromFallingBlocks && this.blockState.getBlock() instanceof net.minecraft.world.level.block.GameMasterBlock ? null : blockData).map(CompoundTag::copy).orElse(null); // Paper - Filter bad block entity nbt data from falling blocks
         this.cancelDrop = input.getBooleanOr("CancelDrop", false);
         this.autoExpire = input.getBooleanOr("Paper.AutoExpire", true); // Paper - Expand FallingBlock API
+        this.mergeData.count = input.getIntOr("merge_count", 1); // Sakura - merge cannon entities; load from nbt
     }
 
     public void setHurtsEntities(float fallDamagePerDistance, int fallDamageMax) {
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index f7b29e5343182ad26c0b9c15a023f5947ef08cdc..35e378db13ecc5c108961b1535866d00e03bc4a6 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -34,7 +34,7 @@ import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
 // CraftBukkit end
 
-public class PrimedTnt extends Entity implements TraceableEntity {
+public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sakura.entity.merge.MergeableEntity { // Sakura - merge cannon entities
     private static final EntityDataAccessor<Integer> DATA_FUSE_ID = SynchedEntityData.defineId(PrimedTnt.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<BlockState> DATA_BLOCK_STATE_ID = SynchedEntityData.defineId(PrimedTnt.class, EntityDataSerializers.BLOCK_STATE);
     private static final short DEFAULT_FUSE_TIME = 80;
@@ -59,11 +59,48 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     public float explosionPower = 4.0F;
     public boolean isIncendiary = false; // CraftBukkit
 
+    // Sakura start - merge cannon entities
+    private final me.samsuik.sakura.entity.merge.MergeEntityData mergeData = new me.samsuik.sakura.entity.merge.MergeEntityData(this);
+
+    @Override
+    public final me.samsuik.sakura.entity.merge.MergeEntityData getMergeEntityData() {
+        return this.mergeData;
+    }
+
+    @Override
+    public final boolean isSafeToMergeInto(final me.samsuik.sakura.entity.merge.MergeableEntity entity, final boolean ticksLived) {
+        return entity instanceof PrimedTnt tnt
+            && tnt.getFuse() + 1 == this.getFuse()
+            // required to prevent issues with powdered snow
+            && (tnt.entityState().fallDistance() == this.fallDistance
+            || tnt.entityState().fallDistance() > 2.5f && this.fallDistance > 2.5f);
+    }
+
+    @Override
+    public final void respawnEntity(int count) {
+        final PrimedTnt tnt = new PrimedTnt(EntityType.TNT, this.level());
+        tnt.updateBukkitHandle(this); // update handle for plugins
+        while (count-- > 1) {
+            this.setFuse(100); // Prevent unwanted explosions while ticking
+
+            // Cause an explosion to affect this entity
+            tnt.setPos(this.position());
+            tnt.setDeltaMovement(this.getDeltaMovement());
+            this.entityState().apply(this);
+            tnt.explode();
+            this.storeEntityState();
+
+            this.tick();
+        }
+    }
+    // Sakura end - merge cannon entities
+
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level level) {
         super(type, level);
         this.blocksBuilding = true;
         this.isPrimedTNT = true; // Sakura - client visibility settings
         this.loadChunks = level.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
+        this.mergeData.mergeLevel = level.sakuraConfig().cannons.mergeLevel; // Sakura - merge cannon entities
     }
 
     public PrimedTnt(Level level, double x, double y, double z, @Nullable LivingEntity owner) {
@@ -143,6 +180,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         if (i <= 0) {
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
             //this.discard();
+            this.tryToRespawnEntity(); // Sakura - merge cannon entities
             if (!this.level().isClientSide()) {
                 this.explode();
             }
@@ -202,6 +240,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         }
 
         EntityReference.store(this.owner, output, "owner");
+        // Sakura start - merge cannon entities; save to nbt
+        if (this.mergeData.count != 1) {
+            output.putInt("merge_count", this.mergeData.count);
+        }
+        // Sakura end - merge cannon entities; save to nbt
     }
 
     @Override
@@ -210,6 +253,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         this.setBlockState(input.read("block_state", BlockState.CODEC).orElse(DEFAULT_BLOCK_STATE));
         this.explosionPower = Mth.clamp(input.getFloatOr("explosion_power", 4.0F), 0.0F, 128.0F);
         this.owner = EntityReference.read(input, "owner");
+        this.mergeData.count = input.getIntOr("merge_count", 1); // Sakura - merge cannon entities; load from nbt
     }
 
     @Override
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 443605c2b64f390d56af19e85076b5a4eb92103f..cada2b6c89a5c78d29c5ce8b9a34cadfefb2c34a 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -834,6 +834,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.scheduler.LevelTickScheduler levelTickScheduler = new me.samsuik.sakura.scheduler.LevelTickScheduler(this);
     public final me.samsuik.sakura.block_change.BlockStateChangeTracker blockStateChangeTracker = new me.samsuik.sakura.block_change.BlockStateChangeTracker(this);
     // Sakura end - track block changes and tick scheduler
+    public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
 
     protected Level(
         WritableLevelData levelData,
diff --git a/net/minecraft/world/level/block/BasePressurePlateBlock.java b/net/minecraft/world/level/block/BasePressurePlateBlock.java
index 7135a6bc87f90fafa94d0783aaf7b42d66449010..77f30569214d97d5de1ded3cf20491d70a5be405 100644
--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -92,7 +92,7 @@ public abstract class BasePressurePlateBlock extends Block {
     }
 
     private void checkPressed(@Nullable Entity entity, Level level, BlockPos pos, BlockState state, int currentSignal) {
-        int signalStrength = this.getSignalStrength(level, pos);
+        int signalStrength = this.getSignalStrength(level, pos, currentSignal == 0); // Sakura - merge cannon entities
         boolean flag = currentSignal > 0;
         boolean flag1 = signalStrength > 0;
 
@@ -162,6 +162,12 @@ public abstract class BasePressurePlateBlock extends Block {
         // CraftBukkit end
     }
 
+    // Sakura start - merge cannon entities
+    protected int getSignalStrength(final Level world, final BlockPos pos, final boolean entityInside) {
+        return this.getSignalStrength(world, pos);
+    }
+    // Sakura end - merge cannon entities
+
     protected abstract int getSignalStrength(Level level, BlockPos pos);
 
     protected abstract int getSignalForState(BlockState state);
diff --git a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
index 5e095919828e89d12f2676b3c544842a81e047a1..e45ceec307a4944f40d044af0369d0c598a4e9ae 100644
--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -39,6 +39,13 @@ public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
 
     @Override
     protected int getSignalStrength(Level level, BlockPos pos) {
+        // Sakura start - merge cannon entities
+        return this.getSignalStrength(level, pos, false);
+    }
+
+    @Override
+    protected final int getSignalStrength(final Level level, final BlockPos pos, final boolean entityInside) {
+        // Sakura end - merge cannon entities
         // CraftBukkit start
         // int min = Math.min(getEntityCount(level, TOUCH_AABB.move(pos), Entity.class), this.maxWeight);
         int min = 0;
@@ -54,7 +61,7 @@ public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
 
             // We only want to block turning the plate on if all events are cancelled
             if (!cancellable.isCancelled()) {
-                min++;
+                min += !entityInside && entity instanceof me.samsuik.sakura.entity.merge.MergeableEntity mergeEntity ? mergeEntity.getMergeEntityData().count : 1; // Sakura - merge cannon entities
             }
         }
         // CraftBukkit end
