From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Mon, 22 Apr 2024 23:01:26 +0100
Subject: [PATCH] Replace explosion density cache


diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index db77eb6e71360ee8cb6837a1d3c57181f7a653b5..ebe8275d28ddc42de0d94461b7fe62f115e098a3 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -9,5 +9,6 @@ import org.jspecify.annotations.NullMarked;
 public final class ScheduleSakuraTasks {
     public static void schedule(final LevelTickScheduler scheduler) {
         scheduler.endless((tick, level, cb) -> level.mergeHandler.expire(tick), 200); // Sakura - merge cannon entities
+        scheduler.endless((tick, level, cb) -> level.densityCache.expire(tick), 0); // Sakura - explosion density cache
     }
 }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 30c62cdd2ac4ed8b59473004a3116d231c43ecce..ae7ecb98f996bef1ead72b66e9cf0019cd05d0d3 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -845,6 +845,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     // Sakura end - implement state watchers and tick scheduler
     public final me.samsuik.sakura.player.visibility.MinimalVisibility minimalVisibility = new me.samsuik.sakura.player.visibility.MinimalVisibility(); // Sakura - client visibility settings
     public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
+    public final me.samsuik.sakura.explosion.density.BlockDensityCache densityCache = new me.samsuik.sakura.explosion.density.BlockDensityCache(this); // Sakura - optimise explosion density cache
 
     protected Level(
         final WritableLevelData levelData,
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index a3d98f9f7d7a3762a3e99d313982ea18d838f9d3..13e4e5089c74be3e253c5e338ae7e56b36e01b5e 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -288,7 +288,12 @@ public class ServerExplosion implements Explosion {
                             Math.fma(dz, diffZ, offZ)
                     );
 
-                    if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
+                    // Sakura start - replace density cache
+                    final float density = this.level.densityCache.getKnownDensity(from);
+                    if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
+                        missedRays += (int) density;
+                    } else if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
+                        // Sakura end - replace density cache
                         ++missedRays;
                     }
                 }
@@ -346,8 +351,17 @@ public class ServerExplosion implements Explosion {
                         double y = Mth.lerp(yy, bb.minY, bb.maxY);
                         double z = Mth.lerp(zz, bb.minZ, bb.maxZ);
                         Vec3 from = new Vec3(x + xOffset, y, z + zOffset);
-                        if (entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType()
-                            == HitResult.Type.MISS) {
+
+                        // Sakura start - replace density cache
+                        final net.minecraft.world.phys.HitResult.Type hitResult;
+                        final float density = entity.level().densityCache.getKnownDensity(center);
+                        if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
+                            hitResult = density != 0.0f ? net.minecraft.world.phys.HitResult.Type.MISS : net.minecraft.world.phys.HitResult.Type.BLOCK;
+                        } else {
+                            hitResult = entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                        }
+                        if (hitResult == HitResult.Type.MISS) {
+                            // Sakura end - replace density cache
                             hits++;
                         }
 
@@ -594,6 +608,11 @@ public class ServerExplosion implements Explosion {
             return;
         }
         // CraftBukkit end
+        // Sakura start - explosion density cache
+        if (!targetBlocks.isEmpty() && !this.level.sakuraConfig().cannons.mechanics.useBrokenPaperExplosionBehaviour(this.level)) {
+            this.level.densityCache.invalidate();
+        }
+        // Sakura end - explosion density cache
 
         for (BlockPos pos : targetBlocks) {
             // CraftBukkit start - TNTPrimeEvent
@@ -753,14 +772,12 @@ public class ServerExplosion implements Explosion {
 
     // Paper start - Optimize explosions
     protected float getBlockDensity(Vec3 vec3d, Entity entity) {
-        if (!this.level.paperConfig().environment.optimizeExplosions) {
-            return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
-        }
-        CacheKey key = new CacheKey(this, entity.getBoundingBox());
-        Float blockDensity = this.level.explosionDensityCache.get(key);
-        if (blockDensity == null) {
+        // Sakura start - replace density cache
+        float blockDensity = this.level.densityCache.getBlockDensity(vec3d, entity);
+        if (blockDensity == me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
             blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
-            this.level.explosionDensityCache.put(key, blockDensity);
+            this.level.densityCache.putDensity(vec3d, entity, blockDensity);
+            // Sakura end - replace density cache
         }
 
         return blockDensity;
diff --git a/net/minecraft/world/level/block/BasePressurePlateBlock.java b/net/minecraft/world/level/block/BasePressurePlateBlock.java
index 19d6bea66edcbeadc2cf4b015acb40862177f54b..44669c625db63c8454e85d349970bb81719c41da 100644
--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -115,6 +115,11 @@ public abstract class BasePressurePlateBlock extends Block {
         // CraftBukkit end
         if (oldSignal != signal) {
             BlockState newState = this.setSignalForState(state, signal);
+            // Sakura start - explosion density cache
+            if (!level.sakuraConfig().cannons.mechanics.useBrokenPaperExplosionBehaviour(level)) {
+                level.densityCache.invalidate();
+            }
+            // Sakura end - explosion density cache
             level.setBlock(pos, newState, Block.UPDATE_CLIENTS);
             this.updateNeighbours(level, pos);
             level.setBlocksDirty(pos, state, newState);
diff --git a/net/minecraft/world/level/block/TripWireHookBlock.java b/net/minecraft/world/level/block/TripWireHookBlock.java
index 2c8a4febdbc0a8010e786293f5b110236757aa60..2125418779dbeff02813f328d3cc84093d214f05 100644
--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -173,6 +173,11 @@ public class TripWireHookBlock extends Block {
                 if (!cancelledReceiverHook) { // always trigger two events even when the first hook current change is cancelled
                 // Paper end - Call BlockRedstoneEvent
                 Direction opposite = direction.getOpposite();
+                // Sakura start - explosion density cache
+                if (!level.sakuraConfig().cannons.mechanics.useBrokenPaperExplosionBehaviour(level)) {
+                    level.densityCache.invalidate();
+                }
+                // Sakura end - explosion density cache
                 level.setBlock(testPosx, newState.setValue(FACING, opposite), Block.UPDATE_ALL);
                 notifyNeighbors(block, level, testPosx, opposite);
                 if (!level.getBlockState(pos).is(Blocks.TRIPWIRE_HOOK)) {
diff --git a/net/minecraft/world/phys/AABB.java b/net/minecraft/world/phys/AABB.java
index 259a19cbcf9a42c5f48e5d03448729283baf36a7..4cf1658798d18c9fb2237414e795c5398927f825 100644
--- a/net/minecraft/world/phys/AABB.java
+++ b/net/minecraft/world/phys/AABB.java
@@ -19,6 +19,30 @@ public class AABB {
     public final double maxY;
     public final double maxZ;
 
+    // Sakura start - explosion density cache
+    public final boolean containsInclusive(final AABB other) {
+        return this.minX <= other.minX && this.maxX >= other.maxX
+            && this.minY <= other.minY && this.maxY >= other.maxY
+            && this.minZ <= other.minZ && this.maxZ >= other.maxZ;
+    }
+
+    public final boolean containsInclusive(final Vec3 pos) {
+        return this.minX <= pos.x && this.maxX >= pos.x
+            && this.minY <= pos.y && this.maxY >= pos.y
+            && this.minZ <= pos.z && this.maxZ >= pos.z;
+    }
+
+    public final AABB expand(final Vec3 pos) {
+        final double minX = Math.min(this.minX, pos.x);
+        final double minY = Math.min(this.minY, pos.y);
+        final double minZ = Math.min(this.minZ, pos.z);
+        final double maxX = Math.max(this.maxX, pos.x);
+        final double maxY = Math.max(this.maxY, pos.y);
+        final double maxZ = Math.max(this.maxZ, pos.z);
+        return new AABB(minX, minY, minZ, maxX, maxY, maxZ);
+    }
+    // Sakura end - explosion density cache
+
     public AABB(final double minX, final double minY, final double minZ, final double maxX, final double maxY, final double maxZ) {
         this.minX = Math.min(minX, maxX);
         this.minY = Math.min(minY, maxY);
