From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 27 Jun 2024 17:02:32 +0100
Subject: [PATCH] Add maxSearch to getEntities


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 6087d66f6ba9c1389dd7997f53cf867f07a6744d..5db28a841fc00366cb3934945b665325a8ead896 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -83,6 +83,7 @@ public final class ChunkEntitySlices {
         return this.allEntities.getSectionEntities(sectionIndex);
     }
     // Sakura end - optimise explosions
+    private final java.util.Random random = new java.util.Random(); // Sakura - add maxSearch to getEntities
 
     public ChunkEntitySlices(final Level world, final int chunkX, final int chunkZ, final FullChunkStatus status,
                              final ChunkData chunkData, final int minSection, final int maxSection) { // inclusive, inclusive
@@ -348,7 +349,14 @@ public final class ChunkEntitySlices {
 
     public boolean getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
                                final int maxCount) {
-        return this.allEntities.getEntitiesLimited(except, box, into, predicate, maxCount);
+        // Sakura start - add maxSearch to getEntities
+        return this.getEntities(except, box, into, predicate, maxCount, Integer.MAX_VALUE);
+    }
+
+    public boolean getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
+                               final int maxCount, final int maxSearch) {
+        return this.allEntities.getEntitiesLimited(except, box, into, predicate, maxCount, maxSearch);
+        // Sakura end - add maxSearch to getEntities
     }
 
     public <T extends Entity> void getEntities(final EntityType<?> type, final AABB box, final List<? super T> into,
@@ -586,6 +594,13 @@ public final class ChunkEntitySlices {
 
         public boolean getEntitiesLimited(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
                                           final int maxCount) {
+            // Sakura start - add maxSearch to getEntities
+            return this.getEntitiesLimited(except, box, into, predicate, maxCount, Integer.MAX_VALUE);
+        }
+
+        public boolean getEntitiesLimited(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
+                                          final int maxCount, final int maxSearch) {
+            // Sakura end - add maxSearch to getEntities
             if (this.count == 0) {
                 return false;
             }
@@ -607,8 +622,14 @@ public final class ChunkEntitySlices {
 
                 final Entity[] storage = list.storage;
 
-                for (int i = 0, len = Math.min(storage.length, list.size()); i < len; ++i) {
-                    final Entity entity = storage[i];
+                // Sakura start - add maxSearch to getEntities
+                final int len = Math.min(storage.length, list.size());
+                final int offset = this.slices.random.nextInt(len);
+                for (int i = 0; i < len; ++i) {
+                    final int pos = (i + offset) % len;
+                    final Entity entity = storage[pos];
+                    if (i > maxSearch) break;
+                    // Sakura end - add maxSearch to getEntities
 
                     if (entity == null || entity == except || !entity.getBoundingBox().intersects(box)) {
                         continue;
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index 328784ba08236553e49d62320b79f7222292412f..c9dee57eacb353262f51f15dd9d0f66f21bdb50b 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -730,6 +730,13 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
 
     public void getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
                             final int maxCount) {
+        // Sakura start - add maxSearch to getEntities
+        this.getEntities(except, box, into, predicate, maxCount, Integer.MAX_VALUE);
+    }
+
+    public void getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
+                            final int maxCount, final int maxSearch) {
+        // Sakura end - add maxSearch to getEntities
         final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
         final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
         final int maxChunkX = (Mth.floor(box.maxX) + 2) >> 4;
@@ -761,7 +768,7 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
                             continue;
                         }
 
-                        if (chunk.getEntities(except, box, into, predicate, maxCount)) {
+                        if (chunk.getEntities(except, box, into, predicate, maxCount, maxSearch)) { // Sakura - add maxSearch to getEntities
                             return;
                         }
                     }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 68233cbf5f120e17fd0349cb796e8614de1c5a34..b47d57ba870acda65aea5936bb7bc3b57ccfbe72 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1820,10 +1820,18 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.getEntities(type, bb, selector, output, Integer.MAX_VALUE);
     }
 
-    // Paper start - rewrite chunk system
     public <T extends Entity> void getEntities(final EntityTypeTest<Entity, T> entityTypeTest,
                                                final AABB boundingBox, final Predicate<? super T> predicate,
                                                final List<? super T> into, final int maxCount) {
+        // Sakura start - add maxSearch to getEntities
+        this.getEntities(entityTypeTest, boundingBox, predicate, into, maxCount, Integer.MAX_VALUE);
+    }
+
+    // Paper start - rewrite chunk system
+    public <T extends Entity> void getEntities(final EntityTypeTest<Entity, T> entityTypeTest,
+                                               final AABB boundingBox, final Predicate<? super T> predicate,
+                                               final List<? super T> into, final int maxCount, final int maxSearch) {
+        // Sakura end - add maxSearch to getEntities
         Profiler.get().incrementCounter("getEntities");
 
         if (entityTypeTest instanceof net.minecraft.world.entity.EntityType<T> byType) {
@@ -1840,7 +1848,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
         if (entityTypeTest == null) {
             if (maxCount != Integer.MAX_VALUE) {
-                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)this).moonrise$getEntityLookup().getEntities((Entity)null, boundingBox, (List)into, (Predicate)predicate, maxCount);
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)this).moonrise$getEntityLookup().getEntities((Entity)null, boundingBox, (List)into, (Predicate)predicate, maxCount, maxSearch); // Sakura - add maxSearch to getEntities
                 ca.spottedleaf.moonrise.common.PlatformHooks.get().addToGetEntities((Level)(Object)this, entityTypeTest, boundingBox, predicate, into, maxCount);
                 return;
             } else {
