From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 3 May 2024 15:04:31 +0100
Subject: [PATCH] Specialised Explosions


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 2d7e47ee49012d87438b388d8829d29966228e09..6c2afecc8eee49ef327a8ef753eba74d35f47950 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2007,7 +2007,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             case STANDARD -> Explosion.BlockInteraction.DESTROY; // CraftBukkit - handle custom explosion type
         };
         Vec3 center = new Vec3(x, y, z);
-        ServerExplosion explosion = new ServerExplosion(this, source, damageSource, damageCalculator, center, r, fire, blockInteraction);
+        // Sakura start - specialised explosions
+        final ServerExplosion explosion;
+        if (source instanceof net.minecraft.world.entity.item.PrimedTnt tnt) {
+            explosion = new me.samsuik.sakura.explosion.TntExplosion(this, tnt, damageSource, damageCalculator, center, r, fire, blockInteraction);
+        } else {
+            explosion = new ServerExplosion(this, source, damageSource, damageCalculator, center, r, fire, blockInteraction);
+        }
+        // Sakura end - specialised explosions
         if (configurator != null) configurator.accept(explosion); // Paper - Allow explosions to damage source
         int blockCount = explosion.explode();
         // CraftBukkit start
@@ -2015,6 +2022,29 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             return explosion;
         }
         // CraftBukkit end
+        // Sakura start - specialised explosions
+        if (explosion instanceof me.samsuik.sakura.explosion.SpecialisedExplosion<?> specialisedExplosion) {
+            me.samsuik.sakura.explosion.ExplosionToSend explosionToSend;
+            while ((explosionToSend = specialisedExplosion.getExplosionsToSend().poll()) != null) {
+                this.notifyPlayersOfExplosion(explosion, explosionToSend.position(), r, explosionToSend.blocksDestroyed(), smallExplosionParticles, largeExplosionParticles, explosionSound, blockParticles);
+            }
+        }
+
+        this.notifyPlayersOfExplosion(explosion, center, r, blockCount, smallExplosionParticles, largeExplosionParticles, explosionSound, blockParticles);
+        return explosion;
+    }
+
+    private void notifyPlayersOfExplosion(
+        final ServerExplosion explosion,
+        final Vec3 center,
+        final float r,
+        final int blockCount,
+        final ParticleOptions smallExplosionParticles,
+        final ParticleOptions largeExplosionParticles,
+        final Holder<SoundEvent> explosionSound,
+        final WeightedList<ExplosionParticleInfo> blockParticles
+    ) {
+        // Sakura end - specialised explosions
         ParticleOptions explosionParticle = explosion.isSmall() ? smallExplosionParticles : largeExplosionParticles;
 
         for (ServerPlayer player : this.players) {
@@ -2042,7 +2072,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
         }
 
-        return explosion; // CraftBukkit
+        explosion.getHitPlayers().clear(); // Sakura - specialised explosions
     }
 
     private Explosion.BlockInteraction getDestroyType(final GameRule<Boolean> gameRule) {
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 51a7be19f8ac748490467d4fa32c851a9b280304..bbe8df26077ae4e4259aed37b212e679a1d7e03f 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -80,20 +80,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
 
     @Override
     public final void respawnEntity(int count) {
-        final PrimedTnt tnt = new PrimedTnt(EntityType.TNT, this.level());
-        tnt.updateBukkitHandle(this); // update handle for plugins
-        while (count-- > 1) {
-            this.setFuse(100); // Prevent unwanted explosions while ticking
-
-            // Cause an explosion to affect this entity
-            tnt.setPos(this.position());
-            tnt.setDeltaMovement(this.getDeltaMovement());
-            this.entityState().apply(this);
-            tnt.explode();
-            this.storeEntityState();
-
-            this.tick();
-        }
+        this.mergeData.count = count; // Sakura - specialised explosions
     }
     // Sakura end - merge cannon entities
 
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 39a7c8d280d6e6cee8c5c9aecc8d98128c9216a2..546a3abe25a138beed1c16f8d18d01383022bfa4 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -319,6 +319,38 @@ public class ServerExplosion implements Explosion {
         return true;
     }
     // Sakura end - optimise explosion protected regions
+    // Sakura start - specialised explosions
+    protected final void createBlockCache() {
+        // Paper start - collision optimisations
+        this.blockCache = new it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<>();
+        this.chunkPosCache = new long[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
+        java.util.Arrays.fill(this.chunkPosCache, ChunkPos.INVALID_CHUNK_POS);
+        this.chunkCache = new net.minecraft.world.level.chunk.LevelChunk[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
+        this.directMappedBlockCache = new ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH];
+        this.mutablePos = new BlockPos.MutableBlockPos();
+        // Paper end - collision optimisations
+    }
+
+    protected final void clearBlockCache() {
+        // Paper start - collision optimisations
+        this.blockCache = null;
+        this.chunkPosCache = null;
+        this.chunkCache = null;
+        this.directMappedBlockCache = null;
+        this.mutablePos = null;
+        // Paper end - collision optimisations
+    }
+
+    protected final void markBlocksInCacheAsExplodable(final List<BlockPos> explodedPositions) {
+        for (final BlockPos explodedPos : explodedPositions) {
+            final ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cache = this.blockCache.get(explodedPos.asLong());
+            // May be null if the blockCache is cleared then retrieved from the recent block cache
+            if (cache != null) {
+                cache.shouldExplode = null;
+            }
+        }
+    }
+    // Sakura end - specialised explosions
 
     public ServerExplosion(
         final ServerLevel level,
@@ -512,17 +544,29 @@ public class ServerExplosion implements Explosion {
     }
 
     private void hurtEntities() {
-        if (!(this.radius < 1.0E-5F)) {
-            float doubleRadius = this.radius * 2.0F;
-            int x0 = Mth.floor(this.center.x - doubleRadius - 1.0);
-            int x1 = Mth.floor(this.center.x + doubleRadius + 1.0);
-            int y0 = Mth.floor(this.center.y - doubleRadius - 1.0);
-            int y1 = Mth.floor(this.center.y + doubleRadius + 1.0);
-            int z0 = Mth.floor(this.center.z - doubleRadius - 1.0);
-            int z1 = Mth.floor(this.center.z + doubleRadius + 1.0);
-            List<Entity> list = this.level.getEntities(this.excludeSourceFromDamage ? this.source : null, new AABB(x0, y0, z0, x1, y1, z1), entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
-
-            for (Entity entity : list) { // Paper - used in loop
+        // Sakura start - specialised explosions
+        final float explosionRadius = this.radius * 2.0F;
+        if (explosionRadius < 1.0e-5f) {
+            return;
+        }
+        for (final Entity entity : this.level.getEntities(null, this.getExplosionBounds(explosionRadius))) {
+            this.impactEntity(explosionRadius, entity);
+        }
+    }
+
+    private AABB getExplosionBounds(final float doubleRadius) {
+        final int minX = Mth.floor(this.center.x - doubleRadius - 1.0);
+        final int maxX = Mth.floor(this.center.x + doubleRadius + 1.0);
+        final int minY = Mth.floor(this.center.y - doubleRadius - 1.0);
+        final int maxY = Mth.floor(this.center.y + doubleRadius + 1.0);
+        final int minZ = Mth.floor(this.center.z - doubleRadius - 1.0);
+        final int maxZ = Mth.floor(this.center.z + doubleRadius + 1.0);
+        return new AABB(minX, minY, minZ, maxX, maxY, maxZ);
+    }
+
+    protected final void impactEntity(final float doubleRadius, final Entity entity) {
+        if (entity.isAlive() && !entity.isSpectator() && (!this.excludeSourceFromDamage || entity != this.source)) { // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
+            { // Sakura end - specialised explosions
                 if (!entity.ignoreExplosion(this)) {
                     double dist = Math.sqrt(entity.distanceToSqr(this.center)) / doubleRadius;
                     if (!(dist > 1.0)) {
@@ -541,15 +585,16 @@ public class ServerExplosion implements Explosion {
                             // - Damaging EnderDragon does nothing
                             // - EnderDragon hitbox always covers the other parts and is therefore always present
                             if (entity instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragonPart) {
-                                continue;
+                                return; // Sakura - specialised explosions
                             }
 
                             entity.lastDamageCancelled = false;
 
                             if (entity instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon enderDragon) {
+                                final AABB explosionBounds = this.getExplosionBounds(doubleRadius); // Sakura - specialised explosions
                                 for (net.minecraft.world.entity.boss.enderdragon.EnderDragonPart dragonPart : enderDragon.getSubEntities()) {
                                     // Calculate damage separately for each EntityComplexPart
-                                    if (list.contains(dragonPart)) {
+                                    if (explosionBounds.intersects(dragonPart.getBoundingBox())) { // Sakura - specialised explosions
                                         dragonPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, dragonPart, exposure));
                                     }
                                 }
@@ -558,7 +603,7 @@ public class ServerExplosion implements Explosion {
                             }
 
                             if (entity.lastDamageCancelled) { // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Skip entity if damage event was cancelled
-                                continue;
+                                return; // Sakura - specialised explosions
                             }
                             // CraftBukkit end
                         }
@@ -580,7 +625,10 @@ public class ServerExplosion implements Explosion {
                         if (entity.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
                             projectile.setOwner(this.damageSource.getEntity());
                         } else if (entity instanceof Player player && !player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Option to disable explosion knockback
-                            this.hitPlayers.put(player, knockback);
+                            // Sakura start - specialised explosions; tally player velocity
+                            final Vec3 finalKnockback = knockback;
+                            this.hitPlayers.compute(player, (p, v) -> v != null ? v.add(finalKnockback) : finalKnockback);
+                            // Sakura end - specialised explosions; tally player velocity
                         }
 
                         entity.onExplosionHit(this.source);
@@ -683,14 +731,7 @@ public class ServerExplosion implements Explosion {
     }
 
     public int explode() {
-        // Paper start - collision optimisations
-        this.blockCache = new it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<>();
-        this.chunkPosCache = new long[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
-        java.util.Arrays.fill(this.chunkPosCache, ChunkPos.INVALID_CHUNK_POS);
-        this.chunkCache = new net.minecraft.world.level.chunk.LevelChunk[CHUNK_CACHE_WIDTH * CHUNK_CACHE_WIDTH];
-        this.directMappedBlockCache = new ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH * BLOCK_EXPLOSION_CACHE_WIDTH];
-        this.mutablePos = new BlockPos.MutableBlockPos();
-        // Paper end - collision optimisations
+        this.createBlockCache(); // Sakura - specialised explosions
         this.level.gameEvent(this.source, GameEvent.EXPLODE, this.center);
         List<BlockPos> toBlow = this.calculateExplodedPositions();
         this.hurtEntities();
@@ -705,13 +746,7 @@ public class ServerExplosion implements Explosion {
             this.createFire(toBlow);
         }
 
-        // Paper start - collision optimisations
-        this.blockCache = null;
-        this.chunkPosCache = null;
-        this.chunkCache = null;
-        this.directMappedBlockCache = null;
-        this.mutablePos = null;
-        // Paper end - collision optimisations
+        this.clearBlockCache(); // Sakura - specialised explosions
         return toBlow.size();
     }
 
