From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 23 Sep 2021 18:50:13 +0100
Subject: [PATCH] Use maxEntityCollision limit for entity retrieval


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index b2e6fbe1c7ae3ae59c17b4f8c06b30185b95e000..924408367d999ee6bda34699fce7bd5b100f253b 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3951,7 +3951,17 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             return;
         }
         // Paper end - don't run getEntities if we're not going to use its result
-        List<Entity> pushableEntities = this.level().getPushableEntities(this, this.getBoundingBox());
+        // Sakura start - use maxEntityCollision limit for entity retrieval
+        //List<Entity> pushableEntities = this.level().getPushableEntities(this, this.getBoundingBox());
+        final int limit = Math.max(i, maxEntityCollisions);
+        final int search = limit * limit;
+        final List<Entity> pushableEntities = new java.util.ArrayList<>(4);
+        this.level().getEntities(
+            net.minecraft.world.level.entity.EntityTypeTest.forClass(Entity.class),
+            this.getBoundingBox(), EntitySelector.pushableBy(this),
+            pushableEntities, limit, search
+        );
+        // Sakura end - use maxEntityCollision limit for entity retrieval
         if (!pushableEntities.isEmpty()) {
             if (this.level() instanceof ServerLevel serverLevel) {
                 // Paper - don't run getEntities if we're not going to use its result; moved up
