From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 30 Nov 2023 15:54:49 +0000
Subject: [PATCH] Reduce entity tracker player updates


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index f16166dd781f70d159a3120017f1ae0348ffc26b..3c3cc60d1345aa85b713a2dabb8fcc42535e0f49 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1000,7 +1000,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 }
 
                 UpdateInterval updateInterval = type.hasUpdateInterval() ? UpdateInterval.periodic(type.updateInterval()) : UpdateInterval.NEVER;
-                ChunkMap.TrackedEntity trackedEntity = new ChunkMap.TrackedEntity(entity, range, updateInterval, type.trackDeltas());
+                ChunkMap.TrackedEntity trackedEntity = new ChunkMap.TrackedEntity(entity, range, updateInterval, type.trackDeltas(), type.updateInterval()); // Sakura start - reduce entity tracker player updates
                 this.entityMap.put(entity.getId(), trackedEntity);
                 // Paper start - optimise entity tracker
                 if (((ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity)entity).moonrise$getTrackedEntity() != null) {
@@ -1104,7 +1104,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 if (tracker == null) {
                     continue;
                 }
-                ((ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerTrackedEntity)tracker).moonrise$tick(((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)entity).moonrise$getChunkData().nearbyPlayers);
+                // Sakura start - reduce entity tracker player updates
+                if (tracker.shouldUpdatePlayers()) {
+                    ((ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerTrackedEntity)tracker).moonrise$tick(((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)entity).moonrise$getChunkData().nearbyPlayers);
+                }
+                // Sakura end - reduce entity tracker player updates
                 if (((ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerTrackedEntity)tracker).moonrise$hasPlayers()
                     || ((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)entity).moonrise$getChunkStatus().isOrAfter(FullChunkStatus.ENTITY_TICKING)) {
                     tracker.serverEntity.sendChanges();
@@ -1347,12 +1351,41 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             return !this.seenBy.isEmpty();
         }
         // Paper end - optimise entity tracker
+        // Sakura start - reduce entity tracker player updates
+        private final int updatePlayerInterval;
+        private net.minecraft.world.phys.Vec3 lastUpdatedPosition;
+
+        public final boolean shouldUpdatePlayers() {
+            // Always update players otherwise they sometimes turn invisible when teleporting
+            if (this.entity instanceof net.minecraft.world.entity.player.Player || this.entity.tickCount % this.updatePlayerInterval == 0) {
+                return true;
+            }
+
+            final net.minecraft.world.phys.Vec3 lastPosition = this.lastUpdatedPosition;
+            this.lastUpdatedPosition = this.entity.position();
+            return this.entity.position().distanceToSqr(lastPosition) >= (double) this.range / 2.0;
+        }
 
         public TrackedEntity(final Entity entity, final int range, final UpdateInterval updateInterval, final boolean trackDelta) {
+            this(entity, range, updateInterval, trackDelta, 0);
+        }
+
+        public TrackedEntity(
+            final Entity entity,
+            final int range,
+            final UpdateInterval updateInterval,
+            final boolean trackDelta,
+            final int updatePeriod
+        ) {
+            // Sakura end - reduce entity tracker player updates
             this.serverEntity = new ServerEntity(ChunkMap.this.level, entity, updateInterval, trackDelta, this, this.seenBy); // Paper
             this.entity = entity;
             this.range = range;
             this.lastSectionPos = SectionPos.of(entity);
+            // Sakura start - reduce entity tracker player updates
+            this.updatePlayerInterval = Math.clamp(updatePeriod, 1, 20);
+            this.lastUpdatedPosition = entity.position();
+            // Sakura end - reduce entity tracker player updates
         }
 
         @Override
