From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Nov 2023 17:03:08 +0000
Subject: [PATCH] Configure cannon physics


diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index c093f2295d6a0e1880897fe66bdcd64649aa667e..ce13de98e45782a7042e6f232431605a64adb668 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -1774,6 +1774,13 @@ public final class CollisionUtil {
     }
 
     public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb, final List<AABB> potentialCollisions) {
+        // Sakura start - configure server mechanics
+        return performAABBCollisions(moveVector, axisalignedbb, potentialCollisions, null);
+    }
+    public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                             final List<AABB> potentialCollisions,
+                                             final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        // Sakura end - configure server mechanics
         double x = moveVector.x;
         double y = moveVector.y;
         double z = moveVector.z;
@@ -1785,7 +1792,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, potentialCollisions);
@@ -1811,9 +1818,21 @@ public final class CollisionUtil {
     public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
                                          final List<VoxelShape> voxels,
                                          final List<AABB> aabbs) {
+        // Sakura start - configure server mechanics
+        return performCollisions(moveVector, axisalignedbb, voxels, aabbs, null);
+    }
+
+    public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                         final List<VoxelShape> voxels,
+                                         final List<AABB> aabbs,
+                                         final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        if (mechanicsTarget != null && mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.PAPER) && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2, me.samsuik.sakura.mechanics.MechanicVersion.v1_20_2)) {
+            me.samsuik.sakura.mechanics.EntityBehaviour.convertVoxelsIntoAABBs(moveVector, axisalignedbb, voxels, aabbs);
+        }
+        // Sakura end - configure server mechanics
         if (voxels.isEmpty()) {
             // fast track only AABBs
-            return performAABBCollisions(moveVector, axisalignedbb, aabbs);
+            return performAABBCollisions(moveVector, axisalignedbb, aabbs, mechanicsTarget); // Sakura - configure server mechanics
         }
 
         double x = moveVector.x;
@@ -1828,7 +1847,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, aabbs);
@@ -1886,6 +1905,7 @@ public final class CollisionUtil {
     public static final int COLLISION_FLAG_CHECK_BORDER = 1 << 2;
     public static final int COLLISION_FLAG_CHECK_ONLY = 1 << 3;
     public static final int COLLISION_FLAG_ADD_TICKET = 1 << 4; // Sakura - load chunks on movement
+    public static final int COLLISION_FLAG_SKIP_EDGES = 1 << 5; // Sakura - configure server mechanics
 
     public static boolean getCollisionsForBlocksOrWorldBorder(final Level world, final Entity entity, final AABB aabb,
                                                               final List<VoxelShape> intoVoxel, final List<AABB> intoAABB,
@@ -1939,6 +1959,7 @@ public final class CollisionUtil {
 
         final boolean loadChunks = (collisionFlags & COLLISION_FLAG_LOAD_CHUNKS) != 0;
         final boolean addTicket  = (collisionFlags & COLLISION_FLAG_ADD_TICKET) != 0; // Sakura - load chunks on movement
+        final boolean skipEdges  = (collisionFlags & COLLISION_FLAG_SKIP_EDGES) != 0; // Sakura - configure server mechanics
         final ChunkSource chunkSource = world.getChunkSource();
 
         for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
@@ -1981,7 +2002,7 @@ public final class CollisionUtil {
                         continue;
                     }
 
-                    final boolean hasSpecial = ((BlockCountingChunkSection)section).moonrise$hasSpecialCollidingBlocks();
+                    final boolean hasSpecial = !skipEdges && ((BlockCountingChunkSection)section).moonrise$hasSpecialCollidingBlocks(); // Sakura - configure server mechanics
                     final int sectionAdjust = !hasSpecial ? 1 : 0;
 
                     final PalettedContainer<BlockState> blocks = section.states;
diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index 4c456c2d1b42ecfcecb22ba2bf3819cab68206b8..7d2df500c90e0451ec2cf117f44413b3593fdd62 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -11,5 +11,6 @@ public final class ScheduleSakuraTasks {
         scheduler.endless((tick, level, cb) -> level.explosionPositions.clear(), 0); // Sakura - client visibility settings
         scheduler.endless((tick, level, cb) -> level.mergeHandler.expire(tick), 200); // Sakura - merge cannon entities
         scheduler.endless((tick, level, cb) -> level.densityCache.expire(tick), 0); // Sakura - explosion density cache
+        scheduler.endless((tick, level, cb) -> level.dispenseRelocationHandler.clear(), 0); // Sakura - configure server mechanics
     }
 }
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 751232d09f5570a23223356a6572090e7d61139b..4bd8940827ef74045cd009383a1f9132e04b8efc 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -407,6 +407,22 @@ public interface DispenseItemBehavior {
                         }
                     }
 
+                    // Sakura start - configure server mechanics
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicTarget = level.localConfig().at(target).mechanicsTarget;
+                    if (!mechanicTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.VANILLA)
+                        && mechanicTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)
+                        && !mechanicTarget.isLegacy()) {
+                        event.setVelocity(event.getVelocity().setY(event.getVelocity().getY() + 0.5));
+                    }
+
+                    if (mechanicTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+                        final BlockPos originalSpawnPos = org.bukkit.craftbukkit.util.CraftVector.toBlockPos(event.getVelocity());
+                        final BlockPos spawnPos = level.dispenseRelocationHandler.relocatePosition(originalSpawnPos, source.pos(), source.state());
+                        final net.minecraft.world.phys.Vec3 centeredPos = net.minecraft.world.phys.Vec3.atBottomCenterOf(spawnPos);
+
+                        event.setVelocity(org.bukkit.craftbukkit.util.CraftVector.toBukkit(centeredPos));
+                    }
+                    // Sakura end - configure server mechanics
                     PrimedTnt tnt = new PrimedTnt(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), null);
                     // CraftBukkit end
                     level.addFreshEntity(tnt);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 66b50348888fa3c32f6ab0d0167b7cc6336ee3ef..b04646182f2d37b7f3d1784695d1b2687aa2d9ca 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1377,7 +1377,7 @@ public abstract class Entity
         this.finalMovementsThisTick.clear();
         this.finalMovementsThisTick.addAll(this.movementThisTick);
         this.movementThisTick.clear();
-        if (this.finalMovementsThisTick.isEmpty()) {
+        if (this.finalMovementsThisTick.isEmpty() || this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
             this.finalMovementsThisTick.add(new Entity.Movement(this.oldPosition(), this.position()));
         } else if (this.finalMovementsThisTick.getLast().to.distanceToSqr(this.position()) > 9.9999994E-11F) {
             this.finalMovementsThisTick.add(new Entity.Movement(this.finalMovementsThisTick.getLast().to, this.position()));
@@ -1788,7 +1788,7 @@ public abstract class Entity
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER  | this.getExtraCollisionFlags(), null // Sakura - load chunks on movement
         );
         potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
+        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB, this.mechanicsTarget); // Sakura - configure server mechanics
 
         final boolean collidedX = collided.x != movement.x;
         final boolean collidedY = collided.y != movement.y;
diff --git a/net/minecraft/world/entity/Entity.javapscLxj b/net/minecraft/world/entity/Entity.javapscLxj
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index a66247beee27f9b521b2d167df3228da8a334624..089ca387316a0ddda7c56c11a3da023ba63d9b43 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -129,6 +129,18 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public final double distanceToSqr(final Vec3 pos) {
+        if (!this.mechanicsTarget.isLegacy()) {
+            return super.distanceToSqr(pos);
+        }
+        final double x = this.getX() - pos.x;
+        final double y = this.getEyeY() - pos.y;
+        final double z = this.getZ() - pos.z;
+        return x * x + y * y + z * z;
+    }
+    // Sakura end - configure server mechanics
 
     public FallingBlockEntity(final EntityType<? extends FallingBlockEntity> type, final Level level) {
         super(type, level);
@@ -149,12 +161,16 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         this.yo = y;
         this.zo = z;
         this.setStartPos(this.blockPosition());
+        // Sakura start - configure server mechanics
+        this.mechanicsTarget = level.localConfig().at(this.blockPosition()).mechanicsTarget;
+        this.eyeHeight = this.mechanicsTarget.isLegacy() ? 0.49f : this.eyeHeight;
+        // Sakura end - configure server mechanics
     }
 
     // Sakura start - falling block height parity api
     @Override
     public final double getEyeY() {
-        return this.heightParity ? this.getY() : super.getEyeY();
+        return this.heightParity && !this.mechanicsTarget.isLegacy() ? this.getY() : super.getEyeY(); // Sakura - configure server mechanics
     }
     // Sakura end - falling block height parity api
     // Sakura start - falling block stacking restrictions
@@ -188,7 +204,11 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             state.hasProperty(BlockStateProperties.WATERLOGGED) ? state.setValue(BlockStateProperties.WATERLOGGED, false) : state
         );
         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.getFluidState().createLegacyBlock())) return entity; // CraftBukkit
-        level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        // Sakura start - configure server mechanics
+        if (entity.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        }
+        // Sakura end - configure server mechanics
         level.addFreshEntity(entity);
         return entity;
     }
@@ -232,7 +252,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -241,6 +261,12 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             Block block = this.blockState.getBlock();
+            // Sakura start - configure server mechanics
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+            if (this.time == 0 && mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                me.samsuik.sakura.mechanics.FallingBlockBehaviour.removeBlockOnFall(this, block);
+            }
+            // Sakura end - configure server mechanics
             this.time++;
             this.applyGravity();
             this.move(MoverType.SELF, this.getDeltaMovement());
@@ -255,9 +281,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             }
             // Paper end - Configurable falling blocks height nerf
             this.handlePortal();
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                this.setDeltaMovement(this.getDeltaMovement().scale(0.98F));
+            }
+            // Sakura end - configure server mechanics
             if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
                 // Sakura start - fix falling block floating point errors
-                BlockPos blockPos = serverLevel.sakuraConfig().cannons.mechanics.fallingBlockFloatingPointFix
+                BlockPos blockPos = serverLevel.sakuraConfig().cannons.mechanics.fallingBlockFloatingPointFix || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17) // Sakura - configure server mechanics
                     ? this.patchedBlockPosition()
                     : this.blockPosition();
                 // Sakura end - fix falling block floating point errors
@@ -287,8 +318,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                     }
                 } else {
                     BlockState currentState = this.level().getBlockState(pos);
-                    this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
-                    if (!currentState.is(Blocks.MOVING_PISTON)) {
+                    // Sakura start - configure server mechanics
+                    final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                        ? 0.7f
+                        : 0.7;
+                    this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+                    // todo: sakura-update: review flag1
+                    if (!currentState.is(Blocks.MOVING_PISTON) && (flag1 || me.samsuik.sakura.mechanics.FallingBlockBehaviour.isAbleToStackOnBlock(this, mechanicsTarget))) {
+                        // Sakura end - configure server mechanics
                         if (!this.cancelDrop) {
                             boolean mayReplace = currentState.canBeReplaced(
                                 new DirectionalPlaceContext(this.level(), pos, Direction.DOWN, ItemStack.EMPTY, Direction.UP)
@@ -359,7 +396,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                 }
             }
 
-            this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                    ? 0.98f
+                    : 0.98;
+                this.setDeltaMovement(this.getDeltaMovement().scale(drag));
+            }
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 86769ddbab6598a1beaeb6580f95a23e98f869c5..193d7a1c5b624b48161c4bfde346a65b330c2972 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -89,6 +89,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public final double getEyeY() {
+        return this.mechanicsTarget.isLegacy() ? super.getEyeY() : this.getY();
+    }
+
+    @Override
+    public final double distanceToSqr(final net.minecraft.world.phys.Vec3 pos) {
+        if (!this.mechanicsTarget.isLegacy()) {
+            return super.distanceToSqr(pos);
+        }
+        final double x = this.getX() - pos.x;
+        final double y = this.getEyeY() - pos.y;
+        final double z = this.getZ() - pos.z;
+        return x * x + y * y + z * z;
+    }
+    // Sakura end - configure server mechanics
 
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
@@ -114,6 +131,13 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             case Y -> this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
         }
         // Sakura end - configure cannon mechanics
+        // Sakura start - configure server mechanics
+        this.mechanicsTarget = level.localConfig().at(this.blockPosition()).mechanicsTarget;
+        this.eyeHeight = this.mechanicsTarget.isLegacy() ? 0.49f : this.eyeHeight;
+        if (this.mechanicsTarget.isLegacy()) {
+            this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
+        }
+        // Sakura end - configure server mechanics
     }
 
     // Sakura start - optimise tnt fluid state
@@ -149,7 +173,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -165,14 +189,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             return;
         }
         // Paper end - Configurable TNT height nerf
-        this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+        final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+            ? 0.98f
+            : 0.98;
+        this.setDeltaMovement(this.getDeltaMovement().scale(drag));
         if (this.onGround()) {
-            this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
+            final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                ? 0.7f
+                : 0.7;
+            this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+            // Sakura end - configure server mechanics
         }
 
         int fuse = this.getFuse() - 1;
         this.setFuse(fuse);
-        if (fuse <= 0) {
+        if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9) ? (fuse < 0) : (fuse <= 0)) { // Sakura - configure server mechanics
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
             //this.discard();
             this.tryToRespawnEntity(); // Sakura - merge cannon entities
@@ -211,13 +244,19 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
                 return;
             }
             // CraftBukkit end
+            // Sakura start - configure server mechanics
+            // Undocumented on the minecraft wiki but the explosion position did change in 1.10
+            final double explosionY = this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10)
+                ? this.getY() + (double) 0.49f
+                : this.getY(0.0625D);
+            // Sakura end - configure server mechanics
             this.level()
                 .explode(
                     this,
                     Explosion.getDefaultDamageSource(this.level(), this),
                     this.usedPortal ? USED_PORTAL_DAMAGE_CALCULATOR : null,
                     this.getX(),
-                    this.getY(0.0625),
+                    explosionY, // Sakura - configure server mechanics
                     this.getZ(),
                     event.getRadius(), // CraftBukkit
                     event.getFire(), // CraftBukkit
@@ -306,7 +345,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
     // Paper start - Option to prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater  && super.isPushedByFluid(); // Sakura - configure cannon mechanics
+        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater && !this.mechanicsTarget.isLegacy() && super.isPushedByFluid(); // Sakura - configure server mechanics // Sakura - configure cannon mechanics
     }
     // Paper end - Option to prevent TNT from moving in water
 }
diff --git a/net/minecraft/world/level/BlockGetter.java b/net/minecraft/world/level/BlockGetter.java
index f9073223aad937a9bd27a9a8be8817de32f967f6..4c9eb6ee78727e21925f47d56d260bf7bc17302d 100644
--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -207,8 +207,23 @@ public interface BlockGetter extends LevelHeightAccessor {
     }
 
     static boolean forEachBlockIntersectedBetween(final Vec3 from, final Vec3 to, final AABB aabbAtTarget, final BlockGetter.BlockStepVisitor visitor) {
+        // Sakura start - configure server mechanics
+        return forEachBlockIntersectedBetween(from, to, aabbAtTarget, me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest(), visitor);
+    }
+
+    static boolean forEachBlockIntersectedBetween(
+        final Vec3 from,
+        final Vec3 to,
+        final AABB aabbAtTarget,
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget,
+        final BlockGetter.BlockStepVisitor visitor
+    ) {
         Vec3 travel = to.subtract(from);
-        if (travel.lengthSqr() < Mth.square(1.0E-5F)) {
+        final double movedThreshold = mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)
+            ? 1.0E-5F
+            : 0.99999F;
+        if (travel.lengthSqr() < Mth.square(movedThreshold) || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            // Sakura end - configure server mechanics
             for (BlockPos blockPos : BlockPos.betweenClosed(aabbAtTarget)) {
                 if (!visitor.visit(blockPos, 0)) {
                     return false;
@@ -219,15 +234,22 @@ public interface BlockGetter extends LevelHeightAccessor {
         } else {
             LongSet visitedBlocks = new LongOpenHashSet();
 
-            for (BlockPos blockPosx : BlockPos.betweenCornersInDirection(aabbAtTarget.move(travel.scale(-1.0)), travel)) {
-                if (!visitor.visit(blockPosx, 0)) {
-                    return false;
-                }
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                for (BlockPos blockPosx : BlockPos.betweenCornersInDirection(aabbAtTarget.move(travel.scale(-1.0)), travel)) {
+                    if (!visitor.visit(blockPosx, 0)) {
+                        return false;
+                    }
 
-                visitedBlocks.add(blockPosx.asLong());
+                    visitedBlocks.add(blockPosx.asLong());
+                }
             }
 
             int iterations = addCollisionsAlongTravel(visitedBlocks, travel, aabbAtTarget, visitor);
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+                iterations = 0;
+            }
+            // Sakura end - configure server mechanics
             if (iterations < 0) {
                 return false;
             } else {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 446ca8787231e40de2a6c62b0c4bf2aef9367455..1b2eaab42a615bf596011399794280ed41dfc970 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -846,6 +846,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
     public final me.samsuik.sakura.explosion.density.BlockDensityCache densityCache = new me.samsuik.sakura.explosion.density.BlockDensityCache(this); // Sakura - optimise explosion density cache
     public final me.samsuik.sakura.explosion.durable.DurableBlockManager durabilityManager = new me.samsuik.sakura.explosion.durable.DurableBlockManager(); // Sakura - explosion durable blocks
+    public final me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler dispenseRelocationHandler = new me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler(this); // Sakura - configure server mechanics
 
     protected Level(
         final WritableLevelData levelData,
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 061b23137baa3a96fed1f0031b188b1f14e15fdc..d08bf878a53de5815982a827071eb8e0eabc7605 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -370,6 +370,37 @@ public class ServerExplosion implements Explosion {
         return this.damageCalculator.getBlockExplosionResistance(this, this.level, pos, blockState, fluidState);
     }
     // Sakura end - explosion durable blocks
+    // Sakura start - configure server mechanics
+    protected final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+
+    private float sakura$getSeenPercent(final Vec3 explosionPos, final Entity entity) {
+        if (this.mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+            return this.getSeenPercentSingleRaycast(explosionPos, entity);
+        } else if (this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) {
+            return this.getSeenFraction(explosionPos, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+        } else {
+            return getSeenPercent(explosionPos, entity);
+        }
+    }
+
+    private float getSeenPercentSingleRaycast(final Vec3 explosionPos, final Entity entity) {
+        final AABB bounds = entity.getBoundingBox();
+        final double middle = 1.0 / (bounds.getYsize() * 2.0 + 1.0);
+        final Vec3 point = new Vec3(entity.getX(), entity.getY() + middle, entity.getZ());
+
+        final HitResult.Type hitResult;
+        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), point, explosionPos);
+        } else {
+            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                ? ClipContext.Block.COLLIDER
+                : ClipContext.Block.OUTLINE;
+            hitResult = entity.level().clip(new ClipContext(point, explosionPos, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+        }
+
+        return hitResult == HitResult.Type.MISS ? 1.0f : 0.0f;
+    }
+    // Sakura end - configure server mechanics
 
     public ServerExplosion(
         final ServerLevel level,
@@ -394,6 +425,7 @@ public class ServerExplosion implements Explosion {
         this.yield = Double.isFinite(this.yield) ? this.yield : 0; // Paper - Don't allow infinite default yields
         // Paper end - add yield
         this.consistentExplosionRadius = level.localConfig().at(this.center).consistentExplosionRadius; // Sakura - consistent explosion radius
+        this.mechanicsTarget = source != null ? source.mechanicsTarget() : level.localConfig().at(this.center).mechanicsTarget; // Sakura - configure server mechanics
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(final @Nullable Entity source) {
@@ -424,8 +456,15 @@ public class ServerExplosion implements Explosion {
                         final float density = entity.level().densityCache.getKnownDensity(center);
                         if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
                             hitResult = density != 0.0f ? net.minecraft.world.phys.HitResult.Type.MISS : net.minecraft.world.phys.HitResult.Type.BLOCK;
+                            // Sakura start - configure server mechanics
+                        } else if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+                            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), vec3, explosionVector);
                         } else {
-                            hitResult = entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                                ? ClipContext.Block.COLLIDER
+                                : ClipContext.Block.OUTLINE;
+                            hitResult = entity.level().clip(new ClipContext(from, center, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+                            // Sakura end - configure server mechanics
                         }
                         if (hitResult == HitResult.Type.MISS) {
                             // Sakura end - replace density cache
@@ -523,6 +562,15 @@ public class ServerExplosion implements Explosion {
                 }
 
                 if (cachedBlock.outOfWorld) {
+                    // Sakura start - configure server mechanics
+                    if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                        power -= 0.22500001F;
+                        currX += incX;
+                        currY += incY;
+                        currZ += incZ;
+                        continue;
+                    }
+                    // Sakura end - configure server mechanics
                     break;
                 }
                 final BlockState iblockdata = cachedBlock.blockState;
@@ -590,7 +638,27 @@ public class ServerExplosion implements Explosion {
                     double dist = Math.sqrt(entity.distanceToSqr(this.center)) / doubleRadius;
                     if (!(dist > 1.0)) {
                         Vec3 entityOrigin = entity instanceof PrimedTnt ? entity.position() : entity.getEyePosition();
-                        Vec3 direction = entityOrigin.subtract(this.center).normalize();
+                        // Sakura start - configure server mechanics
+                        final double dirX = entityOrigin.x - this.center.x;
+                        final double dirY = entityOrigin.y - this.center.y;
+                        final double dirZ = entityOrigin.z - this.center.z;
+                        double magnitude = Math.sqrt(dirX * dirX + dirY * dirY + dirZ * dirZ);
+
+                        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                            dist = (float) dist;
+                            magnitude = (float) magnitude;
+                        }
+
+                        final Vec3 direction;
+                        if (magnitude < 1.0e-5f && this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                            direction = Vec3.ZERO;
+                        } else {
+                            final double normalisedX = dirX / magnitude;
+                            final double normalisedY = dirY / magnitude;
+                            final double normalisedZ = dirZ / magnitude;
+                            direction = new Vec3(normalisedX, normalisedY, normalisedZ);
+                        }
+                        // Sakura end - configure server mechanics
                         boolean shouldDamageEntity = this.damageCalculator.shouldDamageEntity(this, entity);
                         float knockbackMultiplier = this.damageCalculator.getKnockbackMultiplier(entity);
                         float exposure = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : this.getBlockDensity(this.center, entity); // Paper - Optimize explosions
@@ -858,7 +926,7 @@ public class ServerExplosion implements Explosion {
         // Sakura start - replace density cache
         float blockDensity = this.level.densityCache.getBlockDensity(vec3d, entity);
         if (blockDensity == me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
-            blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+            blockDensity = this.sakura$getSeenPercent(vec3d, entity); // Sakura - configure server mechanics
             this.level.densityCache.putDensity(vec3d, entity, blockDensity);
             // Sakura end - replace density cache
         }
diff --git a/net/minecraft/world/level/block/FallingBlock.java b/net/minecraft/world/level/block/FallingBlock.java
index f42cc2807d6caf3ecf53e5352390ce9ea94e7b16..d6c5ccffdd6000c2e0364d7b0e6b4a61f16c24c2 100644
--- a/net/minecraft/world/level/block/FallingBlock.java
+++ b/net/minecraft/world/level/block/FallingBlock.java
@@ -22,6 +22,22 @@ public abstract class FallingBlock extends Block implements Fallable {
         super(properties);
     }
 
+    // Sakura start - configure server mechanics
+    @Override
+    public void neighborChanged(
+        final BlockState state,
+        final Level level,
+        final BlockPos pos,
+        final Block sourceBlock,
+        final net.minecraft.world.level.redstone.Orientation wireOrientation,
+        final boolean notify
+    ) {
+        if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.scheduleTick(pos, this, this.getDelayAfterPlace());
+        }
+    }
+    // Sakura end - configure server mechanics
+
     @Override
     protected abstract MapCodec<? extends FallingBlock> codec();
 
diff --git a/net/minecraft/world/level/block/FenceGateBlock.java b/net/minecraft/world/level/block/FenceGateBlock.java
index 63400332eb64250d457bd36140bca419d4781ccf..fdf397959741b4859c145c1b654d0888ac9271c1 100644
--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -199,8 +199,15 @@ public class FenceGateBlock extends HorizontalDirectionalBlock {
                 hasPower = eventRedstone.getNewCurrent() > 0;
             }
             // CraftBukkit end
-            if (state.getValue(POWERED) != hasPower) {
-                level.setBlock(pos, state.setValue(POWERED, hasPower).setValue(OPEN, hasPower), Block.UPDATE_CLIENTS);
+            // Sakura start - configure server mechanics
+            final boolean pre1_10_0 = level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10);
+            final boolean powered = state.getValue(POWERED);
+            if (pre1_10_0 ? (hasPower || block.defaultBlockState().isSignalSource()) : powered != hasPower) {
+                final boolean openGate = pre1_10_0 && (hasPower == powered || state.getValue(OPEN) != powered)
+                    ? state.getValue(OPEN)
+                    : hasPower;
+                level.setBlock(pos, state.setValue(POWERED, hasPower).setValue(OPEN, openGate), Block.UPDATE_CLIENTS);
+                // Sakura end - configure server mechanics
                 if (state.getValue(OPEN) != hasPower) {
                     level.playSound(
                         null,
diff --git a/net/minecraft/world/level/block/HoneyBlock.java b/net/minecraft/world/level/block/HoneyBlock.java
index e517f58719607cb98043872415c8c423c8f85faa..e27cf447981c16a1435b08e6715a9b6b516e5b5c 100644
--- a/net/minecraft/world/level/block/HoneyBlock.java
+++ b/net/minecraft/world/level/block/HoneyBlock.java
@@ -80,11 +80,19 @@ public class HoneyBlock extends HalfTransparentBlock {
         super.entityInside(state, level, pos, entity, effectApplier, isPrecise);
     }
 
-    private static double getOldDeltaY(final double deltaY) {
+    // Sakura start - configure server mechanics
+    private static double getOldDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
         return deltaY / 0.98F + 0.08;
     }
 
-    private static double getNewDeltaY(final double deltaY) {
+    private static double getNewDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
+        // Sakura end - configure server mechanics
         return (deltaY - 0.08) * 0.98F;
     }
 
@@ -93,7 +101,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             return false;
         } else if (entity.getY() > pos.getY() + 0.9375 - 1.0E-7) {
             return false;
-        } else if (getOldDeltaY(entity.getDeltaMovement().y) >= -0.08) {
+        } else if (getOldDeltaY(entity.getDeltaMovement().y, entity) >= -0.08) { // Sakura - configure server mechanics
             return false;
         } else {
             double dx = Math.abs(pos.getX() + 0.5 - entity.getX());
@@ -111,11 +119,15 @@ public class HoneyBlock extends HalfTransparentBlock {
 
     private void doSlideMovement(final Entity entity) {
         Vec3 deltaMovement = entity.getDeltaMovement();
-        if (getOldDeltaY(entity.getDeltaMovement().y) < -0.13) {
-            double horizontalReductionFactor = -0.05 / getOldDeltaY(entity.getDeltaMovement().y);
-            entity.setDeltaMovement(new Vec3(deltaMovement.x * horizontalReductionFactor, getNewDeltaY(-0.05), deltaMovement.z * horizontalReductionFactor));
+        // Sakura start - configure server mechanics
+        final double oldDeltaY = getOldDeltaY(deltaMovement.y, entity);
+        final double newDeltaY = getNewDeltaY(-0.05, entity);
+        if (oldDeltaY < -0.13) {
+            double horizontalReductionFactor = -0.05 / oldDeltaY;
+            entity.setDeltaMovement(new Vec3(deltaMovement.x * horizontalReductionFactor, newDeltaY, deltaMovement.z * horizontalReductionFactor));
         } else {
-            entity.setDeltaMovement(new Vec3(deltaMovement.x, getNewDeltaY(-0.05), deltaMovement.z));
+            entity.setDeltaMovement(new Vec3(deltaMovement.x, newDeltaY, deltaMovement.z));
+            // Sakura end - configure server mechanics
         }
 
         entity.resetFallDistance();
diff --git a/net/minecraft/world/level/block/LadderBlock.java b/net/minecraft/world/level/block/LadderBlock.java
index 9496fdb50c13ea5ce30220244164e47c395f034f..a65ae7fa22a79e0619bcbd147c1927ea4460a1df 100644
--- a/net/minecraft/world/level/block/LadderBlock.java
+++ b/net/minecraft/world/level/block/LadderBlock.java
@@ -33,6 +33,15 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
         return CODEC;
     }
 
+    // Sakura start - configure server mechanics
+    public static final Map<Direction, VoxelShape> LEGACY_SHAPES = Shapes.rotateHorizontal(Block.boxZ(16.0, 14.0, 16.0));
+
+    @Override
+    public boolean hasDynamicShape() {
+        return true;
+    }
+    // Sakura end - configure server mechanics
+
     protected LadderBlock(final BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, false));
@@ -40,6 +49,11 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     protected VoxelShape getShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
+        // Sakura start - configure server mechanics
+        if (level instanceof net.minecraft.world.level.Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return LEGACY_SHAPES.get(state.getValue(FACING));
+        }
+        // Sakura end - configure server mechanics
         return SHAPES.get(state.getValue(FACING));
     }
 
diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index 268b915a1476ce37e0e7afd6e756196d032477f7..3c8f110ab6869c8534aa684b89222caf12100abe 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -246,7 +246,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                 }
                 // Sakura end - configure fluid ticking outside the world border
                 if (level.getFluidState(neighbourPos).is(FluidTags.WATER)) {
-                    Block convertToBlock = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    // Sakura start - configure server mechanics
+                    final FluidState fluidState = state.getFluidState();
+                    final Block convertToBlock = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    if (convertToBlock == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.LegacyBlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
+                        return true;
+                    }
+                    // Sakura end - configure server mechanics
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, convertToBlock.defaultBlockState(), Block.UPDATE_ALL)) {
                         this.fizz(level, pos);
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index fea48c065cb86187be195b6e62cc26b0e9dee868..7a6087f02848d93c626e39aa26a7ab6077cf8ebe 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -332,6 +332,10 @@ public class RedStoneWireBlock extends Block {
     ) {
         if (useExperimentalEvaluator(level)) {
             new ExperimentalRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.isLegacy()) {
+            new me.samsuik.sakura.redstone.PandaRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
+            // Sakura end - configure server mechanics
         } else {
             this.evaluator.updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
         }
@@ -563,7 +567,7 @@ public class RedStoneWireBlock extends Block {
     protected InteractionResult useWithoutItem(
         final BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult
     ) {
-        if (!player.getAbilities().mayBuild) {
+        if (!player.getAbilities().mayBuild || isCross(state) && level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) { // Sakura - configure server mechanics
             return InteractionResult.PASS;
         } else {
             if (isCross(state) || isDot(state)) {
diff --git a/net/minecraft/world/level/block/piston/MovingPistonBlock.java b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
index 2b6ad989e1e1424d0283bae39e5dae33eba44923..0eb40e2d41550d328eb66ae30299fc40b0147b47 100644
--- a/net/minecraft/world/level/block/piston/MovingPistonBlock.java
+++ b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
@@ -105,6 +105,16 @@ public class MovingPistonBlock extends BaseEntityBlock {
     @Override
     protected VoxelShape getCollisionShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
         PistonMovingBlockEntity blockEntity = this.getBlockEntity(level, pos);
+        // Sakura start - configure server mechanics
+        if (blockEntity != null && level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            final VoxelShape shape = blockEntity.getCollisionShapeFromProgress(level, pos);
+            if (context.isAbove(shape, pos, false)) {
+                return shape;
+            } else {
+                return blockEntity.getMovedState().getCollisionShape(level, pos);
+            }
+        }
+        // Sakura end - configure server mechanics
         return blockEntity != null ? blockEntity.getCollisionShape(level, pos) : Shapes.empty();
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 016d56f9de0f9fa78ff12543d588304b690d4f1b..5868fd196bfb05fe126cc7340940b26e5585f3e2 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -121,6 +121,11 @@ public class PistonBaseBlock extends DirectionalBlock {
                 event = 2;
             }
 
+            // Sakura start - configure server mechanics
+            if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                level.setBlock(pos, state.setValue(PistonBaseBlock.EXTENDED, false), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE);
+            }
+            // Sakura end - configure server mechanics
             level.blockEvent(pos, this, event, direction.get3DDataValue());
         }
     }
diff --git a/net/minecraft/world/level/block/piston/PistonHeadBlock.java b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
index f8caa0b3c21766dbad73c524c1680346e4257188..5703513b1be21ccc898f6103e01f0174f3de26d9 100644
--- a/net/minecraft/world/level/block/piston/PistonHeadBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
@@ -105,6 +105,11 @@ public class PistonHeadBlock extends DirectionalBlock {
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
         BlockState base = level.getBlockState(pos.relative(state.getValue(FACING).getOpposite()));
+        // Sakura start - configure server mechanics
+        if (level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return this.isFittingBase(state, state);
+        }
+        // Sakura end - configure server mechanics
         return this.isFittingBase(state, base) || base.is(Blocks.MOVING_PISTON) && base.getValue(FACING) == state.getValue(FACING);
     }
 
@@ -118,6 +123,10 @@ public class PistonHeadBlock extends DirectionalBlock {
                 block,
                 ExperimentalRedstoneUtils.withFront(orientation, state.getValue(FACING).getOpposite())
             );
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            level.setBlock(pos, Blocks.AIR.defaultBlockState(), 19);
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.javaK4VUQu b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.javaK4VUQu
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 8b9417ca6ac57587dfce553aef61b2a663f7d292..ef00b1771ec433c9495fa5255f0dd24ac9ec0495 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -186,6 +186,11 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
+        // Sakura start - configure server mechanics
+        if (level instanceof ServerLevel serverLevel && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+            return false;
+        }
+        // Sakura end - configure server mechanics
         return state.getHeight(level, pos) >= 0.44444445F && other.is(FluidTags.WATER);
     }
 
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index d0a62bad5592c3ef8bd2ba5bc9719fad66c4edf0..9fc0ae6acf9137f20c6bb6aac9a465f9941c9e72 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -123,7 +123,13 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
-        return direction == Direction.DOWN && !other.is(FluidTags.WATER);
+        // Sakura start - configure server mechanics
+        final boolean canReplace = direction == Direction.DOWN
+            || level instanceof ServerLevel serverLevel
+            && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Before 1.13 lava could replace water
+        return canReplace && !other.is(FluidTags.WATER);
+        // Sakura end - configure server mechanics
     }
 
     @Override
