From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Nov 2023 17:03:08 +0000
Subject: [PATCH] Configure cannon physics


diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index 535cee71d3b286387e1e7c2cad5faeb32d27b68b..d591a09ba202c22e79268d3ef40a2c93ec193130 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -1774,6 +1774,13 @@ public final class CollisionUtil {
     }
 
     public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb, final List<AABB> potentialCollisions) {
+        // Sakura start - configure server mechanics
+        return performAABBCollisions(moveVector, axisalignedbb, potentialCollisions, null);
+    }
+    public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                             final List<AABB> potentialCollisions,
+                                             final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        // Sakura end - configure server mechanics
         double x = moveVector.x;
         double y = moveVector.y;
         double z = moveVector.z;
@@ -1785,7 +1792,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, potentialCollisions);
@@ -1811,9 +1818,21 @@ public final class CollisionUtil {
     public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
                                          final List<VoxelShape> voxels,
                                          final List<AABB> aabbs) {
+        // Sakura start - configure server mechanics
+        return performCollisions(moveVector, axisalignedbb, voxels, aabbs, null);
+    }
+
+    public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                         final List<VoxelShape> voxels,
+                                         final List<AABB> aabbs,
+                                         final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        if (mechanicsTarget != null && mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.PAPER) && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2, me.samsuik.sakura.mechanics.MechanicVersion.v1_20_2)) {
+            me.samsuik.sakura.mechanics.EntityBehaviour.convertVoxelsIntoAABBs(moveVector, axisalignedbb, voxels, aabbs);
+        }
+        // Sakura end - configure server mechanics
         if (voxels.isEmpty()) {
             // fast track only AABBs
-            return performAABBCollisions(moveVector, axisalignedbb, aabbs);
+            return performAABBCollisions(moveVector, axisalignedbb, aabbs, mechanicsTarget); // Sakura - configure server mechanics
         }
 
         double x = moveVector.x;
@@ -1828,7 +1847,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, aabbs);
@@ -1886,6 +1905,7 @@ public final class CollisionUtil {
     public static final int COLLISION_FLAG_CHECK_BORDER = 1 << 2;
     public static final int COLLISION_FLAG_CHECK_ONLY = 1 << 3;
     public static final int COLLISION_FLAG_ADD_TICKET = 1 << 4; // Sakura - load chunks on movement
+    public static final int COLLISION_FLAG_SKIP_EDGES = 1 << 5; // Sakura - configure server mechanics
 
     public static boolean getCollisionsForBlocksOrWorldBorder(final Level world, final Entity entity, final AABB aabb,
                                                               final List<VoxelShape> intoVoxel, final List<AABB> intoAABB,
@@ -1939,6 +1959,7 @@ public final class CollisionUtil {
 
         final boolean loadChunks = (collisionFlags & COLLISION_FLAG_LOAD_CHUNKS) != 0;
         final boolean addTicket  = (collisionFlags & COLLISION_FLAG_ADD_TICKET) != 0; // Sakura - load chunks on movement
+        final boolean skipEdges  = (collisionFlags & COLLISION_FLAG_SKIP_EDGES) != 0; // Sakura - configure server mechanics
         final ChunkSource chunkSource = world.getChunkSource();
 
         for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
@@ -1981,7 +2002,7 @@ public final class CollisionUtil {
                         continue;
                     }
 
-                    final boolean hasSpecial = ((BlockCountingChunkSection)section).moonrise$hasSpecialCollidingBlocks();
+                    final boolean hasSpecial = !skipEdges && ((BlockCountingChunkSection)section).moonrise$hasSpecialCollidingBlocks(); // Sakura - configure server mechanics
                     final int sectionAdjust = !hasSpecial ? 1 : 0;
 
                     final PalettedContainer<BlockState> blocks = section.states;
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index bfefb5031544caa59230f0073e8880c2b39ebf4d..4b4255eec2b57593e6ee2f0d9c70c2572f0e35df 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -477,6 +477,22 @@ public interface DispenseItemBehavior {
                         }
                     }
 
+                    // Sakura start - configure server mechanics
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicTarget = serverLevel.localConfig().at(blockPos).mechanicsTarget;
+                    if (!mechanicTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.VANILLA)
+                        && mechanicTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)
+                        && !mechanicTarget.isLegacy()) {
+                        event.setVelocity(event.getVelocity().setY(event.getVelocity().getY() + 0.5));
+                    }
+
+                    if (mechanicTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+                        final BlockPos originalSpawnPos = org.bukkit.craftbukkit.util.CraftVector.toBlockPos(event.getVelocity());
+                        final BlockPos spawnPos = serverLevel.dispenseRelocationHandler.relocatePosition(originalSpawnPos, blockSource.pos(), blockSource.state());
+                        final net.minecraft.world.phys.Vec3 centeredPos = net.minecraft.world.phys.Vec3.atBottomCenterOf(spawnPos);
+
+                        event.setVelocity(org.bukkit.craftbukkit.util.CraftVector.toBukkit(centeredPos));
+                    }
+                    // Sakura end - configure server mechanics
                     PrimedTnt primedTnt = new PrimedTnt(serverLevel, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), null);
                     // CraftBukkit end
                     serverLevel.addFreshEntity(primedTnt);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index d6866c115d139e8aaab508f202fc03ffa37ed96b..b5023c8bcaf4b68cc1515cd812635a2621461e22 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -728,6 +728,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.levelTickScheduler.repeatingTask(this.explosionPositions::clear, 0); // Sakura - client visibility settings
         this.levelTickScheduler.repeatingTask(this.mergeHandler::expire, 200); // Sakura - merge cannon entities
         this.levelTickScheduler.repeatingTask(this.densityCache::expire, 0); // Sakura - explosion density cache
+        this.levelTickScheduler.repeatingTask(this.dispenseRelocationHandler::clear, 0); // Sakura - configure server mechanics
     }
 
     // Paper start
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9b6db1cc0b4c70a83c2272705b8753239bd3f29d..9b4cbc136158fab3f082829f9d52eeca6ba190d4 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -588,6 +588,13 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
     }
     // Sakura end - merge cannon entities
+    // Sakura start - configure server mechanics
+    protected me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest();
+
+    public final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget() {
+        return this.mechanicsTarget;
+    }
+    // Sakura end - configure server mechanics
 
     public Entity(EntityType<?> type, Level level) {
         this.type = type;
@@ -1130,7 +1137,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     protected void checkSupportingBlock(boolean onGround, @Nullable Vec3 movement) {
-        if (onGround) {
+        if (onGround && this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_20)) { // Sakura - configure server mechanics
             AABB boundingBox = this.getBoundingBox();
             AABB aabb = new AABB(boundingBox.minX, boundingBox.minY - 1.0E-6, boundingBox.minZ, boundingBox.maxX, boundingBox.minY, boundingBox.maxZ);
             Optional<BlockPos> optional = this.level.findSupportingBlock(this, aabb);
@@ -1184,7 +1191,10 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             this.verticalCollisionBelow = false;
             this.minorHorizontalCollision = false;
         } else {
-            if (type == MoverType.PISTON) {
+            // Sakura start - configure server mechanics
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+            if (type == MoverType.PISTON && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                // Sakura end - configure server mechanics
                 // Paper start - EAR 2
                 this.activatedTick = Math.max(this.activatedTick, net.minecraft.server.MinecraftServer.currentTick + 20);
                 this.activatedImmunityTick = Math.max(this.activatedImmunityTick, net.minecraft.server.MinecraftServer.currentTick + 20);
@@ -1198,7 +1208,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             ProfilerFiller profilerFiller = Profiler.get();
             profilerFiller.push("move");
             if (this.stuckSpeedMultiplier.lengthSqr() > 1.0E-7) {
-                if (type != MoverType.PISTON) {
+                if (type != MoverType.PISTON || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_10)) { // Sakura - configure server mechanics
                     movement = movement.multiply(this.stuckSpeedMultiplier);
                 }
 
@@ -1216,10 +1226,17 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             movement = this.maybeBackOffFromEdge(movement, type);
             Vec3 vec3 = this.collide(movement);
             double d = vec3.lengthSqr();
-            if (d > 1.0E-7 || movement.lengthSqr() - d < 1.0E-7) {
-                if (this.fallDistance != 0.0 && d >= 1.0 && !this.isFallingBlock) { // Sakura - optimise cannon entity movement
-                    double min = Math.min(vec3.length(), 8.0);
-                    Vec3 vec31 = this.position().add(vec3.normalize().scale(min));
+            // Sakura start - configure server mechanics
+            if (me.samsuik.sakura.mechanics.EntityBehaviour.canMoveEntity(d, movement, mechanicsTarget)) {
+                if (this.fallDistance != 0.0 && d >= 1.0 && !this.isFallingBlock && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) { // Sakura - optimise cannon entity movement
+                    final Vec3 vec31;
+                    if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                        double min = Math.min(Math.sqrt(d), 8.0);
+                        vec31 = this.position().add(vec3.normalize().scale(min));
+                    } else {
+                        vec31 = this.position().add(vec3);
+                    }
+                    // Sakura end - configure server mechanics
                     BlockHitResult blockHitResult = this.level()
                         .clip(new ClipContext(this.position(), vec31, ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this));
                     if (blockHitResult.getType() != HitResult.Type.MISS) {
@@ -1228,9 +1245,15 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 }
 
                 Vec3 vec32 = this.position();
-                Vec3 vec33 = vec32.add(vec3);
-                this.addMovementThisTick(new Entity.Movement(vec32, vec33, movement));
-                this.setPos(vec33);
+                // Sakura start - configure server mechanics
+                if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_6)) {
+                    me.samsuik.sakura.mechanics.EntityBehaviour.changeEntityPosition(this, vec32, vec3, mechanicsTarget);
+                } else {
+                    Vec3 vec33 = vec32.add(vec3);
+                    this.addMovementThisTick(new Entity.Movement(vec32, vec33, movement));
+                    this.setPos(vec33);
+                }
+                // Sakura end - configure server mechanics
             }
 
             profilerFiller.pop();
@@ -1261,6 +1284,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             } else {
                 if (this.horizontalCollision) {
                     Vec3 deltaMovement = this.getDeltaMovement();
+                    // Sakura start - configure server mechanics
+                    if (flag && flag1 && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_14, me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                        flag = false;
+                    }
+                    // Sakura end - configure server mechanics
                     this.setDeltaMovement(flag ? 0.0 : deltaMovement.x, deltaMovement.y, flag1 ? 0.0 : deltaMovement.z);
                 }
 
@@ -1350,7 +1378,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         this.finalMovementsThisTick.clear();
         this.finalMovementsThisTick.addAll(this.movementThisTick);
         this.movementThisTick.clear();
-        if (this.finalMovementsThisTick.isEmpty()) {
+        if (this.finalMovementsThisTick.isEmpty() || this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
             this.finalMovementsThisTick.add(new Entity.Movement(this.oldPosition(), this.position()));
         } else if (this.finalMovementsThisTick.getLast().to.distanceToSqr(this.position()) > 9.9999994E-11F) {
             this.finalMovementsThisTick.add(new Entity.Movement(this.finalMovementsThisTick.getLast().to, this.position()));
@@ -1586,7 +1614,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         final List<AABB> potentialCollisionsBB = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(4);
         final AABB currBoundingBox = this.getBoundingBox();
 
-        if (movement.lengthSqr() >= 12.0) { // axis scan on large movement
+        if (movement.lengthSqr() >= 12.0 || this.mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) { // axis scan on large movement // Sakura - configure server mechanics
             return this.collideAxisScan(movement, currBoundingBox, potentialCollisionsVoxel, potentialCollisionsBB);
         } else {
             return this.collideCube(movement, currBoundingBox, potentialCollisionsVoxel, potentialCollisionsBB);
@@ -1605,8 +1633,18 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             bb = currBoundingBox.expandTowards(movement.x, movement.y, movement.z);
         }
 
-        this.collectCollisions(bb, voxelList, bbList, ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER);
-        return ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currBoundingBox, voxelList, bbList);
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+        if (mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.VANILLA)) {
+            // Fallback to the slow vanilla movement method if we are targeting vanilla mechanics.
+            // Float alignment is a bit wonky with the moonrise optimised collisions.
+            voxelList.addAll(this.level.getEntityCollisions(this, currBoundingBox));
+            return Entity.collideBoundingBox(this, movement, currBoundingBox, this.level, voxelList);
+        } else {
+            this.collectCollisions(bb, voxelList, bbList, ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER);
+            return ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currBoundingBox, voxelList, bbList, this.mechanicsTarget);
+        }
+        // Sakura end - configure server mechanics
     }
 
     private Vec3 collideAxisScan(final Vec3 movement, AABB currBoundingBox, final List<VoxelShape> voxelList, final List<AABB> bbList) {
@@ -1614,7 +1652,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         double y = movement.y;
         double z = movement.z;
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (y != 0.0) {
             y = this.scanY(currBoundingBox, y, voxelList, bbList);
@@ -1665,7 +1703,12 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     private double scanY(final AABB currBoundingBox, double y, final List<VoxelShape> voxelList, final List<AABB> bbList) {
         final AABB scanBox = cutBoundingBoxY(currBoundingBox, y);
-        this.collectCollisions(scanBox, voxelList, bbList, 0);
+        // Sakura start - configure server mechanics
+        final int flags = this.mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)
+            ? ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_SKIP_EDGES
+            : 0;
+        this.collectCollisions(scanBox, voxelList, bbList, flags);
+        // Sakura end - configure server mechanics
         y = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performAABBCollisionsY(currBoundingBox, y, bbList);
         return ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performVoxelCollisionsY(currBoundingBox, y, voxelList);
     }
@@ -1736,7 +1779,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER  | this.getExtraCollisionFlags(), null // Sakura - load chunks on movement
         );
         potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
+        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB, this.mechanicsTarget); // Sakura - configure server mechanics
 
         final boolean collidedX = collided.x != movement.x;
         final boolean collidedY = collided.y != movement.y;
@@ -1859,25 +1902,37 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     private void checkInsideBlocks(List<Entity.Movement> movements, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector) {
         if (this.isAffectedByBlocks()) {
             LongSet set = this.visitedBlocks;
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget; // Sakura - configure server mechanics
 
             for (Entity.Movement movement : movements) {
                 Vec3 vec3 = movement.from;
                 Vec3 vec31 = movement.to().subtract(movement.from());
                 int i = 16;
-                if (movement.axisDependentOriginalMovement().isPresent() && vec31.lengthSqr() > 0.0) {
-                    for (Direction.Axis axis : Direction.axisStepOrder(movement.axisDependentOriginalMovement().get())) {
+                // Sakura start - configure server mechanics
+                final Optional<Vec3> axisDependentMovement = movement.axisDependentOriginalMovement();
+                if (axisDependentMovement.isPresent() && vec31.lengthSqr() > 0.0 && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+                    final Vec3 stepOrderVec = mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)
+                        ? axisDependentMovement.get()
+                        : vec31;
+
+                    // The axis order here may differ from the order during movement in versions below 1.21.9
+                    for (Direction.Axis axis : Direction.axisStepOrder(stepOrderVec)) {
                         double d = vec31.get(axis);
                         if (d != 0.0) {
                             Vec3 vec32 = vec3.relative(axis.getPositive(), d);
-                            i -= this.checkInsideBlocks(vec3, vec32, stepBasedCollector, set, i);
+                            final int stepsTaken = this.checkInsideBlocks(vec3, vec32, stepBasedCollector, set, i);
+                            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                                i -= stepsTaken;
+                            }
                             vec3 = vec32;
                         }
                     }
-                } else {
+                } else if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
                     i -= this.checkInsideBlocks(movement.from(), movement.to(), stepBasedCollector, set, 16);
                 }
 
-                if (i <= 0) {
+                if (i <= 0 || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+                    // Sakura end - configure server mechanics
                     this.checkInsideBlocks(movement.to(), movement.to(), stepBasedCollector, set, 1);
                 }
             }
@@ -1887,8 +1942,20 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     private int checkInsideBlocks(Vec3 from, Vec3 to, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector, LongSet visited, int maxSteps) {
-        AABB aabb = this.makeBoundingBox(to).deflate(1.0E-5F);
-        boolean flag = from.distanceToSqr(to) > Mth.square(0.9999900000002526);
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+        final double margin;
+        if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            margin = 1.0e-5f;
+        } else if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_19_3)) {
+            margin = 1.0e-7f;
+        } else {
+            margin = 0.001;
+        }
+        final AABB aabb = this.makeBoundingBox(to).deflate(margin);
+        boolean flag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_10)
+            || from.distanceToSqr(to) > Mth.square(0.9999900000002526);
+        // Sakura end - configure server mechanics
         boolean flag1 = this.level instanceof ServerLevel serverLevel
             && serverLevel.getServer().debugSubscribers().hasAnySubscriberFor(DebugSubscriptions.ENTITY_BLOCK_INTERSECTIONS);
         AtomicInteger atomicInteger = new AtomicInteger();
@@ -1896,6 +1963,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             from,
             to,
             aabb,
+            mechanicsTarget, // Sakura - configure server mechanics
             (pos, index) -> {
                 if (!this.isAlive()) {
                     return false;
@@ -1915,7 +1983,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                         boolean flag2 = entityInsideCollisionShape == Shapes.block()
                             || this.collidedWithShapeMovingFrom(from, to, entityInsideCollisionShape.move(new Vec3(pos)).toAabbs());
                         boolean flag3 = this.collidedWithFluid(blockState.getFluidState(), pos, from, to);
-                        if ((flag2 || flag3) && visited.add(pos.asLong())) {
+                        if (me.samsuik.sakura.mechanics.EntityBehaviour.canCheckInsideBlock(pos, flag2 || flag3, visited, mechanicsTarget)) { // Sakura - configure server mechanics
                             if (flag2) {
                                 try {
                                     boolean flag4 = flag || aabb.intersects(pos);
@@ -1971,6 +2039,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean collidedWithShapeMovingFrom(Vec3 from, Vec3 to, List<AABB> boxes) {
+        // Sakura - configure server mechanics; this was changed in 1.21.4
         AABB aabb = this.makeBoundingBox(from);
         Vec3 vec3 = to.subtract(from);
         return aabb.collidedAlongVector(vec3, boxes);
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 3a29066910bd3b068beb56d807abf1b82a3ffa8a..f530dc91e0e4468e01b490c150ed257f4aacc499 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -129,6 +129,25 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public final double distanceToSqr(final Vec3 vector) {
+        if (!this.mechanicsTarget.isLegacy()) {
+            return super.distanceToSqr(vector);
+        }
+        final double x = this.getX() - vector.x;
+        final double y = this.getEyeY() - vector.y;
+        final double z = this.getZ() - vector.z;
+        return x * x + y * y + z * z;
+    }
+
+    private BlockPos patchedBlockPosition() {
+        // Mitigates MC-261789 - falling blocks breaking when landing on a block below y 0
+        // This fix allows some tech that uses "indirect clipping" to still function.
+        // L stackers and midairs may still encounter issues if they shoot too high.
+        return BlockPos.containing(this.getX(), this.getY() + 1.0e-12, this.getZ());
+    }
+    // Sakura end - configure server mechanics
 
     public FallingBlockEntity(EntityType<? extends FallingBlockEntity> type, Level level) {
         super(type, level);
@@ -149,12 +168,16 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         this.yo = y;
         this.zo = z;
         this.setStartPos(this.blockPosition());
+        // Sakura start - configure server mechanics
+        this.mechanicsTarget = level.localConfig().at(this.blockPosition()).mechanicsTarget;
+        this.eyeHeight = this.mechanicsTarget.isLegacy() ? 0.49f : this.eyeHeight;
+        // Sakura end - configure server mechanics
     }
 
     // Sakura start - falling block height parity api
     @Override
     public final double getEyeY() {
-        return this.heightParity ? this.getY() : super.getEyeY();
+        return this.heightParity && !this.mechanicsTarget.isLegacy() ? this.getY() : super.getEyeY(); // Sakura - configure server mechanics
     }
     // Sakura end - falling block height parity api
     // Sakura start - falling block stacking restrictions
@@ -180,7 +203,11 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             state.hasProperty(BlockStateProperties.WATERLOGGED) ? state.setValue(BlockStateProperties.WATERLOGGED, false) : state
         );
         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(fallingBlockEntity, pos, state.getFluidState().createLegacyBlock())) return fallingBlockEntity; // CraftBukkit
-        level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        // Sakura start - configure server mechanics
+        if (fallingBlockEntity.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        }
+        // Sakura end - configure server mechanics
         level.addFreshEntity(fallingBlockEntity);
         return fallingBlockEntity;
     }
@@ -224,7 +251,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -233,6 +260,12 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             Block block = this.blockState.getBlock();
+            // Sakura start - configure server mechanics
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+            if (this.time == 0 && mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                me.samsuik.sakura.mechanics.FallingBlockBehaviour.removeBlockOnFall(this, block);
+            }
+            // Sakura end - configure server mechanics
             this.time++;
             this.applyGravity();
             this.move(MoverType.SELF, this.getDeltaMovement());
@@ -247,8 +280,15 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             }
             // Paper end - Configurable falling blocks height nerf
             this.handlePortal();
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                this.setDeltaMovement(this.getDeltaMovement().scale(0.98F));
+            }
             if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
-                BlockPos blockPos = this.blockPosition();
+                BlockPos blockPos = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)
+                    ? this.patchedBlockPosition() // Mitigate MC-261789
+                    : this.blockPosition();
+                // Sakura end - configure server mechanics
                 boolean flag = this.level().sakuraConfig().cannons.sand.concreteSolidifyInWater && this.blockState.getBlock() instanceof ConcretePowderBlock; // Sakura - configure concrete solidifying in water
                 boolean flag1 = flag && this.level().getFluidState(blockPos).is(FluidTags.WATER);
                 double d = this.getDeltaMovement().lengthSqr();
@@ -275,8 +315,13 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                     }
                 } else {
                     BlockState blockState = this.level().getBlockState(blockPos);
-                    this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
-                    if (!blockState.is(Blocks.MOVING_PISTON)) {
+                    // Sakura start - configure server mechanics
+                    final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                        ? 0.7f
+                        : 0.7;
+                    this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+                    if (!blockState.is(Blocks.MOVING_PISTON) && (flag1 || me.samsuik.sakura.mechanics.FallingBlockBehaviour.isAbleToStackOnBlock(this, mechanicsTarget))) {
+                        // Sakura end - configure server mechanics
                         if (!this.cancelDrop) {
                             boolean canBeReplaced = blockState.canBeReplaced(
                                 new DirectionalPlaceContext(this.level(), blockPos, Direction.DOWN, ItemStack.EMPTY, Direction.UP)
@@ -348,7 +393,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                 }
             }
 
-            this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                    ? 0.98f
+                    : 0.98;
+                this.setDeltaMovement(this.getDeltaMovement().scale(drag));
+            }
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 63c289a15b95e89d7066d0fdd61e6eed246f0311..1a28b537f13eca936a6d25095c8f7602903d6af6 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -87,6 +87,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public final double getEyeY() {
+        return this.mechanicsTarget.isLegacy() ? super.getEyeY() : this.getY();
+    }
+
+    @Override
+    public final double distanceToSqr(final net.minecraft.world.phys.Vec3 vector) {
+        if (!this.mechanicsTarget.isLegacy()) {
+            return super.distanceToSqr(vector);
+        }
+        final double x = this.getX() - vector.x;
+        final double y = this.getEyeY() - vector.y;
+        final double z = this.getZ() - vector.z;
+        return x * x + y * y + z * z;
+    }
+    // Sakura end - configure server mechanics
 
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level level) {
         super(type, level);
@@ -112,6 +129,13 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             case Y -> this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
         }
         // Sakura end - configure cannon mechanics
+        // Sakura start - configure server mechanics
+        this.mechanicsTarget = level.localConfig().at(this.blockPosition()).mechanicsTarget;
+        this.eyeHeight = this.mechanicsTarget.isLegacy() ? 0.49f : this.eyeHeight;
+        if (this.mechanicsTarget.isLegacy()) {
+            this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
+        }
+        // Sakura end - configure server mechanics
     }
 
     // Sakura start - optimise tnt fluid state
@@ -147,7 +171,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -163,14 +187,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             return;
         }
         // Paper end - Configurable TNT height nerf
-        this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+        final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+            ? 0.98f
+            : 0.98;
+        this.setDeltaMovement(this.getDeltaMovement().scale(drag));
         if (this.onGround()) {
-            this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
+            final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                ? 0.7f
+                : 0.7;
+            this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+            // Sakura end - configure server mechanics
         }
 
         int i = this.getFuse() - 1;
         this.setFuse(i);
-        if (i <= 0) {
+        if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9) ? (i < 0) : (i <= 0)) { // Sakura - configure server mechanics
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
             //this.discard();
             this.tryToRespawnEntity(); // Sakura - merge cannon entities
@@ -209,13 +242,19 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
                 return;
             }
             // CraftBukkit end
+            // Sakura start - configure server mechanics
+            // Undocumented on the minecraft wiki but the explosion position did change in 1.10
+            final double explosionY = this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10)
+                ? this.getY() + (double) 0.49f
+                : this.getY(0.0625D);
+            // Sakura end - configure server mechanics
             this.level()
                 .explode(
                     this,
                     Explosion.getDefaultDamageSource(this.level(), this),
                     this.usedPortal ? USED_PORTAL_DAMAGE_CALCULATOR : null,
                     this.getX(),
-                    this.getY(0.0625),
+                    explosionY, // Sakura - configure server mechanics
                     this.getZ(),
                     event.getRadius(), // CraftBukkit
                     event.getFire(), // CraftBukkit
@@ -304,7 +343,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
     // Paper start - Option to prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater  && super.isPushedByFluid(); // Sakura - configure cannon mechanics
+        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater && !this.mechanicsTarget.isLegacy() && super.isPushedByFluid(); // Sakura - configure server mechanics // Sakura - configure cannon mechanics
     }
     // Paper end - Option to prevent TNT from moving in water
 }
diff --git a/net/minecraft/world/level/BlockGetter.java b/net/minecraft/world/level/BlockGetter.java
index 8a21910fe7ed61c903c4f36fced63b179ace4577..2e873467b219e7d8e6daf055cc87e6246f4ea9f8 100644
--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -210,8 +210,23 @@ public interface BlockGetter extends LevelHeightAccessor {
     }
 
     static boolean forEachBlockIntersectedBetween(Vec3 from, Vec3 to, AABB boundingBox, BlockGetter.BlockStepVisitor visitor) {
+        // Sakura start - configure server mechanics
+        return forEachBlockIntersectedBetween(from, to, boundingBox, me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest(), visitor);
+    }
+
+    static boolean forEachBlockIntersectedBetween(
+        final Vec3 from,
+        final Vec3 to,
+        final AABB boundingBox,
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget,
+        final BlockGetter.BlockStepVisitor visitor
+    ) {
         Vec3 vec3 = to.subtract(from);
-        if (vec3.lengthSqr() < Mth.square(1.0E-5F)) {
+        final double movedThreshold = mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)
+            ? 1.0E-5F
+            : 0.99999F;
+        if (vec3.lengthSqr() < Mth.square(movedThreshold) || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            // Sakura end - configure server mechanics
             for (BlockPos blockPos : BlockPos.betweenClosed(boundingBox)) {
                 if (!visitor.visit(blockPos, 0)) {
                     return false;
@@ -222,15 +237,22 @@ public interface BlockGetter extends LevelHeightAccessor {
         } else {
             LongSet set = new LongOpenHashSet();
 
-            for (BlockPos blockPos1 : BlockPos.betweenCornersInDirection(boundingBox.move(vec3.scale(-1.0)), vec3)) {
-                if (!visitor.visit(blockPos1, 0)) {
-                    return false;
-                }
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                for (BlockPos blockPos1 : BlockPos.betweenCornersInDirection(boundingBox.move(vec3.scale(-1.0)), vec3)) {
+                    if (!visitor.visit(blockPos1, 0)) {
+                        return false;
+                    }
 
-                set.add(blockPos1.asLong());
+                    set.add(blockPos1.asLong());
+                }
             }
 
             int i = addCollisionsAlongTravel(set, vec3, boundingBox, visitor);
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+                i = 0;
+            }
+            // Sakura end - configure server mechanics
             if (i < 0) {
                 return false;
             } else {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 4a856355a72e0c67f00f72e11e7a279e0d81daab..74b4cffacb62b556063631e959fe08b14354adc4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -837,6 +837,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
     public final me.samsuik.sakura.explosion.density.BlockDensityCache densityCache = new me.samsuik.sakura.explosion.density.BlockDensityCache(this); // Sakura - optimise explosion density cache
     public final me.samsuik.sakura.explosion.durable.DurableBlockManager durabilityManager = new me.samsuik.sakura.explosion.durable.DurableBlockManager(); // Sakura - explosion durable blocks
+    public final me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler dispenseRelocationHandler = new me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler(this); // Sakura - configure server mechanics
 
     protected Level(
         WritableLevelData levelData,
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index c283e3ac76170259a2a5ee2d28f12d248ad209b6..98464196cce04e5fd1115603d92f2a1c1e400459 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -378,6 +378,37 @@ public class ServerExplosion implements Explosion {
         return this.damageCalculator.getBlockExplosionResistance(this, this.level, pos, blockState, fluidState);
     }
     // Sakura end - explosion durable blocks
+    // Sakura start - configure server mechanics
+    protected final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+
+    private float sakura$getSeenPercent(final Vec3 explosionPos, final Entity entity) {
+        if (this.mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+            return this.getSeenPercentSingleRaycast(explosionPos, entity);
+        } else if (this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) {
+            return this.getSeenFraction(explosionPos, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+        } else {
+            return getSeenPercent(explosionPos, entity);
+        }
+    }
+
+    private float getSeenPercentSingleRaycast(final Vec3 explosionPos, final Entity entity) {
+        final AABB bounds = entity.getBoundingBox();
+        final double middle = 1.0 / (bounds.getYsize() * 2.0 + 1.0);
+        final Vec3 point = new Vec3(entity.getX(), entity.getY() + middle, entity.getZ());
+
+        final HitResult.Type hitResult;
+        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), point, explosionPos);
+        } else {
+            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                ? ClipContext.Block.COLLIDER
+                : ClipContext.Block.OUTLINE;
+            hitResult = entity.level().clip(new ClipContext(point, explosionPos, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+        }
+
+        return hitResult == HitResult.Type.MISS ? 1.0f : 0.0f;
+    }
+    // Sakura end - configure server mechanics
 
     public ServerExplosion(
         ServerLevel level,
@@ -402,6 +433,7 @@ public class ServerExplosion implements Explosion {
         this.yield = Double.isFinite(this.yield) ? this.yield : 0; // Paper - Don't allow infinite default yields
         // Paper end - add yield
         this.consistentExplosionRadius = level.localConfig().at(this.center).consistentExplosionRadius; // Sakura - consistent explosion radius
+        this.mechanicsTarget = source != null ? source.mechanicsTarget() : level.localConfig().at(this.center).mechanicsTarget; // Sakura - configure server mechanics
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity entity) {
@@ -431,8 +463,15 @@ public class ServerExplosion implements Explosion {
                         final float density = entity.level().densityCache.getKnownDensity(vec3);
                         if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
                             hitResult = density != 0.0f ? net.minecraft.world.phys.HitResult.Type.MISS : net.minecraft.world.phys.HitResult.Type.BLOCK;
+                            // Sakura start - configure server mechanics
+                        } else if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+                            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), vec3, explosionVector);
                         } else {
-                            hitResult = entity.level().clip(new ClipContext(vec3, explosionVector, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                                ? ClipContext.Block.COLLIDER
+                                : ClipContext.Block.OUTLINE;
+                            hitResult = entity.level().clip(new ClipContext(vec3, explosionVector, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+                            // Sakura end - configure server mechanics
                         }
                         if (hitResult == HitResult.Type.MISS) {
                             // Sakura end - replace density cache
@@ -530,6 +569,15 @@ public class ServerExplosion implements Explosion {
                 }
 
                 if (cachedBlock.outOfWorld) {
+                    // Sakura start - configure server mechanics
+                    if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                        power -= 0.22500001F;
+                        currX += incX;
+                        currY += incY;
+                        currZ += incZ;
+                        continue;
+                    }
+                    // Sakura end - configure server mechanics
                     break;
                 }
                 final BlockState iblockdata = cachedBlock.blockState;
@@ -597,7 +645,27 @@ public class ServerExplosion implements Explosion {
                     double d = Math.sqrt(entity.distanceToSqr(this.center)) / f;
                     if (!(d > 1.0)) {
                         Vec3 vec3 = entity instanceof PrimedTnt ? entity.position() : entity.getEyePosition();
-                        Vec3 vec31 = vec3.subtract(this.center).normalize();
+                        // Sakura start - configure server mechanics
+                        final double dirX = vec3.x - this.center.x;
+                        final double dirY = vec3.y - this.center.y;
+                        final double dirZ = vec3.z - this.center.z;
+                        double magnitude = Math.sqrt(dirX * dirX + dirY * dirY + dirZ * dirZ);
+
+                        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                            d = (float) d;
+                            magnitude = (float) magnitude;
+                        }
+
+                        final Vec3 vec31;
+                        if (magnitude < 1.0e-5f && this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                            vec31 = Vec3.ZERO;
+                        } else {
+                            final double normalisedX = dirX / magnitude;
+                            final double normalisedY = dirY / magnitude;
+                            final double normalisedZ = dirZ / magnitude;
+                            vec31 = new Vec3(normalisedX, normalisedY, normalisedZ);
+                        }
+                        // Sakura end - configure server mechanics
                         boolean shouldDamageEntity = this.damageCalculator.shouldDamageEntity(this, entity);
                         float knockbackMultiplier = this.damageCalculator.getKnockbackMultiplier(entity);
                         float f1 = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : this.getBlockDensity(this.center, entity); // Paper - Optimize explosions
@@ -868,7 +936,7 @@ public class ServerExplosion implements Explosion {
         // Sakura start - replace density cache
         float blockDensity = this.level.densityCache.getBlockDensity(vec3d, entity);
         if (blockDensity == me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
-            blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+            blockDensity = this.sakura$getSeenPercent(vec3d, entity); // Sakura - configure server mechanics
             this.level.densityCache.putDensity(vec3d, entity, blockDensity);
             // Sakura end - replace density cache
         }
diff --git a/net/minecraft/world/level/block/FallingBlock.java b/net/minecraft/world/level/block/FallingBlock.java
index 24b9ce38bc144febeefc2304242e96243533a01c..ecd93719347d21b9d1dbe6515e1be873515f4732 100644
--- a/net/minecraft/world/level/block/FallingBlock.java
+++ b/net/minecraft/world/level/block/FallingBlock.java
@@ -45,6 +45,22 @@ public abstract class FallingBlock extends Block implements Fallable {
         return super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
     }
 
+    // Sakura start - configure server mechanics
+    @Override
+    public void neighborChanged(
+        final BlockState state,
+        final Level level,
+        final BlockPos pos,
+        final Block sourceBlock,
+        final net.minecraft.world.level.redstone.Orientation wireOrientation,
+        final boolean notify
+    ) {
+        if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.scheduleTick(pos, this, this.getDelayAfterPlace());
+        }
+    }
+    // Sakura end - configure server mechanics
+
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (isFree(level.getBlockState(pos.below())) && pos.getY() >= level.getMinY()) {
diff --git a/net/minecraft/world/level/block/FenceGateBlock.java b/net/minecraft/world/level/block/FenceGateBlock.java
index 48aa8820761806f40761ab18496720af538bbd4d..8171f65bb2acbacddb7c8f15394d675c717080a1 100644
--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -210,8 +210,15 @@ public class FenceGateBlock extends HorizontalDirectionalBlock {
                 hasNeighborSignal = eventRedstone.getNewCurrent() > 0;
             }
             // CraftBukkit end
-            if (state.getValue(POWERED) != hasNeighborSignal) {
-                level.setBlock(pos, state.setValue(POWERED, hasNeighborSignal).setValue(OPEN, hasNeighborSignal), Block.UPDATE_CLIENTS);
+            // Sakura start - configure server mechanics
+            final boolean pre1_10_0 = level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10);
+            final boolean powered = state.getValue(POWERED);
+            if (pre1_10_0 ? (hasNeighborSignal || neighborBlock.defaultBlockState().isSignalSource()) : powered != hasNeighborSignal) {
+                final boolean openGate = pre1_10_0 && (hasNeighborSignal == powered || state.getValue(OPEN) != powered)
+                    ? state.getValue(OPEN)
+                    : hasNeighborSignal;
+                level.setBlock(pos, state.setValue(POWERED, hasNeighborSignal).setValue(OPEN, openGate), Block.UPDATE_CLIENTS);
+                // Sakura end - configure server mechanics
                 if (state.getValue(OPEN) != hasNeighborSignal) {
                     level.playSound(
                         null,
diff --git a/net/minecraft/world/level/block/HoneyBlock.java b/net/minecraft/world/level/block/HoneyBlock.java
index 66628e690453ab417e0e5805057e037e93b12754..869cf33c8ae8fab471f778b0fd0726b80189ebcf 100644
--- a/net/minecraft/world/level/block/HoneyBlock.java
+++ b/net/minecraft/world/level/block/HoneyBlock.java
@@ -72,11 +72,19 @@ public class HoneyBlock extends HalfTransparentBlock {
         super.entityInside(state, level, pos, entity, effectApplier, pastEdges);
     }
 
-    private static double getOldDeltaY(double deltaY) {
+    // Sakura start - configure server mechanics
+    private static double getOldDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
         return deltaY / 0.98F + 0.08;
     }
 
-    private static double getNewDeltaY(double deltaY) {
+    private static double getNewDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
+        // Sakura end - configure server mechanics
         return (deltaY - 0.08) * 0.98F;
     }
 
@@ -85,7 +93,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             return false;
         } else if (entity.getY() > pos.getY() + 0.9375 - 1.0E-7) {
             return false;
-        } else if (getOldDeltaY(entity.getDeltaMovement().y) >= -0.08) {
+        } else if (getOldDeltaY(entity.getDeltaMovement().y, entity) >= -0.08) { // Sakura - configure server mechanics
             return false;
         } else {
             double abs = Math.abs(pos.getX() + 0.5 - entity.getX());
@@ -103,11 +111,13 @@ public class HoneyBlock extends HalfTransparentBlock {
 
     private void doSlideMovement(Entity entity) {
         Vec3 deltaMovement = entity.getDeltaMovement();
-        if (getOldDeltaY(entity.getDeltaMovement().y) < -0.13) {
-            double d = -0.05 / getOldDeltaY(entity.getDeltaMovement().y);
-            entity.setDeltaMovement(new Vec3(deltaMovement.x * d, getNewDeltaY(-0.05), deltaMovement.z * d));
+        // Sakura start - configure server mechanics
+        if (getOldDeltaY(entity.getDeltaMovement().y, entity) < -0.13) {
+            double d = -0.05 / getOldDeltaY(entity.getDeltaMovement().y, entity);
+            entity.setDeltaMovement(new Vec3(deltaMovement.x * d, getNewDeltaY(-0.05, entity), deltaMovement.z * d));
         } else {
-            entity.setDeltaMovement(new Vec3(deltaMovement.x, getNewDeltaY(-0.05), deltaMovement.z));
+            entity.setDeltaMovement(new Vec3(deltaMovement.x, getNewDeltaY(-0.05, entity), deltaMovement.z));
+            // Sakura end - configure server mechanics
         }
 
         entity.resetFallDistance();
diff --git a/net/minecraft/world/level/block/LadderBlock.java b/net/minecraft/world/level/block/LadderBlock.java
index 349f29bc3ec49903582323a57c11577bac87a4d6..bd1cdb801724148dc77c878543093fa8ad90a6d4 100644
--- a/net/minecraft/world/level/block/LadderBlock.java
+++ b/net/minecraft/world/level/block/LadderBlock.java
@@ -33,6 +33,15 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
         return CODEC;
     }
 
+    // Sakura start - configure server mechanics
+    public static final Map<Direction, VoxelShape> LEGACY_SHAPES = Shapes.rotateHorizontal(Block.boxZ(16.0, 14.0, 16.0));
+
+    @Override
+    public boolean hasDynamicShape() {
+        return true;
+    }
+    // Sakura end - configure server mechanics
+
     protected LadderBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, false));
@@ -40,6 +49,11 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
+        // Sakura start - configure server mechanics
+        if (level instanceof net.minecraft.world.level.Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return LEGACY_SHAPES.get(state.getValue(FACING));
+        }
+        // Sakura end - configure server mechanics
         return SHAPES.get(state.getValue(FACING));
     }
 
diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index 1e53bca808849563322ec9a8b57e8965729794ad..685d4624a254789f8809e7101342a169c34acd49 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -201,7 +201,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                 }
                 // Sakura end - configure fluid ticking outside the world border
                 if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
-                    Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    // Sakura start - configure server mechanics
+                    final FluidState fluidState = state.getFluidState();
+                    final Block block = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    if (block == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.LegacyBlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
+                        return true;
+                    }
+                    // Sakura end - configure server mechanics
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState(), 3)) {
                         this.fizz(level, pos);
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 4477cb0b78092edebf3c4ac1208d72f3d48fd44a..8c88c8bd57ded43ed5268f4ec3ec156da601ee7a 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -321,6 +321,10 @@ public class RedStoneWireBlock extends Block {
     private void updatePowerStrength(Level level, BlockPos pos, BlockState state, @Nullable Orientation orientation, boolean updateShape) {
         if (useExperimentalEvaluator(level)) {
             new ExperimentalRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, updateShape);
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.isLegacy()) {
+            new me.samsuik.sakura.redstone.PandaRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, updateShape);
+            // Sakura end - configure server mechanics
         } else {
             this.evaluator.updatePowerStrength(level, pos, state, orientation, updateShape);
         }
@@ -541,7 +545,7 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
-        if (!player.getAbilities().mayBuild) {
+        if (!player.getAbilities().mayBuild || isCross(state) && level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) { // Sakura - configure server mechanics
             return InteractionResult.PASS;
         } else {
             if (isCross(state) || isDot(state)) {
diff --git a/net/minecraft/world/level/block/WaterlilyBlock.java b/net/minecraft/world/level/block/WaterlilyBlock.java
index 3e8d3e49bced11e3a1349a8354ae7111c97f155e..415f8e08f149ec92d99a8412c6f325e89d80b35d 100644
--- a/net/minecraft/world/level/block/WaterlilyBlock.java
+++ b/net/minecraft/world/level/block/WaterlilyBlock.java
@@ -24,6 +24,15 @@ public class WaterlilyBlock extends VegetationBlock {
         return CODEC;
     }
 
+    // Sakura start - configure server mechanics
+    protected static final VoxelShape LEGACY_SHAPE = Block.column(16.0, 0.0, 0.25);
+
+    @Override
+    public boolean hasDynamicShape() {
+        return true;
+    }
+    // Sakura end - configure server mechanics
+
     protected WaterlilyBlock(BlockBehaviour.Properties properties) {
         super(properties);
     }
@@ -44,6 +53,11 @@ public class WaterlilyBlock extends VegetationBlock {
 
     @Override
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
+        // Sakura start - configure server mechanics
+        if (level instanceof net.minecraft.world.level.Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return LEGACY_SHAPE;
+        }
+        // Sakura end - configure server mechanics
         return SHAPE;
     }
 
diff --git a/net/minecraft/world/level/block/piston/MovingPistonBlock.java b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
index 58b78e6b9a43605940420d8cb0f1163ee3f0a5ff..c6cffeb660983c9602599f4b8fa375e900979ecc 100644
--- a/net/minecraft/world/level/block/piston/MovingPistonBlock.java
+++ b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
@@ -98,6 +98,16 @@ public class MovingPistonBlock extends BaseEntityBlock {
     @Override
     protected VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
         PistonMovingBlockEntity blockEntity = this.getBlockEntity(level, pos);
+        // Sakura start - configure server mechanics
+        if (blockEntity != null && level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            final VoxelShape shape = blockEntity.getCollisionShapeFromProgress(level, pos);
+            if (context.isAbove(shape, pos, false)) {
+                return shape;
+            } else {
+                return blockEntity.getMovedState().getCollisionShape(level, pos);
+            }
+        }
+        // Sakura end - configure server mechanics
         return blockEntity != null ? blockEntity.getCollisionShape(level, pos) : Shapes.empty();
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 82e3b145bafb4bc7353244e8f058763cf0bdd03d..214a036f946aab660bd1211855741cea61e9bf03 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -123,6 +123,11 @@ public class PistonBaseBlock extends DirectionalBlock {
                 i = 2;
             }
 
+            // Sakura start - configure server mechanics
+            if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                level.setBlock(pos, state.setValue(PistonBaseBlock.EXTENDED, false), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE);
+            }
+            // Sakura end - configure server mechanics
             level.blockEvent(pos, this, i, direction.get3DDataValue());
         }
     }
diff --git a/net/minecraft/world/level/block/piston/PistonHeadBlock.java b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
index b369e80a4fc371a3762e34e7d4d180b86d1d190f..71e45da209c5331fefb683aef9ba2f30d39fda46 100644
--- a/net/minecraft/world/level/block/piston/PistonHeadBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
@@ -105,6 +105,11 @@ public class PistonHeadBlock extends DirectionalBlock {
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
         BlockState blockState = level.getBlockState(pos.relative(state.getValue(FACING).getOpposite()));
+        // Sakura start - configure server mechanics
+        if (level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return this.isFittingBase(state, blockState);
+        }
+        // Sakura end - configure server mechanics
         return this.isFittingBase(state, blockState) || blockState.is(Blocks.MOVING_PISTON) && blockState.getValue(FACING) == state.getValue(FACING);
     }
 
@@ -116,6 +121,10 @@ public class PistonHeadBlock extends DirectionalBlock {
                 neighborBlock,
                 ExperimentalRedstoneUtils.withFront(orientation, state.getValue(FACING).getOpposite())
             );
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            level.setBlock(pos, Blocks.AIR.defaultBlockState(), 19);
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index 9383e91846e1119dc11259235f586e9e25c7e51a..6973d307a250d8ef1632fa630a0f8a12e08bf926 100644
--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -64,6 +64,165 @@ public class PistonMovingBlockEntity extends BlockEntity {
         this.isSourcePiston = isSourcePiston;
     }
 
+    // Sakura start - configure server mechanics
+    @org.jspecify.annotations.Nullable
+    private AABB getBoundsFromProgress(
+        final BlockGetter level,
+        final BlockPos pos,
+        final BlockState state,
+        final float progress,
+        final Direction dir,
+        final boolean absolute
+    ) {
+        if (state.is(Blocks.MOVING_PISTON) || state.isAir()) {
+            return null;
+        }
+
+        VoxelShape shape = this.movedState.getCollisionShape(level, pos);
+        if (shape.isEmpty()) {
+            return null;
+        }
+
+        if (absolute) {
+            shape = shape.move(pos.getX(), pos.getY(), pos.getZ());
+        }
+
+        final AABB bounds = shape.bounds();
+        double minX = bounds.minX;
+        double minY = bounds.minY;
+        double minZ = bounds.minZ;
+        double maxX = bounds.maxX;
+        double maxY = bounds.maxY;
+        double maxZ = bounds.maxZ;
+
+        if (dir.getStepX() < 0) {
+            minX -= (float) dir.getStepX() * progress;
+        } else {
+            maxX -= (float) dir.getStepX() * progress;
+        }
+
+        if (dir.getStepY() < 0) {
+            minY -= (float) dir.getStepY() * progress;
+        } else {
+            maxY -= (float) dir.getStepY() * progress;
+        }
+
+        if (dir.getStepZ() < 0) {
+            minZ -= (float) dir.getStepZ() * progress;
+        } else {
+            maxZ -= (float) dir.getStepZ() * progress;
+        }
+
+        return this.fixZeroWidthBB(new AABB(minX, minY, minZ, maxX, maxY, maxZ), dir);
+    }
+
+    private AABB fixZeroWidthBB(AABB bb, final Direction dir) {
+        // Legacy behaviour relied on entities being able to collide with zero width shapes
+        // This is no longer possible, so we have to create a difference here for it to work
+        final double expandX = bb.getXsize() == 0.0 ? 1.0e-5 * dir.getStepX() : 0.0;
+        final double expandY = bb.getYsize() == 0.0 ? 1.0e-5 * dir.getStepY() : 0.0;
+        final double expandZ = bb.getZsize() == 0.0 ? 1.0e-5 * dir.getStepZ() : 0.0;
+
+        if (expandX != 0 || expandY != 0 || expandZ != 0) {
+            bb = bb.expandTowards(expandX, expandY, expandZ);
+        }
+
+        return bb;
+    }
+
+    public final VoxelShape getCollisionShapeFromProgress(final BlockGetter level, final BlockPos pos) {
+        float progress = this.getProgress(0.0f);
+        if (this.extending) {
+            progress = 1.0F - progress;
+        }
+
+        final AABB bb = this.getBoundsFromProgress(level, pos, this.movedState, progress, this.direction, false);
+        return bb == null ? Shapes.empty() : Shapes.create(bb);
+    }
+
+    private void moveEntities(final Level level, final float relativeProgress) {
+        float remaining = this.progress;
+        if (this.extending) {
+            remaining = 1.0F - remaining;
+        } else {
+            --remaining;
+        }
+
+        final Direction direction = this.direction;
+        final AABB bb = this.getBoundsFromProgress(level, this.worldPosition, this.movedState, remaining, direction, true);
+        if (bb == null || bb.getSize() == 0.0) {
+            return;
+        }
+
+        for (final Entity entity : level.getEntities(null, bb)) {
+            if (this.movedState.is(Blocks.SLIME_BLOCK) && this.extending) {
+                final Vec3 movement = entity.getDeltaMovement();
+                double moveX = movement.x;
+                double moveY = movement.y;
+                double moveZ = movement.z;
+
+                switch (direction.getAxis()) {
+                    case X -> moveX = direction.getStepX();
+                    case Y -> moveY = direction.getStepY();
+                    case Z -> moveZ = direction.getStepZ();
+                }
+
+                entity.setDeltaMovement(moveX, moveY, moveZ);
+            } else {
+                final Vec3 pistonMovement = new Vec3(
+                    relativeProgress * (float) direction.getStepX(),
+                    relativeProgress * (float) direction.getStepY(),
+                    relativeProgress * (float) direction.getStepZ()
+                );
+                entity.move(MoverType.PISTON, pistonMovement);
+                entity.applyEffectsFromBlocks(entity.position(), entity.position());
+            }
+        }
+    }
+
+    private static void moveEntityByPistonInDirection(final Direction direction, final Entity entity, final AABB movingBB) {
+        final AABB entityBB = entity.getBoundingBox();
+        double moveX = 0.0;
+        double moveY = 0.0;
+        double moveZ = 0.0;
+
+        switch (direction.getAxis()) {
+            case X -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveX = movingBB.maxX - entityBB.minX;
+                } else {
+                    moveX = entityBB.maxX - movingBB.minX;
+                }
+                moveX += 0.01D;
+            }
+            case Y -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveY = movingBB.maxY - entityBB.minY;
+                } else {
+                    moveY = entityBB.maxY - movingBB.minY;
+                }
+                moveY += 0.01D;
+            }
+            case Z -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveZ = movingBB.maxZ - entityBB.minZ;
+                } else {
+                    moveZ = entityBB.maxZ - movingBB.minZ;
+                }
+                moveZ += 0.01D;
+            }
+        }
+
+        final Vec3 pistonMovement = new Vec3(
+            moveX * direction.getStepX(),
+            moveY * direction.getStepY(),
+            moveZ * direction.getStepZ()
+        );
+        entity.move(MoverType.PISTON, pistonMovement);
+        entity.applyEffectsFromBlocks(entity.position(), entity.position());
+    }
+    // Sakura end - configure server mechanics
+
     @Override
     public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
         return this.saveCustomOnly(registries);
@@ -168,6 +327,12 @@ public class PistonMovingBlockEntity extends BlockEntity {
 
                     double d4 = 0.0;
 
+                    // Sakura start - configure server mechanics
+                    if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                        moveEntityByPistonInDirection(movementDirection, entity, aabb);
+                        return;
+                    }
+                    // Sakura end - configure server mechanics
                     for (AABB aabb1 : list) {
                         AABB movementArea = PistonMath.getMovementArea(moveByPositionAndProgress(pos, aabb1, piston), movementDirection, d);
                         AABB boundingBox = entity.getBoundingBox();
@@ -195,6 +360,11 @@ public class PistonMovingBlockEntity extends BlockEntity {
         NOCLIP.set(noClipDirection);
         Vec3 vec3 = entity.position();
         entity.move(MoverType.PISTON, new Vec3(progress * direction.getStepX(), progress * direction.getStepY(), progress * direction.getStepZ()));
+        // Sakura start - configure server mechanics
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+            vec3 = entity.oldPosition();
+        }
+        // Sakura end - configure server mechanics
         entity.applyEffectsFromBlocks(vec3, entity.position());
         entity.removeLatestMovementRecording();
         NOCLIP.set(null);
@@ -307,12 +477,21 @@ public class PistonMovingBlockEntity extends BlockEntity {
     }
 
     public static void tick(Level level, BlockPos pos, BlockState state, PistonMovingBlockEntity blockEntity) {
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget; // Sakura - configure server mechanics
         blockEntity.lastTicked = level.getGameTime();
         blockEntity.progressO = blockEntity.progress;
         if (blockEntity.progressO >= 1.0F) {
             if (level.isClientSide() && blockEntity.deathTicks < 5) {
                 blockEntity.deathTicks++;
             } else {
+                // Sakura start - configure server mechanics
+                if (mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_9, me.samsuik.sakura.mechanics.MechanicVersion.v1_10)) {
+                    moveCollidedEntities(level, pos, 1.0f, blockEntity);
+                    moveStuckEntities(level, pos, 1.0f, blockEntity);
+                } else if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                    blockEntity.moveEntities(level, 0.25f);
+                }
+                // Sakura end - configure server mechanics
                 level.removeBlockEntity(pos);
                 blockEntity.setRemoved();
                 if (level.getBlockState(pos).is(Blocks.MOVING_PISTON)) {
@@ -334,12 +513,22 @@ public class PistonMovingBlockEntity extends BlockEntity {
             }
         } else {
             float f = blockEntity.progress + 0.5F;
-            moveCollidedEntities(level, pos, f, blockEntity);
-            moveStuckEntities(level, pos, f, blockEntity);
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                moveCollidedEntities(level, pos, f, blockEntity);
+                moveStuckEntities(level, pos, f, blockEntity);
+            }
             blockEntity.progress = f;
             if (blockEntity.progress >= 1.0F) {
                 blockEntity.progress = 1.0F;
             }
+            if (mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_9, me.samsuik.sakura.mechanics.MechanicVersion.v1_10)) {
+                moveCollidedEntities(level, pos, f, blockEntity);
+                moveStuckEntities(level, pos, f, blockEntity);
+            } else if (blockEntity.extending && mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                blockEntity.moveEntities(level, blockEntity.progress - blockEntity.progressO + 0.0625f);
+            }
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 44ad677901e5ce09a753ae1f3e4081a7bff8b521..914d72254aaf66e03c9cea3cb332141453a555bb 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -185,6 +185,11 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter level, BlockPos pos, Fluid fluid, Direction direction) {
+        // Sakura start - configure server mechanics
+        if (level instanceof ServerLevel serverLevel && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+            return false;
+        }
+        // Sakura end - configure server mechanics
         return fluidState.getHeight(level, pos) >= 0.44444445F && fluid.is(FluidTags.WATER);
     }
 
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index 45ade22e8f69609598725a94a0581251cf3f384f..5f91f490c927f16071fc866efb1cd58ee18b3632 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -123,7 +123,13 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter level, BlockPos pos, Fluid fluid, Direction direction) {
-        return direction == Direction.DOWN && !fluid.is(FluidTags.WATER);
+        // Sakura start - configure server mechanics
+        final boolean canReplace = direction == Direction.DOWN
+            || level instanceof ServerLevel serverLevel
+            && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Before 1.13 lava could replace water
+        return canReplace && !fluid.is(FluidTags.WATER);
+        // Sakura end - configure server mechanics
     }
 
     @Override
