From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 2 Dec 2023 15:14:15 +0000
Subject: [PATCH] Legacy lava block formation


diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index f7d122b1a8deee9892735df05a1873b5a7e03f4f..87ce9be33c28735a9137181fde1f8c01871c1608 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -229,7 +229,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                     // Sakura start - configure server mechanics
                     final FluidState fluidState = state.getFluidState();
                     final Block convertToBlock = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
-                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    // Sakura start - legacy lava block formation
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+                    if (level.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                        mechanicsTarget = me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.legacy();
+                    } else {
+                        mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    }
+                    // Sakura end - legacy lava block formation
                     if (convertToBlock == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.BlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
                         return true;
                     }
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 50eb2bf672aa04642be85d26491070135b4a0fe9..f386f7dc322dd189c465e4526996568b7e30a58b 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -186,9 +186,16 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
-        // Sakura start - configure server mechanics
-        if (level instanceof ServerLevel serverLevel && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
-            return false;
+        // Sakura start - configure server mechanics & legacy lava block formation
+        if (level instanceof ServerLevel serverLevel) {
+            if (serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                return false;
+            }
+
+            if (serverLevel.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                return false;
+            }
+            // Sakura end - legacy lava block formation
         }
         // Sakura end - configure server mechanics
         return state.getHeight(level, pos) >= 0.44444445F && other.is(FluidTags.WATER);
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index 9fc0ae6acf9137f20c6bb6aac9a465f9941c9e72..6ab44418ac6e1a1acf49ab13cbec5d88e9d87f11 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -123,10 +123,18 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
-        // Sakura start - configure server mechanics
-        final boolean canReplace = direction == Direction.DOWN
-            || level instanceof ServerLevel serverLevel
-            && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Sakura start - configure server mechanics & legacy lava block formation
+        boolean canReplace = false;
+        if (direction == Direction.DOWN) {
+            canReplace = true;
+        } else if (level instanceof ServerLevel serverLevel) {
+            if (serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                canReplace = true;
+            } else if (serverLevel.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                canReplace = true;
+            }
+        }
+        // Sakura end - legacy lava block formation
         // Before 1.13 lava could replace water
         return canReplace && !other.is(FluidTags.WATER);
         // Sakura end - configure server mechanics
