From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Nov 2023 17:03:08 +0000
Subject: [PATCH] Configure server mechanics


diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index c093f2295d6a0e1880897fe66bdcd64649aa667e..0cb1906c33589592ab42bef31569d55b5ffb9fd4 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -1774,6 +1774,13 @@ public final class CollisionUtil {
     }
 
     public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb, final List<AABB> potentialCollisions) {
+        // Sakura start - configure server mechanics
+        return performAABBCollisions(moveVector, axisalignedbb, potentialCollisions, null);
+    }
+    public static Vec3 performAABBCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                             final List<AABB> potentialCollisions,
+                                             final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        // Sakura end - configure server mechanics
         double x = moveVector.x;
         double y = moveVector.y;
         double z = moveVector.z;
@@ -1785,7 +1792,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, potentialCollisions);
@@ -1811,9 +1818,21 @@ public final class CollisionUtil {
     public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
                                          final List<VoxelShape> voxels,
                                          final List<AABB> aabbs) {
+        // Sakura start - configure server mechanics
+        return performCollisions(moveVector, axisalignedbb, voxels, aabbs, null);
+    }
+
+    public static Vec3 performCollisions(final Vec3 moveVector, AABB axisalignedbb,
+                                         final List<VoxelShape> voxels,
+                                         final List<AABB> aabbs,
+                                         final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget) {
+        if (mechanicsTarget != null && mechanicsTarget.isPaperOrDerivative() && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2, me.samsuik.sakura.mechanics.MechanicVersion.v1_20_2)) {
+            me.samsuik.sakura.mechanics.EntityBehaviour.convertVoxelsIntoAABBs(moveVector, axisalignedbb, voxels, aabbs);
+        }
+        // Sakura end - configure server mechanics
         if (voxels.isEmpty()) {
             // fast track only AABBs
-            return performAABBCollisions(moveVector, axisalignedbb, aabbs);
+            return performAABBCollisions(moveVector, axisalignedbb, aabbs, mechanicsTarget); // Sakura - configure server mechanics
         }
 
         double x = moveVector.x;
@@ -1828,7 +1847,7 @@ public final class CollisionUtil {
             }
         }
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, mechanicsTarget); // Sakura - configure server mechanics
 
         if (xSmaller && z != 0.0) {
             z = performAABBCollisionsZ(axisalignedbb, z, aabbs);
@@ -1937,6 +1956,13 @@ public final class CollisionUtil {
         final int minChunkZ = minBlockZ >> 4;
         final int maxChunkZ = maxBlockZ >> 4;
 
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = entity != null
+            ? entity.mechanicsTarget()
+            : me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest();
+        final boolean brokenFloatAlignment = mechanicsTarget.isPaperOrDerivative() && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2, me.samsuik.sakura.mechanics.MechanicVersion.v1_21_11);
+        // Sakura end - configure server mechanics
+
         final boolean loadChunks = (collisionFlags & COLLISION_FLAG_LOAD_CHUNKS) != 0;
         final boolean addTicket  = (collisionFlags & COLLISION_FLAG_ADD_TICKET) != 0; // Sakura - load chunks on movement
         final ChunkSource chunkSource = world.getChunkSource();
@@ -2030,7 +2056,7 @@ public final class CollisionUtil {
 
                                         // Not using the epsilon violates the collision rules but this is what Vanilla does.
                                         // Check BlockCollisions.class
-                                        if (blockCollision == Shapes.block()) {
+                                        if (!brokenFloatAlignment && blockCollision == Shapes.block()) { // Sakura - configure server mechanics
                                             if (!aabb.intersects(singleAABB)) {
                                                 continue;
                                             }
diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index 4c456c2d1b42ecfcecb22ba2bf3819cab68206b8..7d2df500c90e0451ec2cf117f44413b3593fdd62 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -11,5 +11,6 @@ public final class ScheduleSakuraTasks {
         scheduler.endless((tick, level, cb) -> level.explosionPositions.clear(), 0); // Sakura - client visibility settings
         scheduler.endless((tick, level, cb) -> level.mergeHandler.expire(tick), 200); // Sakura - merge cannon entities
         scheduler.endless((tick, level, cb) -> level.densityCache.expire(tick), 0); // Sakura - explosion density cache
+        scheduler.endless((tick, level, cb) -> level.dispenseRelocationHandler.clear(), 0); // Sakura - configure server mechanics
     }
 }
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 751232d09f5570a23223356a6572090e7d61139b..b83a62296008e8fa3452f2bb0d03701b0cf03b0d 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -407,6 +407,22 @@ public interface DispenseItemBehavior {
                         }
                     }
 
+                    // Sakura start - configure server mechanics
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicTarget = level.localConfig().at(target).mechanicsTarget;
+                    if (!mechanicTarget.isVanilla()
+                        && mechanicTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)
+                        && !mechanicTarget.isLegacy()) {
+                        event.setVelocity(event.getVelocity().setY(event.getVelocity().getY() + 0.5));
+                    }
+
+                    if (mechanicTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+                        final BlockPos originalSpawnPos = org.bukkit.craftbukkit.util.CraftVector.toBlockPos(event.getVelocity());
+                        final BlockPos spawnPos = level.dispenseRelocationHandler.relocatePosition(originalSpawnPos, source.pos(), source.state());
+                        final net.minecraft.world.phys.Vec3 centeredPos = net.minecraft.world.phys.Vec3.atBottomCenterOf(spawnPos);
+
+                        event.setVelocity(org.bukkit.craftbukkit.util.CraftVector.toBukkit(centeredPos));
+                    }
+                    // Sakura end - configure server mechanics
                     PrimedTnt tnt = new PrimedTnt(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), null);
                     // CraftBukkit end
                     level.addFreshEntity(tnt);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 66b50348888fa3c32f6ab0d0167b7cc6336ee3ef..3f4b0daf98c5c2bdd35420b18d987ccdf46bbacd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -602,6 +602,17 @@ public abstract class Entity
         }
     }
     // Sakura end - merge cannon entities
+    // Sakura start - configure server mechanics
+    private me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest();
+
+    public final void updateMechanicsTarget() {
+        this.mechanicsTarget = this.level.localConfig().at(this.blockPosition).mechanicsTarget;
+    }
+
+    public final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget() {
+        return this.mechanicsTarget;
+    }
+    // Sakura end - configure server mechanics
 
     public Entity(final EntityType<?> type, final Level level) {
         this.type = type;
@@ -1149,7 +1160,7 @@ public abstract class Entity
     }
 
     protected void checkSupportingBlock(final boolean onGround, final @Nullable Vec3 movement) {
-        if (onGround) {
+        if (onGround && this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_20)) { // Sakura - configure server mechanics
             AABB boundingBox = this.getBoundingBox();
             AABB testArea = new AABB(boundingBox.minX, boundingBox.minY - 1.0E-6, boundingBox.minZ, boundingBox.maxX, boundingBox.minY, boundingBox.maxZ);
             Optional<BlockPos> supportingBlock = this.level.findSupportingBlock(this, testArea);
@@ -1203,7 +1214,10 @@ public abstract class Entity
             this.verticalCollisionBelow = false;
             this.minorHorizontalCollision = false;
         } else {
-            if (moverType == MoverType.PISTON) {
+            // Sakura start - configure server mechanics
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+            if (moverType == MoverType.PISTON && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                // Sakura end - configure server mechanics
                 delta = this.limitPistonMovement(delta);
                 // Paper start - EAR 2
                 this.activatedTick = Math.max(this.activatedTick, net.minecraft.server.MinecraftServer.currentTick + 20);
@@ -1217,7 +1231,7 @@ public abstract class Entity
             ProfilerFiller profiler = Profiler.get();
             profiler.push("move");
             if (this.stuckSpeedMultiplier.lengthSqr() > 1.0E-7) {
-                if (moverType != MoverType.PISTON) {
+                if (moverType != MoverType.PISTON || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_10)) { // Sakura - configure server mechanics
                     delta = delta.multiply(this.stuckSpeedMultiplier);
                 }
 
@@ -1235,10 +1249,22 @@ public abstract class Entity
             delta = this.maybeBackOffFromEdge(delta, moverType);
             Vec3 movement = this.collide(delta);
             double movementLength = movement.lengthSqr();
-            if (movementLength > 1.0E-7 || delta.lengthSqr() - movementLength < 1.0E-7) {
-                if (this.fallDistance != 0.0 && movementLength >= 1.0) {
-                    double checkDistance = Math.min(movement.length(), 8.0);
-                    Vec3 checkTo = this.position().add(movement.normalize().scale(checkDistance));
+            // Sakura start - configure server mechanics
+            if (me.samsuik.sakura.mechanics.EntityBehaviour.canMoveEntity(movementLength, movement, mechanicsTarget)) {
+                if (this.fallDistance != 0.0 && movementLength >= 1.0 && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                    final Vec3 checkTo;
+                    if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                        final double movementDist = Math.sqrt(movementLength);
+                        final double checkDistance = Math.min(movementDist, 8.0);
+                        final double checkX = (movement.x() / movementDist) * checkDistance;
+                        final double checkY = (movement.y() / movementDist) * checkDistance;
+                        final double checkZ = (movement.z() / movementDist) * checkDistance;
+
+                        checkTo = this.position().add(checkX, checkY, checkZ);
+                    } else {
+                        checkTo = this.position().add(movement);
+                    }
+                    // Sakura end - configure server mechanics
                     BlockHitResult hitResult = this.level()
                         .clip(new ClipContext(this.position(), checkTo, ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this));
                     if (hitResult.getType() != HitResult.Type.MISS) {
@@ -1247,9 +1273,15 @@ public abstract class Entity
                 }
 
                 Vec3 pos = this.position();
-                Vec3 newPosition = pos.add(movement);
-                this.addMovementThisTick(new Entity.Movement(pos, newPosition, delta));
-                this.setPos(newPosition);
+                // Sakura start - configure server mechanics
+                if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_6)) {
+                    me.samsuik.sakura.mechanics.EntityBehaviour.changeEntityPosition(this, pos, delta, mechanicsTarget);
+                } else {
+                    Vec3 newPosition = pos.add(movement);
+                    this.addMovementThisTick(new Entity.Movement(pos, newPosition, delta));
+                    this.setPos(newPosition);
+                }
+                // Sakura end - configure server mechanics
             }
 
             profiler.pop();
@@ -1280,6 +1312,11 @@ public abstract class Entity
             } else {
                 if (this.horizontalCollision) {
                     Vec3 vec3 = this.getDeltaMovement();
+                    // Sakura start - configure server mechanics
+                    if (xCollision && zCollision && mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_14, me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                        xCollision = false;
+                    }
+                    // Sakura end - configure server mechanics
                     this.setDeltaMovement(xCollision ? 0.0 : vec3.x, vec3.y, zCollision ? 0.0 : vec3.z);
                 }
 
@@ -1377,7 +1414,7 @@ public abstract class Entity
         this.finalMovementsThisTick.clear();
         this.finalMovementsThisTick.addAll(this.movementThisTick);
         this.movementThisTick.clear();
-        if (this.finalMovementsThisTick.isEmpty()) {
+        if (this.finalMovementsThisTick.isEmpty() || this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) { // Sakura - configure server mechanics
             this.finalMovementsThisTick.add(new Entity.Movement(this.oldPosition(), this.position()));
         } else if (this.finalMovementsThisTick.getLast().to.distanceToSqr(this.position()) > 9.9999994E-11F) {
             this.finalMovementsThisTick.add(new Entity.Movement(this.finalMovementsThisTick.getLast().to, this.position()));
@@ -1659,7 +1696,7 @@ public abstract class Entity
         double y = movement.y;
         double z = movement.z;
 
-        final boolean xSmaller = Math.abs(x) < Math.abs(z);
+        final boolean xSmaller = me.samsuik.sakura.mechanics.EntityBehaviour.prioritiseXFirst(x, z, this.mechanicsTarget); // Sakura - configure server mechanics
         if (y != 0.0) {
             y = this.scanY(currentBox, y, collisionsVoxel, collisionsBB);
             if (y != 0.0) {
@@ -1788,7 +1825,7 @@ public abstract class Entity
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER  | this.getExtraCollisionFlags(), null // Sakura - load chunks on movement
         );
         potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
+        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB, this.mechanicsTarget); // Sakura - configure server mechanics
 
         final boolean collidedX = collided.x != movement.x;
         final boolean collidedY = collided.y != movement.y;
@@ -1917,25 +1954,37 @@ public abstract class Entity
     private void checkInsideBlocks(final List<Entity.Movement> movements, final InsideBlockEffectApplier.StepBasedCollector effectCollector) {
         if (this.isAffectedByBlocks()) {
             LongSet visitedBlocks = this.visitedBlocks;
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget; // Sakura - configure server mechanics
 
             for (Entity.Movement movement : movements) {
                 Vec3 pos = movement.from;
                 Vec3 delta = movement.to().subtract(movement.from());
                 int maxMovementIterations = 16;
-                if (movement.axisDependentOriginalMovement().isPresent() && delta.lengthSqr() > 0.0) {
-                    for (Direction.Axis axis : Direction.axisStepOrder(movement.axisDependentOriginalMovement().get())) {
+                // Sakura start - configure server mechanics
+                final Optional<Vec3> axisDependentMovement = movement.axisDependentOriginalMovement();
+                if (axisDependentMovement.isPresent() && delta.lengthSqr() > 0.0 && mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+                    final Vec3 stepOrderVec = mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)
+                        ? axisDependentMovement.get()
+                        : delta;
+
+                    // The axis order here may differ from the order during movement in versions below 1.21.9
+                    for (Direction.Axis axis : Direction.axisStepOrder(stepOrderVec)) {
                         double axisMove = delta.get(axis);
                         if (axisMove != 0.0) {
                             Vec3 to = pos.relative(axis.getPositive(), axisMove);
-                            maxMovementIterations -= this.checkInsideBlocks(pos, to, effectCollector, visitedBlocks, maxMovementIterations);
+                            final int stepsTaken = this.checkInsideBlocks(pos, to, effectCollector, visitedBlocks, maxMovementIterations);
+                            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                                maxMovementIterations -= stepsTaken;
+                            }
                             pos = to;
                         }
                     }
-                } else {
+                } else if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
                     maxMovementIterations -= this.checkInsideBlocks(movement.from(), movement.to(), effectCollector, visitedBlocks, 16);
                 }
 
-                if (maxMovementIterations <= 0) {
+                if (maxMovementIterations <= 0 || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+                    // Sakura end - configure server mechanics
                     this.checkInsideBlocks(movement.to(), movement.to(), effectCollector, visitedBlocks, 1);
                 }
             }
@@ -1951,8 +2000,20 @@ public abstract class Entity
         final LongSet visitedBlocks,
         final int maxMovementIterations
     ) {
-        AABB deflatedBoundingBoxAtTarget = this.makeBoundingBox(to).deflate(1.0E-5F);
-        boolean movedFar = from.distanceToSqr(to) > Mth.square(0.9999900000002526);
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget;
+        final double margin;
+        if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            margin = 1.0e-5f;
+        } else if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_19_3)) {
+            margin = 1.0e-7f;
+        } else {
+            margin = 0.001;
+        }
+        AABB deflatedBoundingBoxAtTarget = this.makeBoundingBox(to).deflate(margin);
+        boolean movedFar = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_10)
+            || from.distanceToSqr(to) > Mth.square(0.9999900000002526);
+        // Sakura end - configure server mechanics
         boolean debugEntityBlockIntersections = this.level instanceof ServerLevel serverLevel
             && serverLevel.getServer().debugSubscribers().hasAnySubscriberFor(DebugSubscriptions.ENTITY_BLOCK_INTERSECTIONS);
         AtomicInteger iterations = new AtomicInteger();
@@ -1960,6 +2021,7 @@ public abstract class Entity
             from,
             to,
             deflatedBoundingBoxAtTarget,
+            mechanicsTarget, // Sakura - configure server mechanics
             (blockIntersection, iteration) -> {
                 if (!this.isAlive()) {
                     return false;
@@ -1979,7 +2041,7 @@ public abstract class Entity
                         boolean insideBlock = intersectShape == Shapes.block()
                             || this.collidedWithShapeMovingFrom(from, to, intersectShape.move(new Vec3(blockIntersection)).toAabbs());
                         boolean insideFluid = this.collidedWithFluid(state.getFluidState(), blockIntersection, from, to);
-                        if ((insideBlock || insideFluid) && visitedBlocks.add(blockIntersection.asLong())) {
+                        if (me.samsuik.sakura.mechanics.EntityBehaviour.canCheckInsideBlock(blockIntersection, insideBlock || insideFluid, visitedBlocks, mechanicsTarget)) { // Sakura - configure server mechanics
                             if (insideBlock) {
                                 try {
                                     boolean isPrecise = movedFar || deflatedBoundingBoxAtTarget.intersects(blockIntersection);
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index dfdec5e6eb5ae23e88a6bc9ac92319f886384d1a..d49b899eb3fa842b4f2e3e8554c7d4fe3571af54 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -77,7 +77,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
     // Sakura start - falling block height parity api
     @Override
     public double getEyeY() {
-        return this.heightParity ? this.getY() : super.getEyeY();
+        return this.heightParity && !this.mechanicsTarget().isLegacy() ? this.getY() : super.getEyeY(); // Sakura - configure server mechanics
     }
     // Sakura end - falling block height parity api
     // Sakura start - falling block stacking restrictions
@@ -156,6 +156,18 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public double distanceToSqr(final Vec3 pos) {
+        if (!this.mechanicsTarget().isLegacy()) {
+            return super.distanceToSqr(pos);
+        }
+        final double x = this.getX() - pos.x;
+        final double y = this.getEyeY() - pos.y;
+        final double z = this.getZ() - pos.z;
+        return x * x + y * y + z * z;
+    }
+    // Sakura end - configure server mechanics
 
     public FallingBlockEntity(final EntityType<? extends FallingBlockEntity> type, final Level level) {
         super(type, level);
@@ -164,6 +176,10 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         this.isFallingBlock = true; // Sakura - client visibility settings
         this.loadChunks = level.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
         this.mergeData.mergeLevel = level.sakuraConfig().cannons.mergeLevel; // Sakura - merge cannon entities
+        // Sakura start - configure server mechanics
+        this.updateMechanicsTarget();
+        this.eyeHeight = this.mechanicsTarget().isLegacy() ? 0.49f : this.eyeHeight;
+        // Sakura end - configure server mechanics
     }
 
     public FallingBlockEntity(final Level level, final double x, final double y, final double z, final BlockState blockState) {
@@ -187,7 +203,11 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             state.hasProperty(BlockStateProperties.WATERLOGGED) ? state.setValue(BlockStateProperties.WATERLOGGED, false) : state
         );
         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.getFluidState().createLegacyBlock())) return entity; // CraftBukkit
-        level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        // Sakura start - configure server mechanics
+        if (entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
+        }
+        // Sakura end - configure server mechanics
         level.addFreshEntity(entity);
         return entity;
     }
@@ -231,7 +251,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -240,6 +260,12 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             Block block = this.blockState.getBlock();
+            // Sakura start - configure server mechanics
+            final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget();
+            if (this.time == 0 && mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+                me.samsuik.sakura.mechanics.FallingBlockBehaviour.removeBlockOnFall(this, block);
+            }
+            // Sakura end - configure server mechanics
             this.time++;
             this.applyGravity();
             this.move(MoverType.SELF, this.getDeltaMovement());
@@ -254,9 +280,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             }
             // Paper end - Configurable falling blocks height nerf
             this.handlePortal();
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                this.setDeltaMovement(this.getDeltaMovement().scale(0.98F));
+            }
+            // Sakura end - configure server mechanics
             if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
                 // Sakura start - fix falling block floating point errors
-                BlockPos pos = serverLevel.sakuraConfig().cannons.mechanics.fallingBlockFloatingPointFix
+                BlockPos pos = serverLevel.sakuraConfig().cannons.mechanics.fallingBlockFloatingPointFix || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17) // Sakura - configure server mechanics
                     ? this.patchedBlockPosition()
                     : this.blockPosition();
                 // Sakura end - fix falling block floating point errors
@@ -286,8 +317,13 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                     }
                 } else {
                     BlockState currentState = this.level().getBlockState(pos);
-                    this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
-                    if (!currentState.is(Blocks.MOVING_PISTON)) {
+                    // Sakura start - configure server mechanics
+                    final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                        ? 0.7f
+                        : 0.7;
+                    this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+                    if (!currentState.is(Blocks.MOVING_PISTON) && (isStuckInWater || me.samsuik.sakura.mechanics.FallingBlockBehaviour.isAbleToStackOnBlock(this, mechanicsTarget))) {
+                        // Sakura end - configure server mechanics
                         if (!this.cancelDrop) {
                             boolean mayReplace = currentState.canBeReplaced(
                                 new DirectionalPlaceContext(this.level(), pos, Direction.DOWN, ItemStack.EMPTY, Direction.UP)
@@ -358,7 +394,14 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
                 }
             }
 
-            this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_12)) {
+                final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                    ? 0.98f
+                    : 0.98;
+                this.setDeltaMovement(this.getDeltaMovement().scale(drag));
+            }
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 96650e19466f15eccde8fc6ef27771648653bfc1..0bee019f161703bab07a300eeea959d71a19db5c 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -89,6 +89,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
         return this.sakura$collide(movement);
     }
     // Sakura end - optimise cannon entity movement
+    // Sakura start - configure server mechanics
+    @Override
+    public final double getEyeY() {
+        return this.mechanicsTarget().isLegacy() ? super.getEyeY() : this.getY();
+    }
+
+    @Override
+    public final double distanceToSqr(final net.minecraft.world.phys.Vec3 pos) {
+        if (!this.mechanicsTarget().isLegacy()) {
+            return super.distanceToSqr(pos);
+        }
+        final double x = this.getX() - pos.x;
+        final double y = this.getEyeY() - pos.y;
+        final double z = this.getZ() - pos.z;
+        return x * x + y * y + z * z;
+    }
+    // Sakura end - configure server mechanics
 
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
@@ -114,6 +131,13 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             case Y -> this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
         }
         // Sakura end - configure cannon mechanics
+        // Sakura start - configure server mechanics
+        this.updateMechanicsTarget();
+        this.eyeHeight = this.mechanicsTarget().isLegacy() ? 0.49f : this.eyeHeight;
+        if (this.mechanicsTarget().isLegacy()) {
+            this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
+        }
+        // Sakura end - configure server mechanics
     }
 
     @Override
@@ -134,7 +158,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
 
     @Override
     protected double getDefaultGravity() {
-        return 0.04;
+        return this.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14) ? 0.04f : 0.04; // Sakura - configure server mechanics
     }
 
     @Override
@@ -150,14 +174,23 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
             return;
         }
         // Paper end - Configurable TNT height nerf
-        this.setDeltaMovement(this.getDeltaMovement().scale(0.98));
+        // Sakura start - configure server mechanics
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = this.mechanicsTarget();
+        final double drag = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+            ? 0.98f
+            : 0.98;
+        this.setDeltaMovement(this.getDeltaMovement().scale(drag));
         if (this.onGround()) {
-            this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
+            final double friction = mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)
+                ? 0.7f
+                : 0.7;
+            this.setDeltaMovement(this.getDeltaMovement().multiply(friction, -0.5, friction));
+            // Sakura end - configure server mechanics
         }
 
         int fuse = this.getFuse() - 1;
         this.setFuse(fuse);
-        if (fuse <= 0) {
+        if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9) ? (fuse < 0) : (fuse <= 0)) { // Sakura - configure server mechanics
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
             //this.discard();
             this.tryToRespawnEntity(); // Sakura - merge cannon entities
@@ -196,13 +229,19 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
                 return;
             }
             // CraftBukkit end
+            // Sakura start - configure server mechanics
+            // Undocumented on the minecraft wiki but the explosion position did change in 1.10
+            final double explosionY = this.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10)
+                ? this.getY() + (double) 0.49f
+                : this.getY(0.0625D);
+            // Sakura end - configure server mechanics
             this.level()
                 .explode(
                     this,
                     Explosion.getDefaultDamageSource(this.level(), this),
                     this.usedPortal ? USED_PORTAL_DAMAGE_CALCULATOR : null,
                     this.getX(),
-                    this.getY(0.0625),
+                    explosionY, // Sakura - configure server mechanics
                     this.getZ(),
                     event.getRadius(), // CraftBukkit
                     event.getFire(), // CraftBukkit
@@ -291,7 +330,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
     // Paper start - Option to prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater  && super.isPushedByFluid(); // Sakura - configure cannon mechanics
+        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater && !this.mechanicsTarget().isLegacy() && super.isPushedByFluid(); // Sakura - configure server mechanics // Sakura - configure cannon mechanics
     }
     // Paper end - Option to prevent TNT from moving in water
 }
diff --git a/net/minecraft/world/level/BlockGetter.java b/net/minecraft/world/level/BlockGetter.java
index f9073223aad937a9bd27a9a8be8817de32f967f6..4c9eb6ee78727e21925f47d56d260bf7bc17302d 100644
--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -207,8 +207,23 @@ public interface BlockGetter extends LevelHeightAccessor {
     }
 
     static boolean forEachBlockIntersectedBetween(final Vec3 from, final Vec3 to, final AABB aabbAtTarget, final BlockGetter.BlockStepVisitor visitor) {
+        // Sakura start - configure server mechanics
+        return forEachBlockIntersectedBetween(from, to, aabbAtTarget, me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.latest(), visitor);
+    }
+
+    static boolean forEachBlockIntersectedBetween(
+        final Vec3 from,
+        final Vec3 to,
+        final AABB aabbAtTarget,
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget,
+        final BlockGetter.BlockStepVisitor visitor
+    ) {
         Vec3 travel = to.subtract(from);
-        if (travel.lengthSqr() < Mth.square(1.0E-5F)) {
+        final double movedThreshold = mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)
+            ? 1.0E-5F
+            : 0.99999F;
+        if (travel.lengthSqr() < Mth.square(movedThreshold) || mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            // Sakura end - configure server mechanics
             for (BlockPos blockPos : BlockPos.betweenClosed(aabbAtTarget)) {
                 if (!visitor.visit(blockPos, 0)) {
                     return false;
@@ -219,15 +234,22 @@ public interface BlockGetter extends LevelHeightAccessor {
         } else {
             LongSet visitedBlocks = new LongOpenHashSet();
 
-            for (BlockPos blockPosx : BlockPos.betweenCornersInDirection(aabbAtTarget.move(travel.scale(-1.0)), travel)) {
-                if (!visitor.visit(blockPosx, 0)) {
-                    return false;
-                }
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                for (BlockPos blockPosx : BlockPos.betweenCornersInDirection(aabbAtTarget.move(travel.scale(-1.0)), travel)) {
+                    if (!visitor.visit(blockPosx, 0)) {
+                        return false;
+                    }
 
-                visitedBlocks.add(blockPosx.asLong());
+                    visitedBlocks.add(blockPosx.asLong());
+                }
             }
 
             int iterations = addCollisionsAlongTravel(visitedBlocks, travel, aabbAtTarget, visitor);
+            if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+                iterations = 0;
+            }
+            // Sakura end - configure server mechanics
             if (iterations < 0) {
                 return false;
             } else {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 446ca8787231e40de2a6c62b0c4bf2aef9367455..1b2eaab42a615bf596011399794280ed41dfc970 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -846,6 +846,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.entity.merge.EntityMergeHandler mergeHandler = new me.samsuik.sakura.entity.merge.EntityMergeHandler(); // Sakura - merge cannon entities
     public final me.samsuik.sakura.explosion.density.BlockDensityCache densityCache = new me.samsuik.sakura.explosion.density.BlockDensityCache(this); // Sakura - optimise explosion density cache
     public final me.samsuik.sakura.explosion.durable.DurableBlockManager durabilityManager = new me.samsuik.sakura.explosion.durable.DurableBlockManager(); // Sakura - explosion durable blocks
+    public final me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler dispenseRelocationHandler = new me.samsuik.sakura.entity.dispensing.DispenseRelocationHandler(this); // Sakura - configure server mechanics
 
     protected Level(
         final WritableLevelData levelData,
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 35e787d257c78e922831ef5fbc906340a34e0dd2..c3bf4b8fdc910ef31818af2f77c918c27da33f6a 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -370,6 +370,37 @@ public class ServerExplosion implements Explosion {
         return this.damageCalculator.getBlockExplosionResistance(this, this.level, pos, blockState, fluidState);
     }
     // Sakura end - explosion durable blocks
+    // Sakura start - configure server mechanics
+    protected final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+
+    private float sakura$getSeenPercent(final Vec3 explosionPos, final Entity entity) {
+        if (this.mechanicsTarget.isServerType(me.samsuik.sakura.mechanics.ServerType.SAKE)) {
+            return this.getSeenPercentSingleRaycast(explosionPos, entity);
+        } else if (this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) {
+            return this.getSeenFraction(explosionPos, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+        } else {
+            return getSeenPercent(explosionPos, entity);
+        }
+    }
+
+    private float getSeenPercentSingleRaycast(final Vec3 explosionPos, final Entity entity) {
+        final AABB bounds = entity.getBoundingBox();
+        final double middle = 1.0 / (bounds.getYsize() * 2.0 + 1.0);
+        final Vec3 point = new Vec3(entity.getX(), entity.getY() + middle, entity.getZ());
+
+        final HitResult.Type hitResult;
+        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), point, explosionPos);
+        } else {
+            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                ? ClipContext.Block.COLLIDER
+                : ClipContext.Block.OUTLINE;
+            hitResult = entity.level().clip(new ClipContext(point, explosionPos, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+        }
+
+        return hitResult == HitResult.Type.MISS ? 1.0f : 0.0f;
+    }
+    // Sakura end - configure server mechanics
 
     public ServerExplosion(
         final ServerLevel level,
@@ -394,6 +425,7 @@ public class ServerExplosion implements Explosion {
         this.yield = Double.isFinite(this.yield) ? this.yield : 0; // Paper - Don't allow infinite default yields
         // Paper end - add yield
         this.consistentExplosionRadius = level.localConfig().at(this.center).consistentExplosionRadius; // Sakura - consistent explosion radius
+        this.mechanicsTarget = source != null ? source.mechanicsTarget() : level.localConfig().at(this.center).mechanicsTarget; // Sakura - configure server mechanics
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(final @Nullable Entity source) {
@@ -424,8 +456,15 @@ public class ServerExplosion implements Explosion {
                         final float density = entity.level().densityCache.getKnownDensity(center);
                         if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
                             hitResult = density != 0.0f ? net.minecraft.world.phys.HitResult.Type.MISS : net.minecraft.world.phys.HitResult.Type.BLOCK;
+                            // Sakura start - configure server mechanics
+                        } else if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_14)) {
+                            hitResult = me.samsuik.sakura.mechanics.LegacyExplosionBlockClipping.clip(entity.level(), from, center);
                         } else {
-                            hitResult = entity.level().clip(new ClipContext(from, center, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                            final ClipContext.Block blockContext = entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)
+                                ? ClipContext.Block.COLLIDER
+                                : ClipContext.Block.OUTLINE;
+                            hitResult = entity.level().clip(new ClipContext(from, center, blockContext, ClipContext.Fluid.NONE, entity)).getType();
+                            // Sakura end - configure server mechanics
                         }
                         if (hitResult == HitResult.Type.MISS) {
                             // Sakura end - replace density cache
@@ -523,6 +562,15 @@ public class ServerExplosion implements Explosion {
                 }
 
                 if (cachedBlock.outOfWorld) {
+                    // Sakura start - configure server mechanics
+                    if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                        power -= 0.22500001F;
+                        currX += incX;
+                        currY += incY;
+                        currZ += incZ;
+                        continue;
+                    }
+                    // Sakura end - configure server mechanics
                     break;
                 }
                 final BlockState iblockdata = cachedBlock.blockState;
@@ -590,7 +638,27 @@ public class ServerExplosion implements Explosion {
                     double dist = Math.sqrt(entity.distanceToSqr(this.center)) / doubleRadius;
                     if (!(dist > 1.0)) {
                         Vec3 entityOrigin = entity instanceof PrimedTnt ? entity.position() : entity.getEyePosition();
-                        Vec3 direction = entityOrigin.subtract(this.center).normalize();
+                        // Sakura start - configure server mechanics
+                        final double dirX = entityOrigin.x - this.center.x;
+                        final double dirY = entityOrigin.y - this.center.y;
+                        final double dirZ = entityOrigin.z - this.center.z;
+                        double magnitude = Math.sqrt(dirX * dirX + dirY * dirY + dirZ * dirZ);
+
+                        if (this.mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_17)) {
+                            dist = (float) dist;
+                            magnitude = (float) magnitude;
+                        }
+
+                        final Vec3 direction;
+                        if (magnitude < 1.0e-5f && this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_9)) {
+                            direction = Vec3.ZERO;
+                        } else {
+                            final double normalisedX = dirX / magnitude;
+                            final double normalisedY = dirY / magnitude;
+                            final double normalisedZ = dirZ / magnitude;
+                            direction = new Vec3(normalisedX, normalisedY, normalisedZ);
+                        }
+                        // Sakura end - configure server mechanics
                         boolean shouldDamageEntity = this.damageCalculator.shouldDamageEntity(this, entity);
                         float knockbackMultiplier = this.damageCalculator.getKnockbackMultiplier(entity);
                         float exposure = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : this.getBlockDensity(this.center, entity); // Paper - Optimize explosions
@@ -858,7 +926,7 @@ public class ServerExplosion implements Explosion {
         // Sakura start - replace density cache
         float blockDensity = this.level.densityCache.getBlockDensity(vec3d, entity);
         if (blockDensity == me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
-            blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
+            blockDensity = this.sakura$getSeenPercent(vec3d, entity); // Sakura - configure server mechanics
             this.level.densityCache.putDensity(vec3d, entity, blockDensity);
             // Sakura end - replace density cache
         }
diff --git a/net/minecraft/world/level/block/FallingBlock.java b/net/minecraft/world/level/block/FallingBlock.java
index f42cc2807d6caf3ecf53e5352390ce9ea94e7b16..d6c5ccffdd6000c2e0364d7b0e6b4a61f16c24c2 100644
--- a/net/minecraft/world/level/block/FallingBlock.java
+++ b/net/minecraft/world/level/block/FallingBlock.java
@@ -22,6 +22,22 @@ public abstract class FallingBlock extends Block implements Fallable {
         super(properties);
     }
 
+    // Sakura start - configure server mechanics
+    @Override
+    public void neighborChanged(
+        final BlockState state,
+        final Level level,
+        final BlockPos pos,
+        final Block sourceBlock,
+        final net.minecraft.world.level.redstone.Orientation wireOrientation,
+        final boolean notify
+    ) {
+        if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
+            level.scheduleTick(pos, this, this.getDelayAfterPlace());
+        }
+    }
+    // Sakura end - configure server mechanics
+
     @Override
     protected abstract MapCodec<? extends FallingBlock> codec();
 
diff --git a/net/minecraft/world/level/block/FenceGateBlock.java b/net/minecraft/world/level/block/FenceGateBlock.java
index 63400332eb64250d457bd36140bca419d4781ccf..fdf397959741b4859c145c1b654d0888ac9271c1 100644
--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -199,8 +199,15 @@ public class FenceGateBlock extends HorizontalDirectionalBlock {
                 hasPower = eventRedstone.getNewCurrent() > 0;
             }
             // CraftBukkit end
-            if (state.getValue(POWERED) != hasPower) {
-                level.setBlock(pos, state.setValue(POWERED, hasPower).setValue(OPEN, hasPower), Block.UPDATE_CLIENTS);
+            // Sakura start - configure server mechanics
+            final boolean pre1_10_0 = level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_10);
+            final boolean powered = state.getValue(POWERED);
+            if (pre1_10_0 ? (hasPower || block.defaultBlockState().isSignalSource()) : powered != hasPower) {
+                final boolean openGate = pre1_10_0 && (hasPower == powered || state.getValue(OPEN) != powered)
+                    ? state.getValue(OPEN)
+                    : hasPower;
+                level.setBlock(pos, state.setValue(POWERED, hasPower).setValue(OPEN, openGate), Block.UPDATE_CLIENTS);
+                // Sakura end - configure server mechanics
                 if (state.getValue(OPEN) != hasPower) {
                     level.playSound(
                         null,
diff --git a/net/minecraft/world/level/block/HoneyBlock.java b/net/minecraft/world/level/block/HoneyBlock.java
index e517f58719607cb98043872415c8c423c8f85faa..e27cf447981c16a1435b08e6715a9b6b516e5b5c 100644
--- a/net/minecraft/world/level/block/HoneyBlock.java
+++ b/net/minecraft/world/level/block/HoneyBlock.java
@@ -80,11 +80,19 @@ public class HoneyBlock extends HalfTransparentBlock {
         super.entityInside(state, level, pos, entity, effectApplier, isPrecise);
     }
 
-    private static double getOldDeltaY(final double deltaY) {
+    // Sakura start - configure server mechanics
+    private static double getOldDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
         return deltaY / 0.98F + 0.08;
     }
 
-    private static double getNewDeltaY(final double deltaY) {
+    private static double getNewDeltaY(final double deltaY, final Entity entity) {
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_2)) {
+            return deltaY;
+        }
+        // Sakura end - configure server mechanics
         return (deltaY - 0.08) * 0.98F;
     }
 
@@ -93,7 +101,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             return false;
         } else if (entity.getY() > pos.getY() + 0.9375 - 1.0E-7) {
             return false;
-        } else if (getOldDeltaY(entity.getDeltaMovement().y) >= -0.08) {
+        } else if (getOldDeltaY(entity.getDeltaMovement().y, entity) >= -0.08) { // Sakura - configure server mechanics
             return false;
         } else {
             double dx = Math.abs(pos.getX() + 0.5 - entity.getX());
@@ -111,11 +119,15 @@ public class HoneyBlock extends HalfTransparentBlock {
 
     private void doSlideMovement(final Entity entity) {
         Vec3 deltaMovement = entity.getDeltaMovement();
-        if (getOldDeltaY(entity.getDeltaMovement().y) < -0.13) {
-            double horizontalReductionFactor = -0.05 / getOldDeltaY(entity.getDeltaMovement().y);
-            entity.setDeltaMovement(new Vec3(deltaMovement.x * horizontalReductionFactor, getNewDeltaY(-0.05), deltaMovement.z * horizontalReductionFactor));
+        // Sakura start - configure server mechanics
+        final double oldDeltaY = getOldDeltaY(deltaMovement.y, entity);
+        final double newDeltaY = getNewDeltaY(-0.05, entity);
+        if (oldDeltaY < -0.13) {
+            double horizontalReductionFactor = -0.05 / oldDeltaY;
+            entity.setDeltaMovement(new Vec3(deltaMovement.x * horizontalReductionFactor, newDeltaY, deltaMovement.z * horizontalReductionFactor));
         } else {
-            entity.setDeltaMovement(new Vec3(deltaMovement.x, getNewDeltaY(-0.05), deltaMovement.z));
+            entity.setDeltaMovement(new Vec3(deltaMovement.x, newDeltaY, deltaMovement.z));
+            // Sakura end - configure server mechanics
         }
 
         entity.resetFallDistance();
diff --git a/net/minecraft/world/level/block/LadderBlock.java b/net/minecraft/world/level/block/LadderBlock.java
index 9496fdb50c13ea5ce30220244164e47c395f034f..a65ae7fa22a79e0619bcbd147c1927ea4460a1df 100644
--- a/net/minecraft/world/level/block/LadderBlock.java
+++ b/net/minecraft/world/level/block/LadderBlock.java
@@ -33,6 +33,15 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
         return CODEC;
     }
 
+    // Sakura start - configure server mechanics
+    public static final Map<Direction, VoxelShape> LEGACY_SHAPES = Shapes.rotateHorizontal(Block.boxZ(16.0, 14.0, 16.0));
+
+    @Override
+    public boolean hasDynamicShape() {
+        return true;
+    }
+    // Sakura end - configure server mechanics
+
     protected LadderBlock(final BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, false));
@@ -40,6 +49,11 @@ public class LadderBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     protected VoxelShape getShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
+        // Sakura start - configure server mechanics
+        if (level instanceof net.minecraft.world.level.Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return LEGACY_SHAPES.get(state.getValue(FACING));
+        }
+        // Sakura end - configure server mechanics
         return SHAPES.get(state.getValue(FACING));
     }
 
diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index fb5e469aa2c98f53e3b2de28d7708169152f2677..9ea42a3caba83d3404f85cc07c8aa939f89eac23 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -246,7 +246,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                 }
                 // Sakura end - configure fluid ticking outside the world border
                 if (level.getFluidState(neighbourPos).is(FluidTags.WATER)) {
-                    Block convertToBlock = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    // Sakura start - configure server mechanics
+                    final FluidState fluidState = state.getFluidState();
+                    final Block convertToBlock = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    if (convertToBlock == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.LegacyBlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
+                        return true;
+                    }
+                    // Sakura end - configure server mechanics
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, convertToBlock.defaultBlockState(), Block.UPDATE_ALL)) {
                         this.fizz(level, pos);
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index fea48c065cb86187be195b6e62cc26b0e9dee868..7a6087f02848d93c626e39aa26a7ab6077cf8ebe 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -332,6 +332,10 @@ public class RedStoneWireBlock extends Block {
     ) {
         if (useExperimentalEvaluator(level)) {
             new ExperimentalRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.isLegacy()) {
+            new me.samsuik.sakura.redstone.PandaRedstoneWireEvaluator(this).updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
+            // Sakura end - configure server mechanics
         } else {
             this.evaluator.updatePowerStrength(level, pos, state, orientation, shapeUpdateWiresAroundInitialPosition);
         }
@@ -563,7 +567,7 @@ public class RedStoneWireBlock extends Block {
     protected InteractionResult useWithoutItem(
         final BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult
     ) {
-        if (!player.getAbilities().mayBuild) {
+        if (!player.getAbilities().mayBuild || isCross(state) && level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) { // Sakura - configure server mechanics
             return InteractionResult.PASS;
         } else {
             if (isCross(state) || isDot(state)) {
diff --git a/net/minecraft/world/level/block/piston/MovingPistonBlock.java b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
index 2b6ad989e1e1424d0283bae39e5dae33eba44923..0eb40e2d41550d328eb66ae30299fc40b0147b47 100644
--- a/net/minecraft/world/level/block/piston/MovingPistonBlock.java
+++ b/net/minecraft/world/level/block/piston/MovingPistonBlock.java
@@ -105,6 +105,16 @@ public class MovingPistonBlock extends BaseEntityBlock {
     @Override
     protected VoxelShape getCollisionShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
         PistonMovingBlockEntity blockEntity = this.getBlockEntity(level, pos);
+        // Sakura start - configure server mechanics
+        if (blockEntity != null && level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            final VoxelShape shape = blockEntity.getCollisionShapeFromProgress(level, pos);
+            if (context.isAbove(shape, pos, false)) {
+                return shape;
+            } else {
+                return blockEntity.getMovedState().getCollisionShape(level, pos);
+            }
+        }
+        // Sakura end - configure server mechanics
         return blockEntity != null ? blockEntity.getCollisionShape(level, pos) : Shapes.empty();
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index a0992572a6bd4d1c9745195dbbefbf99959f6829..5a3be369119353ec5cf26b2e4e7f95b2abd396ee 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -121,6 +121,11 @@ public class PistonBaseBlock extends DirectionalBlock {
                 event = 2;
             }
 
+            // Sakura start - configure server mechanics
+            if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                level.setBlock(pos, state.setValue(PistonBaseBlock.EXTENDED, false), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE);
+            }
+            // Sakura end - configure server mechanics
             level.blockEvent(pos, this, event, direction.get3DDataValue());
         }
     }
diff --git a/net/minecraft/world/level/block/piston/PistonHeadBlock.java b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
index f8caa0b3c21766dbad73c524c1680346e4257188..5703513b1be21ccc898f6103e01f0174f3de26d9 100644
--- a/net/minecraft/world/level/block/piston/PistonHeadBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonHeadBlock.java
@@ -105,6 +105,11 @@ public class PistonHeadBlock extends DirectionalBlock {
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
         BlockState base = level.getBlockState(pos.relative(state.getValue(FACING).getOpposite()));
+        // Sakura start - configure server mechanics
+        if (level instanceof Level gameLevel && gameLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            return this.isFittingBase(state, state);
+        }
+        // Sakura end - configure server mechanics
         return this.isFittingBase(state, base) || base.is(Blocks.MOVING_PISTON) && base.getValue(FACING) == state.getValue(FACING);
     }
 
@@ -118,6 +123,10 @@ public class PistonHeadBlock extends DirectionalBlock {
                 block,
                 ExperimentalRedstoneUtils.withFront(orientation, state.getValue(FACING).getOpposite())
             );
+            // Sakura start - configure server mechanics
+        } else if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+            level.setBlock(pos, Blocks.AIR.defaultBlockState(), 19);
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index a1328b8bdf91e260cdaa7203d42cef88258afad8..30433fa880e3a68b3ef3dec7b808c083c65032c4 100644
--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -51,6 +51,164 @@ public class PistonMovingBlockEntity extends BlockEntity {
     private @org.jspecify.annotations.Nullable Direction shapeDirection = null;
     private float shapeProgress = Float.MIN_VALUE;
     // Sakura end - cache moving block entity collision shape
+    // Sakura start - configure server mechanics
+    @org.jspecify.annotations.Nullable
+    private AABB getBoundsFromProgress(
+        final BlockGetter level,
+        final BlockPos pos,
+        final BlockState state,
+        final float progress,
+        final Direction dir,
+        final boolean absolute
+    ) {
+        if (state.is(Blocks.MOVING_PISTON) || state.isAir()) {
+            return null;
+        }
+
+        VoxelShape shape = this.movedState.getCollisionShape(level, pos);
+        if (shape.isEmpty()) {
+            return null;
+        }
+
+        if (absolute) {
+            shape = shape.move(pos.getX(), pos.getY(), pos.getZ());
+        }
+
+        final AABB bounds = shape.bounds();
+        double minX = bounds.minX;
+        double minY = bounds.minY;
+        double minZ = bounds.minZ;
+        double maxX = bounds.maxX;
+        double maxY = bounds.maxY;
+        double maxZ = bounds.maxZ;
+
+        if (dir.getStepX() < 0) {
+            minX -= (float) dir.getStepX() * progress;
+        } else {
+            maxX -= (float) dir.getStepX() * progress;
+        }
+
+        if (dir.getStepY() < 0) {
+            minY -= (float) dir.getStepY() * progress;
+        } else {
+            maxY -= (float) dir.getStepY() * progress;
+        }
+
+        if (dir.getStepZ() < 0) {
+            minZ -= (float) dir.getStepZ() * progress;
+        } else {
+            maxZ -= (float) dir.getStepZ() * progress;
+        }
+
+        return this.fixZeroWidthBB(new AABB(minX, minY, minZ, maxX, maxY, maxZ), dir);
+    }
+
+    private AABB fixZeroWidthBB(AABB bb, final Direction dir) {
+        // Legacy behaviour relied on entities being able to collide with zero width shapes
+        // This is no longer possible, so we have to create a difference here for it to work
+        final double expandX = bb.getXsize() == 0.0 ? 1.0e-5 * dir.getStepX() : 0.0;
+        final double expandY = bb.getYsize() == 0.0 ? 1.0e-5 * dir.getStepY() : 0.0;
+        final double expandZ = bb.getZsize() == 0.0 ? 1.0e-5 * dir.getStepZ() : 0.0;
+
+        if (expandX != 0 || expandY != 0 || expandZ != 0) {
+            bb = bb.expandTowards(expandX, expandY, expandZ);
+        }
+
+        return bb;
+    }
+
+    public final VoxelShape getCollisionShapeFromProgress(final BlockGetter level, final BlockPos pos) {
+        float progress = this.getProgress(0.0f);
+        if (this.extending) {
+            progress = 1.0F - progress;
+        }
+
+        final AABB bb = this.getBoundsFromProgress(level, pos, this.movedState, progress, this.direction, false);
+        return bb == null ? Shapes.empty() : Shapes.create(bb);
+    }
+
+    private void moveEntities(final Level level, final float relativeProgress) {
+        float remaining = this.progress;
+        if (this.extending) {
+            remaining = 1.0F - remaining;
+        } else {
+            --remaining;
+        }
+
+        final Direction direction = this.direction;
+        final AABB bb = this.getBoundsFromProgress(level, this.worldPosition, this.movedState, remaining, direction, true);
+        if (bb == null || bb.getSize() == 0.0) {
+            return;
+        }
+
+        for (final Entity entity : level.getEntities(null, bb)) {
+            if (this.movedState.is(Blocks.SLIME_BLOCK) && this.extending) {
+                final Vec3 movement = entity.getDeltaMovement();
+                double moveX = movement.x;
+                double moveY = movement.y;
+                double moveZ = movement.z;
+
+                switch (direction.getAxis()) {
+                    case X -> moveX = direction.getStepX();
+                    case Y -> moveY = direction.getStepY();
+                    case Z -> moveZ = direction.getStepZ();
+                }
+
+                entity.setDeltaMovement(moveX, moveY, moveZ);
+            } else {
+                final Vec3 pistonMovement = new Vec3(
+                    relativeProgress * (float) direction.getStepX(),
+                    relativeProgress * (float) direction.getStepY(),
+                    relativeProgress * (float) direction.getStepZ()
+                );
+                entity.move(MoverType.PISTON, pistonMovement);
+                entity.applyEffectsFromBlocks(entity.position(), entity.position());
+            }
+        }
+    }
+
+    private static void moveEntityByPistonInDirection(final Direction direction, final Entity entity, final AABB movingBB) {
+        final AABB entityBB = entity.getBoundingBox();
+        double moveX = 0.0;
+        double moveY = 0.0;
+        double moveZ = 0.0;
+
+        switch (direction.getAxis()) {
+            case X -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveX = movingBB.maxX - entityBB.minX;
+                } else {
+                    moveX = entityBB.maxX - movingBB.minX;
+                }
+                moveX += 0.01D;
+            }
+            case Y -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveY = movingBB.maxY - entityBB.minY;
+                } else {
+                    moveY = entityBB.maxY - movingBB.minY;
+                }
+                moveY += 0.01D;
+            }
+            case Z -> {
+                if (direction.getAxisDirection() == Direction.AxisDirection.POSITIVE) {
+                    moveZ = movingBB.maxZ - entityBB.minZ;
+                } else {
+                    moveZ = entityBB.maxZ - movingBB.minZ;
+                }
+                moveZ += 0.01D;
+            }
+        }
+
+        final Vec3 pistonMovement = new Vec3(
+            moveX * direction.getStepX(),
+            moveY * direction.getStepY(),
+            moveZ * direction.getStepZ()
+        );
+        entity.move(MoverType.PISTON, pistonMovement);
+        entity.applyEffectsFromBlocks(entity.position(), entity.position());
+    }
+    // Sakura end - configure server mechanics
 
     public PistonMovingBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityType.PISTON, worldPosition, blockState);
@@ -175,6 +333,12 @@ public class PistonMovingBlockEntity extends BlockEntity {
 
                     double delta = 0.0;
 
+                    // Sakura start - configure server mechanics
+                    if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                        moveEntityByPistonInDirection(movement, entity, aabb);
+                        return;
+                    }
+                    // Sakura end - configure server mechanics
                     for (AABB shapeAabb : shapeAabbs) {
                         AABB movingAABB = PistonMath.getMovementArea(moveByPositionAndProgress(pos, shapeAabb, self), movement, deltaProgress);
                         AABB entityAabb = entity.getBoundingBox();
@@ -202,6 +366,11 @@ public class PistonMovingBlockEntity extends BlockEntity {
         NOCLIP.set(pistonDirection);
         Vec3 previousPos = entity.position();
         entity.move(MoverType.PISTON, new Vec3(delta * movement.getStepX(), delta * movement.getStepY(), delta * movement.getStepZ()));
+        // Sakura start - configure server mechanics
+        if (entity.mechanicsTarget().before(me.samsuik.sakura.mechanics.MechanicVersion.v1_21_5)) {
+            previousPos = entity.position();
+        }
+        // Sakura end - configure server mechanics
         entity.applyEffectsFromBlocks(previousPos, entity.position());
         entity.removeLatestMovementRecording();
         NOCLIP.set(null);
@@ -317,12 +486,21 @@ public class PistonMovingBlockEntity extends BlockEntity {
     }
 
     public static void tick(final Level level, final BlockPos pos, final BlockState state, final PistonMovingBlockEntity entity) {
+        final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget; // Sakura - configure server mechanics
         entity.lastTicked = level.getGameTime();
         entity.progressO = entity.progress;
         if (entity.progressO >= 1.0F) {
             if (level.isClientSide() && entity.deathTicks < 5) {
                 entity.deathTicks++;
             } else {
+                // Sakura start - configure server mechanics
+                if (mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_9, me.samsuik.sakura.mechanics.MechanicVersion.v1_10)) {
+                    moveCollidedEntities(level, pos, 1.0f, entity);
+                    moveStuckEntities(level, pos, 1.0f, entity);
+                } else if (mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                    entity.moveEntities(level, 0.25f);
+                }
+                // Sakura end - configure server mechanics
                 level.removeBlockEntity(pos);
                 entity.setRemoved();
                 if (level.getBlockState(pos).is(Blocks.MOVING_PISTON)) {
@@ -342,12 +520,22 @@ public class PistonMovingBlockEntity extends BlockEntity {
             }
         } else {
             float newProgress = entity.progress + 0.5F;
-            moveCollidedEntities(level, pos, newProgress, entity);
-            moveStuckEntities(level, pos, newProgress, entity);
+            // Sakura start - configure server mechanics
+            if (mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_11)) {
+                moveCollidedEntities(level, pos, newProgress, entity);
+                moveStuckEntities(level, pos, newProgress, entity);
+            }
             entity.progress = newProgress;
             if (entity.progress >= 1.0F) {
                 entity.progress = 1.0F;
             }
+            if (mechanicsTarget.between(me.samsuik.sakura.mechanics.MechanicVersion.v1_9, me.samsuik.sakura.mechanics.MechanicVersion.v1_10)) {
+                moveCollidedEntities(level, pos, newProgress, entity);
+                moveStuckEntities(level, pos, newProgress, entity);
+            } else if (entity.extending && mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_9)) {
+                entity.moveEntities(level, entity.progress - entity.progressO + 0.0625f);
+            }
+            // Sakura end - configure server mechanics
         }
     }
 
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index cffde590c457e7be053c268524fd9c7163cee9ad..b48a5e38f8c497486b1425315cdfa8c25a482769 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -186,6 +186,11 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
+        // Sakura start - configure server mechanics
+        if (level instanceof ServerLevel serverLevel && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+            return false;
+        }
+        // Sakura end - configure server mechanics
         return state.getHeight(level, pos) >= 0.44444445F && other.is(FluidTags.WATER);
     }
 
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index d0a62bad5592c3ef8bd2ba5bc9719fad66c4edf0..9fc0ae6acf9137f20c6bb6aac9a465f9941c9e72 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -123,7 +123,13 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(final FluidState state, final BlockGetter level, final BlockPos pos, final Fluid other, final Direction direction) {
-        return direction == Direction.DOWN && !other.is(FluidTags.WATER);
+        // Sakura start - configure server mechanics
+        final boolean canReplace = direction == Direction.DOWN
+            || level instanceof ServerLevel serverLevel
+            && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Before 1.13 lava could replace water
+        return canReplace && !other.is(FluidTags.WATER);
+        // Sakura end - configure server mechanics
     }
 
     @Override
