From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 20 Feb 2024 19:16:16 +0000
Subject: [PATCH] Add entity travel distance limits


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 65e09f28d0527bca1a6f8582194166fbda1d3d58..c64521af6f2fcf000072d8dd2f33d45d235083c0 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1398,6 +1398,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             } else {entity.inactiveTick();} // Paper - EAR 2
         }
         // Sakura end - optimise cannon entities
+        // Sakura start - entity travel distance limits
+        if (entity.isPastTravelDistanceLimit()) {
+            entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+        }
+        // Sakura end - entity travel distance limits
         profiler.pop();
 
         for (Entity passenger : entity.getPassengers()) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 6c7566b1974055072df03c77dfefe0fd76231f8d..b781c49989d8de8b8faa22da6ca113fba676f28e 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -173,6 +173,18 @@ public abstract class Entity
     static boolean isLevelAtLeast(ValueInput input, int level) {
         return input.getIntOr("Bukkit.updateLevel", CURRENT_LEVEL) >= level;
     }
+    // Sakura start - entity travel distance limits
+    private final double travelDistanceLimit;
+
+    public final boolean isPastTravelDistanceLimit() {
+        if (this.origin == null) {
+            return false;
+        }
+        final double x = Math.pow(this.origin.x() - this.position.x(), 2.0);
+        final double z = Math.pow(this.origin.z() - this.position.z(), 2.0);
+        return Math.max(x, z) >= this.travelDistanceLimit;
+    }
+    // Sakura end - entity travel distance limits
 
     // Paper start - Share random for entities to make them more random
     public static RandomSource SHARED_RANDOM = new RandomRandomSource();
@@ -620,6 +632,7 @@ public abstract class Entity
         this.setPos(0.0, 0.0, 0.0);
         this.eyeHeight = this.dimensions.eyeHeight();
         this.despawnTime = level == null || type == EntityType.PLAYER ? -1 : level.paperConfig().entities.spawning.despawnTime.getOrDefault(type, io.papermc.paper.configuration.type.number.IntOr.Disabled.DISABLED).or(-1); // Paper - entity despawn time limit
+        this.travelDistanceLimit = level == null ? Integer.MAX_VALUE : Math.pow(level.sakuraConfig().entity.chunkTravelLimit.getOrDefault(type, Integer.MAX_VALUE) * 16.0, 2); // Sakura - entity travel distance limits
     }
 
     public boolean isColliding(final BlockPos pos, final BlockState state) {
