From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 9 Aug 2024 20:43:53 +0100
Subject: [PATCH] Configurable left shooting and adjusting limits


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 49c063c9672535b54391ba664f21cd0e86f3659d..555f63ecf6ce7ec5ec2b26830f27d1d65a7537ae 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -182,6 +182,47 @@ public abstract class Entity
         return Math.max(x, z) >= this.travelDistanceLimit;
     }
     // Sakura end - entity travel distance limits
+    // Sakura start - configurable left shooting and adjusting limits
+    protected final void limitLeftShooting() {
+        final Vec3 movement = this.getDeltaMovement();
+        final int threshold = this.level.sakuraConfig().cannons.restrictions.leftShootingThreshold.or(-1);
+        if (threshold > 0 && (movement.x != 0.0 || movement.z != 0.0) && this.origin != null) {
+            final double travelledX = Math.abs(this.getX() - this.origin.x());
+            final double travelledZ = Math.abs(this.getZ() - this.origin.z());
+            final boolean xSmaller = travelledX < travelledZ; // intended // todo is it?
+
+            // Once entities have travelled past the threshold changing direction is restricted.
+            if (xSmaller && travelledX > threshold) {
+                this.setDeltaMovement(movement.multiply(1.0, 1.0, 0.0)); // limit z
+            } else if (!xSmaller && travelledZ > threshold) {
+                this.setDeltaMovement(movement.multiply(0.0, 1.0, 1.0)); // limit x
+            }
+        }
+    }
+
+    private void limitAdjustMovement(final AABB currBoundingBox, final double dir, final boolean xAdjust, final List<VoxelShape> shapes) {
+        final int adjustDistance = this.level.sakuraConfig().cannons.restrictions.maxAdjustDistance.or(-1);
+        if (adjustDistance > 0 && Math.abs(dir) > adjustDistance) {
+            double minX = Double.NEGATIVE_INFINITY;
+            double minZ = Double.NEGATIVE_INFINITY;
+            double maxX = Double.POSITIVE_INFINITY;
+            double maxZ = Double.POSITIVE_INFINITY;
+            if (xAdjust) { // limit x adjust
+                minX = Math.floor(currBoundingBox.minX) - adjustDistance;
+                maxX = Math.floor(currBoundingBox.maxX) + adjustDistance + 1;
+            } else {       // limit z adjust
+                minZ = Math.floor(currBoundingBox.minZ) - adjustDistance;
+                maxZ = Math.floor(currBoundingBox.maxZ) + adjustDistance + 1;
+            }
+
+            final VoxelShape safeSpace = Shapes.box(
+                minX, Double.NEGATIVE_INFINITY, minZ,
+                maxX, Double.POSITIVE_INFINITY, maxZ
+            );
+            shapes.add(Shapes.join(Shapes.INFINITY, safeSpace, BooleanOp.ONLY_FIRST));
+        }
+    }
+    // Sakura end - configurable left shooting and adjusting limits
 
     // Paper start - Share random for entities to make them more random
     public static RandomSource SHARED_RANDOM = new RandomRandomSource();
@@ -1758,6 +1799,7 @@ public abstract class Entity
         }
 
         if (xSmaller && z != 0.0) {
+            this.limitAdjustMovement(currentBox, z, false, collisionsVoxel); // Sakura - configurable left shooting and adjusting limits
             z = this.scanZ(currentBox, z, collisionsVoxel, collisionsBB);
             if (z != 0.0) {
                 currentBox = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.offsetZ(currentBox, z);
@@ -1765,6 +1807,11 @@ public abstract class Entity
         }
 
         if (x != 0.0) {
+            // Sakura start - configurable left shooting and adjusting limits
+            if (!xSmaller) {
+                this.limitAdjustMovement(currentBox, x, true, collisionsVoxel);
+            }
+            // Sakura end - configurable left shooting and adjusting limits
             x = this.scanX(currentBox, x, collisionsVoxel, collisionsBB);
             if (x != 0.0) {
                 currentBox = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.offsetX(currentBox, x);
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index f0ab75cf3795d28894bc769bb37fc2bc8c79e119..647a3446136c924f53e45061061eec5500ae81a9 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -292,6 +292,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             // Sakura end - configure server mechanics
             this.time++;
             this.applyGravity();
+            this.limitLeftShooting(); // Sakura - configurable left shooting and adjusting limits
             this.move(MoverType.SELF, this.getDeltaMovement());
             this.applyEffectsFromBlocks();
             // Paper start - Configurable falling blocks height nerf
