From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Mon, 12 Aug 2024 15:35:57 +0100
Subject: [PATCH] Optimise hopper ticking


diff --git a/me/samsuik/sakura/container/SlotCountingItemList.java b/me/samsuik/sakura/container/SlotCountingItemList.java
new file mode 100644
index 0000000000000000000000000000000000000000..822e29472b94c77ca93963110a69fab6872f4032
--- /dev/null
+++ b/me/samsuik/sakura/container/SlotCountingItemList.java
@@ -0,0 +1,134 @@
+package me.samsuik.sakura.container;
+
+import me.samsuik.sakura.tracking.ChangeListener;
+import net.minecraft.core.NonNullList;
+import net.minecraft.world.item.ItemStack;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Item list that keeps track of empty and full slots.
+ */
+@NullMarked
+public final class SlotCountingItemList extends NonNullList<ItemStack> {
+    public static SlotCountingItemList withSize(final int size, final ItemStack defaultItem) {
+        final ItemStack[] items = new ItemStack[size];
+        Arrays.fill(items, defaultItem);
+        return new SlotCountingItemList(Arrays.asList(items), defaultItem);
+    }
+
+    public static SlotCountingItemList copy(final NonNullList<ItemStack> items) {
+        final SlotCountingItemList itemList = withSize(items.size(), items.defaultValue);
+        for (int index = 0; index < items.size(); ++index) {
+            itemList.set(index, items.get(index));
+        }
+
+        return itemList;
+    }
+
+    private final @Nullable ItemWatcher[] watchers;
+    private int emptySlots;
+    private int fullSlots;
+
+    SlotCountingItemList(final List<ItemStack> items, final @Nullable ItemStack defaultItem) {
+        super(items, defaultItem);
+        this.watchers = new ItemWatcher[items.size()];
+        this.recalculate();
+    }
+
+    private void recalculate() {
+        this.emptySlots = this.size();
+        this.fullSlots = 0;
+
+        for (int slot = 0; slot < this.size(); ++slot) {
+            this.updateItem(slot, ItemStack.EMPTY, this.get(slot));
+        }
+    }
+
+    private void updateItem(final int index, final ItemStack oldItemStack, final ItemStack newItemStack) {
+        final ItemWatcher watcher = this.watchers[index];
+        if (watcher != null) {
+            oldItemStack.unwatch(watcher);
+        }
+
+        if (!newItemStack.isEmpty()) {
+            this.watchers[index] = new ItemWatcher(newItemStack);
+        }
+
+        final boolean oldEmpty = oldItemStack.isEmpty();
+        final boolean oldFull  = oldItemStack.getCount() == oldItemStack.getMaxStackSize();
+        this.updateCounts(oldEmpty, oldFull, newItemStack);
+    }
+
+    private void updateCounts(final boolean oldItemEmpty, final boolean oldItemFull, final ItemStack itemStack) {
+        if (oldItemEmpty != itemStack.isEmpty()) {
+            this.emptySlots += oldItemEmpty ? -1 : 1;
+        }
+
+        if (oldItemFull != (itemStack.getCount() == itemStack.getMaxStackSize())) {
+            this.fullSlots += oldItemFull ? -1 : 1;
+        }
+    }
+
+    public boolean isEmpty() {
+        return this.emptySlots == this.size();
+    }
+
+    public boolean isFull() {
+        return this.fullSlots == this.size();
+    }
+
+    public boolean hasSpaceAvailable() {
+        return this.fullSlots != this.size();
+    }
+
+    @Override
+    public ItemStack set(final int index, final ItemStack newItemStack) {
+        final ItemStack oldItemStack = super.set(index, newItemStack);
+        this.updateItem(index, oldItemStack, newItemStack);
+
+        return oldItemStack;
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        this.recalculate();
+    }
+
+    @Override
+    public boolean add(final ItemStack itemStack) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ItemStack remove(final int index) {
+        throw new UnsupportedOperationException();
+    }
+
+    private class ItemWatcher implements ChangeListener<ItemStack> {
+        private final ItemStack itemStack;
+        private boolean empty;
+        private boolean full;
+
+        private ItemWatcher(final ItemStack itemStack) {
+            this.itemStack = itemStack;
+            this.update(itemStack);
+            itemStack.watch(this);
+        }
+
+        private void update(final ItemStack itemStack) {
+            this.empty = itemStack.isEmpty();
+            this.full = itemStack.getCount() == this.itemStack.getMaxStackSize();
+        }
+
+        @Override
+        public void changed(final ItemStack itemStack) {
+            SlotCountingItemList.this.updateCounts(this.empty, this.full, this.itemStack);
+            this.update(itemStack);
+        }
+    }
+}
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 1b2eaab42a615bf596011399794280ed41dfc970..01057cd781d769a20aa12a3976774bd9809e26a9 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1527,7 +1527,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         // Paper end - Fix MC-117075 use removeAll
             if (ticker.isRemoved()) {
                 toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
-            } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
+            } else if (tickBlockEntities && ticker.isBlockEntityActive() && this.shouldTickBlocksAt(ticker.getPos())) { // Sakura - optimise hopper ticking
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 if ((++tickedEntities & 7) == 0) {
diff --git a/net/minecraft/world/level/block/HopperBlock.java b/net/minecraft/world/level/block/HopperBlock.java
index e032b5289abe413009fe7c9068e71e788fab512a..a84587e9ee015c1272cd67dbbbe043a1424416d5 100644
--- a/net/minecraft/world/level/block/HopperBlock.java
+++ b/net/minecraft/world/level/block/HopperBlock.java
@@ -123,6 +123,12 @@ public class HopperBlock extends BaseEntityBlock {
     private void checkPoweredState(final Level level, final BlockPos pos, final BlockState state) {
         boolean shouldBeOn = !level.hasNeighborSignal(pos);
         if (shouldBeOn != state.getValue(ENABLED)) {
+            // Sakura start - optimise hopper ticking
+            final BlockEntity blockEntity = level.getBlockEntity(pos);
+            if (blockEntity instanceof HopperBlockEntity hbe && level.sakuraConfig().technical.optimiseIdleHopperTicking) {
+                hbe.setBlockEntityTicking(shouldBeOn);
+            }
+            // Sakura end - optimise hopper ticking
             level.setBlock(pos, state.setValue(ENABLED, shouldBeOn), Block.UPDATE_CLIENTS);
         }
     }
diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index 865f08c92946010a36947b7766b957f4c0517285..57addfc524eb769247dc84da404dfcf29950dd57 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -54,6 +54,31 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
 
+    // Sakura start - optimise hopper ticking
+    private boolean blockEntityTicking = true;
+    private int idleTickCount = 0;
+
+    public final boolean isBlockEntityActive() {
+        if (!this.blockEntityTicking && this.idleTickCount >= 1200) {
+            this.setBlockEntityTicking(true);
+        }
+
+        this.idleTickCount++;
+        return this.blockEntityTicking;
+    }
+
+    public final void setBlockEntityTicking(final boolean blockEntityTicking) {
+        if (blockEntityTicking) {
+            this.catchupTicks(this.idleTickCount);
+        }
+
+        this.idleTickCount = 0;
+        this.blockEntityTicking = blockEntityTicking;
+    }
+
+    protected void catchupTicks(final int ticks) {}
+    // Sakura end - optimise hopper ticking
+
     public BlockEntity(final BlockEntityType<?> type, final BlockPos worldPosition, final BlockState blockState) {
         this.type = type;
         this.worldPosition = worldPosition.immutable();
diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 085ed0eecc47cf9c0fa9fc26b3f7b8522610a941..2f1c4c85fbe936436d7bf160c0987bab95b23478 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -28,13 +28,13 @@ import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import org.jspecify.annotations.Nullable;
 
-public class HopperBlockEntity extends RandomizableContainerBlockEntity implements Hopper {
+public class HopperBlockEntity extends RandomizableContainerBlockEntity implements Hopper, me.samsuik.sakura.tracking.ChangeListener<Container> { // Sakura - optimise hopper ticking
     public static final int MOVE_ITEM_SPEED = 8;
     public static final int HOPPER_CONTAINER_SIZE = 5;
     private static final int[][] CACHED_SLOTS = new int[54][];
     private static final int NO_COOLDOWN_TIME = -1;
     private static final Component DEFAULT_NAME = Component.translatable("container.hopper");
-    private NonNullList<ItemStack> items = NonNullList.withSize(5, ItemStack.EMPTY);
+    private me.samsuik.sakura.container.SlotCountingItemList items = me.samsuik.sakura.container.SlotCountingItemList.withSize(5, ItemStack.EMPTY); // Sakura - optimise hopper ticking
     public int cooldownTime = -1;
     private long tickedGameTime;
     private Direction facing;
@@ -72,6 +72,119 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         this.maxStack = size;
     }
     // CraftBukkit end
+    // Sakura start - optimise hopper ticking
+    private static final int SOURCE_CONTAINER   = 1 << 0;
+    private static final int ATTACHED_CONTAINER = 1 << 1;
+    private int knownContainers = 0;
+    private final @Nullable Container[] containers = new Container[2];
+
+    @Override
+    public final void changed(final Container container) {
+        this.setBlockEntityTicking(true);
+    }
+
+    @Override
+    public final void removed(final Container container) {
+        this.unwatchContainers();
+        this.setBlockEntityTicking(true);
+    }
+
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        this.setBlockEntityTicking(true);
+    }
+
+    @Override
+    public void setRemoved() {
+        this.unwatchContainers();
+        super.setRemoved();
+    }
+
+    @Override
+    protected final void catchupTicks(int ticks) {
+        this.cooldownTime -= ticks;
+    }
+
+    private void unwatchContainers() {
+        for (final Container container : this.containers) {
+            if (container != null) {
+                container.unwatch(this);
+            }
+        }
+
+        this.knownContainers = 0;
+    }
+
+    private void waitForChange(final int fullState) {
+        if ((fullState == HOPPER_IS_FULL || (this.knownContainers & SOURCE_CONTAINER) != 0)
+            && (fullState == HOPPER_EMPTY || (this.knownContainers & ATTACHED_CONTAINER) != 0)) {
+            this.setBlockEntityTicking(false);
+        }
+    }
+
+    private static @Nullable Container sakura$getSourceContainer(final Level level, final Hopper hopper, final BlockPos pos, final BlockState state) {
+        if (!(hopper instanceof HopperBlockEntity hbe) || org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            return getSourceContainer(level, hopper, pos, state);
+        }
+
+        Container sourceContainer = hbe.containers[0];
+        if ((hbe.knownContainers & SOURCE_CONTAINER) != 0 && isValidContainer(sourceContainer)) {
+            return sourceContainer;
+        }
+
+        if (sourceContainer != null) {
+            sourceContainer.unwatch(hbe);
+        }
+
+        sourceContainer = getSourceContainer(level, hopper, pos, state);
+        hbe.listenForContainerChanges(sourceContainer, SOURCE_CONTAINER);
+        hbe.containers[0] = sourceContainer;
+
+        if (sourceContainer != null) {
+            sourceContainer.watch(hbe);
+        }
+
+        return sourceContainer;
+    }
+
+    private static @Nullable Container sakura$getAttachedContainer(final Level level, final BlockPos pos, final HopperBlockEntity hbe) {
+        if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            return getAttachedContainer(level, pos, hbe);
+        }
+
+        Container attachedContainer = hbe.containers[1];
+        if ((hbe.knownContainers & ATTACHED_CONTAINER) != 0 && isValidContainer(attachedContainer)) {
+            return attachedContainer;
+        }
+
+        if (attachedContainer != null) {
+            attachedContainer.unwatch(hbe);
+        }
+
+        attachedContainer = getAttachedContainer(level, pos, hbe);
+        hbe.listenForContainerChanges(attachedContainer, ATTACHED_CONTAINER);
+        hbe.containers[1] = attachedContainer;
+
+        if (attachedContainer != null) {
+            attachedContainer.watch(hbe);
+        }
+
+        return attachedContainer;
+    }
+
+    private void listenForContainerChanges(final @Nullable Container container, final int type) {
+        if (isValidContainer(container)) {
+            this.knownContainers |= type; // set
+        } else if ((this.knownContainers & type) != 0) {
+            this.knownContainers ^= type; // unset
+        }
+    }
+
+    private static boolean isValidContainer(@Nullable final Container container) {
+        return container != null && !container.sakura$isRemoved();
+    }
+    // Sakura end - optimise hopper ticking
 
     public HopperBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityType.HOPPER, worldPosition, blockState);
@@ -81,7 +194,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     @Override
     protected void loadAdditional(final ValueInput input) {
         super.loadAdditional(input);
-        this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
+        this.items = me.samsuik.sakura.container.SlotCountingItemList.withSize(this.getContainerSize(), ItemStack.EMPTY); // Sakura - optimise hopper ticking
         if (!this.tryLoadLootTable(input)) {
             ContainerHelper.loadAllItems(input, this.items);
         }
@@ -150,32 +263,15 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static int getFullState(final HopperBlockEntity hopper) {
         hopper.unpackLootTable(null);
 
-        final List<ItemStack> hopperItems = hopper.items;
-
-        boolean empty = true;
-        boolean full = true;
-
-        for (int i = 0, len = hopperItems.size(); i < len; ++i) {
-            final ItemStack stack = hopperItems.get(i);
-            if (stack.isEmpty()) {
-                full = false;
-                continue;
-            }
-
-            if (!full) {
-                // can't be full
-                return HOPPER_HAS_ITEMS;
-            }
-
-            empty = false;
-
-            if (stack.getCount() != stack.getMaxStackSize()) {
-                // can't be full or empty
-                return HOPPER_HAS_ITEMS;
-            }
+        // Sakura start - optimise hopper ticking
+        if (hopper.items.isEmpty()) {
+            return HOPPER_EMPTY;
+        } else if (hopper.items.isFull()) {
+            return HOPPER_IS_FULL;
+        } else {
+            return HOPPER_HAS_ITEMS;
         }
-
-        return empty ? HOPPER_EMPTY : (full ? HOPPER_IS_FULL : HOPPER_HAS_ITEMS);
+        // Sakura end - optimise hopper ticking
     }
     // Paper end - Perf: Optimize Hoppers
 
@@ -201,6 +297,12 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     setChanged(level, pos, state);
                     return true;
                 }
+
+                // Sakura start - optimise hopper ticking
+                if (level.sakuraConfig().technical.optimiseIdleHopperTicking) {
+                    entity.waitForChange(fullState);
+                }
+                // Sakura end - optimise hopper ticking
             }
 
             return false;
@@ -418,7 +520,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(final Level level, final BlockPos blockPos, final HopperBlockEntity self) {
-        Container container = getAttachedContainer(level, blockPos, self);
+        Container container = sakura$getAttachedContainer(level, blockPos, self); // Sakura - optimise hopper ticking
         if (container == null) {
             return false;
         } else {
@@ -478,7 +580,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(final Level level, final Hopper hopper) {
         BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockState = level.getBlockState(blockPos);
-        Container container = getSourceContainer(level, hopper, blockPos, blockState);
+        Container container = sakura$getSourceContainer(level, hopper, blockPos, blockState); // Sakura - optimise hopper ticking
         if (container != null) {
             Direction direction = Direction.DOWN;
             skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
@@ -761,7 +863,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
     @Override
     protected void setItems(final NonNullList<ItemStack> items) {
-        this.items = items;
+        this.items = me.samsuik.sakura.container.SlotCountingItemList.copy(items); // Sakura - optimise hopper ticking
     }
 
     public static void entityInside(final Level level, final BlockPos pos, final BlockState blockState, final Entity entity, final HopperBlockEntity hopper) {
diff --git a/net/minecraft/world/level/block/entity/TickingBlockEntity.java b/net/minecraft/world/level/block/entity/TickingBlockEntity.java
index 28e3b73507b988f7234cbf29c4024c88180d0aef..a0d247aa883553708c4b92158232425593d50534 100644
--- a/net/minecraft/world/level/block/entity/TickingBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TickingBlockEntity.java
@@ -10,4 +10,10 @@ public interface TickingBlockEntity {
     BlockPos getPos();
 
     String getType();
+
+    // Sakura start - optimise hopper ticking
+    default boolean isBlockEntityActive() {
+        return true;
+    }
+    // Sakura end - optimise hopper ticking
 }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 21710c5acb1a5f12c303954b5d1249462cf5366d..58f441b9e7e18de37f53c23951f88f1b4a11206d 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -1002,6 +1002,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
             return this.blockEntity.typeHolder().getRegisteredName();
         }
 
+        // Sakura start - optimise hopper ticking
+        @Override
+        public final boolean isBlockEntityActive() {
+            return this.blockEntity.isBlockEntityActive();
+        }
+        // Sakura end - optimise hopper ticking
+
         @Override
         public String toString() {
             return "Level ticker for " + this.getType() + "@" + this.getPos();
@@ -1050,6 +1057,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
             return this.ticker.getType();
         }
 
+        // Sakura start - optimise hopper ticking
+        @Override
+        public final boolean isBlockEntityActive() {
+            return this.ticker.isBlockEntityActive();
+        }
+        // Sakura end - optimise hopper ticking
+
         @Override
         public String toString() {
             return this.ticker + " <wrapped>";
