From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 8 Nov 2024 19:35:49 +0000
Subject: [PATCH] Optimise check inside blocks and traverse blocks


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index ee7da44161fee45ce3de26d7e990eab3a3f5ff6e..b804964e620c49130c89d67e0371f06881b95eca 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1936,6 +1936,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     private void checkInsideBlocks(List<Entity.Movement> movements, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector) {
         if (this.isAffectedByBlocks()) {
             LongSet set = this.visitedBlocks;
+            final net.minecraft.world.level.chunk.ChunkAccess[] chunkCache = new net.minecraft.world.level.chunk.ChunkAccess[4]; // Sakura - optimise check inside blocks
 
             for (Entity.Movement movement : movements) {
                 Vec3 vec3 = movement.from;
@@ -1945,12 +1946,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         double d = vec31.get(axis);
                         if (d != 0.0) {
                             Vec3 vec32 = vec3.relative(axis.getPositive(), d);
-                            this.checkInsideBlocks(vec3, vec32, stepBasedCollector, set);
+                            this.checkInsideBlocks(vec3, vec32, stepBasedCollector, set, chunkCache); // Sakura - optimise check inside blocks
                             vec3 = vec32;
                         }
                     }
                 } else {
-                    this.checkInsideBlocks(movement.from(), movement.to(), stepBasedCollector, set);
+                    this.checkInsideBlocks(movement.from(), movement.to(), stepBasedCollector, set, chunkCache); // Sakura - optimise check inside blocks
                 }
             }
 
@@ -1958,7 +1959,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
-    private void checkInsideBlocks(Vec3 vec3, Vec3 vec31, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector, LongSet set) {
+    // Sakura start - optimise check inside blocks
+    private void checkInsideBlocks(Vec3 vec3, Vec3 vec31, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector,
+                                   LongSet set, net.minecraft.world.level.chunk.ChunkAccess[] chunkCache) {
+        // Sakura end - optimise check inside blocks
         // Sakura start - configure cannon physics
         double margin = this.physics.afterOrEqual(1_21_2) ? 1.0E-5f : this.physics.afterOrEqual(1_19_3) ? 1.0E-7 : 0.001;
         AABB aabb = this.makeBoundingBox(vec31).deflate(margin);
@@ -1972,7 +1976,20 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 if (!this.isAlive()) {
                     return false;
                 } else {
-                    BlockState blockState = this.level().getBlockState(pos);
+                    // Sakura start - optimise check inside blocks
+                    final int chunkX = pos.getX() >> 4;
+                    final int chunkZ = pos.getZ() >> 4;
+                    final int chunkKey = ((chunkX << 2) | chunkZ) & 3;
+                    net.minecraft.world.level.chunk.ChunkAccess chunk = chunkCache[chunkKey];
+                    if (chunk == null || chunk.locX != chunkX || chunk.locZ != chunkZ) {
+                        chunk = this.level().getChunkIfLoadedImmediately(chunkX, chunkZ);
+                        if (chunk == null) {
+                            return true;
+                        }
+                        chunkCache[chunkKey] = chunk;
+                    }
+                    final BlockState blockState = chunk.getBlockState(pos);
+                    // Sakura end - optimise check inside blocks
                     if (blockState.isAir()) {
                         this.debugBlockIntersection(pos, false, false);
                         return true;
diff --git a/net/minecraft/world/level/BlockGetter.java b/net/minecraft/world/level/BlockGetter.java
index 67c9393133f4509abf1bd352fbbc8e21dbb116e2..9e235b8bec7ed8da7c0cb099c47c3b23fbccb9b4 100644
--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -221,7 +221,7 @@ public interface BlockGetter extends LevelHeightAccessor {
         Vec3 vec3 = to.subtract(from);
         if (physics != null && physics.before(1_21_2) || vec3.lengthSqr() < Mth.square(0.99999F)) {
             // Sakura end - configure cannon physics
-            for (BlockPos blockPos : BlockPos.betweenClosed(boundingBox)) {
+            for (BlockPos blockPos : me.samsuik.sakura.utils.BlockPosIterator.iterable(boundingBox)) { // Sakura - optimise check inside blocks
                 if (!visitor.visit(blockPos, 0)) {
                     return false;
                 }
@@ -229,6 +229,20 @@ public interface BlockGetter extends LevelHeightAccessor {
 
             return true;
         } else {
+            // Sakura start - optimise check inside blocks
+            final boolean xZero = vec3.x() == 0.0;
+            final boolean yZero = vec3.y() == 0.0;
+            final boolean zZero = vec3.z() == 0.0;
+            if (xZero && yZero || yZero && zZero || xZero && zZero) {
+                int blockIndex = 0;
+                for (final BlockPos blockPos : me.samsuik.sakura.utils.BlockPosIterator.traverseArea(vec3, boundingBox)) {
+                    if (!visitor.visit(blockPos, blockIndex++)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            // Sakura end - optimise check inside blocks
             LongSet set = new LongOpenHashSet();
             Vec3 minPosition = boundingBox.getMinPosition();
             Vec3 vec31 = minPosition.subtract(vec3);
@@ -236,7 +250,7 @@ public interface BlockGetter extends LevelHeightAccessor {
             if (i < 0) {
                 return false;
             } else {
-                for (BlockPos blockPos1 : BlockPos.betweenClosed(boundingBox)) {
+                for (BlockPos blockPos1 : me.samsuik.sakura.utils.BlockPosIterator.iterable(boundingBox)) { // Sakura - optimise check inside blocks
                     if (!set.contains(blockPos1.asLong()) && !visitor.visit(blockPos1, i + 1)) {
                         return false;
                     }
