From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 25 Jun 2025 17:57:44 +0100
Subject: [PATCH] Protect blocks above a configured Y-level from explosions


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index a789ed6cb7bdd114a20743e171069cc3044a9fa3..14d561207dbbc44e0ac718f7a41909fb96787327 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -374,6 +374,11 @@ public class ServerExplosion implements Explosion {
             final Block block = blockState.getBlock();
             final me.samsuik.sakura.explosion.durable.DurableMaterial material = this.durableMaterials.get(block);
 
+            // Sakura start - protect blocks above a configured Y-level from explosions
+            if (this.level.sakuraConfig().cannons.explosion.protectBlocksAboveY.test(val -> pos.getY() >= val)) {
+                return Optional.of(net.minecraft.world.level.block.Blocks.BARRIER.getExplosionResistance());
+            }
+            // Sakura end - protect blocks above a configured Y-level from explosions
             if (material != null && material.replaceBlastResistance() && pos.getY() > this.level.getMinY()) {
                 return Optional.of(material.resistance());
             }
