From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 29 Mar 2026 20:25:30 +0100
Subject: [PATCH] Introduce limit on instant block falling


diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index fe1294fa83e1a3c42cbbd762d531b18b2a8cfe89..92b01f8941aec04a3c2cef369d7e6c1eeac11579 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -11,5 +11,6 @@ public final class ScheduleSakuraTasks {
         scheduler.endless((tick, level) -> level.sakuraEntityManager.tick(tick), 0);
         scheduler.endless((tick, level) -> level.obstructionCache.tick(tick), 0);
         scheduler.endless((tick, level) -> level.dispenseRelocationHandler.clear(), 0);
+        scheduler.endless((tick, level) -> level.fallingBlockColumnLimits.clear(), 0); // Sakura - limit instant block falling
     }
 }
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 82d8be7524a4e0aa08e5cc335150cf62e5e90efc..c7a6e7bd1beb4666b2f376bd15b59346d0fdcc0a 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -227,6 +227,13 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             state.hasProperty(BlockStateProperties.WATERLOGGED) ? state.setValue(BlockStateProperties.WATERLOGGED, false) : state
         );
         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.getFluidState().createLegacyBlock())) return entity; // CraftBukkit
+
+        // Sakura start - limit instant block falling
+        final int instantBlockFallingLimit = level.sakuraConfig().cannons.restrictions.instantBlockFallingLimit.or(-1);
+        if (instantBlockFallingLimit > 1 && level.fallingBlockColumnLimits.count(pos) >= instantBlockFallingLimit && level instanceof ServerLevel serverLevel) {
+            serverLevel.getBlockTicks().clearArea(new net.minecraft.world.level.levelgen.structure.BoundingBox(pos.above()));
+        }
+        // Sakura end - limit instant block falling
         // Sakura start - configure server mechanics
         if (entity.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
             level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 56c65f424b314a0a62593d3a14b18551c5991bb0..919fbe437885e88dea4eae5ed4ca22f6042d09f9 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -855,6 +855,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.mechanics.sake.DispenseRelocationHandler dispenseRelocationHandler = new me.samsuik.sakura.mechanics.sake.DispenseRelocationHandler(this); // Sakura - configure server mechanics
     public final me.samsuik.sakura.explosion.DurableBlockManager durabilityManager = new me.samsuik.sakura.explosion.DurableBlockManager(); // Sakura - explosion durable blocks
     public final me.samsuik.sakura.redstone.RedstoneWireCache redstoneWireCache = new me.samsuik.sakura.redstone.RedstoneWireCache(); // Sakura - cache vanilla and eigencraft wires
+    public final me.samsuik.sakura.entity.FallingBlockColumnLimits fallingBlockColumnLimits = new me.samsuik.sakura.entity.FallingBlockColumnLimits(); // Sakura - limit instant block falling
 
     protected Level(
         final WritableLevelData levelData,
