From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 25 Jun 2025 17:57:44 +0100
Subject: [PATCH] Protect blocks above a configured Y-level from explosions


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 71401708e65c9b06f30d124757fecf5f61eff957..d6ef61484019095399bb2ba3fcbd7aac30cff435 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -417,6 +417,11 @@ public class ServerExplosion implements Explosion {
                 return Optional.of(Blocks.BARRIER.getExplosionResistance());
             }
             // Sakura end - protect scaffolding from creepers
+            // Sakura start - protect blocks above a configured Y-level from explosions
+            if (this.level.sakuraConfig().cannons.explosion.protectBlocksAboveY.test(val -> pos.getY() >= val)) {
+                return Optional.of(net.minecraft.world.level.block.Blocks.BARRIER.getExplosionResistance());
+            }
+            // Sakura end - protect blocks above a configured Y-level from explosions
         }
 
         return this.damageCalculator.getBlockExplosionResistance(this, this.level, pos, blockState, fluidState);
