From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 25 Jun 2025 17:57:44 +0100
Subject: [PATCH] Protect blocks above a configured Y-level from explosions


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index de051d43cec6b48cafe054a4ff89d6fdcd971bca..51fa70aa5de941c1174b815622df65f530a0f479 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -357,6 +357,11 @@ public class ServerExplosion implements Explosion {
             final Block block = blockState.getBlock();
             final me.samsuik.sakura.explosion.durable.DurableMaterial material = this.level.localConfig().at(pos).durableMaterials.get(block);
 
+            // Sakura start - protect blocks above a configured Y-level from explosions
+            if (this.level.sakuraConfig().cannons.explosion.protectBlocksAboveY.test(val -> pos.getY() >= val)) {
+                return Optional.of(net.minecraft.world.level.block.Blocks.BARRIER.getExplosionResistance());
+            }
+            // Sakura end - protect blocks above a configured Y-level from explosions
             if (material != null && material.replaceBlastResistance() && pos.getY() > this.level.getMinY()) {
                 return Optional.of(material.resistance());
             }
