From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 29 Mar 2026 20:25:30 +0100
Subject: [PATCH] Introduce limit on instant block falling


diff --git a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
index 7d2df500c90e0451ec2cf117f44413b3593fdd62..fef94d0a0cce8c0ca0a0a73e7328e26f206e50e7 100644
--- a/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
+++ b/me/samsuik/sakura/scheduler/ScheduleSakuraTasks.java
@@ -12,5 +12,6 @@ public final class ScheduleSakuraTasks {
         scheduler.endless((tick, level, cb) -> level.mergeHandler.expire(tick), 200); // Sakura - merge cannon entities
         scheduler.endless((tick, level, cb) -> level.densityCache.expire(tick), 0); // Sakura - explosion density cache
         scheduler.endless((tick, level, cb) -> level.dispenseRelocationHandler.clear(), 0); // Sakura - configure server mechanics
+        scheduler.endless((tick, level, cb) -> level.fallingBlockColumnLimits.clear(), 0); // Sakura - limit instant block falling
     }
 }
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index ac2effcb0cafa7f9f2a84ddd4d8cb28c302a7085..12a0f328482430827379742cc3230eaa6d7c2a72 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -203,6 +203,13 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
             state.hasProperty(BlockStateProperties.WATERLOGGED) ? state.setValue(BlockStateProperties.WATERLOGGED, false) : state
         );
         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.getFluidState().createLegacyBlock())) return entity; // CraftBukkit
+
+        // Sakura start - limit instant block falling
+        final int instantBlockFallLimit = level.sakuraConfig().cannons.restrictions.instantBlockFallLimit.or(-1);
+        if (instantBlockFallLimit > 1 && level.fallingBlockColumnLimits.count(pos) >= instantBlockFallLimit && level instanceof ServerLevel serverLevel) {
+            serverLevel.getBlockTicks().clearArea(new net.minecraft.world.level.levelgen.structure.BoundingBox(pos.above()));
+        }
+        // Sakura end - limit instant block falling
         // Sakura start - configure server mechanics
         if (entity.mechanicsTarget().atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_18_2)) {
             level.setBlock(pos, state.getFluidState().createLegacyBlock(), Block.UPDATE_ALL);
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 7e9ad3d37bf23da11e47a3ab59945fb11c1ce2d7..aed148d8d206f7122d25d99fedab6062369e3a81 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -853,6 +853,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public final me.samsuik.sakura.explosion.DurableBlockManager durabilityManager = new me.samsuik.sakura.explosion.DurableBlockManager(); // Sakura - explosion durable blocks
     public final me.samsuik.sakura.mechanics.sake.DispenseRelocationHandler dispenseRelocationHandler = new me.samsuik.sakura.mechanics.sake.DispenseRelocationHandler(this); // Sakura - configure server mechanics
     public final me.samsuik.sakura.redstone.RedstoneWireCache redstoneWireCache = new me.samsuik.sakura.redstone.RedstoneWireCache(); // Sakura - cache vanilla and eigencraft wires
+    public final me.samsuik.sakura.entity.FallingBlockColumnLimits fallingBlockColumnLimits = new me.samsuik.sakura.entity.FallingBlockColumnLimits(); // Sakura - limit instant block falling
 
     protected Level(
         final WritableLevelData levelData,
