From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 25 Sep 2025 11:26:32 +0100
Subject: [PATCH] fixup! Configure cannon physics


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index ba4979657e65bfc97bb597e7aaa6ce6d030c1346..a88add750945a1a374d248ba6b2eb0688d9bec18 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -975,16 +975,6 @@ public class ServerExplosion implements Explosion {
         return blockDensity;
     }
 
-    // Sakura start - configure server mechanics
-    private float sakura_getSeenPercent(final Vec3 vec3d, final Entity entity) {
-        if (this.mechanicsTarget.atLeast(me.samsuik.sakura.mechanics.MechanicVersion.v1_16)) {
-            return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
-        } else {
-            return getSeenPercent(vec3d, entity);
-        }
-    }
-    // Sakura end - configure server mechanics
-
     static class CacheKey {
         private final Level world;
         private final double posX, posY, posZ;
