--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -66,7 +_,7 @@
         }
 
         boolean naturalRegen = level.getGameRules().get(GameRules.NATURAL_HEALTH_REGENERATION);
-        if (naturalRegen && this.saturationLevel > 0.0F && player.isHurt() && this.foodLevel >= FoodConstants.MAX_FOOD) {
+        if (naturalRegen && this.saturationLevel > 0.0F && player.isHurt() && this.foodLevel >= FoodConstants.MAX_FOOD && player.level().sakuraConfig().players.combat.fastHealthRegen) { // Sakura - configure fast health regen
             this.tickTimer++;
             if (this.tickTimer >= this.saturatedRegenRate) { // CraftBukkit
                 float saturationSpent = Math.min(this.saturationLevel, 6.0F);
