--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -173,7 +_,7 @@
             .defaultOptions(PaperConfigurations::defaultOptions);
     }
 
-    private static ConfigurationOptions defaultOptions(ConfigurationOptions options) {
+    public static ConfigurationOptions defaultOptions(ConfigurationOptions options) { // Sakura - sakura configuration files
         return options.serializers(builder -> builder
             .register(MapSerializer.TYPE, new MapSerializer(false))
             .register(new EnumValueSerializer())
@@ -470,7 +_,7 @@
     }
 
     // Symlinks are not correctly checked in createDirectories
-    static void createDirectoriesSymlinkAware(Path path) throws IOException {
+    public static void createDirectoriesSymlinkAware(Path path) throws IOException { // Sakura - package-protected -> public
         if (!Files.isDirectory(path)) {
             Files.createDirectories(path);
         }
