--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -179,7 +_,7 @@
             .defaultOptions(PaperConfigurations::defaultOptions);
     }
 
-    private static ConfigurationOptions defaultOptions(ConfigurationOptions options) {
+    public static ConfigurationOptions defaultOptions(ConfigurationOptions options) { // Sakura - sakura configuration files; private -> public
         return options.serializers(builder -> builder
             .register(MapSerializer.TYPE, new MapSerializer(false))
             .register(new EnumValueSerializer())
@@ -341,7 +_,7 @@
         }
     }
 
-    private static List<Definition<? extends Annotation, ?, ? extends FieldProcessor.Factory<?, ?>>> defaultFieldProcessors() {
+    public static List<Definition<? extends Annotation, ?, ? extends FieldProcessor.Factory<?, ?>>> defaultFieldProcessors() { // Sakura - sakura configuration files; private -> public
         return List.of(
             MergeMap.DEFINITION
         );
@@ -482,7 +_,7 @@
     }
 
     // Symlinks are not correctly checked in createDirectories
-    static void createDirectoriesSymlinkAware(Path path) throws IOException {
+    public static void createDirectoriesSymlinkAware(Path path) throws IOException { // Sakura - sakura configuration files; package-protected -> public
         if (!Files.isDirectory(path)) {
             Files.createDirectories(path);
         }
