--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -980,6 +_,12 @@
         victim.expToDrop = event.getDroppedExp();
 
         Runnable dropAll = () -> {
+            // Sakura start - mobs insert items into hoppers on death
+            if (level.sakuraConfig().entity.insertItemsIntoHoppersOnDeath) {
+                me.samsuik.sakura.container.InsertItemsIntoHoppers.onMobDeath(level, entity.getHandle(), drops);
+            }
+            // Sakura end - mobs insert items into hoppers on death
+
             dropAllItems(drops, item -> victim.spawnAtLocation(level, item));
         };
 
