--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -52,6 +_,19 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        // Sakura start - use random chance for crop growth
+        if (level.sakuraConfig().environment.crops.useRandomChanceToGrow) {
+            final int modifier = level.spigotConfig.cactusModifier;
+            if (random.nextFloat() >= modifier / (100.0f * 16)) {
+                return;
+            }
+            // set crop age to max so it grows right away
+            state = state.setValue(CactusBlock.AGE, AGE.max);
+        }
+        this.ageAndGrow(state, level, pos, random);
+    }
+    private void ageAndGrow(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Sakura end - use random chance for crop growth
         BlockPos blockPos = pos.above();
         if (level.isEmptyBlock(blockPos)) {
             int i = 1;
@@ -75,6 +_,11 @@
                 }
                 // Paper end
                 BlockState blockState = state.setValue(AGE, 0);
+                // Sakura start - use random chance for crop growth; fix cactus growing next to a block and not breaking
+                if (level.sakuraConfig().environment.crops.useRandomChanceToGrow) {
+                    level.neighborShapeChanged(Direction.UP, blockPos, pos, state, 4, 1);
+                }
+                // Sakura end - use random chance for crop growth; fix cactus growing next to a block and not breaking
                 level.setBlock(pos, blockState, 260);
                 level.neighborChanged(blockState, blockPos, this, null, false);
             }
