--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -169,7 +_,7 @@
 
     // Paper start - Configurable speed for water flowing over lava
     public int getFlowSpeed(Level level, BlockPos pos) {
-        if (net.minecraft.core.registries.BuiltInRegistries.FLUID.wrapAsHolder(this.fluid).is(FluidTags.WATER)) {
+        if (this.fluid.getTickDelay(level) != level.paperConfig().environment.waterOverLavaFlowSpeed && net.minecraft.core.registries.BuiltInRegistries.FLUID.wrapAsHolder(this.fluid).is(FluidTags.WATER)) { // Sakura - avoid expensive lava search
             if (
                 isLava(level, pos.north()) ||
                 isLava(level, pos.south()) ||
@@ -179,7 +_,7 @@
                 return level.paperConfig().environment.waterOverLavaFlowSpeed;
             }
         }
-        return this.fluid.getTickDelay(level);
+        return this.fluid.getTickDelay(level, pos); // Sakura - lava flow speed api
     }
 
     private static boolean isLava(Level level, BlockPos pos) {
@@ -240,6 +_,11 @@
 
             for (Direction direction : POSSIBLE_FLOW_DIRECTIONS) {
                 BlockPos neighbourPos = pos.relative(direction.getOpposite());
+                // Sakura start - configure fluid ticking outside the world border
+                if (level.sakuraConfig().environment.disableFluidsFlowingThroughTheWorldBorder && !level.getWorldBorder().isWithinBounds(blockPos)) {
+                    continue;
+                }
+                // Sakura end - configure fluid ticking outside the world border
                 if (level.getFluidState(neighbourPos).is(FluidTags.WATER)) {
                     Block convertToBlock = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
                     // CraftBukkit start
