--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -48,6 +_,20 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Sakura start - use random chance for crop growth
+        BlockState agedState = state;
+        if (level.sakuraConfig().environment.crops.useRandomChanceToGrow) {
+            final int modifier = level.spigotConfig.caneModifier;
+            if (random.nextFloat() >= modifier / (100.0f * 16)) {
+                return;
+            }
+            // set crop age to max so it grows right away
+            agedState = state.setValue(SugarCaneBlock.AGE, AGE.max);
+        }
+        this.ageAndGrow(agedState, level, pos, random);
+    }
+    private void ageAndGrow(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Sakura end - use random chance for crop growth
         if (level.isEmptyBlock(pos.above())) {
             int height = 1;
 
