--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -48,6 +_,19 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        // Sakura start - use random chance for crop growth
+        if (level.sakuraConfig().environment.crops.useRandomChanceToGrow) {
+            final int modifier = level.spigotConfig.caneModifier;
+            if (random.nextFloat() >= modifier / (100.0f * 16)) {
+                return;
+            }
+            // set crop age to max so it grows right away
+            state = state.setValue(SugarCaneBlock.AGE, AGE.max);
+        }
+        this.ageAndGrow(state, level, pos, random);
+    }
+    private void ageAndGrow(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Sakura end - use random chance for crop growth
         if (level.isEmptyBlock(pos.above())) {
             int i = 1;
 
@@ -61,6 +_,7 @@
                 if (ageValue >= 15 || (modifier != 100 && random.nextFloat() < (modifier / (100.0F * 16)))) { // Spigot - SPIGOT-7159: Better modifier resolution
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos.above(), this.defaultBlockState(), 3); // CraftBukkit
                     level.setBlock(pos, state.setValue(AGE, 0), Block.UPDATE_NONE);
+                    // Sakura - use random chance for crop growth; conflict on change
                 } else if (modifier == 100 || random.nextFloat() < (modifier / (100.0F * 16))) { // Spigot - SPIGOT-7159: Better modifier resolution
                     level.setBlock(pos, state.setValue(AGE, ageValue + 1), Block.UPDATE_NONE);
                 }
