--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -238,7 +_,7 @@
         if (shouldBePressed) {
             level.scheduleTick(new BlockPos(pos), this, 10);
         } else if (wasPressed) {
-            level.scheduleTick(new BlockPos(pos), this, 0);
+            level.scheduleTick(new BlockPos(pos), this, 1); // Sakura - MC-310372; Tripwires depowering breaks cross-chunk block tick processing order
         }
     }
 
