--- a/net/minecraft/world/level/block/WebBlock.java
+++ b/net/minecraft/world/level/block/WebBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
+import net.minecraft.world.phys.shapes.EntityCollisionContext;
 
 public class WebBlock extends Block {
     public static final MapCodec<WebBlock> CODEC = simpleCodec(WebBlock::new);
@@ -22,6 +_,31 @@
     public WebBlock(BlockBehaviour.Properties properties) {
         super(properties);
     }
+
+    // Sakura start - configure players colliding with cobwebs
+    @Override
+    protected final net.minecraft.world.phys.shapes.VoxelShape getCollisionShape(
+        final BlockState state,
+        final net.minecraft.world.level.BlockGetter blockGetter,
+        final BlockPos pos,
+        final net.minecraft.world.phys.shapes.CollisionContext context
+    ) {
+        if (blockGetter instanceof Level level
+            && level.sakuraConfig().players.collideWithCobwebs
+            && context instanceof EntityCollisionContext entityContext
+            && entityContext.getEntity() instanceof net.minecraft.world.entity.player.Player) {
+            final BlockPos blockPosAbove = pos.above();
+            final BlockState stateAbove = level.getBlockState(blockPosAbove);
+
+            // If the block above is a full block then this cobweb should have a solid collision (for players)
+            if (stateAbove.getCollisionShape(level, blockPosAbove, context).moonrise$isFullBlock()) {
+                return net.minecraft.world.phys.shapes.Shapes.block();
+            }
+        }
+
+        return super.getCollisionShape(state, blockGetter, pos, context);
+    }
+    // Sakura end - configure players colliding with cobwebs
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
