--- a/net/minecraft/world/level/block/WebBlock.java
+++ b/net/minecraft/world/level/block/WebBlock.java
@@ -23,6 +_,31 @@
         super(properties);
     }
 
+    // Sakura start - configure players colliding with cobwebs
+    @Override
+    protected final net.minecraft.world.phys.shapes.VoxelShape getCollisionShape(
+        final BlockState state,
+        final net.minecraft.world.level.BlockGetter blockGetter,
+        final BlockPos pos,
+        final net.minecraft.world.phys.shapes.CollisionContext context
+    ) {
+        if (blockGetter instanceof final Level gameLevel
+            && gameLevel.sakuraConfig().players.collideWithCobwebs
+            && context instanceof net.minecraft.world.phys.shapes.EntityCollisionContext entityContext
+            && entityContext.getEntity() instanceof net.minecraft.world.entity.player.Player) {
+            final BlockPos blockPosAbove = pos.above();
+            final BlockState stateAbove = gameLevel.getBlockState(blockPosAbove);
+
+            // If the block above is a full block then this cobweb should have a solid collision (for players)
+            if (stateAbove.getCollisionShape(gameLevel, blockPosAbove, context).moonrise$isFullBlock()) {
+                return net.minecraft.world.phys.shapes.Shapes.block();
+            }
+        }
+
+        return super.getCollisionShape(state, blockGetter, pos, context);
+    }
+    // Sakura end - configure players colliding with cobwebs
+
     @Override
     protected void entityInside(
         final BlockState state,
