--- a/net/minecraft/world/entity/animal/IronGolem.java
+++ b/net/minecraft/world/entity/animal/IronGolem.java
@@ -229,6 +_,13 @@
         }
     }
 
+    // Sakura start - configure iron golems taking fall damage
+    @Override
+    protected final boolean isFallDamageImmune() {
+        return !this.level().sakuraConfig().entity.ironGolemsTakeFalldamage && this.isFallDamageImmune();
+    }
+    // Sakura end - configure iron golems taking fall damage
+
     public int getAttackAnimationTick() {
         return this.attackAnimationTick;
     }
