--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -184,6 +_,36 @@
         return (org.bukkit.craftbukkit.entity.CraftHumanEntity) super.getBukkitEntity();
     }
     // CraftBukkit end
+    private long lastSprintKnockback = -1; // Sakura - configure entity knockback
+    public long lastLungeMillis = 0; // Sakura - configure lunge enchantment
+    // Sakura start - player poses shrink collision box & prevent players glitching through blocks
+    @Override
+    public net.minecraft.world.entity.EntityDimensions getDefaultDimensions(final Pose pose) {
+        final net.minecraft.world.entity.EntityDimensions dimensions = super.getDefaultDimensions(pose);
+        if (!this.level().sakuraConfig().players.posesShrinkCollisionBox && dimensions.height() != Avatar.STANDING_DIMENSIONS.height()) {
+            return STANDING_DIMENSIONS;
+        }
+        return dimensions;
+    }
+
+    public final boolean isInsideAnyCollision() {
+        return this.insideAnyCollisionAt(this.level(), this.position(), Pose.STANDING);
+    }
+
+    public final boolean insideAnyCollisionAt(final Level level, final Vec3 position) {
+        final Pose pose = this.getVehicle() != null || !this.level().sakuraConfig().players.posesShrinkCollisionBox
+            ? Pose.STANDING
+            : Pose.SWIMMING;
+        return this.insideAnyCollisionAt(level, position, pose);
+    }
+
+    public final boolean insideAnyCollisionAt(final Level level, final Vec3 position, final Pose pose) {
+        final net.minecraft.world.entity.EntityDimensions entityPoseDimensions = this.getDimensions(pose);
+        final AABB poseCollisionBox = entityPoseDimensions.makeBoundingBox(position);
+        return !level.noCollision(this, poseCollisionBox.deflate(1.0e-7))
+            || !level.getWorldBorder().isWithinBounds(poseCollisionBox);
+    }
+    // Sakura end - prevent players glitching through blocks & prevent players glitching through blocks
 
     public Player(final Level level, final GameProfile gameProfile) {
         super(EntityType.PLAYER, level);
@@ -356,7 +_,7 @@
         return new ItemCooldowns();
     }
 
-    protected void updatePlayerPose() {
+    public void updatePlayerPose() {
         if (this.canPlayerFitWithinBlocksAndEntitiesWhen(Pose.SWIMMING)) {
             Pose desiredPose = this.getDesiredPose();
             Pose actualPose;
@@ -699,6 +_,10 @@
     public boolean isInvulnerableTo(final ServerLevel level, final DamageSource source) {
         if (super.isInvulnerableTo(level, source)) {
             return true;
+            // Sakura start - allow disabling explosions hurting players
+        } else if (!this.level().sakuraConfig().cannons.explosion.explosionsHurtPlayers && source.is(DamageTypeTags.IS_EXPLOSION)) {
+            return true;
+            // Sakura end - allow disabling explosions hurting players
         } else if (source.is(DamageTypeTags.IS_DROWNING)) {
             return !level.getGameRules().get(GameRules.DROWNING_DAMAGE);
         } else if (source.is(DamageTypeTags.IS_FALL)) {
@@ -736,6 +_,12 @@
                     }
                 }
 
+                // Sakura start - configure the max damage a player can take
+                final double maxDamage = level.sakuraConfig().players.combat.maxDamage.or(-1.0);
+                if (damage > 0.0 && maxDamage >= 0.0) {
+                    damage = Math.min(damage, (float) maxDamage);
+                }
+                // Sakura end - configure the max damage a player can take
                 // return damage != 0.0F && super.hurtServer(level, source, damage);
                 // CraftBukkit start - Don't filter out 0 damage
                 boolean damaged = super.hurtServer(level, source, damage);
@@ -1020,19 +_,28 @@
 
         if (playerAttackEntityEvent.callEvent() && willAttack) { // Logic moved to willAttack local variable.
         // Paper end - PlayerAttackEntityEvent
-            float baseDamage = this.isAutoSpinAttack() ? this.autoSpinAttackDmg : (float)this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+            float baseDamage = this.isAutoSpinAttack() ? this.autoSpinAttackDmg : this.getAttackDamageFromAttributes(); // Sakura - legacy combat mechanics
             ItemStack attackingItemStack = this.getWeaponItem();
             DamageSource damageSource = this.createAttackSource(attackingItemStack); final DamageSource dmgSourceFinal = damageSource; // Paper - damage events
             float attackStrengthScale = this.getAttackStrengthScale(0.5F);
+            // Sakura start - legacy combat mechanics
+            if (this.level().sakuraConfig().players.combat.legacyCombatMechanics) {
+                attackStrengthScale = 1.0f;
+            }
             float magicBoost = attackStrengthScale * (this.getEnchantedDamage(entity, baseDamage, damageSource) - baseDamage);
-            baseDamage *= this.baseDamageScaleFactor();
+            if (!this.level().sakuraConfig().players.combat.legacyCombatMechanics) {
+                baseDamage *= this.baseDamageScaleFactor();
+            } else if (magicBoost != 0.0) {
+                magicBoost += me.samsuik.sakura.player.combat.CombatUtil.calculateLegacySharpnessDamage(this, attackingItemStack, damageSource);
+            }
+            // Sakura end - legacy combat mechanics
             this.onAttack(entity); // Paper
             final float dmgFinal = magicBoost; // Paper - damage events
             if (!this.deflectProjectile(entity, () -> !org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(entity, dmgSourceFinal, dmgFinal, false))) {
                 if (baseDamage > 0.0F || magicBoost > 0.0F) {
                     boolean fullStrengthAttack = attackStrengthScale > 0.9F;
                     boolean knockbackAttack;
-                    if (this.isSprinting() && fullStrengthAttack) {
+                    if (this.isSprinting() && (!this.level().sakuraConfig().players.knockback.sprinting.requireFullAttack || fullStrengthAttack)) { // Sakura - configure entity knockback
                         this.playServerSideSound(SoundEvents.PLAYER_ATTACK_KNOCKBACK);
                         knockbackAttack = true;
                     } else {
@@ -1057,8 +_,11 @@
                     Vec3 oldMovement = entity.getDeltaMovement();
                     boolean wasHurt = entity.hurtOrSimulate(damageSource, totalDamage);
                     if (wasHurt) {
-                        this.causeExtraKnockback(entity, this.getKnockback(entity, damageSource) + (knockbackAttack ? 0.5F : 0.0F), oldMovement);
-                        if (sweepAttack) {
+                        // Sakura start - configure entity knockback; extra sprinting knockback & allow disabling sweep attacks
+                        final float extraKnockbackToApply = this.level().sakuraConfig().players.combat.legacyCombatMechanics ? 1.0f : 0.5f;
+                        this.causeExtraKnockback(entity, this.getKnockback(entity, damageSource) + (knockbackAttack ? extraKnockbackToApply : 0.0F), oldMovement, !knockbackAttack);
+                        if (sweepAttack && this.level().sakuraConfig().players.combat.allowSweepAttacks) {
+                            // Sakura end - configure entity knockback; extra sprinting knockback & allow disabling sweep attacks
                             this.doSweepAttack(entity, baseDamage, damageSource, attackStrengthScale);
                         }
 
@@ -1109,7 +_,7 @@
             && !this.isMobilityRestricted()
             && !this.isPassenger()
             && entity instanceof LivingEntity
-            && !this.isSprinting();
+            && (this.level().sakuraConfig().players.combat.legacyCombatMechanics || !this.isSprinting()); // Sakura - legacy combat mechanics
     }
 
     private boolean isSweepAttack(final boolean fullStrengthAttack, final boolean criticalAttack, final boolean knockbackAttack) {
@@ -1150,7 +_,7 @@
         if (entity instanceof LivingEntity) {
             float actualDamage = oldLivingEntityHealth - ((LivingEntity)entity).getHealth();
             this.awardStat(Stats.DAMAGE_DEALT, Math.round(actualDamage * 10.0F));
-            if (this.level() instanceof ServerLevel && actualDamage > 2.0F) {
+            if (this.level() instanceof ServerLevel && actualDamage > 2.0F && !this.level().sakuraConfig().players.combat.oldSoundsAndParticleEffects) { // Sakura - old combat sounds and particles
                 int count = (int)(actualDamage * 0.5);
                 ((ServerLevel)this.level())
                     .sendParticles(ParticleTypes.DAMAGE_INDICATOR, entity.getX(), entity.getY(0.5), entity.getZ(), count, 0.1, 0.0, 0.1, 0.2);
@@ -1192,8 +_,32 @@
 
     @Override
     public void causeExtraKnockback(final Entity entity, final float knockbackAmount, final Vec3 oldMovement) {
+        // Sakura start - configure entity knockback; extra sprinting knockback
+        this.causeExtraKnockback(entity, knockbackAmount, oldMovement, true);
+    }
+
+    private void causeExtraKnockback(
+        final Entity entity,
+        float knockbackAmount,
+        final Vec3 oldMovement,
+        final boolean bypassKnockbackDelay
+    ) {
         if (knockbackAmount > 0.0F) {
-            if (entity instanceof LivingEntity livingTarget) {
+            final long millis = System.currentTimeMillis();
+            final long sinceLastKnockback = millis - this.lastSprintKnockback;
+            if (!bypassKnockbackDelay) {
+                double knockbackToApply = 0.0;
+                if (sinceLastKnockback >= this.level().sakuraConfig().players.knockback.sprinting.knockbackDelay.value().orElse(0)) {
+                    knockbackToApply = this.level().sakuraConfig().players.knockback.sprinting.extraKnockback;
+                    this.lastSprintKnockback = millis;
+                }
+                knockbackAmount = (knockbackAmount - 1.0f) + ((float) knockbackToApply * 2.0f);
+            }
+            //noinspection StatementWithEmptyBody
+            if (knockbackAmount == 0.0f) {
+                // required
+            } else if (entity instanceof LivingEntity livingTarget) {
+                // Sakura end - configure entity knockback; extra sprinting knockback
                 livingTarget.knockback(knockbackAmount, Mth.sin(this.getYRot() * Mth.DEG_TO_RAD), -Mth.cos(this.getYRot() * Mth.DEG_TO_RAD), this, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.ENTITY_ATTACK); // Paper - knockback events
             } else {
                 entity.push(-Mth.sin(this.getYRot() * Mth.DEG_TO_RAD) * knockbackAmount, 0.1, Mth.cos(this.getYRot() * Mth.DEG_TO_RAD) * knockbackAmount, this); // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
@@ -1252,7 +_,15 @@
                     nearby.lastDamageCancelled = false;
                     if (nearby.hurtServer(serverLevel, damageSource.knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.ENTITY_SWEEP_ATTACK), enchantedDamage) && !nearby.lastDamageCancelled) {
                         // Paper end - Only apply knockback if the event is not canceled
-                        nearby.knockback(0.4F, Mth.sin(this.getYRot() * Mth.DEG_TO_RAD), -Mth.cos(this.getYRot() * Mth.DEG_TO_RAD), this, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.SWEEP_ATTACK); // Paper - knockback events
+                        // Sakura start - configure entity knockback
+                        nearby.knockback(
+                            (float) this.level().sakuraConfig().players.knockback.sweepingEdgeKnockback,
+                            Mth.sin(this.getYRot() * (float) (Math.PI / 180.0)),
+                            -Mth.cos(this.getYRot() * (float) (Math.PI / 180.0)),
+                            this,
+                            io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.SWEEP_ATTACK // Paper - knockback events
+                        );
+                        // Sakura end - configure entity knockback
                         EnchantmentHelper.doPostAttackEffects(serverLevel, nearby, damageSource);
                     }
                 }
@@ -1281,6 +_,7 @@
         return 0.2F + attackStrengthScale * attackStrengthScale * 0.8F;
     }
 
+    /// 
     @Override
     public boolean stabAttack(
         final EquipmentSlot slot, final Entity target, float baseDamage, final boolean dealsDamage, final boolean dealsKnockback, final boolean dismounts
@@ -1678,6 +_,11 @@
 
     // Paper start - send while respecting visibility
     private static void sendSoundEffect(Player fromEntity, double x, double y, double z, SoundEvent soundEffect, SoundSource soundCategory, float volume, float pitch) {
+        // Sakura start - old combat sounds and particles
+        if (fromEntity.level().sakuraConfig().players.combat.oldSoundsAndParticleEffects) {
+            return;
+        }
+        // Sakura end - old combat sounds and particles
         fromEntity.level().playSound(fromEntity, x, y, z, soundEffect, soundCategory, volume, pitch); // This will not send the effect to the entity itself
         if (fromEntity instanceof ServerPlayer serverPlayer) {
             serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundSoundPacket(net.minecraft.core.registries.BuiltInRegistries.SOUND_EVENT.wrapAsHolder(soundEffect), soundCategory, x, y, z, volume, pitch, fromEntity.random.nextLong()));
