--- a/net/minecraft/world/item/component/KineticWeapon.java
+++ b/net/minecraft/world/item/component/KineticWeapon.java
@@ -108,7 +_,7 @@
             double attackerSpeedProjection = attackerLookVector.dot(getMotion(livingEntity));
             float actionFactor = livingEntity instanceof Player ? 1.0F : 0.2F;
             AttackRange attackRange = livingEntity.getAttackRangeWith(stack);
-            double baseMobDamage = livingEntity.getAttributeBaseValue(Attributes.ATTACK_DAMAGE);
+            double attributeBaseValue = entity.getAttackDamageFromAttributes(); // Sakura - legacy combat mechanics; add support for modifying the spear attack damage
             boolean affected = false;
 
             for (EntityHitResult hitResult : ProjectileUtil.getHitEntitiesAlong(
