--- a/net/minecraft/world/item/component/KineticWeapon.java
+++ b/net/minecraft/world/item/component/KineticWeapon.java
@@ -108,7 +_,7 @@
             double d = lookAngle.dot(getMotion(entity));
             float f = entity instanceof Player ? 1.0F : 0.2F;
             AttackRange attackRange = entity.entityAttackRange();
-            double attributeBaseValue = entity.getAttributeBaseValue(Attributes.ATTACK_DAMAGE);
+            double attributeBaseValue = entity.getAttackDamageFromAttributes(); // Sakura - legacy combat mechanics; add support for modifying the spear attack damage
             boolean flag = false;
 
             for (EntityHitResult entityHitResult : ProjectileUtil.getHitEntitiesAlong(
