--- a/net/minecraft/world/item/component/PiercingWeapon.java
+++ b/net/minecraft/world/item/component/PiercingWeapon.java
@@ -71,7 +_,7 @@
     }
 
     public void attack(final LivingEntity attacker, final EquipmentSlot hand) {
-        float damage = (float)attacker.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float damage = attacker.getAttackDamageFromAttributes(); // Sakura - legacy combat mechanics; add support for modifying the spear attack damage
         ItemStack weaponItem = attacker.getItemBySlot(hand);
         AttackRange attackRange = attacker.getAttackRangeWith(weaponItem);
         boolean hitSomething = false;
