--- a/net/minecraft/world/item/component/PiercingWeapon.java
+++ b/net/minecraft/world/item/component/PiercingWeapon.java
@@ -71,7 +_,7 @@
     }
 
     public void attack(LivingEntity entity, EquipmentSlot slot) {
-        float f = (float)entity.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float f = entity.getAttackDamageFromAttributes(); // Sakura - legacy combat mechanics; add support for modifying the spear attack damage
         AttackRange attackRange = entity.entityAttackRange();
         boolean flag = false;
 
