--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -276,7 +_,7 @@
         if (this.getUnhappyCounter() > 0) {
             this.setUnhappyCounter(this.getUnhappyCounter() - 1);
         }
-        if (this.isEffectiveAi()) {
+        if (this.isEffectiveAi() && this.aware) { // Sakura - fix nerfed mobs ticking goals while inactive
             if (this.level().spigotConfig.tickInactiveVillagers) {
                 this.customServerAiStep(this.level().getMinecraftWorld());
             } else {
