--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -74,8 +_,55 @@
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
 
+    // Sakura start - configure cannon mechanics
+    public boolean heightParity;
+    public boolean floatingPointFix;
+
+    @Override
+    public double getEyeY() {
+        return this.heightParity ? this.getY() : super.getEyeY();
+    }
+
+    private static boolean isFallingBlockInBounds(final Level level, final BlockPos pos) {
+        // Prevent stacking one block below build limit
+        if (!level.sakuraConfig().cannons.sand.canStackAtWorldHeight && pos.getY() >= level.getMaxY() - 1) {
+            return false;
+        }
+
+        final Vec3 center = Vec3.atBottomCenterOf(pos);
+        // Prevent stacking directly against the world border
+        return level.sakuraConfig().cannons.sand.canStackAgainstBorder
+            || level.getWorldBorder().isWithinBounds(center.x(), center.z(), -1.0);
+    }
+
+    private BlockPos patchedBlockPosition() {
+        // Mitigates MC-261789 - falling blocks breaking when landing on a block below y 0
+        // This fix allows some tech that uses "indirect clipping" to still function.
+        // L stackers and midairs may still encounter issues if they shoot too high.
+        return BlockPos.containing(this.getX(), this.getY() + 1.0e-12, this.getZ());
+    }
+
+    private boolean canReplaceBlock(final BlockState state) {
+        return this.level().sakuraConfig().cannons.sand.replaceBlocksWithoutCollision
+            && state.moonrise$emptyCollisionShape()
+            && this.getBlockStateOn().is(this.blockState.getBlock());
+    }
+
+    @Override
+    public final void sakura$addedToWorld() {
+        final me.samsuik.sakura.configuration.local.CachedLocalConfiguration config = this.level().localConfig().at(this.blockPosition());
+        if (!this.heightParity) { // default
+            this.heightParity = config.heightParity;
+        }
+
+        this.floatingPointFix = config.floatingPointFix;
+    }
+    // Sakura end - configure cannon mechanics
+
     public FallingBlockEntity(final EntityType<? extends FallingBlockEntity> type, final Level level) {
         super(type, level);
+        this.dropItem = level.sakuraConfig().cannons.sand.dropItems; // Sakura - configure falling blocks dropping items
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     public FallingBlockEntity(final Level level, final double x, final double y, final double z, final BlockState blockState) {
@@ -167,8 +_,8 @@
             // Paper end - Configurable falling blocks height nerf
             this.handlePortal();
             if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
-                BlockPos pos = this.blockPosition();
-                boolean isConcrete = this.blockState.getBlock() instanceof ConcretePowderBlock;
+                BlockPos pos = this.floatingPointFix ? this.patchedBlockPosition() : this.blockPosition(); // Sakura - fix falling block floating point errors
+                boolean isConcrete = serverLevel.sakuraConfig().cannons.sand.concreteSolidifyInWater && this.blockState.getBlock() instanceof ConcretePowderBlock; // Sakura - configure concrete solidifying in water
                 boolean isStuckInWater = isConcrete && this.level().getFluidState(pos).is(FluidTags.WATER);
                 double moveVec = this.getDeltaMovement().lengthSqr();
                 if (isConcrete && moveVec > 1.0) {
@@ -184,7 +_,7 @@
                     }
                 }
 
-                if (!this.onGround() && !isStuckInWater) {
+                if (!this.onGround() && !isStuckInWater || serverLevel.sakuraConfig().cannons.sand.despawnInsideMovingPistons && this.autoExpire && this.time > 600) { // Sakura - allow falling blocks to despawn inside moving pistons
                     if ((this.time > 100 && this.autoExpire) && (pos.getY() <= this.level().getMinY() || pos.getY() > this.level().getMaxY()) || (this.time > 600 && this.autoExpire)) { // Paper - Expand FallingBlock API
                         if (this.dropItem && serverLevel.getGameRules().get(GameRules.ENTITY_DROPS)) {
                             this.spawnAtLocation(serverLevel, block);
@@ -200,9 +_,10 @@
                             boolean mayReplace = currentState.canBeReplaced(
                                 new DirectionalPlaceContext(this.level(), pos, Direction.DOWN, ItemStack.EMPTY, Direction.UP)
                             );
+                            mayReplace = mayReplace || this.canReplaceBlock(currentState); // Sakura - replace blocks without collision
                             boolean wouldContinueFalling = FallingBlock.isFree(this.level().getBlockState(pos.below())) && (!isConcrete || !isStuckInWater);
                             boolean wouldSurvive = this.blockState.canSurvive(this.level(), pos) && !wouldContinueFalling;
-                            if (mayReplace && wouldSurvive) {
+                            if (mayReplace && wouldSurvive && isFallingBlockInBounds(serverLevel, pos)) { // Sakura - falling block stacking restrictions
                                 if (this.blockState.hasProperty(BlockStateProperties.WATERLOGGED) && this.level().getFluidState(pos).is(Fluids.WATER)) {
                                     this.blockState = this.blockState.setValue(BlockStateProperties.WATERLOGGED, true);
                                 }
@@ -245,6 +_,10 @@
                                     this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
                                     this.callOnBrokenAfterFall(block, pos);
                                     this.spawnAtLocation(serverLevel, block);
+                                    // Sakura start - fix falling blocks staying alive when entity drops are disabled
+                                } else {
+                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
+                                    // Sakura end - fix falling blocks staying alive when entity drops are disabled
                                 }
                             } else {
                                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
