--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -370,6 +_,11 @@
 
     @Override
     public boolean ignoreExplosion(Explosion explosion) {
+        // Sakura start - add list of items that ignore explosions
+        if (this.level().sakuraConfig().entity.items.explosionResistantItems.contains(this.getItem().getItem()) != this.level().sakuraConfig().entity.items.useWhitelistForExplosionResistantItems) {
+            return true;
+        }
+        // Sakura end - add list of items that ignore explosions
         return !explosion.shouldAffectBlocklikeEntities() || super.ignoreExplosion(explosion);
     }
 
