--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -368,6 +_,13 @@
 
     @Override
     public boolean ignoreExplosion(Explosion explosion) {
+        // Sakura start - add list of items that ignore explosions
+        final java.util.Set<Item> blastResistantTypes = this.level().sakuraConfig().entity.items.blastResistant.items;
+        final boolean whitelist = this.level().sakuraConfig().entity.items.blastResistant.whitelistOverBlacklist;
+        if (blastResistantTypes.contains(this.getItem().getItem()) != whitelist) {
+            return true;
+        }
+        // Sakura end - add list of items that ignore explosions
         return !explosion.shouldAffectBlocklikeEntities() || super.ignoreExplosion(explosion);
     }
 
