--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -64,9 +_,32 @@
     public float explosionPower = 4.0F;
     public boolean isIncendiary = false; // CraftBukkit
 
+    // Sakura start - configure cannon mechanics
+    public int heightLimit;
+
+    @Override
+    public final void sakura$addedToWorld() {
+        final me.samsuik.sakura.configuration.local.CachedLocalConfiguration config = this.level().localConfig().at(this.blockPosition());
+        switch (config.tntSpread) {
+            case NONE -> this.setDeltaMovement(0.0, 0.0, 0.0);
+            case Y -> this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
+        }
+
+        if (this.pushedByFluid) { // default
+            this.pushedByFluid = config.tntFlowsInWater;
+        }
+
+        this.heightLimit = config.heightLimit;
+        if (this.heightLimit == Integer.MIN_VALUE) {
+            this.heightLimit = this.level().paperConfig().fixes.tntEntityHeightNerf.or(Integer.MIN_VALUE);
+        }
+    }
+    // Sakura end - configure cannon mechanics
+
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
         this.blocksBuilding = true;
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     public PrimedTnt(final Level level, final double x, final double y, final double z, final @Nullable LivingEntity owner) {
@@ -104,13 +_,13 @@
 
     @Override
     public void tick() {
-        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().spigotConfig.currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot
+        // Sakura - remove max tnt per tick
         this.handlePortal();
         this.applyGravity();
         this.move(MoverType.SELF, this.getDeltaMovement());
         this.applyEffectsFromBlocks();
         // Paper start - Configurable TNT height nerf
-        if (this.level().paperConfig().fixes.tntEntityHeightNerf.test(v -> this.getY() > v)) {
+        if (this.heightLimit != Integer.MIN_VALUE && this.getY() > this.heightLimit) { // Sakura - configure cannon mechanics
             this.discard(EntityRemoveEvent.Cause.OUT_OF_WORLD);
             return;
         }
@@ -136,8 +_,16 @@
                 this.level().addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5, this.getZ(), 0.0, 0.0, 0.0);
             }
         }
+        // Sakura start - configure force position updates
+        if (this.level().sakuraConfig().cannons.tnt.forcePositionUpdates) {
+            this.forcePositionUpdateInWater();
+        }
+    }
+
+    private void forcePositionUpdateInWater() {
+        // Sakura end - configure force position updates
         // Paper start - Option to prevent TNT from moving in water
-        if (!this.isRemoved() && this.wasTouchingWater && this.level().paperConfig().fixes.preventTntFromMovingInWater) {
+        if (!this.isRemoved() && this.wasTouchingWater && !this.isPushedByFluid()) { // Sakura - correctly check if this entity is pushed by water
             this.hurtMarked = true;
             this.needsSync = true;
         }
