--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -67,6 +_,7 @@
     public PrimedTnt(final EntityType<? extends PrimedTnt> type, final Level level) {
         super(type, level);
         this.blocksBuilding = true;
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     public PrimedTnt(final Level level, final double x, final double y, final double z, final @Nullable LivingEntity owner) {
@@ -79,6 +_,12 @@
         this.yo = y;
         this.zo = z;
         this.owner = EntityReference.of(owner);
+        // Sakura start - configure cannon mechanics
+        switch (level.sakuraConfig().cannons.mechanics.tntSpread) {
+            case NONE -> this.setDeltaMovement(0.0, 0.0, 0.0);
+            case Y -> this.setDeltaMovement(this.getDeltaMovement().multiply(0.0, 1.0, 0.0));
+        }
+        // Sakura end - configure cannon mechanics
     }
 
     @Override
@@ -104,7 +_,7 @@
 
     @Override
     public void tick() {
-        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().spigotConfig.currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot
+        // Sakura - remove max tnt per tick
         this.handlePortal();
         this.applyGravity();
         this.move(MoverType.SELF, this.getDeltaMovement());
@@ -136,8 +_,16 @@
                 this.level().addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5, this.getZ(), 0.0, 0.0, 0.0);
             }
         }
+        // Sakura start - configure force position updates
+        if (this.level().sakuraConfig().cannons.tnt.forcePositionUpdates) {
+            this.forcePositionUpdateInWater();
+        }
+    }
+
+    private void forcePositionUpdateInWater() {
+        // Sakura end - configure force position updates
         // Paper start - Option to prevent TNT from moving in water
-        if (!this.isRemoved() && this.wasTouchingWater && this.level().paperConfig().fixes.preventTntFromMovingInWater) {
+        if (!this.isRemoved() && this.wasTouchingWater && !this.isPushedByFluid()) { // Sakura - correctly check if this entity is pushed by water
             this.hurtMarked = true;
             this.needsSync = true;
         }
@@ -241,7 +_,7 @@
     // Paper start - Option to prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && super.isPushedByFluid();
+        return !this.level().paperConfig().fixes.preventTntFromMovingInWater && this.level().sakuraConfig().cannons.mechanics.tntFlowsInWater  && super.isPushedByFluid(); // Sakura - configure cannon mechanics
     }
     // Paper end - Option to prevent TNT from moving in water
 }
