--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -19,6 +_,13 @@
     private Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch;
     private boolean copyOnWrite;
 
+    // Sakura start - modify components sent to the client
+    public <T> void sakura$patchComponent(final DataComponentType<T> prototype, final @Nullable T newValue) {
+        this.ensureMapOwnership();
+        this.patch.put(prototype, Optional.ofNullable(newValue));
+    }
+    // Sakura end - modify components sent to the client
+
     public PatchedDataComponentMap(final DataComponentMap prototype) {
         this(prototype, Reference2ObjectMaps.emptyMap(), true);
     }
