--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -63,7 +_,7 @@
                     return InteractionResult.FAIL;
                 } else {
                     boat.setYRot(player.getYRot());
-                    if (!level.noCollision(boat, boat.getBoundingBox())) {
+                    if (!level.noCollision(boat, boat.getBoundingBox()) || !level.getWorldBorder().isWithinBounds(boat.getBoundingBox())) { // Sakura - fix placing boats outside the world border
                         return InteractionResult.FAIL;
                     } else {
                         if (!level.isClientSide()) {
