--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -65,7 +_,7 @@
             }
 
             boat.setYRot(player.getYRot());
-            if (!level.noCollision(boat, boat.getBoundingBox())) {
+            if (!level.noCollision(boat, boat.getBoundingBox()) || !level.getWorldBorder().isWithinBounds(boat.getBoundingBox())) { // Sakura - fix placing boats outside the world border
                 return InteractionResult.FAIL;
             }
 
