--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -147,7 +_,7 @@
             // CraftBukkit end
             if (!flag2) {
                 return hitResult != null && this.emptyContents(entity, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null, direction, clicked, itemstack, hand); // CraftBukkit
-            } else if (level.dimensionType().ultraWarm() && this.content.is(FluidTags.WATER)) {
+            } else if (!level.sakuraConfig().environment.allowWaterInTheNether && level.dimensionType().ultraWarm() && this.content.is(FluidTags.WATER)) { // Sakura - allow water in the nether
                 int x = pos.getX();
                 int y = pos.getY();
                 int z = pos.getZ();
