--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -63,8 +_,13 @@
                     if (dummyFluid.isEmpty()) return InteractionResult.FAIL; // Don't fire event if the bucket won't be filled.
                     org.bukkit.event.player.PlayerBucketFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketFillEvent(level, player, blockPos, blockPos, playerPovHitResult.getDirection(), itemInHand, dummyFluid.getItem(), hand);
 
-                    if (event.isCancelled()) {
+                    // Sakura start - modify bucket stack size
+                    if (event.isCancelled() || me.samsuik.sakura.configuration.GlobalConfiguration.get().players.bucketStackSize.value().isPresent()) {
                         player.containerMenu.sendAllDataToRemote(); // SPIGOT-4541
+                    }
+
+                    if (event.isCancelled()) {
+                        // Sakura end - modify bucket stack size
                         return InteractionResult.FAIL;
                     }
                     // CraftBukkit end
@@ -148,7 +_,7 @@
             // CraftBukkit end
             if (!flag2) {
                 return hitResult != null && this.emptyContents(entity, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null, direction, clicked, itemstack, hand); // CraftBukkit
-            } else if (level.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, pos) && this.content.is(FluidTags.WATER)) {
+            } else if (!level.sakuraConfig().environment.allowWaterInTheNether && level.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, pos) && this.content.is(FluidTags.WATER)) { // Sakura - allow water in the nether
                 int x = pos.getX();
                 int y = pos.getY();
                 int z = pos.getZ();
