--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -57,6 +_,18 @@
             BlockPattern.BlockPatternMatch match = EndPortalFrameBlock.getOrCreatePortalShape().find(level, pos);
             if (match != null) {
                 BlockPos blockPos = match.getFrontTopLeft().offset(-3, 0, -3);
+                // Sakura start - fix end portal item dupe
+                for (int portalX = 0; portalX < 3; portalX++) {
+                    for (int portalZ = 0; portalZ < 3; portalZ++) {
+                        final BlockPos portalPos = blockPos.offset(portalX, 0, portalZ);
+                        if (!level.getBlockState(portalPos).isAir()) {
+                            // Revert the portal frame
+                            level.setBlock(clickedPos, blockState, Block.UPDATE_CLIENTS);
+                            return InteractionResult.PASS;
+                        }
+                    }
+                }
+                // Sakura end - fix end portal item dupe
 
                 for (int x = 0; x < 3; x++) {
                     for (int z = 0; z < 3; z++) {
