--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -142,6 +_,11 @@
         }
     }
 
+    // Sakura start - modify components sent to the client
+    public void modifyComponentsSentToClient(final net.minecraft.core.component.PatchedDataComponentMap components) {
+    }
+    // Sakura end - modify components sent to the client
+
     @Deprecated
     public Holder.Reference<Item> builtInRegistryHolder() {
         return this.builtInRegistryHolder;
