--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -132,6 +_,15 @@
         return BY_BLOCK.getOrDefault(block, Items.AIR);
     }
 
+    // Sakura start - modify data components
+    public DataComponentMap modifyBaseComponents(final DataComponentMap components) {
+        return components;
+    }
+
+    public void modifyComponentsSentToClient(final net.minecraft.core.component.PatchedDataComponentMap components) {
+    }
+    // Sakura end - modify data components
+
     public Item(final Item.Properties properties) {
         this.descriptionId = properties.effectiveDescriptionId();
         DataComponentInitializers.Initializer<Item> componentInitializer = properties.finalizeInitializer(
@@ -147,6 +_,7 @@
             }
         }
     }
+
 
     @Deprecated
     public Holder.Reference<Item> builtInRegistryHolder() {
