--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -131,6 +_,11 @@
         return BY_BLOCK.getOrDefault(block, Items.AIR);
     }
 
+    // Sakura start - modify components sent to the client
+    public void modifyComponentsSentToClient(final net.minecraft.core.component.PatchedDataComponentMap components) {
+    }
+    // Sakura end - modify components sent to the client
+
     public Item(final Item.Properties properties) {
         this.descriptionId = properties.effectiveDescriptionId();
         DataComponentInitializers.Initializer<Item> componentInitializer = properties.finalizeInitializer(
@@ -146,6 +_,7 @@
             }
         }
     }
+
 
     @Deprecated
     public Holder.Reference<Item> builtInRegistryHolder() {
