--- a/net/minecraft/world/item/SolidBucketItem.java
+++ b/net/minecraft/world/item/SolidBucketItem.java
@@ -27,7 +_,15 @@
         InteractionResult interactionResult = super.useOn(context);
         Player player = context.getPlayer();
         if (interactionResult.consumesAction() && player != null) {
-            player.setItemInHand(context.getHand(), BucketItem.getEmptySuccessItem(context.getItemInHand(), player));
+            // Sakura start - modify bucket stack size
+            final ItemStack itemInHand = context.getItemInHand();
+            final ItemStack emptyBucket = BucketItem.getEmptySuccessItem(itemInHand, player);
+            if (itemInHand.isEmpty()) {
+                player.setItemInHand(context.getHand(), emptyBucket);
+            } else {
+                player.addItem(emptyBucket);
+            }
+            // Sakura end - modify bucket stack size
         }
 
         return interactionResult;
