--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -95,6 +_,13 @@
         } else if (!type.isAllowedInPeaceful(itemStack.get(DataComponents.ENTITY_DATA).getUnsafe()) && level.getDifficulty() == Difficulty.PEACEFUL) { // Paper - check peaceful override
             return InteractionResult.FAIL;
         } else {
+            // Sakura start - prevent placing spawn eggs inside blocks
+            if (level.sakuraConfig().players.preventPlacingSpawnEggsInsideBlocks
+                && !level.noCollision(type.getSpawnAABB(spawnPos.getX() + 0.5, spawnPos.getY(), spawnPos.getZ() + 0.5))) {
+                return InteractionResult.PASS;
+            }
+            // Sakura end - prevent placing spawn eggs inside blocks
+
             if (type.spawn((ServerLevel)level, itemStack, user, spawnPos, EntitySpawnReason.SPAWN_ITEM_USE, tryMoveDown, movedUp) != null) {
                 itemStack.consume(1, user);
                 level.gameEvent(user, GameEvent.ENTITY_PLACE, spawnPos);
