--- a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/AbstractWindCharge.java
@@ -35,6 +_,7 @@
     public AbstractWindCharge(final EntityType<? extends AbstractWindCharge> type, final Level level) {
         super(type, level);
         this.accelerationPower = 0.0;
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     public AbstractWindCharge(
@@ -43,6 +_,7 @@
         super(type, x, y, z, level);
         this.setOwner(owner);
         this.accelerationPower = 0.0;
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     protected AbstractWindCharge(
@@ -50,6 +_,7 @@
     ) {
         super(type, x, y, z, direction, level);
         this.accelerationPower = 0.0;
+        this.cannonEntity = true; // Sakura - add cannon entity field
     }
 
     @Override
