From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d64ddd3c54ff76b74f0b93565c4899b891bcb4ef..1bedf13b2c9007267b5e7d8de11501d2087fe7bc 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -79,6 +79,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     void setTrackingRangeModifier(final double modifier);
     // Sakura end - entity tracking range modifier
+    // Sakura start - client visibility settings api
+    /**
+     * Server-side api to disable sending visual effects to the client.
+     *
+     * @return visibility api
+     */
+    me.samsuik.sakura.player.visibility.VisibilitySettings getVisibility();
+    // Sakura end - client visibility settings api
 
     // Paper start
     @Override
