From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 21 Sep 2021 23:54:25 +0100
Subject: [PATCH] Client Visibility Settings API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8127c961793c5a5332fbb66fa95b8a0bf3fdc4df..3b51a842e85425c84fc7d8e52f13282d2cf6b3b5 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -70,6 +70,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     void setTrackingRangeModifier(double mod);
     // Sakura end - entity tracking range modifier
+    // Sakura start - client visibility settings api
+    /**
+     * Server-side api to disable sending visual effects to the client.
+     *
+     * @return visibility api
+     */
+    me.samsuik.sakura.player.visibility.VisibilitySettings getVisibility();
+    // Sakura end - client visibility settings api
 
     // Paper start
     @Override
