From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 13 Oct 2023 20:01:48 +0100
Subject: [PATCH] isPushedByFluid API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a2a423d4e4c2702ba5967223cab0432dd7d04732..9b5e4db5f258ca30ef9a7bfbd0ff867340199e8f 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -108,6 +108,22 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public boolean isInWater();
 
+    // Sakura start
+    /**
+     * Gets if the entity will be pushed by fluid.
+     *
+     * @return if this entity can be pushed by fluid.
+     */
+    boolean isPushedByFluid();
+
+    /**
+     * Sets if the entity will be pushed by fluid.
+     *
+     * @param state whether entity should be pushed by fluid
+     */
+    void setPushedByFluid(boolean state);
+    // Sakura end
+
     /**
      * Gets the current world this entity resides in
      *
