From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 13 Oct 2023 20:01:48 +0100
Subject: [PATCH] isPushedByFluid API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 1d0fd7ff8449f815a7d980af0b378181ea8bf8d8..e0ef6e5a4664b69a5797afeafac2c6436137cab3 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -111,6 +111,22 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public boolean isInWater();
 
+    // Sakura start
+    /**
+     * Gets if the entity will be pushed by fluid.
+     *
+     * @return if this entity can be pushed by fluid.
+     */
+    boolean isPushedByFluid();
+
+    /**
+     * Sets if the entity will be pushed by fluid.
+     *
+     * @param state whether entity should be pushed by fluid
+     */
+    void setPushedByFluid(boolean state);
+    // Sakura end
+
     /**
      * Gets the current world this entity resides in
      *
