From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 15 Oct 2023 22:53:38 +0100
Subject: [PATCH] Falling Block Parity API


diff --git a/src/main/java/org/bukkit/entity/FallingBlock.java b/src/main/java/org/bukkit/entity/FallingBlock.java
index 3b4c83aab5e646266d91a2f3cb0f57a242941069..8d9909eb6aef5a1b7643e4249e28729543f1d78c 100644
--- a/src/main/java/org/bukkit/entity/FallingBlock.java
+++ b/src/main/java/org/bukkit/entity/FallingBlock.java
@@ -122,4 +122,19 @@ public interface FallingBlock extends Entity, me.samsuik.sakura.entity.merge.Mer
      */
     void shouldAutoExpire(boolean autoExpires);
     // Paper End - Auto expire setting
+    // Sakura start
+    /**
+     * Gets if falling block has height parity
+     *
+     * @return parity
+     */
+    boolean getHeightParity();
+
+    /**
+     * Sets falling block height parity
+     *
+     * @param parity value
+     */
+    void setHeightParity(boolean parity);
+    // Sakura end
 }
