From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 15 Oct 2023 22:53:38 +0100
Subject: [PATCH] Falling Block Parity API


diff --git a/src/main/java/org/bukkit/entity/FallingBlock.java b/src/main/java/org/bukkit/entity/FallingBlock.java
index 2479453b96a75fda2c307261f84c91034e73000d..a8e3a1c34957b7f7e225dc9932edbd300755be23 100644
--- a/src/main/java/org/bukkit/entity/FallingBlock.java
+++ b/src/main/java/org/bukkit/entity/FallingBlock.java
@@ -183,4 +183,19 @@ public interface FallingBlock extends Entity, me.samsuik.sakura.entity.merge.Mer
      */
     void shouldAutoExpire(boolean autoExpires);
     // Paper end - Auto expire setting
+    // Sakura start
+    /**
+     * Gets if falling block has height parity
+     *
+     * @return parity
+     */
+    boolean getHeightParity();
+
+    /**
+     * Sets falling block height parity
+     *
+     * @param parity value
+     */
+    void setHeightParity(boolean parity);
+    // Sakura end
 }
