From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 18 Jun 2024 13:34:39 +0100
Subject: [PATCH] Entity tracking range modifier


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4c2e82a6421722d9a990ff4a216cedc159c7e579..489ad34e24c8b65d8ac7623d38707fb768d3cac1 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -68,6 +68,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     me.samsuik.sakura.player.visibility.Visibility getVisibility();
     // Sakura end
 
+    // Sakura start - entity tracking range modifier
+    double getTrackingRangeModifier();
+
+    void setTrackingRangeModifier(double mod);
+    // Sakura end - entity tracking range modifier
+
     // Paper start
     @Override
     default net.kyori.adventure.identity.@NotNull Identity identity() {
