From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 18 Jun 2024 13:34:39 +0100
Subject: [PATCH] Entity tracking range modifier


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d99d1ad6a2fcd132aefcbe285c77820362672ab6..53c8b97776f308bf78d05fa4939df987cb93f086 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -67,6 +67,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     me.samsuik.sakura.player.visibility.Visibility getVisibility();
     // Sakura end
 
+    // Sakura start - entity tracking range modifier
+    double getTrackingRangeModifier();
+
+    void setTrackingRangeModifier(double mod);
+    // Sakura end - entity tracking range modifier
+
     // Paper start
     @Override
     default net.kyori.adventure.identity.@NotNull Identity identity() {
