From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 5 Dec 2024 22:27:38 +0000
Subject: [PATCH] Add lava flow speed api


diff --git a/src/main/java/me/samsuik/sakura/local/LocalValueKey.java b/src/main/java/me/samsuik/sakura/local/LocalValueKey.java
index 780a02fb451d5b4b98f6895b488b90c75a6507a8..ed5ff8c6262e2c9ab01417859061cba2897b2422 100644
--- a/src/main/java/me/samsuik/sakura/local/LocalValueKey.java
+++ b/src/main/java/me/samsuik/sakura/local/LocalValueKey.java
@@ -33,6 +33,10 @@ public record LocalValueKey<T>(NamespacedKey key, Supplier<T> defaultSupplier) {
         new NamespacedKey("sakura", "redstone-cache"), () -> false
     );
 
+    public static final LocalValueKey<Integer> LAVA_FLOW_SPEED = new LocalValueKey<>(
+        new NamespacedKey("sakura", "lava-flow-speed"), () -> -1
+    );
+
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
