From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 11 Sep 2021 22:23:39 +0100
Subject: [PATCH] Optional Force Position Updates


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index d14128dd5d63e93a3d4aec76b0ee8d0e7d786687..2eb69c7d7843a7b58479049633b838212d003c10 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -129,6 +129,14 @@ public class PrimedTnt extends Entity implements TraceableEntity {
             }
         }
 
+        // Sakura start - configure force position updates
+        if (level().sakuraConfig().cannons.tnt.forcePositionUpdates) {
+            forcePositionUpdate();
+        }
+    }
+
+    private void forcePositionUpdate() {
+        // Sakura end - configure force position updates
         // Paper start - Option to prevent TNT from moving in water
         if (!this.isRemoved() && this.wasTouchingWater && this.level().paperConfig().fixes.preventTntFromMovingInWater) {
             /*
