From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 11 Sep 2021 22:23:39 +0100
Subject: [PATCH] Optional Force Position Updates


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 80e495688b00d1297357350f9abebe5fbbf24cd8..7d2a2d69241344593b691321ba859e5ba246d15d 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -60,7 +60,9 @@ public class PrimedTnt extends Entity {
 
     @Override
     public void tick() {
-        if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
+        if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) {
+            return;
+        } // Spigot
         if (!this.isNoGravity()) {
             this.addDeltaMovement(0.0D, -0.04D, 0.0D); // Sakura - reduce movement allocations
         }
@@ -95,6 +97,14 @@ public class PrimedTnt extends Entity {
             }
         }
 
+        // Sakura start - configure force position updates
+        if (this.level.sakuraConfig().cannons.tnt.forcePositionUpdates) {
+            this.forcePositionUpdate();
+        }
+    }
+
+    private void forcePositionUpdate() {
+        // Sakura end - configure force position updates
         // Paper start - Optional prevent TNT from moving in water
         if (!this.isRemoved() && this.wasTouchingWater && this.level.paperConfig().fixes.preventTntFromMovingInWater) {
             /*
