From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 11 Sep 2021 22:23:39 +0100
Subject: [PATCH] Optional Force Position Updates


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 369cfaeba980d1d98d0dcefffb584dcbbe144b1d..5b4e1d56f63935c5b6506dc94d174e8efebd260f 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -99,6 +99,14 @@ public class PrimedTnt extends Entity implements TraceableEntity {
             }
         }
 
+        // Sakura start - configurable force position updates
+        if (level().sakuraConfig().cannons.tnt.forcePositionUpdates) {
+            forcePositionUpdate();
+        }
+    }
+
+    private void forcePositionUpdate() {
+        // Sakura end
         // Paper start - Optional prevent TNT from moving in water
         if (!this.isRemoved() && this.wasTouchingWater && this.level().paperConfig().fixes.preventTntFromMovingInWater) {
             /*
