From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 15 Oct 2023 22:48:35 +0100
Subject: [PATCH] Cannon Mechanics


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index c2f56308492d1e5d56993e32db8daaba8e081274..00c17b86c69b73d751ca93b65a54c74b6ed8fa6f 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -72,6 +72,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
     public boolean forceTickAfterTeleportToDuplicate;
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
+    public boolean heightParity; // Sakura - configure cannon mechanics
 
     // Sakura start - merge cannon entities
     private final me.samsuik.sakura.entity.merge.MergeEntityData mergeData = new me.samsuik.sakura.entity.merge.MergeEntityData(this);
@@ -132,6 +133,7 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         this.isFallingBlock = true; // Sakura
         this.loadChunks = world.sakuraConfig().cannons.loadChunks; // Sakura - load chunks on movement
         this.mergeData.setMergeLevel(world.sakuraConfig().cannons.mergeLevel); // Sakura - merge cannon entities
+        this.heightParity = world.sakuraConfig().cannons.mechanics.fallingBlockParity; // Sakura - configure cannon mechanics
     }
 
     public FallingBlockEntity(Level world, double x, double y, double z, BlockState block) {
@@ -189,6 +191,13 @@ public class FallingBlockEntity extends Entity implements me.samsuik.sakura.enti
         return !this.isRemoved();
     }
 
+    // Sakura start - configure cannon mechanics
+    @Override
+    public final double getEyeY() {
+        return this.heightParity ? this.getY() : super.getEyeY();
+    }
+    // Sakura end - configure cannon mechanics
+
     @Override
     protected double getDefaultGravity() {
         return 0.04D;
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index cc2f2dbee1d717d54014005ec57e8b864c2a7241..7c2323bf7682254de14ed963e8d57b24dd55ad5f 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -96,6 +96,12 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
         this.yo = y;
         this.zo = z;
         this.owner = igniter;
+        // Sakura start - configure cannon mechanics
+        switch (world.sakuraConfig().cannons.mechanics.tntSpread) {
+            case NONE -> this.multiplyDeltaMovement(0, 0, 0);
+            case Y -> this.multiplyDeltaMovement(0, 1, 0);
+        }
+        // Sakura end - configure cannon mechanics
     }
 
     @Override
@@ -268,7 +274,7 @@ public class PrimedTnt extends Entity implements TraceableEntity, me.samsuik.sak
     // Paper start - Option to prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !level().paperConfig().fixes.preventTntFromMovingInWater && super.isPushedByFluid();
+        return !level().paperConfig().fixes.preventTntFromMovingInWater && level().sakuraConfig().cannons.mechanics.tntFlowsInWater && super.isPushedByFluid(); // Sakura - configure cannon mechanics
     }
     // Paper end - Option to prevent TNT from moving in water
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
index 55f67c2ca07eca0d3e1522eebbb4ce37704bdd04..2de0befc3948a23910cfc609fc25c57bbd0cb030 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingBlock.java
@@ -35,6 +35,17 @@ public class CraftFallingBlock extends CraftEntity implements FallingBlock {
         this.getHandle().getMergeEntityData().setCount(stacked);
     }
     // Sakura end - merge cannon entities
+    // Sakura start - cannon mechanics
+    @Override
+    public void setHeightParity(boolean parity) {
+        this.getHandle().heightParity = parity;
+    }
+
+    @Override
+    public boolean getHeightParity() {
+        return this.getHandle().heightParity;
+    }
+    // Sakura end - cannon mechanics
 
     @Override
     public FallingBlockEntity getHandle() {
