From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 16 Oct 2023 22:57:55 +0100
Subject: [PATCH] Optimise TNT fluid state and pushing


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 94d7a8568ffe9ecedcb3a9a6c2f42d8c4562d472..432601bc7a33a0da2034b3032e827fbdce8f6b6e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2199,7 +2199,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         return this.isInWater() || flag;
     }
 
-    void updateInWaterStateAndDoWaterCurrentPushing() {
+    protected void updateInWaterStateAndDoWaterCurrentPushing() { // Sakura
         Entity entity = this.getVehicle();
 
         if (entity instanceof Boat) {
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 2123f82f881fea63cef197d91c267358c24ff126..3ebc6fabbd9e5f9e71a97fe6153f24116c66f828 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -117,6 +117,19 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     }
     */
     // Sakura end
+    // Sakura start
+    protected boolean updateInWaterStateAndDoFluidPushing() {
+        if (this.isPushedByFluid()) {
+            return super.updateInWaterStateAndDoFluidPushing();
+        } else {
+            // super method also handles lava fluid pushing
+            // we only need to search for water to negate fall distance
+            this.fluidHeight.clear();
+            this.updateInWaterStateAndDoWaterCurrentPushing();
+            return this.isInWater();
+        }
+    }
+    // Sakura end
 
     @Override
     public void tick() {
