From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Thu, 16 Nov 2023 00:59:04 +0000
Subject: [PATCH] Destroy Waterlogged Blocks


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 97ebbb51773d7da046687ec00d16b3e4a8d569d5..40730373a60f2939e8c2ff261f75bd0abf454732 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -285,6 +285,12 @@ public class Explosion {
                                 break;
                             }
 
+                            // Sakura start - destroy water logged blocks
+                            if (!iblockdata.isAir() && this.level.sakuraConfig().cannons.explosion.destroyWaterloggedBlocks) {
+                                fluid = Blocks.AIR.defaultBlockState().getFluidState();
+                            }
+                            // Sakura end
+
                             Optional<Float> optional = this.damageCalculator.getBlockExplosionResistance(this, this.level, blockposition, iblockdata, fluid);
 
                             // Sakura start - durable materials
