From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Thu, 16 Nov 2023 00:59:04 +0000
Subject: [PATCH] Destroy Waterlogged Blocks


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 4abd9fe2c810057beb9d9d5c732d295325b011d7..45e95e9565b9db018b5d7422189b98517e2daac3 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -196,6 +196,12 @@ public class Explosion {
             BlockState blockState = chunk.getBlockStateFinal(x, y, z);
             FluidState fluidState = blockState.getFluidState();
 
+            // Sakura start - destroy water logged blocks
+            if (calculateResistance && !blockState.isAir() && this.level.sakuraConfig().cannons.explosion.destroyWaterloggedBlocks) {
+                fluidState = Blocks.AIR.defaultBlockState().getFluidState();
+            }
+            // Sakura end
+
             Optional<Float> resistance = !calculateResistance ? Optional.empty() : this.damageCalculator.getBlockExplosionResistance((Explosion)(Object)this, this.level, pos, blockState, fluidState);
 
             // Sakura start - durable materials
