From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 21 Nov 2023 10:56:30 +0000
Subject: [PATCH] Remove spigot max tnt per tick


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 92fe040d9b43cb2a2ff8118335eaa007c0828115..e97729f2d8febac1740678840566f8beb4f07cdc 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -100,9 +100,7 @@ public class PrimedTnt extends Entity {
 
     @Override
     public void tick() {
-        if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) {
-            return;
-        } // Spigot
+        // Sakura - remove max tnt per tick
         if (!this.isNoGravity()) {
             this.addDeltaMovement(0.0D, -0.04D, 0.0D); // Sakura - reduce movement allocations
         }
