From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 18 Nov 2023 15:12:14 +0000
Subject: [PATCH] Add option to disable entity ai


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 2439f8d48daca7329049436414f06a36b4b79029..bdf3c81d9721ac06fd66a3f68ef0cc174858391b 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -898,7 +898,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     protected final void serverAiStep() {
         ++this.noActionTime;
         // Paper start - Allow nerfed mobs to jump and float
-        if (!this.aware) {
+        if (!this.aware || this.level().sakuraConfig().entity.disableMobAi) { // Sakura
             if (goalFloat != null) {
                 if (goalFloat.canUse()) goalFloat.tick();
                 this.getJumpControl().tick();
