From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 18 Nov 2023 15:12:14 +0000
Subject: [PATCH] Add option to disable entity ai


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 5af48151159135b869ec4753bbcf79dd257c1570..0b807081ad85ea91ecf397f6c29c6910010f8638 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -897,7 +897,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     protected final void serverAiStep() {
         ++this.noActionTime;
         // Paper start - Allow nerfed mobs to jump and float
-        if (!this.aware) {
+        if (!this.aware || this.level().sakuraConfig().entity.disableMobAi) { // Sakura
             if (goalFloat != null) {
                 if (goalFloat.canUse()) goalFloat.tick();
                 this.getJumpControl().tick();
