From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 21 Nov 2023 11:20:14 +0000
Subject: [PATCH] fixup! Sakura Configuration Files


diff --git a/src/main/java/me/samsuik/sakura/configuration/WorldConfiguration.java b/src/main/java/me/samsuik/sakura/configuration/WorldConfiguration.java
index 4e62a5fbd2e1e674abfe4afb763118b7b0127b67..21310cd375267759f61cffd99dcfb0a6c5263151 100644
--- a/src/main/java/me/samsuik/sakura/configuration/WorldConfiguration.java
+++ b/src/main/java/me/samsuik/sakura/configuration/WorldConfiguration.java
@@ -129,6 +129,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Entity extends ConfigurationPart {
         @Comment("Only modify if you know what you're doing")
         public boolean disableMobAi = false;
+        public boolean waterSensitivity = true;
 
         public Items items = new Items();
         public class Items extends ConfigurationPart {
