From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 26 Nov 2023 17:57:50 +0000
Subject: [PATCH] Treat all collidable blocks as full while moving fast


diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
index 93977027caa8c398ca6f0d2bb0890231adf8bf49..07d3aac679ed4a3d85573706bd664e4d6f6af0ab 100644
--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
+++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
@@ -485,6 +485,7 @@ public final class CollisionUtil {
             chunkProvider = null;
         }
 
+        final boolean fullBlocks = entity != null && entity.treatAllBlocksAsFullWhenMoving(); // Sakura
         for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
             final int minZ = currChunkZ == minChunkZ ? minBlockZ & 15 : 0; // coordinate in chunk
             final int maxZ = currChunkZ == maxChunkZ ? maxBlockZ & 15 : 15; // coordinate in chunk
@@ -539,7 +540,7 @@ public final class CollisionUtil {
                     final int maxY = currChunkY == maxChunkYIterate ? maxYIterate & 15 : 15; // coordinate in chunk
                     final int chunkYGlobalPos = currChunkY << 4;
 
-                    final boolean sectionHasSpecial = section.hasSpecialCollidingBlocks();
+                    final boolean sectionHasSpecial = !fullBlocks && section.hasSpecialCollidingBlocks(); // Sakura
 
                     final int minXIterate;
                     final int maxXIterate;
@@ -651,6 +652,11 @@ public final class CollisionUtil {
                                     }
                                     VoxelShape voxelshape2 = blockData.getCollisionShape(getter, mutablePos, collisionShape);
                                     if (voxelshape2 != Shapes.empty()) {
+                                        // Sakura start - Treat all collidable blocks as full when moving fast
+                                        if (fullBlocks) {
+                                            voxelshape2 = Shapes.block();
+                                        }
+                                        // Sakura end - Treat all collidable blocks as full when moving fast
                                         VoxelShape voxelshape3 = voxelshape2.move((double)blockX, (double)blockY, (double)blockZ);
 
                                         if (predicate != null && !predicate.test(blockData, mutablePos)) {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index fb9d93de6ea371a2fb3423db5022207e3becf7ba..8c1c978cd6cfbe6b987e6cf9a3be4283c78f2b1f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -690,6 +690,18 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.physics;
     }
     // Sakura end - physics version api
+    // Sakura start - Treat all collidable blocks as full when moving fast
+    public final boolean treatAllBlocksAsFullWhenMoving() {
+        if (this.level.sakuraConfig().cannons.treatAllBlocksAsFullWhenMoving && (this.isPrimedTNT || this.isFallingBlock)) {
+            this.syncDeltaMovement();
+            double horizontalMovementSqr = this.movementX*this.movementX + this.movementZ*this.movementZ;
+            if (horizontalMovementSqr > Math.pow(this.level.sakuraConfig().cannons.treatAllBlocksAsFullWhenMovingFasterThan, 2.0)) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Sakura end - Treat all collidable blocks as full when moving fast
 
     public Entity(EntityType<?> type, Level world) {
         this.id = Entity.ENTITY_COUNTER.incrementAndGet();
