From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index c1effe8ecd110ac39103cc7778e7070ac9c33732..68b9b340b6dddcf29f7c3d80600854dd87b67f65 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -623,6 +623,11 @@ public class Explosion {
                 }
             }
 
+            // Sakura start - config for explosions dropping items
+            if (!this.level.sakuraConfig.explosionsDropItems) {
+                objectarraylist.clear();
+            }
+            // Sakura end - config for explosions dropping items
             ObjectListIterator objectlistiterator = objectarraylist.iterator();
 
             while (objectlistiterator.hasNext()) {
