From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 726910c21440c62e35220294e4b639f4a8d242e2..64978e2c3bb5056e85b50a04f2d7fa8f8c76cd4b 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -908,6 +908,11 @@ public class Explosion {
                 this.level.densityCache.clear(-1);
             }
             // Sakura end - explosion density cache
+            // Sakura start - config for explosions dropping items
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end - config for explosions dropping items
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
