From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 8dd53e7c714657fe357757cdbfdfd7182b8a73eb..fda0aeae04276ff9ac3b041895f1344e8c6dd6fe 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -585,6 +585,12 @@ public class Explosion {
                 }
             }
 
+            // Sakura start
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                objectarraylist.clear();
+            }
+            // Sakura end
+
             objectlistiterator = objectarraylist.iterator();
 
             while (objectlistiterator.hasNext()) {
