From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 0b90b05bd538186f8f1d51376560dd6485fdce57..7d68fc98f2fe995391c57a220f7f5986aea07166 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -915,6 +915,12 @@ public class Explosion {
                 });
             }
 
+            // Sakura start
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end
+
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
