From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 97b5854f9ad38308c8abf15a7899b879f70e1271..d31d3042953f629c9fa19b76b6bdd30ad517b686 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -891,6 +891,12 @@ public class Explosion {
                 });
             }
 
+            // Sakura start
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end
+
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
