From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 21 Jan 2024 22:12:30 +0000
Subject: [PATCH] fixup! Optimised Explosions


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index fda0aeae04276ff9ac3b041895f1344e8c6dd6fe..cd5ba771bd11549f9a9c2c03a31479d236f24e02 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -247,9 +247,9 @@ public class Explosion {
             {
                 {
                     {
-                        double d0 = BLOCK_RAYCAST_VECTORS[0];
-                        double d1 = BLOCK_RAYCAST_VECTORS[1];
-                        double d2 = BLOCK_RAYCAST_VECTORS[2];
+                        double d0 = BLOCK_RAYCAST_VECTORS[i];
+                        double d1 = BLOCK_RAYCAST_VECTORS[i + 1];
+                        double d2 = BLOCK_RAYCAST_VECTORS[i + 2];
 
                         float f = this.radius * (0.7F + (this.level.sakuraConfig().cannons.explosion.consistentRadius ? 0.7F : this.level.random.nextFloat()) * 0.6F); // Sakura
                         double d4 = this.x;
