From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index b69b7818a5613bd210ccfd259f01ce025e88a675..0152e7ea9b9aad9b443bef784a2bec5e22e66680 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -952,6 +952,11 @@ public class Explosion {
                 this.level.densityCache.clear(-1);
             }
             // Sakura end - explosion density cache
+            // Sakura start - config for explosions dropping items
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end - config for explosions dropping items
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
