From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index fd53dff917b65b94eceb5796fc80b1a490454cc2..a5ea2a84098e5cef7c82bfb437b6e630727a25c4 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -951,6 +951,11 @@ public class Explosion {
                 this.level.densityCache.clear(-1);
             }
             // Sakura end - explosion density cache
+            // Sakura start - config for explosions dropping items
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end - config for explosions dropping items
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
