From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 11:20:10 +0000
Subject: [PATCH] Add explosions dropping items config


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 2db4da910c67c2ebe8dde88faf775138df1510b9..fe1613301795c46e92494eaa61d0c5ea51a8289e 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -938,6 +938,11 @@ public class Explosion {
                 this.level.densityCache.clear(-1);
             }
             // Sakura end - explosion density cache
+            // Sakura start - config for explosions dropping items
+            if (!level.sakuraConfig().cannons.explosion.explosionsDropItems) {
+                list.clear();
+            }
+            // Sakura end - config for explosions dropping items
             Iterator iterator = list.iterator();
 
             while (iterator.hasNext()) {
