From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 16:56:39 +0000
Subject: [PATCH] Fix paper findSupportingBlock not updating last chunk
 coordinates


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index a93e22d81939cc7fa3247550da91bb39e6d5b111..798c817dd75615458a60c3063174ad7da175eb51 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -896,6 +896,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
 
                 if (chunkDiff != 0) {
                     lastChunk = chunkProvider.getChunkAtIfLoadedImmediately(newChunkX, newChunkZ);
+                    // Sakura start
+                    lastChunkX = newChunkX;
+                    lastChunkZ = newChunkZ;
+                    // Sakura end
                 }
 
                 if (lastChunk == null) {
