From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 24 Dec 2023 16:56:39 +0000
Subject: [PATCH] Fix paper findSupportingBlock not updating last chunk
 coordinates


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index d2f22ad58d10358e7515cb18b570bbcd4ae9e8c6..41b9bda5d378efd51bd6d14322bf6808e52f4400 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -888,6 +888,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
 
                 if (chunkDiff != 0) {
                     lastChunk = chunkProvider.getChunkAtIfLoadedImmediately(newChunkX, newChunkZ);
+                    // Sakura start
+                    lastChunkX = newChunkX;
+                    lastChunkZ = newChunkZ;
+                    // Sakura end
                 }
 
                 if (lastChunk == null) {
