From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 23 Feb 2024 01:49:20 +0000
Subject: [PATCH] Allow disabling sweep attacks


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 77c6b49c89e13a85042bc3dd4d81f8c0a9d14224..180d561ecc88f2a190f1eb9dbe3cfe426d393be9 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1321,7 +1321,7 @@ public abstract class Player extends LivingEntity {
                             // Paper end
                         }
 
-                        if (flag3) {
+                        if (flag3 && this.level.sakuraConfig.allowSweepAttacks) { // Sakura - allow disabling sweep attacks
                             float f4 = 1.0F + EnchantmentHelper.getSweepingDamageRatio(this) * f;
                             List<LivingEntity> list = this.level.getEntitiesOfClass(LivingEntity.class, target.getBoundingBox().inflate(1.0D, 0.25D, 1.0D));
                             Iterator iterator = list.iterator();
