From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Thu, 14 Mar 2024 19:51:20 +0000
Subject: [PATCH] Disable player poses shrinking collision box


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index b32a2388d800f5930274af4d439a477fe03838f9..70e3c8b1082d16cdbb9a4c179c53dafcc59ac1ed 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2289,7 +2289,13 @@ public abstract class Player extends LivingEntity {
 
     @Override
     public EntityDimensions getDefaultDimensions(Pose pose) {
-        return (EntityDimensions) Player.POSES.getOrDefault(pose, Player.STANDING_DIMENSIONS);
+        // Sakura start - player poses shrink collision box
+        EntityDimensions dimensions = (EntityDimensions) Player.POSES.getOrDefault(pose, Player.STANDING_DIMENSIONS);
+        if (!level().sakuraConfig().players.posesShrinkCollisionBox && dimensions.height() == 0.6f) {
+            dimensions = Player.STANDING_DIMENSIONS;
+        }
+        return dimensions;
+        // Sakura end - player poses shrink collision box
     }
 
     @Override
