From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 23 Feb 2024 02:07:03 +0000
Subject: [PATCH] Use old damage reduction with shields


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 241bd21a9500989d22df54827f90e93381c333ed..8ea2e719d0396f7cf96d3d0f943ac68eda250851 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2280,7 +2280,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
             Function<Double, Double> blocking = new Function<Double, Double>() {
                 @Override
                 public Double apply(Double f) {
+                    // Sakura start - old blocking damage reduction
+                    if (!level().sakuraConfig().players.combat.oldBlockingDamageReduction || damagesource.getDirectEntity() instanceof AbstractArrow) {
                     return -((LivingEntity.this.isDamageSourceBlocked(damagesource)) ? f : 0.0);
+                    } else {
+                        return LivingEntity.this.isBlocking() ? f * 0.5 : 0.0;
+                    }
+                    // Sakura end - old blocking damage reduction
                 }
             };
             float blockingModifier = blocking.apply((double) f).floatValue();
