From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Renner <fallenghostttv@gmail.com>
Date: Mon, 5 Aug 2024 22:26:03 -0400
Subject: [PATCH] Configure fluids breaking wool carpet


diff --git a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
index 626b3c04547ca253b254604c8bf5fa004eed806a..bb0aa18fed97cbc8557a2ec6ca26a530cb112ccc 100644
--- a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
@@ -467,7 +467,7 @@ public abstract class FlowingFluid extends Fluid {
         if (block instanceof LiquidBlockContainer ifluidcontainer) {
             return ifluidcontainer.canPlaceLiquid((Player) null, world, pos, state, fluid);
             // Sakura start
-        } else if (world instanceof Level level && !level.sakuraConfig().technical.redstone.fluidsBreakRedstone && state.isSignalSource()) {
+        } else if (world instanceof Level level && !level.sakuraConfig().technical.redstone.fluidsBreakRedstone && (state.isSignalSource() || state.is(BlockTags.WOOL_CARPETS))) {
             return false;
             // Sakura end
         } else {
