From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 7 Mar 2025 19:16:58 +0000
Subject: [PATCH] Configure breaking blocks outside the world border


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 2679c5062a6385466dc707582ada71b361ec7106..8eb9e93427a4a43f8264e8bd73bca1f35ea9085e 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -565,6 +565,11 @@ public class Explosion {
 
     protected final void searchForBlocks(ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache[] blockCache, final ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache initialCache) {
         // Sakura end - optimise paper explosions
+        // Sakura start - configure breaking blocks when outside the world border
+        if (!this.level.sakuraConfig().cannons.explosion.breakBlocksWhenOutsideTheWorldBorder && !this.level.getWorldBorder().isWithinBounds(new Vec3(this.x, this.y, this.z))) {
+            return;
+        }
+        // Sakura end - configure breaking blocks when outside the world border
         for (int ray = 0, len = CACHED_RAYS.length; ray < len;) {
             ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cachedBlock = initialCache;
 
