From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 14 Mar 2025 18:37:26 +0000
Subject: [PATCH] temp


diff --git a/build.gradle.kts b/build.gradle.kts
index d1b4eb9088d5780f56d8df2d4542e6d408c61e6c..fb9250a4e904e75604094694e1dc021e678d1389 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -201,7 +201,7 @@ tasks.registerRunTask("runReobf") {
     classpath(runtimeClasspathWithoutVanillaServer)
 }
 
-tasks.registerRunTask("runDev") {
-    description = "Spin up a non-relocated Mojang-mapped test server"
-    classpath(sourceSets.main.map { it.runtimeClasspath })
-}
+//tasks.registerRunTask("runDev") {
+//    description = "Spin up a non-relocated Mojang-mapped test server"
+//    classpath(sourceSets.main.map { it.runtimeClasspath })
+//}
