--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -423,6 +_,7 @@
         return this.viewDistanceHolder;
     }
     // Paper end - rewrite chunk system
+    public double trackingRangeModifier = 1.0; // Sakura - entity tracking range modifier
 
     public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile, ClientInformation clientInformation) {
         super(level, level.getSharedSpawnPos(), level.getSharedSpawnAngle(), gameProfile);
