--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -190,7 +_,7 @@
 
         for (int slotsSize = this.slots.size(); i < slotsSize; i++) {
             ItemStack slotContents = this.slots.get(i).getItem();
-            itemsToSend.add(slotContents.copy());
+            itemsToSend.add(slotContents.copyForPacket()); // Sakura - modify data components
             this.remoteSlots.get(i).force(slotContents);
         }
 
@@ -203,7 +_,7 @@
         }
 
         if (this.synchronizer != null) {
-            this.synchronizer.sendInitialData(this, itemsToSend, carried.copy(), this.remoteDataSlots.toIntArray());
+            this.synchronizer.sendInitialData(this, itemsToSend, carried.copyForPacket(), this.remoteDataSlots.toIntArray()); // Sakura - modify data components
             this.synchronizer.sendOffHandSlotChange(); // Paper - Sync offhand slot in menus; update player's offhand since the offhand slot is not added to the slots for menus but can be changed by swapping from a menu slot
         }
     }
@@ -237,7 +_,7 @@
         ItemStack carried = this.getCarried();
         this.remoteCarried.force(carried);
         if (this.synchronizer != null) {
-            this.synchronizer.sendCarriedChange(this, carried.copy());
+            this.synchronizer.sendCarriedChange(this, carried.copyForPacket()); // Sakura - modify data components
         }
     }
     // CraftBukkit end
@@ -314,7 +_,7 @@
     public void synchronizeSlotToRemote(final int i, final ItemStack current, final Supplier<ItemStack> currentCopy) {
         if (!this.suppressRemoteUpdates) {
             RemoteSlot remoteExpected = this.remoteSlots.get(i);
-            if (!remoteExpected.matches(current)) {
+            if (!remoteExpected.matches(current) && !remoteExpected.matches(current.copyForPacket())) { // Sakura - modify data components
                 remoteExpected.force(current);
                 if (this.synchronizer != null) {
                     this.synchronizer.sendSlotChange(this, i, currentCopy.get());
@@ -339,9 +_,16 @@
         if (!this.suppressRemoteUpdates) {
             ItemStack carriedItem = this.getCarried();
             if (!this.remoteCarried.matches(carriedItem)) {
+                // Sakura start - modify data components
+                final ItemStack modifiedItem = carriedItem.copyForPacket();
+                if (this.remoteCarried.matches(modifiedItem)) {
+                    return;
+                }
+                // Sakura end - modify data components
+
                 this.remoteCarried.force(carriedItem);
                 if (this.synchronizer != null) {
-                    this.synchronizer.sendCarriedChange(this, carriedItem.copy());
+                    this.synchronizer.sendCarriedChange(this, modifiedItem); // Sakura - modify data components
                 }
             }
         }
