--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -192,7 +_,7 @@
 
         for (int size = this.slots.size(); i < size; i++) {
             ItemStack item = this.slots.get(i).getItem();
-            list.add(item.copy());
+            list.add(item.copyForPacket()); // Sakura - modify components sent to the client
             this.remoteSlots.get(i).force(item);
         }
 
@@ -205,7 +_,7 @@
         }
 
         if (this.synchronizer != null) {
-            this.synchronizer.sendInitialData(this, list, carried.copy(), this.remoteDataSlots.toIntArray());
+            this.synchronizer.sendInitialData(this, list, carried.copyForPacket(), this.remoteDataSlots.toIntArray()); // Sakura - modify components sent to the client
             this.synchronizer.sendOffHandSlotChange(); // Paper - Sync offhand slot in menus; update player's offhand since the offhand slot is not added to the slots for menus but can be changed by swapping from a menu slot
         }
     }
@@ -239,7 +_,7 @@
         ItemStack carried = this.getCarried();
         this.remoteCarried.force(carried);
         if (this.synchronizer != null) {
-            this.synchronizer.sendCarriedChange(this, carried.copy());
+            this.synchronizer.sendCarriedChange(this, carried.copyForPacket()); // Sakura - modify components sent to the client
         }
     }
     // CraftBukkit end
@@ -316,7 +_,7 @@
     public void synchronizeSlotToRemote(int slotIndex, ItemStack stack, Supplier<ItemStack> supplier) {
         if (!this.suppressRemoteUpdates) {
             RemoteSlot remoteSlot = this.remoteSlots.get(slotIndex);
-            if (!remoteSlot.matches(stack)) {
+            if (!remoteSlot.matches(stack) && !remoteSlot.matches(stack.copyForPacket())) { // Sakura - modify components sent to the client
                 remoteSlot.force(stack);
                 if (this.synchronizer != null) {
                     this.synchronizer.sendSlotChange(this, slotIndex, supplier.get());
@@ -341,9 +_,16 @@
         if (!this.suppressRemoteUpdates) {
             ItemStack carried = this.getCarried();
             if (!this.remoteCarried.matches(carried)) {
+                // Sakura start - modify components sent to the client
+                final ItemStack modifiedItem = carried.copyForPacket();
+                if (this.remoteCarried.matches(modifiedItem)) {
+                    return;
+                }
+                // Sakura end - modify components sent to the client
+
                 this.remoteCarried.force(carried);
                 if (this.synchronizer != null) {
-                    this.synchronizer.sendCarriedChange(this, carried.copy());
+                    this.synchronizer.sendCarriedChange(this, modifiedItem); // Sakura - modify components sent to the client
                 }
             }
         }
